From 852860529f24547b662d9383c0eaa821c9efa406 Mon Sep 17 00:00:00 2001
From: Xin Ouyang <Xin.Ouyang@windriver.com>
Date: Thu, 22 Aug 2013 19:18:39 +0800
Subject: [PATCH] efpolicy: fix real path for fstools

Upstream-Status: Inappropriate [configuration]

Signed-off-by: Xin Ouyang <Xin.Ouyang@windriver.com>
---
 policy/modules/system/fstools.fc |   11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/policy/modules/system/fstools.fc b/policy/modules/system/fstools.fc
index 7a46b45..9656352 100644
--- a/policy/modules/system/fstools.fc
+++ b/policy/modules/system/fstools.fc
@@ -1,6 +1,8 @@
 /sbin/badblocks		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/blkid		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
+/sbin/blkid\.util-linux	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/blockdev		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
+/sbin/blockdev\.util-linux	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/cfdisk		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/dosfsck		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/dump		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
@@ -9,9 +11,12 @@
 /sbin/e4fsck		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/e2label		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/fdisk		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
+/sbin/fdisk\.util-linux	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/findfs		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
+/usr/sbin/findfs		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/fsck.*		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/hdparm		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
+/sbin/hdparm\.hdparm	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/install-mbr	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/jfs_.*		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/losetup.*		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
@@ -24,21 +29,27 @@
 /sbin/mkraid		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/mkreiserfs	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/mkswap		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
+/sbin/mkswap\.util-linux	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/parted		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
+/usr/sbin/parted		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/partprobe		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
+/usr/sbin/partprobe		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/partx		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
+/usr/sbin/partx		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/raidautorun	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/raidstart		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/reiserfs(ck|tune)	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/resize.*fs	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/scsi_info		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/sfdisk		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
+/usr/sbin/sfdisk		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/swapoff		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/swapon.*		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/tune2fs		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 
 /usr/bin/partition_uuid	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /usr/bin/raw		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
+/usr/sbin/raw		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /usr/bin/scsi_unique_id	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /usr/bin/syslinux	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 
-- 
1.7.9.5

