From 3615e2d67f402a37ae7333e62b54f1d9d0a3bfd1 Mon Sep 17 00:00:00 2001
From: Xin Ouyang <Xin.Ouyang@windriver.com>
Date: Thu, 22 Aug 2013 19:27:19 +0800
Subject: [PATCH] refpolicy: fix real path for screen

Upstream-Status: Inappropriate [configuration]

Signed-off-by: Xin Ouyang <Xin.Ouyang@windriver.com>
---
 policy/modules/contrib/screen.fc |    1 +
 1 file changed, 1 insertion(+)

diff --git a/policy/modules/contrib/screen.fc b/policy/modules/contrib/screen.fc
index e7c2cf7..49ddca2 100644
--- a/policy/modules/contrib/screen.fc
+++ b/policy/modules/contrib/screen.fc
@@ -3,6 +3,7 @@ HOME_DIR/\.screenrc	--	gen_context(system_u:object_r:screen_home_t,s0)
 HOME_DIR/\.tmux\.conf	--	gen_context(system_u:object_r:screen_home_t,s0)
 
 /usr/bin/screen	--	gen_context(system_u:object_r:screen_exec_t,s0)
+/usr/bin/screen-.*	--	gen_context(system_u:object_r:screen_exec_t,s0)
 /usr/bin/tmux	--	gen_context(system_u:object_r:screen_exec_t,s0)
 
 /var/run/screen(/.*)?	gen_context(system_u:object_r:screen_var_run_t,s0)
-- 
1.7.9.5

