From 845518a6f196e6e8c49ba38791c85e17276920e1 Mon Sep 17 00:00:00 2001
From: Xin Ouyang <Xin.Ouyang@windriver.com>
Date: Thu, 22 Aug 2013 13:37:23 +0800
Subject: [PATCH 3/4] fix update-alternatives for hostname

Upstream-Status: Inappropriate [only for Poky]

Signed-off-by: Xin Ouyang <Xin.Ouyang@windriver.com>
---
 policy/modules/system/hostname.fc |    1 +
 1 file changed, 1 insertion(+)

diff --git a/policy/modules/system/hostname.fc b/policy/modules/system/hostname.fc
index 9dfecf7..4003b6d 100644
--- a/policy/modules/system/hostname.fc
+++ b/policy/modules/system/hostname.fc
@@ -1,2 +1,3 @@
 
 /bin/hostname		--	gen_context(system_u:object_r:hostname_exec_t,s0)
+/bin/hostname\.net-tools	--	gen_context(system_u:object_r:hostname_exec_t,s0)
-- 
1.7.9.5

