From 642fab321a5f1f40495b4ca07f1fca4145024986 Mon Sep 17 00:00:00 2001
From: Xin Ouyang <Xin.Ouyang@windriver.com>
Date: Thu, 22 Aug 2013 19:25:36 +0800
Subject: [PATCH] refpolicy: fix real path for nscd

Upstream-Status: Inappropriate [configuration]

Signed-off-by: Xin Ouyang <Xin.Ouyang@windriver.com>
---
 policy/modules/contrib/nscd.fc |    1 +
 1 file changed, 1 insertion(+)

diff --git a/policy/modules/contrib/nscd.fc b/policy/modules/contrib/nscd.fc
index ba64485..61a6f24 100644
--- a/policy/modules/contrib/nscd.fc
+++ b/policy/modules/contrib/nscd.fc
@@ -1,6 +1,7 @@
 /etc/rc\.d/init\.d/nscd	--	gen_context(system_u:object_r:nscd_initrc_exec_t,s0)
 
 /usr/sbin/nscd	--	gen_context(system_u:object_r:nscd_exec_t,s0)
+/usr/bin/nscd	--	gen_context(system_u:object_r:nscd_exec_t,s0)
 
 /var/cache/nscd(/.*)?	gen_context(system_u:object_r:nscd_var_run_t,s0)
 
-- 
1.7.9.5

