From 843299c135c30b036ed163a10570a1d5efe36ff8 Mon Sep 17 00:00:00 2001
From: Xin Ouyang <Xin.Ouyang@windriver.com>
Date: Thu, 22 Aug 2013 13:37:23 +0800
Subject: [PATCH 1/2] fix xconsole_device_t as a dev_node.

Upstream-Status: Inappropriate [only for Poky]

Signed-off-by: Xin Ouyang <Xin.Ouyang@windriver.com>
---
 policy/modules/services/xserver.te |    1 +
 1 file changed, 1 insertion(+)

diff --git a/policy/modules/services/xserver.te b/policy/modules/services/xserver.te
index 4f6d693..b00f004 100644
--- a/policy/modules/services/xserver.te
+++ b/policy/modules/services/xserver.te
@@ -151,6 +151,7 @@ userdom_user_tmp_file(xauth_tmp_t)
 # this is not actually a device, its a pipe
 type xconsole_device_t;
 files_type(xconsole_device_t)
+dev_node(xconsole_device_t)
 fs_associate_tmpfs(xconsole_device_t)
 files_associate_tmp(xconsole_device_t)
 
-- 
1.7.9.5

