From 86fd93b15b97042bcf5ff1b2d1228e7d64b8cfd1 Mon Sep 17 00:00:00 2001
From: Wenzong Fan <wenzong.fan@windriver.com>
Date: Tue, 7 Jan 2014 22:22:00 -0500
Subject: [PATCH] refpolicy: fix real path for udevd

Upstream-Status: Inappropriate [configuration]

Signed-off-by: Wenzong Fan <wenzong.fan@windriver.com>
---
 policy/modules/system/udev.fc |    1 +
 1 file changed, 1 insertion(+)

diff --git a/policy/modules/system/udev.fc b/policy/modules/system/udev.fc
index 40928d8..36d471d 100644
--- a/policy/modules/system/udev.fc
+++ b/policy/modules/system/udev.fc
@@ -10,6 +10,7 @@
 /etc/udev/scripts/.+ --	gen_context(system_u:object_r:udev_helper_exec_t,s0)
 
 /lib/udev/udev-acl --	gen_context(system_u:object_r:udev_exec_t,s0)
+/lib/udev/udevd    --	gen_context(system_u:object_r:udev_exec_t,s0)
 
 ifdef(`distro_debian',`
 /lib/udev/create_static_nodes -- gen_context(system_u:object_r:udev_exec_t,s0)
-- 
1.7.9.5

