Subject: [PATCH] fix real path for login commands.

Upstream-Status: Inappropriate [only for Poky]

Signed-off-by: Xin Ouyang <Xin.Ouyang@windriver.com>
---
 policy/modules/system/authlogin.fc |    7 ++++---
 1 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/policy/modules/system/authlogin.fc b/policy/modules/system/authlogin.fc
index 28ad538..c8dd17f 100644
--- a/policy/modules/system/authlogin.fc
+++ b/policy/modules/system/authlogin.fc
@@ -1,5 +1,7 @@
 
 /bin/login		--	gen_context(system_u:object_r:login_exec_t,s0)
+/bin/login\.shadow	--	gen_context(system_u:object_r:login_exec_t,s0)
+/bin/login\.tinylogin	--	gen_context(system_u:object_r:login_exec_t,s0)
 
 /etc/\.pwd\.lock	--	gen_context(system_u:object_r:shadow_t,s0)
 /etc/group\.lock	--	gen_context(system_u:object_r:shadow_t,s0)
@@ -9,9 +11,9 @@
 
 /sbin/pam_console_apply	 --	gen_context(system_u:object_r:pam_console_exec_t,s0)
 /sbin/pam_timestamp_check --	gen_context(system_u:object_r:pam_exec_t,s0)
-/sbin/unix_chkpwd	--	gen_context(system_u:object_r:chkpwd_exec_t,s0)
-/sbin/unix_update	--	gen_context(system_u:object_r:updpwd_exec_t,s0)
-/sbin/unix_verify	--	gen_context(system_u:object_r:chkpwd_exec_t,s0)
+/usr/sbin/unix_chkpwd	--	gen_context(system_u:object_r:chkpwd_exec_t,s0)
+/usr/sbin/unix_update	--	gen_context(system_u:object_r:updpwd_exec_t,s0)
+/usr/sbin/unix_verify	--	gen_context(system_u:object_r:chkpwd_exec_t,s0)
 ifdef(`distro_suse', `
 /sbin/unix2_chkpwd	--	gen_context(system_u:object_r:chkpwd_exec_t,s0)
 ')
-- 
1.7.5.4

