Subject: [PATCH] refpolicy: fix real path for su

Upstream-Status: Inappropriate [configuration]

Signed-off-by: Xin Ouyang <Xin.Ouyang@windriver.com>
---
 policy/modules/admin/su.fc | 1 +
 1 file changed, 1 insertion(+)

diff --git a/policy/modules/admin/su.fc b/policy/modules/admin/su.fc
index 688abc2..a563687 100644
--- a/policy/modules/admin/su.fc
+++ b/policy/modules/admin/su.fc
@@ -1,5 +1,6 @@
 
 /bin/su			--	gen_context(system_u:object_r:su_exec_t,s0)
+/usr/bin/su		--	gen_context(system_u:object_r:su_exec_t,s0)
 
 /usr/(local/)?bin/ksu	--	gen_context(system_u:object_r:su_exec_t,s0)
 /usr/bin/kdesu		--	gen_context(system_u:object_r:su_exec_t,s0)
-- 
1.7.11.7

