Subject: [PATCH] refpolicy: fix real path for clock

Upstream-Status: Inappropriate [configuration]

Signed-off-by: Xin Ouyang <Xin.Ouyang@windriver.com>
---
 policy/modules/system/clock.fc | 1 +
 1 file changed, 1 insertion(+)

diff --git a/policy/modules/system/clock.fc b/policy/modules/system/clock.fc
index c5e05ca..a74c40c 100644
--- a/policy/modules/system/clock.fc
+++ b/policy/modules/system/clock.fc
@@ -2,4 +2,5 @@
 /etc/adjtime		--	gen_context(system_u:object_r:adjtime_t,s0)
 
 /sbin/hwclock		--	gen_context(system_u:object_r:hwclock_exec_t,s0)
+/sbin/hwclock\.util-linux	--	gen_context(system_u:object_r:hwclock_exec_t,s0)
 
-- 
1.7.11.7

