Index: refpolicy/policy/modules/contrib/devicekit.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/devicekit.te
+++ refpolicy/policy/modules/contrib/devicekit.te
@@ -234,6 +234,8 @@ kernel_setsched(devicekit_power_t)
 corecmd_exec_bin(devicekit_power_t)
 corecmd_exec_shell(devicekit_power_t)
 
+dbusbouncer_dbus_chat(devicekit_power_t)
+
 dev_read_input(devicekit_power_t)
 dev_read_urand(devicekit_power_t)
 dev_rw_generic_usb_dev(devicekit_power_t)
