Index: refpolicy/policy/modules/system/lvm.fc
===================================================================
--- refpolicy.orig/policy/modules/system/lvm.fc
+++ refpolicy/policy/modules/system/lvm.fc
@@ -57,6 +57,7 @@ ifdef(`distro_gentoo',`
 /sbin/lvdisplay		--	gen_context(system_u:object_r:lvm_exec_t,s0)
 /sbin/lvextend		--	gen_context(system_u:object_r:lvm_exec_t,s0)
 /sbin/lvm		--	gen_context(system_u:object_r:lvm_exec_t,s0)
+/sbin/lvmetad		--	gen_context(system_u:object_r:lvm_exec_t,s0)
 /sbin/lvm\.static	--	gen_context(system_u:object_r:lvm_exec_t,s0)
 /sbin/lvmchange		--	gen_context(system_u:object_r:lvm_exec_t,s0)
 /sbin/lvmdiskscan	--	gen_context(system_u:object_r:lvm_exec_t,s0)
@@ -106,6 +107,7 @@ ifdef(`distro_gentoo',`
 /usr/sbin/clvmd		--	gen_context(system_u:object_r:clvmd_exec_t,s0)
 /usr/sbin/cryptsetup	--	gen_context(system_u:object_r:lvm_exec_t,s0)
 /usr/sbin/lvm		--	gen_context(system_u:object_r:lvm_exec_t,s0)
+/usr/sbin/lvmetad	--	gen_context(system_u:object_r:lvm_exec_t,s0)
 
 #
 # /var
@@ -116,3 +118,6 @@ ifdef(`distro_gentoo',`
 /var/lock/lvm(/.*)?		gen_context(system_u:object_r:lvm_lock_t,s0)
 /var/run/multipathd\.sock -s	gen_context(system_u:object_r:lvm_var_run_t,s0)
 /var/run/dmevent.*		gen_context(system_u:object_r:lvm_var_run_t,s0)
+/var/run/lvm(/.*)?		gen_context(system_u:object_r:lvm_var_run_t,s0)
+/run/lvm(/.*)?			gen_context(system_u:object_r:lvm_var_run_t,s0)
+
