Index: refpolicy/policy/modules/contrib/dbus.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/dbus.te
+++ refpolicy/policy/modules/contrib/dbus.te
@@ -258,4 +258,12 @@ optional_policy(`
 #
 
 allow dbusd_unconfined { system_dbusd_t session_bus_type dbusd_session_bus_client dbusd_system_bus_client }:dbus all_dbus_perms;
-allow { dbusd_session_bus_client dbusd_system_bus_client } dbusd_unconfined:dbus send_msg;
+# This rule was added by upstream refpolicy commit
+# 6bef7a14757124c56fadc08c255e9dd6c29a15f9 in order to allow
+# dbus clients to answer to dbus unconfined domains.  However,
+# this is overly permissive and dangerous, since it exposes
+# dbus unconfined domains (i.e. those domains which are allowed
+# to send to any dbus destination) to any dbus client at all.
+# Disabling this for OpenXT.  Long term solution is to remove
+# unconfined altogether from OpenXT policy - sds.
+#allow { dbusd_session_bus_client dbusd_system_bus_client } dbusd_unconfined:dbus send_msg;
