Index: refpolicy/policy/modules/contrib/hal.if
===================================================================
--- refpolicy.orig/policy/modules/contrib/hal.if
+++ refpolicy/policy/modules/contrib/hal.if
@@ -421,6 +421,24 @@ interface(`hal_manage_pid_dirs',`
 
 ########################################
 ## <summary>
+##     setattr on hald PID files.
+## </summary>
+## <param name="domain">
+##     <summary>
+##     Domain allowed access.
+##     </summary>
+## </param>
+#
+interface(`hal_setattr_pid_dirs',`
+	gen_require(`
+		type hald_var_run_t;
+	')
+
+	allow $1 hald_var_run_t:dir setattr;
+')
+
+########################################
+## <summary>
 ##	Create, read, write, and delete
 ##	hald pid files.
 ## </summary>
@@ -438,3 +456,21 @@ interface(`hal_manage_pid_files',`
 	files_search_pids($1)
 	manage_files_pattern($1, hald_var_run_t, hald_var_run_t)
 ')
+
+########################################
+## <summary>
+##	Set attributes on hal cache directory.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`hal_setattr_cache_dirs',`
+	gen_require(`
+		type hald_cache_t;
+	')
+
+	allow $1 hald_cache_t:dir setattr;
+')
Index: refpolicy/policy/modules/contrib/hal.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/hal.te
+++ refpolicy/policy/modules/contrib/hal.te
@@ -18,7 +18,7 @@ domain_entry_file(hald_acl_t, hald_acl_e
 role system_r types hald_acl_t;
 
 type hald_cache_t;
-files_pid_file(hald_cache_t)
+files_type(hald_cache_t)
 
 type hald_dccm_t, hald_domain;
 type hald_dccm_exec_t;
@@ -136,6 +136,9 @@ dev_rw_power_management(hald_t)
 dev_read_raw_memory(hald_t)
 dev_rw_sysfs(hald_t)
 dev_read_video_dev(hald_t)
+dev_search_xen(hald_t)
+dev_getattr_generic_blk_files(hald_t)
+dev_getattr_xen(hald_t)
 
 domain_use_interactive_fds(hald_t)
 domain_read_all_domains_state(hald_t)
@@ -173,6 +176,7 @@ selinux_compute_create_context(hald_t)
 selinux_compute_relabel_context(hald_t)
 selinux_compute_user_contexts(hald_t)
 
+storage_getattr_blktap(hald_t)
 storage_raw_read_removable_device(hald_t)
 storage_raw_write_removable_device(hald_t)
 storage_raw_read_fixed_disk(hald_t)
