Index: refpolicy/policy/booleans.conf
===================================================================
--- refpolicy.orig/policy/booleans.conf
+++ refpolicy/policy/booleans.conf
@@ -1223,7 +1223,7 @@ console_login = true
 # stack smashing protection.  All domains will
 # be allowed to read from /dev/urandom.
 # 
-global_ssp = false
+global_ssp = true
 
 #
 # Allow email client to various content.
