Index: refpolicy/policy/modules/contrib/dpkg.fc
===================================================================
--- refpolicy.orig/policy/modules/contrib/dpkg.fc
+++ refpolicy/policy/modules/contrib/dpkg.fc
@@ -1,11 +1,11 @@
 /etc/cron\.daily/dpkg	--	gen_context(system_u:object_r:dpkg_exec_t,s0)
 
 /usr/bin/debsums	--	gen_context(system_u:object_r:dpkg_exec_t,s0)
-/usr/bin/dpkg	--	gen_context(system_u:object_r:dpkg_exec_t,s0)
+/usr/bin/(d|o)pkg(-cl|-key)?                   --      gen_context(system_u:object_r:dpkg_exec_t,s0)
 /usr/bin/dselect	--	gen_context(system_u:object_r:dpkg_exec_t,s0)
 
-/var/lib/dpkg(/.*)?	gen_context(system_u:object_r:dpkg_var_lib_t,s0)
-/var/lib/dpkg/(meth)?lock	--	gen_context(system_u:object_r:dpkg_lock_t,s0)
+/var/lib/(d|o)pkg(/.*)?                        gen_context(system_u:object_r:dpkg_var_lib_t,s0)
+/var/lib/(d|o)pkg/(meth)?lock  --      gen_context(system_u:object_r:dpkg_lock_t,s0)
 
 /usr/sbin/dpkg-preconfigure	--	gen_context(system_u:object_r:dpkg_exec_t,s0)
 /usr/sbin/dpkg-reconfigure	--	gen_context(system_u:object_r:dpkg_exec_t,s0)
Index: refpolicy/policy/modules/contrib/dpkg.if
===================================================================
--- refpolicy.orig/policy/modules/contrib/dpkg.if
+++ refpolicy/policy/modules/contrib/dpkg.if
@@ -242,3 +242,23 @@ interface(`dpkg_lock_db',`
 	allow $1 dpkg_var_lib_t:dir list_dir_perms;
 	allow $1 dpkg_lock_t:file manage_file_perms;
 ')
+
+########################################
+## <summary>
+##	Do not audit attempts to lock the dpkg
+##	package database.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	The type of the process performing this action.
+##	</summary>
+## </param>
+#
+interface(`dpkg_dontaudit_lock_db',`
+	gen_require(`
+		type dpkg_lock_t, dpkg_var_lib_t;
+	')
+
+	dontaudit $1 dpkg_var_lib_t:dir list_dir_perms;
+	dontaudit $1 dpkg_lock_t:file manage_file_perms;
+')
Index: refpolicy/policy/modules/contrib/dpkg.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/dpkg.te
+++ refpolicy/policy/modules/contrib/dpkg.te
@@ -80,6 +80,7 @@ fs_tmpfs_filetrans(dpkg_t, dpkg_tmpfs_t,
 allow dpkg_t dpkg_var_lib_t:file mmap_file_perms;
 manage_files_pattern(dpkg_t, dpkg_var_lib_t, dpkg_var_lib_t)
 files_var_lib_filetrans(dpkg_t, dpkg_var_lib_t, dir)
+files_usr_filetrans(dpkg_t, dpkg_var_lib_t, dir)
 
 kernel_read_system_state(dpkg_t)
 kernel_read_kernel_sysctls(dpkg_t)
@@ -139,6 +140,7 @@ auth_dontaudit_read_shadow(dpkg_t)
 
 init_all_labeled_script_domtrans(dpkg_t)
 init_use_script_ptys(dpkg_t)
+init_dontaudit_use_fds(dpkg_t)
 
 libs_exec_ld_so(dpkg_t)
 libs_exec_lib_files(dpkg_t)
