Index: refpolicy/policy/modules/contrib/brctl.fc
===================================================================
--- refpolicy.orig/policy/modules/contrib/brctl.fc
+++ refpolicy/policy/modules/contrib/brctl.fc
@@ -1 +1,2 @@
 /usr/sbin/brctl	--	gen_context(system_u:object_r:brctl_exec_t,s0)
+/usr/sbin/brctl.bridge-utils   --      gen_context(system_u:object_r:brctl_exec_t,s0)
Index: refpolicy/policy/modules/contrib/brctl.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/brctl.te
+++ refpolicy/policy/modules/contrib/brctl.te
@@ -26,7 +26,8 @@ allow brctl_t self:tcp_socket create_soc
 kernel_request_load_module(brctl_t)
 kernel_read_network_state(brctl_t)
 kernel_read_sysctl(brctl_t)
-
+kernel_dontaudit_read_xen_state(brctl_t)
+kernel_dontaudit_write_xen_state(brctl_t)
 corenet_rw_tun_tap_dev(brctl_t)
 
 dev_rw_sysfs(brctl_t)
