Index: refpolicy/policy/modules/roles/staff.te
===================================================================
--- refpolicy.orig/policy/modules/roles/staff.te
+++ refpolicy/policy/modules/roles/staff.te
@@ -40,6 +40,7 @@ optional_policy(`
 
 optional_policy(`
 	ssh_role_template(staff, staff_r, staff_t)
+	ssh_inherit_env(staff_t)
 ')
 
 optional_policy(`
@@ -48,6 +49,7 @@ optional_policy(`
 
 optional_policy(`
 	sysadm_role_change(staff_r)
+	userdom_dontaudit_signal_admin_users(staff_t)
 	userdom_dontaudit_use_user_terminals(staff_t)
 ')
 
