Index: refpolicy/policy/modules/contrib/xen.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/xen.te
+++ refpolicy/policy/modules/contrib/xen.te
@@ -282,6 +282,10 @@ blktap_read_blk_file(xend_t)
 tapdisk_domtrans(xend_t)
 tapctl_domtrans(xend_t)
 
+# vusb domain transition
+vusbc_exec_domtrans(xend_t)
+vusbd_dbus_send(xend_t)
+
 dbd_dbus_chat(xend_t)
 surfman_dbus_chat(xend_t)
 uid_dbus_chat(xend_t)
