Index: refpolicy/policy/modules/system/lvm.te
===================================================================
--- refpolicy.orig/policy/modules/system/lvm.te
+++ refpolicy/policy/modules/system/lvm.te
@@ -413,4 +413,5 @@ optional_policy(`
 	updatemgr_dontaudit_use_fd(lvm_t)
 	updatemgr_dontaudit_rw_fifo_files(lvm_t)
 	updatemgr_dontaudit_rw_stream_sockets(lvm_t)
+	xc_installer_read_tmp_files(lvm_t)
 ')
