From 26f25ed8b41240e0268e2e4472a4dc4564ffe146 Mon Sep 17 00:00:00 2001
From: Andreas Schwab <schwab@redhat.com>
Date: Thu, 10 Jun 2010 16:54:06 +0200
Subject: [PATCH] Don't assume AT_PAGESIZE is always available

---
 sysdeps/unix/sysv/linux/getpagesize.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/sysdeps/unix/sysv/linux/getpagesize.c b/sysdeps/unix/sysv/linux/getpagesize.c
index 0866079..fe9bc8e 100644
--- a/sysdeps/unix/sysv/linux/getpagesize.c
+++ b/sysdeps/unix/sysv/linux/getpagesize.c
@@ -28,7 +28,7 @@
 int
 __getpagesize ()
 {
-#ifdef __ASSUME_AT_PAGESIZE
+#if 0 && defined __ASSUME_AT_PAGESIZE
   assert (GLRO(dl_pagesize) != 0);
   return GLRO(dl_pagesize);
 #else
-- 
1.7.1
