Index: ocaml-4.04.0/configure
===================================================================
--- ocaml-4.04.0.orig/configure
+++ ocaml-4.04.0/configure
@@ -138,6 +138,8 @@ while : ; do
         target_type=$2; shift;;
     -cc*)
         ccoption="$2"; shift;;
+    -mksharedlib)
+        mksharedliboption="$2"; shift;;
     -as)
         asoption="$2"; shift;;
     -aspp)
@@ -700,10 +702,15 @@ if test $with_sharedlibs = "yes"; then
     *-*-linux-gnu|*-*-linux|*-*-freebsd[3-9]*|*-*-freebsd[1-9][0-9]*\
     |*-*-openbsd*|*-*-netbsd*|*-*-dragonfly*|*-*-gnu*|*-*-haiku*)
       sharedcccompopts="-fPIC"
-      mksharedlib="$bytecc -shared"
+      if [[ -z $mksharedliboption ]]; then
+         mksharedlib="$bytecc -shared"
+         mksharedlibrpath="-Wl,-rpath-link,"
+      else
+         mksharedlib="$mksharedliboption"
+         mksharedlibrpath="-rpath "
+      fi
       bytecclinkopts="$bytecclinkopts -Wl,-E"
       byteccrpath="-Wl,-rpath,"
-      mksharedlibrpath="-Wl,-rpath,"
       natdynlinkopts="-Wl,-E"
       shared_libraries_supported=true;;
     alpha*-*-osf*)
Index: ocaml-4.04.0/Makefile
===================================================================
--- ocaml-4.04.0.orig/Makefile
+++ ocaml-4.04.0/Makefile
@@ -270,7 +270,11 @@ installopt:
 	for i in $(OTHERLIBRARIES); \
 	  do (cd otherlibs/$$i; $(MAKE) installopt) || exit $$?; done
 	if test -f ocamlopt.opt ; then $(MAKE) installoptopt; else \
-	   cd $(INSTALL_BINDIR); ln -sf ocamlopt.byte$(EXE) ocamlopt$(EXE); fi
+	   cd $(INSTALL_BINDIR); \
+           ln -sf ocamlc.byte$(EXE) ocamlc$(EXE); \
+           ln -sf ocamlopt.byte$(EXE) ocamlopt$(EXE); \
+           ln -sf ocamllex.byte$(EXE) ocamllex$(EXE); \
+        fi
 	cd tools; $(MAKE) installopt
 
 installoptopt:
