diff --git a/configure b/configure
index aa514f1..3d6a9c6 100755
--- a/configure
+++ b/configure
@@ -75,6 +75,8 @@ while : ; do
         host_type=$2; shift;;
     -cc*)
         ccoption="$2"; shift;;
+    -mksharedlib)
+	mksharedliboption="$2"; shift;;
     -as)
         asoption="$2"; shift;;
     -aspp)
@@ -514,10 +516,15 @@ if test $withsharedlibs = "yes"; then
     *-*-linux-gnu|*-*-linux|*-*-freebsd[3-9]*|*-*-openbsd*|*-*-netbsd*|*-*-gnu*)
       cmxs="cmxs"
       sharedcccompopts="-fPIC"
-      mksharedlib="$bytecc -shared"
+      if [[ -z $mksharedliboption ]]; then
+	  mksharedlib="$bytecc -shared"
+	  mksharedlibrpath="-Wl,-rpath-link,"
+      else
+	  mksharedlib="$mksharedliboption"
+	  mksharedlibrpath="-rpath "
+      fi
       bytecclinkopts="$bytecclinkopts -Wl,-E"
       byteccrpath="-Wl,-rpath,"
-      mksharedlibrpath="-Wl,-rpath,"
       natdynlinkopts="-Wl,-E"
       shared_libraries_supported=true;;
     alpha*-*-osf*)
