commit 4ab6b5ce77b7fa93fb7038fd3b97540dcb6a74c9
Author: Damien Doligez <damien.doligez@inria.fr>
Date:   Wed Sep 28 11:28:00 2016 +0200

    PR#7378: Manual major GC fails to compact the heap

diff --git a/byterun/gc_ctrl.c b/byterun/gc_ctrl.c
index a0864d1..4e3f833 100644
--- a/byterun/gc_ctrl.c
+++ b/byterun/gc_ctrl.c
@@ -423,7 +423,7 @@ static void test_and_compact (void)
   caml_gc_message (0x200, "Estimated overhead (lower bound) = %"
                           ARCH_INTNAT_PRINTF_FORMAT "u%%\n",
                    (uintnat) fp);
-  if (fp >= caml_percent_max && caml_stat_heap_chunks > 1){
+  if (fp >= caml_percent_max){
     caml_gc_message (0x200, "Automatic compaction triggered.\n", 0);
     caml_compact_heap ();
   }
