Index: git/e2fsck/problem.c
===================================================================
--- git.orig/e2fsck/problem.c
+++ git/e2fsck/problem.c
@@ -341,12 +341,14 @@ static struct e2fsck_problem problem_tab
 	/* Last mount time is in the future */
 	{ PR_0_FUTURE_SB_LAST_MOUNT,
 	  N_("@S last mount time (%t,\n\tnow = %T) is in the future.\n"),
-	  PROMPT_FIX, PR_PREEN_OK | PR_NO_OK },
+	  PROMPT_NONE, PR_PREEN_OK
+	},
 
 	/* Last write time is in the future */
 	{ PR_0_FUTURE_SB_LAST_WRITE,
 	  N_("@S last write time (%t,\n\tnow = %T) is in the future.\n"),
-	  PROMPT_FIX, PR_PREEN_OK | PR_NO_OK },
+	  PROMPT_NONE, PR_PREEN_OK
+	},
 
 	{ PR_0_EXTERNAL_JOURNAL_HINT,
 	  N_("@S hint for external superblock @s %X.  "),
@@ -390,13 +392,13 @@ static struct e2fsck_problem problem_tab
 	{ PR_0_FUTURE_SB_LAST_MOUNT_FUDGED,
 	  N_("@S last mount time is in the future.\n\t(by less than a day, "
 	     "probably due to the hardware clock being incorrectly set)\n"),
-	  PROMPT_NONE, PR_PREEN_OK | PR_NO_OK },
+	  PROMPT_NONE, PR_PREEN_OK },
 
 	/* Last write time is in the future (fudged) */
 	{ PR_0_FUTURE_SB_LAST_WRITE_FUDGED,
 	  N_("@S last write time is in the future.\n\t(by less than a day, "
 	     "probably due to the hardware clock being incorrectly set)\n"),
-	  PROMPT_NONE, PR_PREEN_OK | PR_NO_OK },
+	  PROMPT_NONE, PR_PREEN_OK },
 
 	/* Block group checksum (latch question) is invalid. */
 	{ PR_0_GDT_CSUM_LATCH,
