diff --git a/configure b/configure
index 3c72fa0..0a184e2 100755
--- a/configure
+++ b/configure
@@ -1355,7 +1355,8 @@ fi
 # xen probe
 
 if test "$xen" != "no" ; then
-  xen_libs="-lxenstore -lxenctrl -lxenguest"
+  # We need to add -ldl and -lpthread because XenStore depends on it
+  xen_libs="-lxenstore -lxenctrl -lxenguest -ldl -lpthread"
 
   # First we test whether Xen headers and libraries are available.
   # If no, we are done and there is no Xen support.
