diff --git a/policy/modules/contrib/devicekit.te b/policy/modules/contrib/devicekit.te
index 0e6fbcd..72bdfe5 100644
--- a/policy/modules/contrib/devicekit.te
+++ b/policy/modules/contrib/devicekit.te
@@ -229,6 +229,8 @@ kernel_setsched(devicekit_power_t)
 corecmd_exec_bin(devicekit_power_t)
 corecmd_exec_shell(devicekit_power_t)
 
+dbusbouncer_dbus_chat(devicekit_power_t)
+
 dev_read_input(devicekit_power_t)
 dev_read_urand(devicekit_power_t)
 dev_rw_generic_usb_dev(devicekit_power_t)
