diff --git a/policy/modules/kernel/corenetwork.if.in b/policy/modules/kernel/corenetwork.if.in
index a6a115d..ede3f7c 100644
--- a/policy/modules/kernel/corenetwork.if.in
+++ b/policy/modules/kernel/corenetwork.if.in
@@ -3042,3 +3042,22 @@ interface(`corenet_unconfined',`
 
 	typeattribute $1 corenet_unconfined_type;
 ')
+
+########################################
+## <summary>
+##      Unconfined read access to network objects.
+## </summary>
+## <param name="domain">
+##      <summary>
+##      The domain allowed access.
+##      </summary>
+## </param>
+#
+interface(`corenet_unconfined_readonly',`
+        gen_require(`
+                attribute corenet_unconfined_readonly_type;
+        ')
+
+        typeattribute $1 corenet_unconfined_readonly_type;
+')
+
diff --git a/policy/modules/kernel/corenetwork.te.in b/policy/modules/kernel/corenetwork.te.in
index a674ff6..36eb88c 100644
--- a/policy/modules/kernel/corenetwork.te.in
+++ b/policy/modules/kernel/corenetwork.te.in
@@ -16,6 +16,7 @@ attribute rpc_port_type;
 attribute server_packet_type;
 
 attribute corenet_unconfined_type;
+attribute corenet_unconfined_readonly_type;
 
 type ppp_device_t;
 dev_node(ppp_device_t)
@@ -272,3 +273,10 @@ allow corenet_unconfined_type port_type:udp_socket { send_msg recv_msg };
 # Bind to any network address.
 allow corenet_unconfined_type port_type:{ tcp_socket udp_socket } name_bind;
 allow corenet_unconfined_type node_type:{ tcp_socket udp_socket rawip_socket } node_bind;
+
+########################################
+#
+# Unconfined readonly access to this module
+#
+allow corenet_unconfined_readonly_type port_type:tcp_socket { send_msg recv_msg };
+allow corenet_unconfined_readonly_type port_type:udp_socket { send_msg recv_msg };
