diff --git a/policy/modules/contrib/alsa.fc b/policy/modules/contrib/alsa.fc
index 5de1e01..dc73a65 100644
--- a/policy/modules/contrib/alsa.fc
+++ b/policy/modules/contrib/alsa.fc
@@ -6,17 +6,19 @@ HOME_DIR/\.asoundrc	--	gen_context(system_u:object_r:alsa_home_t,s0)
 /etc/alsa/pcm(/.*)?	gen_context(system_u:object_r:alsa_etc_rw_t,s0)
 /etc/asound(/.*)?	gen_context(system_u:object_r:alsa_etc_rw_t,s0)
 /etc/asound\.state	--	gen_context(system_u:object_r:alsa_etc_rw_t,s0)
+/var/volatile/etc/asound(/.*)?	gen_context(system_u:object_r:alsa_etc_rw_t,s0)
 
 /sbin/alsactl	--	gen_context(system_u:object_r:alsa_exec_t,s0)
 /sbin/salsa	--	gen_context(system_u:object_r:alsa_exec_t,s0)
 
+/tmp/alsactl.dump       --      gen_context(system_u:object_r:alsa_tmp_t,s0)
 /usr/bin/ainit	--	gen_context(system_u:object_r:alsa_exec_t,s0)
 /usr/bin/alsaunmute	--	gen_context(system_u:object_r:alsa_exec_t,s0)
 
 /usr/sbin/alsactl	--	gen_context(system_u:object_r:alsa_exec_t,s0)
 /usr/sbin/salsa	--	gen_context(system_u:object_r:alsa_exec_t,s0)
-
 /usr/share/alsa/alsa\.conf	gen_context(system_u:object_r:alsa_etc_rw_t,s0)
+/usr/share/alsa/cards(/.*)?    gen_context(system_u:object_r:alsa_etc_rw_t,s0)
 /usr/share/alsa/pcm(/.*)?	gen_context(system_u:object_r:alsa_etc_rw_t,s0)
 
 /var/lib/alsa(/.*)?	gen_context(system_u:object_r:alsa_var_lib_t,s0)
diff --git a/policy/modules/contrib/alsa.if b/policy/modules/contrib/alsa.if
index 708b743..3a1e6a6 100644
--- a/policy/modules/contrib/alsa.if
+++ b/policy/modules/contrib/alsa.if
@@ -256,3 +256,24 @@ interface(`alsa_read_lib',`
 	files_search_var_lib($1)
 	read_files_pattern($1, alsa_var_lib_t, alsa_var_lib_t)
 ')
+
+########################################
+## <summary>
+##	Manage alsa tmp files.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`alsa_manage_tmp_files',`
+	gen_require(`
+		type alsa_tmp_t;
+	')
+
+	files_search_tmp($1)
+	allow $1 alsa_tmp_t:dir list_dir_perms;
+	manage_files_pattern($1, alsa_tmp_t, alsa_tmp_t)
+	read_lnk_files_pattern($1, alsa_tmp_t, alsa_tmp_t)
+')
diff --git a/policy/modules/contrib/alsa.te b/policy/modules/contrib/alsa.te
index 901a6d3..5356c16 100644
--- a/policy/modules/contrib/alsa.te
+++ b/policy/modules/contrib/alsa.te
@@ -74,6 +74,8 @@ logging_send_syslog_msg(alsa_t)
 
 miscfiles_read_localization(alsa_t)
 
+seutil_use_newrole_fds(alsa_t)
+
 userdom_manage_unpriv_user_semaphores(alsa_t)
 userdom_manage_unpriv_user_shared_mem(alsa_t)
 userdom_search_user_home_dirs(alsa_t)
