diff --git a/policy/modules/contrib/firstboot.fc b/policy/modules/contrib/firstboot.fc
index 12c782c..ed09d11 100644
--- a/policy/modules/contrib/firstboot.fc
+++ b/policy/modules/contrib/firstboot.fc
@@ -3,3 +3,4 @@
 /usr/sbin/firstboot	--	gen_context(system_u:object_r:firstboot_exec_t,s0)
 
 /usr/share/firstboot/firstboot\.py	--	gen_context(system_u:object_r:firstboot_exec_t,s0)
+/usr/share/xenclient/firstboot\.sh	--	gen_context(system_u:object_r:firstboot_exec_t,s0)
