--- a/policy/modules/roles/sysadm.te
+++ b/policy/modules/roles/sysadm.te
@@ -408,6 +408,10 @@
 ')
 
 optional_policy(`
+	tpmsetup_role(sysadm_r, sysadm_t)
+')
+
+optional_policy(`
 	tpmutil_run(sysadm_t, sysadm_r)
 	tpmutil_sa_run(sysadm_t, sysadm_r)
 ')
--- a/policy/modules/system/lvm.te
+++ b/policy/modules/system/lvm.te
@@ -378,6 +378,11 @@
 	rpm_manage_script_tmp_files(lvm_t)
 ')
 
+# leaked fds?
+optional_policy(`
+	tpmsetup_rw_semaphores(lvm_t)
+')
+
 optional_policy(`
 	udev_read_db(lvm_t)
 ')
