diff -ru xfce4-settings-4.8.3.orig/xfce4-settings-helper/main.c xfce4-settings-4.8.3/xfce4-settings-helper/main.c
--- xfce4-settings-4.8.3.orig/xfce4-settings-helper/main.c	2012-06-26 18:33:45.000000000 +0100
+++ xfce4-settings-4.8.3/xfce4-settings-helper/main.c	2012-06-26 18:41:57.000000000 +0100
@@ -240,14 +240,8 @@
 
     /* create the sub daemons */
 #ifdef HAVE_XRANDR
-    displays_helper = g_object_new (XFCE_TYPE_DISPLAYS_HELPER, NULL);
 #endif
-    pointer_helper = g_object_new (XFCE_TYPE_POINTERS_HELPER, NULL);
-    keyboards_helper = g_object_new (XFCE_TYPE_KEYBOARDS_HELPER, NULL);
-    accessibility_helper = g_object_new (XFCE_TYPE_ACCESSIBILITY_HELPER, NULL);
     shortcuts_helper = g_object_new (XFCE_TYPE_KEYBOARD_SHORTCUTS_HELPER, NULL);
-    keyboard_layout_helper = g_object_new (XFCE_TYPE_KEYBOARD_LAYOUT_HELPER, NULL);
-    workspaces_helper = g_object_new (XFCE_TYPE_WORKSPACES_HELPER, NULL);
 
     if (g_getenv ("XFSETTINGSD_NO_CLIPBOARD") == NULL)
     {
@@ -276,14 +276,8 @@
 
     /* release the sub daemons */
 #ifdef HAVE_XRANDR
-    g_object_unref (G_OBJECT (displays_helper));
 #endif
-    g_object_unref (G_OBJECT (pointer_helper));
-    g_object_unref (G_OBJECT (keyboards_helper));
-    g_object_unref (G_OBJECT (accessibility_helper));
     g_object_unref (G_OBJECT (shortcuts_helper));
-    g_object_unref (G_OBJECT (keyboard_layout_helper));
-    g_object_unref (G_OBJECT (workspaces_helper));
 
     if (G_LIKELY (clipboard_daemon != NULL))
     {
