diff --git a/src/optionroms.c b/src/optionroms.c
index 06db1c1..7065889 100644
--- a/src/optionroms.c
+++ b/src/optionroms.c
@@ -427,6 +427,14 @@ vga_setup(void)
     ScreenAndDebug = romfile_loadint("etc/screen-and-debug", 1);
 
     if (CONFIG_OPTIONROMS_DEPLOYED) {
+        // setup hooks for vga pci devices
+        struct pci_device *pci;
+        foreachpci(pci) {
+            if (!is_pci_vga(pci))
+                continue;
+            vgahook_setup(pci);
+            break;
+        }
         // Option roms are already deployed on the system.
         init_optionrom((void*)BUILD_ROM_START, 0, 1);
     } else {
