diff --git a/src/xen.c b/src/xen.c
index 5dfee9e..a17ed5f 100644
--- a/src/xen.c
+++ b/src/xen.c
@@ -65,7 +65,8 @@ void xen_preinit(void)
 
         dprintf(9, "Found hypervisor signature \"%s\" at %x\n",
                 signature, base);
-        if (strcmp(signature, "XenVMMXenVMM") == 0) {
+        if (strcmp(signature, "XenVMMXenVMM") == 0 ||
+            strcmp(signature, "XciVMMXciVMM") == 0) {
             /* Set debug_io_port first, so the following messages work. */
             DebugOutputPort = 0xe9;
             dprintf(1, "SeaBIOS (version %s)\n\n", VERSION);
