diff --git a/tools/buildversion.sh b/tools/buildversion.sh
index c8c1725..fc9b49a 100755
--- a/tools/buildversion.sh
+++ b/tools/buildversion.sh
@@ -3,15 +3,7 @@
 OUTFILE="$1"
 VAR16MODE="$2"
 
-# Extract version info
-if [ -d .git ]; then
-    VERSION="`git describe --tags --long --dirty`"
-elif [ -f .version ]; then
-    VERSION="`cat .version`"
-else
-    VERSION="?"
-fi
-VERSION="${VERSION}-`date +"%Y%m%d_%H%M%S"`-`hostname`"
+VERSION="1.7.3.1-xenclient"
 echo "Version: ${VERSION}"
 
 # Build header file
