--- a/logrotate.c	2011-05-13 18:01:13.807873831 +0100
+++ b/logrotate.c	2011-05-13 18:08:51.000000000 +0100
@@ -75,7 +75,7 @@
 }
 
 static int runScript(char * logfn, char * script) {
-    int fd;
+    int fd, pid;
     char *filespec;
     int rc;
     char buf[256];
@@ -109,16 +109,27 @@
 
     close(fd);
 
-    if (!fork()) {
+    pid = fork();
+    if (pid < 0) {
+	message(MESS_ERROR, "cannot run script: %s\n", strerror(errno));
+	unlink(filespec);
+	return -1;
+    } else if (!pid) {
 	execlp(filespec, filespec, logfn, NULL);
 	exit(1);
     }
 
-    wait(&rc);
-
+    if (waitpid(pid, &rc, 0) < 0) {
+	message(MESS_ERROR, "waitpid: %s\n", strerror(errno));
+        unlink(filespec);
+        return -1;
+    }
     unlink(filespec);
 
-    return rc;
+    if (WEXITSTATUS(rc)) {
+	message(MESS_ERROR, "script returned %i\n", WEXITSTATUS(rc));
+    }
+    return 0;
 }
 
 static int compressLogFile(char * name, logInfo * log, struct stat *sb) {
