Index: gnome-keyring-2.28.0/daemon/gkr-daemon-ops.c
===================================================================
--- gnome-keyring-2.28.0.orig/daemon/gkr-daemon-ops.c	2009-07-23 21:00:35.000000000 +0100
+++ gnome-keyring-2.28.0/daemon/gkr-daemon-ops.c	2011-08-19 16:08:37.000000000 +0100
@@ -304,16 +304,22 @@
 	is_default = (keyring == gkr_keyrings_get_default ());
 	
 	if (app->display_name && app->pathname) {
-		if (is_default) {
-			/* TRANSLATORS: The default keyring is locked */
-			message = g_markup_printf_escaped (_("The application '%s' (%s) wants access to "
-						           "the default keyring, but it is locked"),
-						           app->display_name, app->pathname);
+		if (!strcmp(app->pathname,"/usr/bin/nm-applet") ||
+		    !strcmp(app->pathname,"/usr/bin/nm-connection-editor")) {
+				message = g_markup_printf_escaped (_("Please enter the master password to access "
+								   "stored wireless credentials"));
 		} else {
-			/* TRANSLATORS: The keyring '%s' is locked */
-			message = g_markup_printf_escaped (_("The application '%s' (%s) wants access to "
-						           "the keyring '%s', but it is locked"),
-						           app->display_name, app->pathname, keyring_name);
+			if (is_default) {
+				/* TRANSLATORS: The default keyring is locked */
+				message = g_markup_printf_escaped (_("The application '%s' (%s) wants access to "
+								   "the default keyring, but it is locked"),
+								   app->display_name, app->pathname);
+			} else {
+				/* TRANSLATORS: The keyring '%s' is locked */
+				message = g_markup_printf_escaped (_("The application '%s' (%s) wants access to "
+								   "the keyring '%s', but it is locked"),
+								   app->display_name, app->pathname, keyring_name);
+			}
 		}
 	} else if (app->display_name) {
 		if (is_default) {
@@ -628,9 +634,17 @@
 	
 	/* Build an appropriate message */
 	if (app->display_name && app->pathname) {
-		message = g_markup_printf_escaped (_("The application '%s' (%s) wants to store a password, but there is no default keyring. "
-					           "To create one, you need to choose the password you wish to use for it."),
-					           app->display_name, app->pathname);
+		if (!strcmp(app->pathname,"/usr/bin/nm-applet") ||
+		    !strcmp(app->pathname,"/usr/bin/nm-connection-editor")) {
+			message = g_markup_printf_escaped (_("If you wish to set a master password to protect your wireless "
+							   "credentials enter it here. You will need to enter the master "
+							   "password at every boot. Otherwise leave the password field blank "
+							   "and click 'create'."), app->display_name, app->pathname);
+		} else {
+			message = g_markup_printf_escaped (_("The application '%s' (%s) wants to store a password, but there is no default keyring. "
+							   "To create one, you need to choose the password you wish to use for it."),
+							   app->display_name, app->pathname);
+		}
 	} else if (app->display_name) {
 		message = g_markup_printf_escaped (_("The application '%s' wants to store a password, but there is no default keyring. "
 					           "To create one, you need to choose the password you wish to use for it."),
