Index: usbutils-0.82/lsusb-t.c
===================================================================
--- usbutils-0.82.orig/lsusb-t.c	2009-04-27 12:39:39.000000000 -0400
+++ usbutils-0.82/lsusb-t.c	2015-04-08 10:43:03.306067884 -0400
@@ -238,11 +238,13 @@
 	if (r >= 0 && r < len) {
 		buf[r] = '\0';
 		r--;
-		while (buf[r] == '\n') {
+		/* From this point forward, r can be/become negative. */
+		/* Making sure it's not before doing anything */
+		while (r >= 0 && buf[r] == '\n') {
 			buf[r] = '\0';
 			r--;
 		}
-		while (r) {
+		while (r >= 0) {
 			if (buf[r] == '\n')
 				buf[r] = ' ';
 			r--;
@@ -255,8 +257,13 @@
 
 static void append_dev_interface(struct usbinterface *i, struct usbinterface *new)
 {
-	while (i->next)
+	while (i->next) {
+		if (i == new)
+			return;
 		i = i->next;
+	}
+	if (i == new)
+		return;
 	i->next = new;
 }
 
@@ -275,8 +282,13 @@
 		if (b->busnum == busnum) {
 			i = b->first_interface;
 			if (i) {
-				while (i->next)
+				while (i->next) {
+					if (i == new)
+						return;
 					i = i->next;
+				}
+				if (i == new)
+					return;
 				i->next = new;
 			} else
 				b->first_interface = new;
