Index: usbutils-0.82/lsusb-t.c
===================================================================
--- usbutils-0.82.orig/lsusb-t.c	2015-04-07 10:06:45.302413675 -0400
+++ usbutils-0.82/lsusb-t.c	2015-04-07 10:10:37.121928297 -0400
@@ -238,11 +238,13 @@
 	if (r >= 0 && r < len) {
 		buf[r] = '\0';
 		r--;
-		while (buf[r] == '\n') {
+		/* From this point forward, r can be/become negative. */
+		/* Making sure it's not before doing anything */
+		while (r >= 0 && buf[r] == '\n') {
 			buf[r] = '\0';
 			r--;
 		}
-		while (r) {
+		while (r >= 0) {
 			if (buf[r] == '\n')
 				buf[r] = ' ';
 			r--;
