--- a/policy/modules/contrib/xen.te
+++ b/policy/modules/contrib/xen.te
@@ -801,3 +801,7 @@ optional_policy(`
 
 
 ')
+
+optional_policy(`
+	bats_manage_tmp_files(xm_t)
+')
--- a/policy/modules/apps/xec.te
+++ b/policy/modules/apps/xec.te
@@ -87,6 +87,7 @@ network_daemon_dbus_chat(xec_t)
 
 xen_use_fds(xec_t)
 ssh_use_fds(xec_t)
+init_use_fds(xec_t)
 
 seutil_use_newrole_fds(xec_t)
 allow xec_t self:capability { dac_override dac_read_search };
