--- a/policy/modules/contrib/xen.te
+++ b/policy/modules/contrib/xen.te
@@ -267,6 +267,8 @@ blktap_read_blk_file(xend_t)
 tapdisk_domtrans(xend_t)
 tapctl_domtrans(xend_t)
 
+dbd_dbus_chat(xend_t)
+
 kernel_read_kernel_sysctls(xend_t)
 kernel_read_system_state(xend_t)
 kernel_write_xen_state(xend_t)
--- a/policy/modules/system/init.te
+++ b/policy/modules/system/init.te
@@ -888,6 +888,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	db_upgrade_domtrans(initrc_t)
+')
+
+optional_policy(`
 	dbus_connect_system_bus(initrc_t)
 	dbus_system_bus_client(initrc_t)
 	dbus_read_config(initrc_t)
--- a/policy/modules/system/unconfined.te
+++ b/policy/modules/system/unconfined.te
@@ -84,6 +84,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	dbd_dbus_chat(unconfined_t)
+')
+
+optional_policy(`
 	dpkg_run(unconfined_t, unconfined_r)
 ')
 
