--- a/policy/modules/system/mount.te
+++ b/policy/modules/system/mount.te
@@ -81,6 +81,14 @@ dev_dontaudit_getattr_memory_dev(mount_t
 dev_getattr_sound_dev(mount_t)
 # Early devtmpfs, before udev relabel
 dev_dontaudit_rw_generic_chr_files(mount_t)
+dev_getattr_generic_blk_files(mount_t)
+dev_getattr_fs(mount_t)
+dontaudit mount_t device_t:blk_file read;
+
+xen_dontaudit_rw_unix_stream_sockets(mount_t)
+files_read_mnt_symlinks(mount_t)
+dev_read_sysfs(mount_t)    # /sys/devices/pci*
+dev_dontaudit_rw_generic_chr_files(mount_t)
 
 domain_use_interactive_fds(mount_t)
 
@@ -101,7 +109,9 @@ files_dontaudit_write_all_mountpoints(mo
 files_dontaudit_setattr_all_mountpoints(mount_t)
 
 fs_getattr_xattr_fs(mount_t)
+fs_getattr_tmpfs(mount_t)
 fs_getattr_cifs(mount_t)
+fs_getattr_dos_fs(mount_t)
 fs_mount_all_fs(mount_t)
 fs_unmount_all_fs(mount_t)
 fs_remount_all_fs(mount_t)
@@ -110,10 +120,14 @@ fs_rw_tmpfs_chr_files(mount_t)
 fs_read_tmpfs_symlinks(mount_t)
 fs_dontaudit_write_tmpfs_dirs(mount_t)
 
+# root fs is labeled tmpfs_t for some reason
+fs_search_tmpfs(mount_t)
+
 mls_file_read_all_levels(mount_t)
 mls_file_write_all_levels(mount_t)
 
 selinux_get_enforce_mode(mount_t)
+selinux_getattr_fs(mount_t)
 
 storage_raw_read_fixed_disk(mount_t)
 storage_raw_write_fixed_disk(mount_t)
@@ -141,6 +155,11 @@ selinux_getattr_fs(mount_t)
 
 userdom_use_all_users_fds(mount_t)
 
+networkmanager_mounton(mount_t)
+
+dev_rw_loop_device(mount_t)
+blktap_rw_blk_file(mount_t)
+
 ifdef(`distro_redhat',`
 	optional_policy(`
 		auth_read_pam_console_data(mount_t)
@@ -218,7 +237,6 @@ optional_policy(`
 optional_policy(`
 	samba_run_smbmount(mount_t, mount_roles)
 ')
-
 ########################################
 #
 # Unconfined mount local policy
@@ -227,4 +245,11 @@ optional_policy(`
 optional_policy(`
 	files_etc_filetrans_etc_runtime(unconfined_mount_t, file)
 	unconfined_domain(unconfined_mount_t)
+	unconfined_run_to(unconfined_mount_t, mount_exec_t)
+')
+
+# leaked FDs from input-server
+optional_policy(`
+	kernel_dontaudit_read_xen_state(mount_t)
+	kernel_dontaudit_write_xen_state(mount_t)
 ')
