--- a/policy/modules/kernel/terminal.te
+++ b/policy/modules/kernel/terminal.te
@@ -29,7 +29,7 @@ files_mountpoint(devpts_t)
 fs_associate_tmpfs(devpts_t)
 fs_xattr_type(devpts_t)
 fs_use_trans devpts gen_context(system_u:object_r:devpts_t,s0);
-
+dev_associate(devpts_t)
 #
 # devtty_t is the type of /dev/tty.
 #
--- a/policy/modules/kernel/terminal.fc
+++ b/policy/modules/kernel/terminal.fc
@@ -25,7 +25,7 @@
 
 /dev/pts		-d	gen_context(system_u:object_r:devpts_t,s0-mls_systemhigh)
 /dev/pts/ptmx		-c	gen_context(system_u:object_r:devpts_t,s0)
-/dev/pts/[0-9]+		-c	gen_context(system_u:object_r:user_devpts_t,s0)
+/dev/pts/[0-9]+		-c	<<none>>
 
 /dev/tts/[^/]*		-c	gen_context(system_u:object_r:tty_device_t,s0)
 
