Index: uim-1.8.6/gtk2/toolbar/common-gtk.c
===================================================================
--- uim-1.8.6.orig/gtk2/toolbar/common-gtk.c
+++ uim-1.8.6/gtk2/toolbar/common-gtk.c
@@ -311,11 +311,12 @@ right_click_menu_activated(GtkMenu *menu
 static gboolean
 right_button_pressed(GtkButton *button, GdkEventButton *event, gpointer data)
 {
+#ifndef OPENXT_BRANDING
   gtk_menu_popup(GTK_MENU(right_click_menu), NULL, NULL,
 		 (GtkMenuPositionFunc)calc_menu_position,
 		 (gpointer)button, event->button,
 		 gtk_get_current_event_time());
-
+#endif /* OPENXT_BRANDING */
   return FALSE;
 }
 
