Index: uim-1.8.6/gtk2/toolbar/standalone-gtk.c
===================================================================
--- uim-1.8.6.orig/gtk2/toolbar/standalone-gtk.c
+++ uim-1.8.6/gtk2/toolbar/standalone-gtk.c
@@ -290,6 +290,9 @@ main(int argc, char *argv[])
   GtkWidget *frame;
 #endif /* OPENXT_BRANDING */
 
+  const char *env_show_toolbar = getenv("UIM_SHOW_TOOLBAR");
+  int show_toolbar = !env_show_toolbar || !strcmp(env_show_toolbar, "1");
+
   setlocale(LC_ALL, "");
   bindtextdomain(PACKAGE, LOCALEDIR);
   textdomain(PACKAGE);
@@ -369,6 +372,9 @@ main(int argc, char *argv[])
 		   G_CALLBACK(size_request_cb), NULL);
 #endif
 
+  if (!show_toolbar)
+    goto skip_toolbar;
+
   gtk_widget_show_all(GTK_WIDGET(window));
 
 #ifdef OPENXT_BRANDING
@@ -409,6 +415,7 @@ main(int argc, char *argv[])
     helper_win_set_position(window, x, y);
   }
 
+skip_toolbar:
   gtk_main();
 
   uim_quit();
