diff --git a/tools/blktap2/control/tap-ctl-attach.c b/tools/blktap2/control/tap-ctl-attach.c
index 3cb933c..863114d 100644
--- a/tools/blktap2/control/tap-ctl-attach.c
+++ b/tools/blktap2/control/tap-ctl-attach.c
@@ -43,7 +43,7 @@ tap_ctl_attach(const int id, const int minor)
 	message.type = TAPDISK_MESSAGE_ATTACH;
 	message.cookie = minor;
 
-	err = tap_ctl_connect_send_and_receive(id, &message, 5);
+	err = tap_ctl_connect_send_and_receive(id, &message, TAPCTL_SERVER_TIMEOUT);
 	if (err)
 		return err;
 
diff --git a/tools/blktap2/control/tap-ctl-close.c b/tools/blktap2/control/tap-ctl-close.c
index 2e5f80b..93d3f36 100644
--- a/tools/blktap2/control/tap-ctl-close.c
+++ b/tools/blktap2/control/tap-ctl-close.c
@@ -46,7 +46,7 @@ __tap_ctl_close(const int id, const int minor, const int force)
 		message.type = TAPDISK_MESSAGE_FORCE_SHUTDOWN;
 	message.cookie = minor;
 
-	err = tap_ctl_connect_send_and_receive(id, &message, 5);
+	err = tap_ctl_connect_send_and_receive(id, &message, TAPCTL_SERVER_TIMEOUT);
 	if (err)
 		return err;
 
diff --git a/tools/blktap2/control/tap-ctl-detach.c b/tools/blktap2/control/tap-ctl-detach.c
index 7d7bbf3..6df1409 100644
--- a/tools/blktap2/control/tap-ctl-detach.c
+++ b/tools/blktap2/control/tap-ctl-detach.c
@@ -43,7 +43,7 @@ tap_ctl_detach(const int id, const int minor)
 	message.type = TAPDISK_MESSAGE_DETACH;
 	message.cookie = minor;
 
-	err = tap_ctl_connect_send_and_receive(id, &message, 5);
+	err = tap_ctl_connect_send_and_receive(id, &message, TAPCTL_SERVER_TIMEOUT);
 	if (err)
 		return err;
 
diff --git a/tools/blktap2/control/tap-ctl-open.c b/tools/blktap2/control/tap-ctl-open.c
index 5961c99..bb00d83 100644
--- a/tools/blktap2/control/tap-ctl-open.c
+++ b/tools/blktap2/control/tap-ctl-open.c
@@ -54,7 +54,7 @@ tap_ctl_open(const int id, const int minor, const char *params)
 		return ENAMETOOLONG;
 	}
 
-	err = tap_ctl_connect_send_and_receive(id, &message, 5);
+	err = tap_ctl_connect_send_and_receive(id, &message, TAPCTL_SERVER_TIMEOUT);
 	if (err)
 		return err;
 
diff --git a/tools/blktap2/control/tap-ctl-pause.c b/tools/blktap2/control/tap-ctl-pause.c
index 5e31a58..9b61525 100644
--- a/tools/blktap2/control/tap-ctl-pause.c
+++ b/tools/blktap2/control/tap-ctl-pause.c
@@ -43,7 +43,7 @@ tap_ctl_pause(const int id, const int minor)
 	message.type = TAPDISK_MESSAGE_PAUSE;
 	message.cookie = minor;
 
-	err = tap_ctl_connect_send_and_receive(id, &message, 5);
+	err = tap_ctl_connect_send_and_receive(id, &message, TAPCTL_SERVER_TIMEOUT);
 	if (err)
 		return err;
 
diff --git a/tools/blktap2/control/tap-ctl-spawn.c b/tools/blktap2/control/tap-ctl-spawn.c
index 31a651e..9b5a0b7 100644
--- a/tools/blktap2/control/tap-ctl-spawn.c
+++ b/tools/blktap2/control/tap-ctl-spawn.c
@@ -90,7 +90,7 @@ tap_ctl_get_pid(const int id)
 	memset(&message, 0, sizeof(message));
 	message.type = TAPDISK_MESSAGE_PID;
 
-	err = tap_ctl_connect_send_and_receive(id, &message, 2);
+	err = tap_ctl_connect_send_and_receive(id, &message, TAPCTL_SERVER_TIMEOUT);
 	if (err)
 		return err;
 
diff --git a/tools/blktap2/control/tap-ctl-unpause.c b/tools/blktap2/control/tap-ctl-unpause.c
index dfb7450..b12764c 100644
--- a/tools/blktap2/control/tap-ctl-unpause.c
+++ b/tools/blktap2/control/tap-ctl-unpause.c
@@ -48,7 +48,7 @@ tap_ctl_unpause(const int id, const int minor, const char *params)
 		strncpy(message.u.params.path, params,
 			sizeof(message.u.params.path) - 1);
 
-	err = tap_ctl_connect_send_and_receive(id, &message, 15);
+	err = tap_ctl_connect_send_and_receive(id, &message, TAPCTL_SERVER_TIMEOUT);
 	if (err)
 		return err;
 
diff --git a/tools/blktap2/control/tap-ctl.h b/tools/blktap2/control/tap-ctl.h
index de0bc29..b58edb5 100644
--- a/tools/blktap2/control/tap-ctl.h
+++ b/tools/blktap2/control/tap-ctl.h
@@ -34,6 +34,8 @@
 
 extern int tap_ctl_debug;
 
+#define TAPCTL_SERVER_TIMEOUT 120
+
 #ifdef TAPCTL
 #define DBG(_f, _a...)				\
 	do {					\
