diff --git a/config/StdGNU.mk b/config/StdGNU.mk
index 25fc594..14036a8 100644
--- a/config/StdGNU.mk
+++ b/config/StdGNU.mk
@@ -1,11 +1,11 @@
 AS         = $(CROSS_COMPILE)as
-LD         = $(CROSS_COMPILE)ld
+LD        ?= $(CROSS_COMPILE)ld
 ifeq ($(clang),y)
 CC         = $(CROSS_COMPILE)clang
 LD_LTO     = $(CROSS_COMPILE)llvm-ld
 else
-CC         = $(CROSS_COMPILE)gcc
-LD_LTO     = $(CROSS_COMPILE)ld
+CC        ?= $(CROSS_COMPILE)gcc
+LD_LTO    ?= $(CROSS_COMPILE)ld
 endif
 CPP        = $(CC) -E
 AR         = $(CROSS_COMPILE)ar
