diff -r 5108979095cd tools/blktap/lib/blktaplib.h
--- a/tools/blktap/lib/blktaplib.h
+++ b/tools/blktap/lib/blktaplib.h
@@ -195,6 +195,8 @@
 	pid_t     pid;
 } msg_pid_t;
 
+#undef READ
+#undef WRITE
 #define READ 0
 #define WRITE 1
 
diff -r 5108979095cd tools/blktap2/include/blktaplib.h
--- a/tools/blktap2/include/blktaplib.h
+++ b/tools/blktap2/include/blktaplib.h
@@ -197,6 +197,8 @@
 	int       uuid_len;
 } msg_lock_t;
 
+#undef READ
+#undef WRITE
 #define READ 0
 #define WRITE 1
 
