diff --git a/xen/drivers/passthrough/iommu.c b/xen/drivers/passthrough/iommu.c
index 6f5ad78..d38fea9 100644
--- a/xen/drivers/passthrough/iommu.c
+++ b/xen/drivers/passthrough/iommu.c
@@ -662,9 +662,11 @@ int iommu_do_domctl(
 
         if ( device_assigned(seg, bus, devfn) )
         {
+#if 0
             printk(XENLOG_G_INFO
                    "%04x:%02x:%02x.%u already assigned, or non-existent\n",
                    seg, bus, PCI_SLOT(devfn), PCI_FUNC(devfn));
+#endif
             ret = -EINVAL;
         }
         break;
