# HG changeset patch
# Parent 57dbee387b97d11b0a964ec44c9c345d250db2ce

diff --git a/tools/hotplug/Linux/vif-bridge b/tools/hotplug/Linux/vif-bridge
index f489519..f37d3e1 100644
--- a/tools/hotplug/Linux/vif-bridge
+++ b/tools/hotplug/Linux/vif-bridge
@@ -101,9 +101,10 @@ case "$command" in
         ;;
 esac
 
-if [ "$type_if" = vif ]; then
-    handle_iptable
-fi
+# DMD takes care of iptables setup for VMs
+# if [ "$type_if" = vif ]; then
+#     handle_iptable
+# fi
 
 call_hooks vif post
 
