diff -r 50419a957a1f xen/drivers/char/console.c
--- a/xen/drivers/char/console.c	Thu Mar 03 13:13:32 2011 -0500
+++ b/xen/drivers/char/console.c	Thu Mar 03 13:18:48 2011 -0500
@@ -199,7 +199,7 @@
     sofar = 0;
 
     c = conringc;
-    if ( op->incremental && ((int32_t)(op->index - c) > 0) )
+    if ( op->incremental )
         c = op->index;
 
     while ( (c != conringp) && (sofar < max) )
@@ -486,9 +486,8 @@
     if ( tm.tm_mday == 0 )
         return;
 
-    snprintf(tstr, sizeof(tstr), "[%04u-%02u-%02u %02u:%02u:%02u] ",
-             1900 + tm.tm_year, tm.tm_mon + 1, tm.tm_mday,
-             tm.tm_hour, tm.tm_min, tm.tm_sec);
+    snprintf(tstr, sizeof(tstr), "[%02u:%02u:%02u] %d",
+             tm.tm_hour, tm.tm_min, tm.tm_sec, conringp);
     __putstr(tstr);
 }
 
