# HG changeset patch
# Parent c465c826a8d948011525d169d615f0746d01c6d4

diff -r c465c826a8d9 tools/blktap2/drivers/tapdisk-vbd.c
--- a/tools/blktap2/drivers/tapdisk-vbd.c	Wed May 11 16:38:50 2011 -0400
+++ b/tools/blktap2/drivers/tapdisk-vbd.c	Thu May 12 11:08:24 2011 -0400
@@ -981,6 +981,10 @@
 	if (!list_empty(&vbd->pending_requests))
 		return -EAGAIN;
 
+	if (td_flag_test(vbd->state, TD_VBD_SHUTDOWN_IN_PROGRESS))
+		return -EAGAIN;
+	td_flag_set(vbd->state, TD_VBD_SHUTDOWN_IN_PROGRESS);
+
 	tapdisk_vbd_kick(vbd);
 	tapdisk_vbd_queue_count(vbd, &new, &pending, &failed, &completed);
 
diff -r c465c826a8d9 tools/blktap2/drivers/tapdisk-vbd.h
--- a/tools/blktap2/drivers/tapdisk-vbd.h	Wed May 11 16:38:50 2011 -0400
+++ b/tools/blktap2/drivers/tapdisk-vbd.h	Thu May 12 11:08:24 2011 -0400
@@ -50,6 +50,7 @@
 #define TD_VBD_LOCKING              0x0080
 #define TD_VBD_RETRY_NEEDED         0x0100
 #define TD_VBD_LOG_DROPPED          0x0200
+#define TD_VBD_SHUTDOWN_IN_PROGRESS 0x0400
 
 typedef struct td_ring              td_ring_t;
 typedef struct td_vbd_request       td_vbd_request_t;
