# HG changeset patch
# Parent 95287936c1959cdaaeb32f71d93629c2b678830a

diff -r 95287936c195 tools/blktap2/drivers/tapdisk-vbd.c
--- a/tools/blktap2/drivers/tapdisk-vbd.c	Thu Jun 23 16:28:08 2011 -0400
+++ b/tools/blktap2/drivers/tapdisk-vbd.c	Fri Jul 01 14:02:52 2011 -0400
@@ -1319,8 +1319,10 @@
 	DBG(TLOG_DBG, "writing req %d, sec 0x%08"PRIx64", res %d to ring\n",
 	    (int)tmp.id, tmp.sector_number, vreq->status);
 
-	if (rsp->status != BLKIF_RSP_OKAY)
+	if (rsp->status != BLKIF_RSP_OKAY) {
 		ERR(EIO, "returning BLKIF_RSP %d", rsp->status);
+		syslog(LOG_ERR, "returning EOI BLKIF_RSP %d", rsp->status);
+	}
 
 	vbd->returned++;
 	vbd->callback(vbd->argument, rsp);
