# HG changeset patch
# Parent e7d8cd396876b7df52a1abd4671a04a6f08294af

diff --git a/tools/firmware/vgabios/biossums.c b/tools/firmware/vgabios/biossums.c
index d5816f4..d65bf6e 100644
--- a/tools/firmware/vgabios/biossums.c
+++ b/tools/firmware/vgabios/biossums.c
@@ -143,6 +143,7 @@ int main(int argc, char* argv[])
       modified = 1;
     } else {
       printf("\nBios checksum at:   0x%4lX\n", offset);
+      printf("Bios length:          0x%4lX\n", bios_len);
       printf("Current checksum:     0x%02X\n", cur_val);
       printf("Calculated checksum:  0x%02X  ", new_val);
       if (cur_val != new_val) {
diff --git a/tools/firmware/vgabios/vbe.h b/tools/firmware/vgabios/vbe.h
index 9e2d178..07b2e92 100644
--- a/tools/firmware/vgabios/vbe.h
+++ b/tools/firmware/vgabios/vbe.h
@@ -278,6 +278,8 @@ typedef struct ModeInfoListItem
   #define VBE_DISPI_BANK_ADDRESS           0xA0000
   #define VBE_DISPI_BANK_SIZE_KB           64
   
+// XXX THESE MAXs DON'T DO ANYTHING
+
   #define VBE_DISPI_MAX_XRES               1920
   #define VBE_DISPI_MAX_YRES               1200
   
diff --git a/tools/firmware/vgabios/vbetables-gen.c b/tools/firmware/vgabios/vbetables-gen.c
index 64f3f0d..6edb2a7 100644
--- a/tools/firmware/vgabios/vbetables-gen.c
+++ b/tools/firmware/vgabios/vbetables-gen.c
@@ -12,71 +12,79 @@ typedef struct {
 } ModeInfo;
 
 ModeInfo modes[] = {
-    /* standard VESA modes */
-{ 640, 400, 8                          , 0x100},
-{ 640, 480, 8                          , 0x101},
-{ 800, 600, 4                          , 0x102},
-{ 800, 600, 8                          , 0x103},
-{ 1024, 768, 4                         , 0x104},
-{ 1024, 768, 8                         , 0x105},
-{ 1280, 1024, 4                        , 0x106},
-{ 1280, 1024, 8                        , 0x107},
-{ 320, 200, 15                       , 0x10D},
-{ 320, 200, 16                        , 0x10E},
-{ 320, 200, 24                        , 0x10F},
-{ 640, 480, 15                       , 0x110},
-{ 640, 480, 16                        , 0x111},
-{ 640, 480, 24                        , 0x112},
-{ 800, 600, 15                       , 0x113},
-{ 800, 600, 16                        , 0x114},
-{ 800, 600, 24                        , 0x115},
-{ 1024, 768, 15                      , 0x116},
-{ 1024, 768, 16                       , 0x117},
-{ 1024, 768, 24                       , 0x118},
-{ 1280, 1024, 15                     , 0x119},
-{ 1280, 1024, 16                      , 0x11A},
-{ 1280, 1024, 24                      , 0x11B},
-{ 1600, 1200, 8                        , 0x11C},
-{ 1600, 1200, 15                     , 0x11D},
-{ 1600, 1200, 16                      , 0x11E},
-{ 1600, 1200, 24                      , 0x11F},
+        /* standard VESA modes */
+{        640,    400,    8,             0x100 },
+{        640,    480,    8,             0x101 },
+{        800,    600,    8,             0x103 },
+{       1024,    768,    8,             0x105 },
+{	1280,	1024,	 8,		0x107 },
+{        320,    200,   16,             0x10E },
+{        320,    200,   24,             0x10F },
+{        640,    480,   16,             0x111 },
+{        640,    480,   24,             0x112 },
+{        800,    600,   16,             0x114 },
+{        800,    600,   24,             0x115 },
+{	1024,	 768,	16,		0x117 },
+{       1024,    768,   24,             0x118 },
+{	1280,	1024,	16,		0x11A },
+{	1280,	1024,	16,		0x11B },
 
-      /* BOCHS/PLE, 86 'own' mode numbers */
-{ 320, 200, 32                       , 0x140},
-{ 640, 400, 32                       , 0x141},
-{ 640, 480, 32                       , 0x142},
-{ 800, 600, 32                       , 0x143},
-{ 1024, 768, 32                      , 0x144},
-{ 1280, 1024, 32                     , 0x145},
-{ 320, 200, 8                        , 0x146},
-{ 1600, 1200, 32                     , 0x147},
-{ 1152, 864, 8                       , 0x148},
-{ 1152, 864, 15                      , 0x149},
-{ 1152, 864, 16                      , 0x14a},
-{ 1152, 864, 24                      , 0x14b},
-{ 1152, 864, 32                      , 0x14c},
-{ 1280, 800, 16                      , 0x178},
-{ 1280, 800, 24                      , 0x179},
-{ 1280, 800, 32                      , 0x17a},
-{ 1280, 960, 16                      , 0x17b},
-{ 1280, 960, 24                      , 0x17c},
-{ 1280, 960, 32                      , 0x17d},
-{ 1440, 900, 16                      , 0x17e},
-{ 1440, 900, 24                      , 0x17f},
-{ 1440, 900, 32                      , 0x180},
-{ 1400, 1050, 16                     , 0x181},
-{ 1400, 1050, 24                     , 0x182},
-{ 1400, 1050, 32                     , 0x183},
-{ 1680, 1050, 16                     , 0x184},
-{ 1680, 1050, 24                     , 0x185},
-{ 1680, 1050, 32                     , 0x186},
-{ 1920, 1200, 16                     , 0x187},
-{ 1920, 1200, 24                     , 0x188},
-{ 1920, 1200, 32                     , 0x189},
-{ 2560, 1600, 16                     , 0x18a},
-{ 2560, 1600, 24                     , 0x18b},
-{ 2560, 1600, 32                     , 0x18c},
-{ 0, },
+	/* "Common" modes (http://en.wikipedia.org/wiki/VESA_BIOS_Extensions) */
+{        640,    480,   32,             0x129 },
+{        800,    600,   32,             0x12E },
+{       1024,    768,   32,             0x138 },
+{       1280,   1024,   32,             0x13D },
+{       1600,   1200,   32,             0x142 },
+{       1152,    864,   32,             0x14c },
+
+        /* ORC VESA modes */
+{       1152,    720,   32,             0x14d },
+{       1152,    768,   32,             0x14e },
+{       1152,    864,   32,             0x14f },
+
+{       1280,    720,   32,             0x150 },
+{       1280,    768,   32,             0x151 },
+{       1280,    800,   32,             0x152 },
+{       1280,    854,   32,             0x153 },
+{       1280,    960,   32,             0x154 },
+
+{       1366,    768,   32,             0x156 },
+
+{       1400,   1050,   32,             0x158 },
+
+{       1440,    768,   32,             0x15a },
+{       1440,    900,   32,             0x15b },
+{       1440,    960,   32,             0x15c },
+{       1440,   1080,   32,             0x15f },
+
+{       1600,    768,   32,             0x160 },
+{       1600,    900,   32,             0x161 },
+{       1600,   1024,   32,             0x162 },
+{       1600,   1200,   32,             0x163 },
+
+{       1680,   1050,   32,             0x169 },
+
+{       1920,   1080,   32,             0x16d },
+
+        /* modes which require 16MB frame buffer */
+{       1920,   1200,   32,             0x170 },
+{       1920,   1400,   32,             0x170 },
+{       1920,   1440,   32,             0x170 },
+
+{       2048,   1107,   32,             0x171 },
+{       2048,   1152,   32,             0x172 },
+{       2048,   1280,   32,             0x173 },
+{       2048,   1536,   32,             0x174 },
+
+{       2560,   1440,   32,             0x175 },
+{       2560,   1600,   32,             0x176 },
+
+        /* modes which require 32MB frame buffer */
+{       2560,   2048,   32,             0x177 },
+{	3840,	2160,	32,		0x178 },
+
+        /* end-of-list */
+{ 0, 0, 0, 0 },
 };
 
 int main(int argc, char **argv)
