diff --git a/tools/firmware/hvmloader/config.h b/tools/firmware/hvmloader/config.h
index 6641197..b03f6c0 100644
--- a/tools/firmware/hvmloader/config.h
+++ b/tools/firmware/hvmloader/config.h
@@ -69,7 +69,7 @@ extern unsigned long pci_mem_start, pci_mem_end;
 /* NB. ACPI_INFO_PHYSICAL_ADDRESS *MUST* match definition in acpi/dsdt.asl! */
 #define ACPI_INFO_PHYSICAL_ADDRESS    0xFC000000
 #define RESERVED_MEMORY_DYNAMIC_START 0xFC001000
-#define RESERVED_MEMORY_DYNAMIC_END   0xFE000000
+#define RESERVED_MEMORY_DYNAMIC_END   0xFD000000
 /*
  * GUEST_RESERVED: Physical address space reserved for guest use.
  * This is not dynamically advertised to guests, so this range must *never*
