diff --git a/tools/include/xen-sys/Linux/privcmd.h b/tools/include/xen-sys/Linux/privcmd.h
index c579e81..e646508 100644
--- a/tools/include/xen-sys/Linux/privcmd.h
+++ b/tools/include/xen-sys/Linux/privcmd.h
@@ -83,6 +83,11 @@ typedef struct privcmd_mmapconsec {
 	unsigned long __user *err_bits;  /* bit array of unmapped pfns */
 } privcmd_mmapconsec_t;
 
+typedef struct privcmd_mmapcacheattr {
+       __u64 addr;
+       int type;
+} privcmd_mmapcacheattr_t;
+
 /*
  * @cmd: IOCTL_PRIVCMD_HYPERCALL
  * @arg: &privcmd_hypercall_t
@@ -98,5 +103,7 @@ typedef struct privcmd_mmapconsec {
 	_IOC(_IOC_NONE, 'P', 4, sizeof(privcmd_mmapbatch_v2_t))
 #define IOCTL_PRIVCMD_MMAPCONSEC					\
 	_IOC(_IOC_NONE, 'P', 5, sizeof(struct privcmd_mmapconsec))
+#define IOCTL_PRIVCMD_MMAPCACHEATTR                             \
+	_IOC(_IOC_NONE, 'P', 200, sizeof(privcmd_mmapcacheattr_t))
 
 #endif /* __LINUX_PUBLIC_PRIVCMD_H__ */
diff --git a/tools/libxc/xc_foreign_memory.c b/tools/libxc/xc_foreign_memory.c
index f43d72c..f0ecbe0 100644
--- a/tools/libxc/xc_foreign_memory.c
+++ b/tools/libxc/xc_foreign_memory.c
@@ -129,6 +129,48 @@ void *xc_map_foreign_consec(xc_interface *xch, uint32_t dom, int prot,
                                                   dom, prot, pfn, num, err_bits);
 }
 
+void *xc_map_foreign_batch_cacheattr(xc_interface *xch, uint32_t dom, int prot,
+                                     xen_pfn_t *arr, int num, int cache_type)
+{
+    int fd = xch->ops_handle;
+    privcmd_mmapbatch_t batch;
+    privcmd_mmapcacheattr_t cacheattr;
+    void *addr;
+    addr = mmap(NULL, num * PAGE_SIZE, prot, MAP_SHARED, fd, 0);
+    if ( addr == MAP_FAILED ) {
+        perror("xc_map_foreign_batch: mmap failed");
+        return NULL;
+    }
+
+    cacheattr.addr = (unsigned long)addr;
+    cacheattr.type = cache_type;
+
+    if ( ioctl(fd, IOCTL_PRIVCMD_MMAPCACHEATTR, &cacheattr) ) {
+        int saved_errno = errno;
+        perror("xc_map_foreign_batch_cacheattr: failed to set cache attributes");
+        (void)munmap(addr, num*PAGE_SIZE);
+        errno = saved_errno;
+        return NULL;
+    }
+
+    batch.num=num;
+    batch.dom=dom;
+    batch.addr=(unsigned long)addr;
+    batch.arr=arr;
+
+    if ( ioctl(fd, IOCTL_PRIVCMD_MMAPBATCH, &batch) < 0 )
+    {
+        int saved_errno = errno;
+        perror("xc_map_foreign_batch_cacheattr: ioctl failed");
+        (void)munmap(addr, num*PAGE_SIZE);
+        errno = saved_errno;
+        return NULL;
+    }
+    return addr;
+
+}
+
+
 /*
  * Local variables:
  * mode: C
diff --git a/tools/libxc/xenctrl.h b/tools/libxc/xenctrl.h
index 9ac3293..f40f37f 100644
--- a/tools/libxc/xenctrl.h
+++ b/tools/libxc/xenctrl.h
@@ -1331,6 +1331,19 @@ void *xc_map_foreign_consec(xc_interface *xch, uint32_t dom, int prot,
 void *xc_map_foreign_bulk(xc_interface *xch, uint32_t dom, int prot,
                           const xen_pfn_t *arr, int *err, unsigned int num);
 
+enum {
+    XC_MAP_CACHEATTR_UC = XEN_DOMCTL_MEM_CACHEATTR_UC,
+    XC_MAP_CACHEATTR_WC = XEN_DOMCTL_MEM_CACHEATTR_WC,
+    XC_MAP_CACHEATTR_WT = XEN_DOMCTL_MEM_CACHEATTR_WT,
+    XC_MAP_CACHEATTR_WP = XEN_DOMCTL_MEM_CACHEATTR_WP,
+    XC_MAP_CACHEATTR_WB = XEN_DOMCTL_MEM_CACHEATTR_WB,
+    XC_MAP_CACHEATTR_UCM = XEN_DOMCTL_MEM_CACHEATTR_UCM
+};
+
+void *xc_map_foreign_batch_cacheattr(xc_interface *xch, uint32_t dom, int prot,
+                                     xen_pfn_t *arr, int num,
+                                     int cache_type);
+
 /**
  * Translates a virtual address in the context of a given domain and
  * vcpu returning the GFN containing the address (that is, an MFN for 
