Index: trousers-0.3.2-1/dist/tcsd.conf.in
===================================================================
--- trousers-0.3.2-1.orig/dist/tcsd.conf.in	2010-03-06 02:00:53.000000000 +0000
+++ trousers-0.3.2-1/dist/tcsd.conf.in	2010-03-06 02:02:04.000000000 +0000
@@ -26,7 +26,7 @@
 # Values: Any absolute directory path
 # Description: Path where the tcsd creates its persistent storage file.
 #
-# system_ps_file = @localstatedir@/lib/tpm/system.data
+system_ps_file = /boot/system/tpm/system.data
 #
 
 # Option: firmware_log_file
