Index: refpolicy/policy/modules/contrib/alsa.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/alsa.te	2015-04-02 19:36:27.471366423 -0400
+++ refpolicy/policy/modules/contrib/alsa.te	2015-04-02 22:05:25.513248752 -0400
@@ -90,3 +90,8 @@
 	statusreport_write_tmp_files(alsa_t)
 	statusreport_getattr_tmp_files(alsa_t)
 ')
+
+xc_config_dir_getattr(alsa_t)
+xc_config_dir_search(alsa_t)
+xc_config_filetrans(alsa_t, alsa_etc_rw_t, { dir file })
+
