--- a/policy/modules/services/dbusbouncer.te	1969-12-31 19:00:00.000000000 -0500
+++ b/policy/modules/services/dbusbouncer.te	2015-01-05 16:03:12.793080030 -0500
@@ -0,0 +1,66 @@
+#############################################################################
+#
+# Copyright (C) 2014 Citrix Systems, Inc.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+#############################################################################
+
+policy_module(dbusbouncer, 0.1)
+
+type dbusbouncer_t;
+type dbusbouncer_exec_t;
+
+init_daemon_domain(dbusbouncer_t, dbusbouncer_exec_t)
+
+type dbusbouncer_var_run_t;
+files_pid_file(dbusbouncer_var_run_t)
+files_pid_filetrans(dbusbouncer_t, dbusbouncer_var_run_t, { file sock_file })
+dbus_var_run_domtrans(dbusbouncer_t, dbusbouncer_var_run_t, sock_file)
+
+type dbusbouncer_etc_t;
+files_type(dbusbouncer_etc_t)
+
+corecmd_exec_bin(dbusbouncer_t)
+kernel_read_system_state(dbusbouncer_t)
+dbus_connect_system_bus(dbusbouncer_t)
+dbus_system_bus_client(dbusbouncer_t)
+xen_dbus_chat(dbusbouncer_t)
+dbd_dbus_chat(dbusbouncer_t)
+hal_dbus_chat(dbusbouncer_t)
+input_server_dbus_chat(dbusbouncer_t)
+networkmanager_dbus_chat(dbusbouncer_t)
+ctxusbd_dbus_chat(dbusbouncer_t)
+updatemgr_dbus_chat(dbusbouncer_t)
+logging_send_syslog_msg(dbusbouncer_t)
+fs_rw_xenfs_files(dbusbouncer_t)
+
+xc_files_rw_v4v_chr(dbusbouncer_t)
+
+xen_stream_connect_xenstore(dbusbouncer_t)
+kernel_write_xen_state(dbusbouncer_t)
+kernel_read_xen_state(dbusbouncer_t)
+
+allow dbusbouncer_t self:capability             { ipc_lock sys_admin };
+allow dbusbouncer_t self:process                { signal };
+allow dbusbouncer_t self:unix_dgram_socket { create_socket_perms };
+allow dbusbouncer_t self:fifo_file manage_fifo_file_perms;
+allow dbusbouncer_t dbusbouncer_var_run_t:sock_file manage_sock_file_perms;
+
+# read config
+files_search_etc(dbusbouncer_t)
+files_search_home(dbusbouncer_t)
+allow dbusbouncer_t dbusbouncer_etc_t:file read_file_perms;
+allow dbusbouncer_t self:capability dac_override;
