Index: refpolicy/policy/modules/contrib/alsa.fc
===================================================================
--- refpolicy.orig/policy/modules/contrib/alsa.fc	2015-04-02 19:36:27.471366423 -0400
+++ refpolicy/policy/modules/contrib/alsa.fc	2015-04-02 19:40:30.223359124 -0400
@@ -7,6 +7,7 @@
 /etc/asound(/.*)?	gen_context(system_u:object_r:alsa_etc_rw_t,s0)
 /etc/asound\.state	--	gen_context(system_u:object_r:alsa_etc_rw_t,s0)
 /var/volatile/etc/asound(/.*)?	gen_context(system_u:object_r:alsa_etc_rw_t,s0)
+/config/asound.state --      gen_context(system_u:object_r:alsa_etc_rw_t,s0)
 
 /sbin/alsactl	--	gen_context(system_u:object_r:alsa_exec_t,s0)
 /sbin/salsa	--	gen_context(system_u:object_r:alsa_exec_t,s0)
