diff --git a/policy/modules/system/authlogin.fc b/policy/modules/system/authlogin.fc
index 54ad19c..a64d15a 100644
--- a/policy/modules/system/authlogin.fc
+++ b/policy/modules/system/authlogin.fc
@@ -3,6 +3,12 @@
 /bin/login\.shadow	--	gen_context(system_u:object_r:login_exec_t,s0)
 /bin/login\.tinylogin	--	gen_context(system_u:object_r:login_exec_t,s0)
 
+/config/etc/\.pwd.lock	--	gen_context(system_u:object_r:shadow_t,s0)
+/config/etc/group\.lock	--	gen_context(system_u:object_r:shadow_t,s0)
+/config/etc/gshadow.*	--	gen_context(system_u:object_r:shadow_t,s0)
+/config/etc/passwd\.lock	--	gen_context(system_u:object_r:shadow_t,s0)
+/config/etc/shadow.*	--	gen_context(system_u:object_r:shadow_t,s0)
+
 /etc/\.pwd\.lock	--	gen_context(system_u:object_r:shadow_t,s0)
 /etc/group\.lock	--	gen_context(system_u:object_r:shadow_t,s0)
 /etc/gshadow.*		--	gen_context(system_u:object_r:shadow_t,s0)
@@ -40,6 +46,7 @@ ifdef(`distro_gentoo', `
 /var/log/syslog		--	gen_context(system_u:object_r:var_log_t,s0)
 /var/log/tallylog	--	gen_context(system_u:object_r:faillog_t,s0)
 /var/log/wtmp.*		--	gen_context(system_u:object_r:wtmp_t,s0)
+/var/volatile/log/wtmp.*	--	gen_context(system_u:object_r:wtmp_t,s0)
 
 /var/run/console(/.*)?	 	gen_context(system_u:object_r:pam_var_console_t,s0)
 /var/run/faillock(/.*)?		gen_context(system_u:object_r:faillog_t,s0)
diff --git a/policy/modules/system/authlogin.if b/policy/modules/system/authlogin.if
index 3efd5b6..043dbd6 100644
--- a/policy/modules/system/authlogin.if
+++ b/policy/modules/system/authlogin.if
@@ -1002,6 +1002,24 @@ interface(`auth_dontaudit_read_pam_pid',`
 	dontaudit $1 pam_var_run_t:file { getattr read };
 ')
 
+#######################################
+## <summary>
+##	Set attributes on PAM PID directories.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`auth_setattr_pam_pid_dirs',`
+	gen_require(`
+		type pam_var_run_t;
+	')
+
+	allow $1 pam_var_run_t:dir setattr;
+')
+
 ########################################
 ## <summary>
 ##	Delete pam PID files.
@@ -1570,6 +1588,7 @@ interface(`auth_setattr_login_records',`
 		type wtmp_t;
 	')
 
+	allow $1 wtmp_t:lnk_file read_lnk_file_perms;
 	allow $1 wtmp_t:file setattr;
 	logging_search_logs($1)
 ')
@@ -1592,6 +1611,7 @@ interface(`auth_read_login_records',`
 
 	logging_search_logs($1)
 	allow $1 wtmp_t:file read_file_perms;
+	allow $1 wtmp_t:lnk_file read_lnk_file_perms;
 ')
 
 ########################################
@@ -1610,7 +1630,7 @@ interface(`auth_dontaudit_read_login_records',`
 	gen_require(`
 		type wtmp_t;
 	')
-
+	dontaudit $1 wtmp_t:lnk_file read_lnk_file_perms;
 	dontaudit $1 wtmp_t:file read_file_perms;
 ')
 
@@ -1630,6 +1650,7 @@ interface(`auth_dontaudit_write_login_records',`
 		type wtmp_t;
 	')
 
+	dontaudit $1 wtmp_t:lnk_file read_lnk_file_perms;
 	dontaudit $1 wtmp_t:file write;
 ')
 
@@ -1648,6 +1669,7 @@ interface(`auth_append_login_records',`
 		type wtmp_t;
 	')
 
+	allow $1 wtmp_t:lnk_file read_lnk_file_perms;
 	allow $1 wtmp_t:file append_file_perms;
 	logging_search_logs($1)
 ')
@@ -1667,6 +1689,7 @@ interface(`auth_write_login_records',`
 		type wtmp_t;
 	')
 
+	allow $1 wtmp_t:lnk_file read_lnk_file_perms;
 	allow $1 wtmp_t:file { write_file_perms lock };
 ')
 
@@ -1685,6 +1708,7 @@ interface(`auth_rw_login_records',`
 		type wtmp_t;
 	')
 
+	allow $1 wtmp_t:lnk_file read_lnk_file_perms;
 	allow $1 wtmp_t:file rw_file_perms;
 	logging_search_logs($1)
 ')
@@ -1705,6 +1729,7 @@ interface(`auth_log_filetrans_login_records',`
 		type wtmp_t;
 	')
 
+	allow $1 wtmp_t:lnk_file read_lnk_file_perms;
 	logging_log_filetrans($1, wtmp_t, file)
 ')
 
@@ -1724,6 +1749,7 @@ interface(`auth_manage_login_records',`
 		type wtmp_t;
 	')
 
+	allow $1 wtmp_t:lnk_file read_lnk_file_perms;
 	logging_rw_generic_log_dirs($1)
 	allow $1 wtmp_t:file manage_file_perms;
 ')
