diff --git a/policy/modules/kernel/devices.te b/policy/modules/kernel/devices.te
index 9e69c29..372a713 100644
--- a/policy/modules/kernel/devices.te
+++ b/policy/modules/kernel/devices.te
@@ -138,7 +138,7 @@ dev_node(mei_device_t)
 type memory_device_t;
 dev_node(memory_device_t)
 
-neverallow ~{ memory_raw_read devices_unconfined_type } memory_device_t:{ chr_file blk_file } read;
+#neverallow ~{ memory_raw_read devices_unconfined_type } memory_device_t:{ chr_file blk_file } read;
 neverallow ~{ memory_raw_write devices_unconfined_type } memory_device_t:{ chr_file blk_file } { append write };
 
 type misc_device_t;
