Index: refpolicy/policy/modules/contrib/apm.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/apm.te
+++ refpolicy/policy/modules/contrib/apm.te
@@ -64,6 +64,7 @@ dontaudit apmd_t self:capability { setui
 allow apmd_t self:process { signal_perms getsession };
 allow apmd_t self:fifo_file rw_fifo_file_perms;
 allow apmd_t self:netlink_socket create_socket_perms;
+allow apmd_t self:netlink_generic_socket create_socket_perms;
 allow apmd_t self:unix_stream_socket { accept listen };
 
 allow apmd_t apmd_lock_t:file manage_file_perms;
@@ -100,6 +101,8 @@ dev_rw_apm_bios(apmd_t)
 dev_rw_sysfs(apmd_t)
 dev_dontaudit_getattr_all_chr_files(apmd_t)
 dev_dontaudit_getattr_all_blk_files(apmd_t)
+# Early devtmpfs, before udev relabel
+dev_dontaudit_rw_generic_chr_files(apmd_t)
 
 files_exec_etc_files(apmd_t)
 files_read_etc_runtime_files(apmd_t)
