--- refpolicy/policy/modules/contrib/tpmutil.fc	2017-01-11 11:07:30.101100033 -0500
+++ refpolicy/policy/modules/contrib/tpmutil.fc	2017-01-11 11:08:15.653293716 -0500
@@ -23,3 +23,4 @@
 
 /usr/(s)?bin/tpm(nv)?_[a-z]+	--	gen_context(system_u:object_r:tpmutil_exec_t,s0)
 /usr/bin/tpm(nv)?_[a-z]+_sa	--	gen_context(system_u:object_r:tpmutil_sa_exec_t,s0)
+/usr/(s)?bin/tpm2?_[a-z]+  --  gen_context(system_u:object_r:tpmutil_exec_t,s0)
