commit fcba88e086900157a9e50ec3e162c3b2df2e087a
Author: Jean Guyader <jean.guyader@eu.citrix.com>
Date:   Mon May 9 14:08:58 2011 +0100

    [grub] Branding....

diff --git a/boot/i386/pc/boot.S b/boot/i386/pc/boot.S
index 257f904..98724a4 100644
--- a/boot/i386/pc/boot.S
+++ b/boot/i386/pc/boot.S
@@ -385,7 +385,7 @@ LOCAL(general_error):
 LOCAL(stop):
 	jmp	LOCAL(stop)
 
-notification_string:	.asciz "GRUB "
+notification_string:	.asciz "     "
 geometry_error_string:	.asciz "Geom"
 hd_probe_error_string:	.asciz "Hard Disk"
 read_error_string:	.asciz "Read"
diff --git a/boot/i386/pc/diskboot.S b/boot/i386/pc/diskboot.S
index 92f223f..9d3b16a 100644
--- a/boot/i386/pc/diskboot.S
+++ b/boot/i386/pc/diskboot.S
@@ -320,9 +320,9 @@ LOCAL(general_error):
 /* go here when you need to stop the machine hard after an error condition */
 LOCAL(stop):	jmp	LOCAL(stop)
 
-notification_string:	.asciz "loading"
+notification_string:	.asciz "       "
 
-notification_step:	.asciz "."
+notification_step:	.asciz " "
 notification_done:	.asciz "\r\n"
 
 geometry_error_string:	.asciz "Geom"
diff --git a/kern/main.c b/kern/main.c
index e7566fa..26cf732 100644
--- a/kern/main.c
+++ b/kern/main.c
@@ -154,10 +154,12 @@ grub_main (void)
   /* First of all, initialize the machine.  */
   grub_machine_init ();
 
+#if 0
   /* Hello.  */
   grub_setcolorstate (GRUB_TERM_COLOR_HIGHLIGHT);
   grub_printf ("Welcome to GRUB!\n\n");
   grub_setcolorstate (GRUB_TERM_COLOR_STANDARD);
+#endif
 
   /* Load pre-loaded modules and free the space.  */
   grub_register_exported_symbols ();
diff --git a/normal/menu.c b/normal/menu.c
index 07951da..1868228 100644
--- a/normal/menu.c
+++ b/normal/menu.c
@@ -492,9 +492,11 @@ static void
 notify_booting (grub_menu_entry_t entry,
 		void *userdata __attribute__((unused)))
 {
+#if 0
   grub_printf ("  ");
   grub_printf_ (N_("Booting \'%s\'"), entry->title);
   grub_printf ("\n\n");
+#endif
 }
 
 /* Callback invoked when a default menu entry executed because of a timeout
