diff --git a/hs/DBus/Connection.hs b/hs/DBus/Connection.hs
index ba86ee3..9f824c7 100644
--- a/hs/DBus/Connection.hs
+++ b/hs/DBus/Connection.hs
@@ -20,6 +20,7 @@
 {-# LANGUAGE DeriveDataTypeable #-}
 module DBus.Connection
         (   Connection
+          , Transport (..)
           , connectionAddress
           , connectionUUID
 
@@ -27,6 +28,7 @@ module DBus.Connection
 
           , connect
           , connectFirst
+          , connectUsingTransport
 
           , connectionClose
 
@@ -198,6 +200,14 @@ connect mechanism a = do
         serialMVar <- C.newMVar M.firstSerial
         return $ Connection a t serialMVar readLock uuid
 
+connectUsingTransport :: Auth.Mechanism -> Transport -> IO Connection
+connectUsingTransport mechanism t = do
+        let getByte = L.head `fmap` transportRecv t 1
+        uuid <- Auth.authenticate mechanism (transportSend t) getByte
+        readLock <- C.newMVar ()
+        serialMVar <- C.newMVar M.firstSerial
+        return $ Connection undefined t serialMVar readLock uuid
+
 -- | Try to open a connection to various addresses, returning the first
 -- connection which could be successfully opened.
 
