Index: tboot-1.9.5/Config.mk
===================================================================
--- tboot-1.9.5.orig/Config.mk
+++ tboot-1.9.5/Config.mk
@@ -45,20 +45,16 @@ CFLAGS_WARN       = -Wall -Wformat-secur
 	            $(call cc-option,$(CC),-Wlogical-op,) \
 	            -Wno-missing-field-initializers
 
-AS         = as
-LD         = ld
-CC         = gcc
-CPP        = cpp
-AR         = ar
-RANLIB     = ranlib
-NM         = nm
-STRIP      = strip
-OBJCOPY    = objcopy
-OBJDUMP    = objdump
-
-ifeq ($(debug),n)
-INSTALL_STRIP = -s
-endif
+AS         ?= as
+LD         ?= ld
+CC         ?= gcc
+CPP        ?= cpp
+AR         ?= ar
+RANLIB     ?= ranlib
+NM         ?= nm
+STRIP      ?= strip
+OBJCOPY    ?= objcopy
+OBJDUMP    ?= objdump
 
 INSTALL      = install
 INSTALL_DIR  = $(INSTALL) -d -m0755 -p
