Index: tboot-1.9.5/include/tboot.h
===================================================================
--- tboot-1.9.5.orig/include/tboot.h
+++ tboot-1.9.5/include/tboot.h
@@ -152,6 +152,15 @@ static inline bool tboot_in_measured_env
     return (g_tboot_shared != NULL);
 }
 
+#if __WORDSIZE == 64
+# define __PRI64_PREFIX	"l"
+# define __PRIPTR_PREFIX	"l"
+#else
+# define __PRI64_PREFIX	"ll"
+# define __PRIPTR_PREFIX
+#endif
+#define PRIu64		__PRI64_PREFIX "u"
+
 static inline void print_tboot_shared(const tboot_shared_t *tboot_shared)
 {
     printk(TBOOT_DETA"tboot_shared data:\n");
@@ -165,8 +174,8 @@ static inline void print_tboot_shared(co
     printk(TBOOT_DETA"\t flags: 0x%8.8x\n", tboot_shared->flags);
     printk(TBOOT_DETA"\t ap_wake_addr: 0x%08x\n", (uint32_t)tboot_shared->ap_wake_addr);
     printk(TBOOT_DETA"\t ap_wake_trigger: %u\n", tboot_shared->ap_wake_trigger);
-    printk(TBOOT_DETA"\t evt_log_region: 0x%llx\n", tboot_shared->evt_log_region);
-    printk(TBOOT_DETA"\t evt_log_size: 0x%llx\n", tboot_shared->evt_log_size);
+    printk(TBOOT_DETA"\t evt_log_region: 0x%"PRIu64"\n", tboot_shared->evt_log_region);
+    printk(TBOOT_DETA"\t evt_log_size: 0x%"PRIu64"\n", tboot_shared->evt_log_size);
 }
 
 #endif    /* __TBOOT_H__ */

