Index: tboot-1.9.5/tboot/common/cmdline.c
===================================================================
--- tboot-1.9.5.orig/tboot/common/cmdline.c
+++ tboot-1.9.5/tboot/common/cmdline.c
@@ -526,6 +526,11 @@ void get_tboot_extpol(void)
         g_tpm->extpol = TB_EXTPOL_FIXED;
         g_tpm->cur_alg = TB_HALG_SM3;
     }
+
+    if (g_tpm->major == TPM20_VER_MAJOR) {
+        g_tpm->extpol = TB_EXTPOL_FIXED;
+        g_tpm->cur_alg = TB_HALG_SHA256;
+    }
 }
 
 /*
