From 899298bab2d469514091a41b2597f898cf7aad2b Mon Sep 17 00:00:00 2001
From: Tamas K Lengyel <tamas@tklengyel.com>
Date: Mon, 30 Oct 2017 15:18:12 -0600
Subject: [PATCH 2/6] Make EFI_INCLUDE path configurable during make

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 48f4e7c..15c2794 100644
--- a/Makefile
+++ b/Makefile
@@ -40,7 +40,6 @@ OBJCOPY_GTE224	= $(shell expr `$(OBJCOPY) --version |grep ^"GNU objcopy" | sed '
 
 SUBDIRS		= $(TOPDIR)/Cryptlib $(TOPDIR)/lib
 
-EFI_INCLUDE	:= /usr/include/efi
 EFI_INCLUDES	= -nostdinc -I$(TOPDIR)/Cryptlib -I$(TOPDIR)/Cryptlib/Include \
 		  -I$(EFI_INCLUDE) -I$(EFI_INCLUDE)/$(ARCH) -I$(EFI_INCLUDE)/protocol \
 		  -I$(TOPDIR)/include -iquote $(TOPDIR) -iquote $(shell pwd)
@@ -117,6 +116,7 @@ endif
 
 FORMAT		?= --target efi-app-$(ARCH)
 EFI_PATH	?= $(LIBDIR)/gnuefi
+EFI_INCLUDE	?= /usr/include/efi
 
 MMSTEM		?= mm$(ARCH_SUFFIX)
 MMNAME		= $(MMSTEM).efi
-- 
2.1.4

