From 6fd9327ccde185c547635279ec2d1211a35e4368 Mon Sep 17 00:00:00 2001
From: Tamas K Lengyel <tamas@tklengyel.com>
Date: Sun, 31 Dec 2017 11:03:26 -0700
Subject: [PATCH 4/6] Don't overwrite LoadOptions if NULL

---
 shim.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/shim.c b/shim.c
index ea186ed..7723119 100644
--- a/shim.c
+++ b/shim.c
@@ -1501,8 +1501,10 @@ static EFI_STATUS handle_image (void *data, unsigned int datasize,
 	li->ImageSize = context.ImageSize;
 
 	/* Pass the load options to the second stage loader */
-	li->LoadOptions = load_options;
-	li->LoadOptionsSize = load_options_size;
+	if ( load_options ) {
+		li->LoadOptions = load_options;
+		li->LoadOptionsSize = load_options_size;
+	}
 
 	if (!found_entry_point) {
 		perror(L"Entry point is not within sections\n");
-- 
2.1.4

