--- a/configure
+++ b/configure
@@ -1447,7 +1447,8 @@ fi
 # xen probe
 
 if test "$xen" != "no" ; then
-  xen_libs="-lxenstore -lxenctrl -lxenguest"
+  # We need to add -ldl and -lpthread because XenStore depends on it
+  xen_libs="-lxenstore -lxenctrl -lxenguest -ldl -lpthread"
 
   # First we test whether Xen headers and libraries are available.
   # If no, we are done and there is no Xen support.
@@ -2192,7 +2193,7 @@ else
   fi
   mkdir -p pixman/pixman
   pixman_cflags="-I\$(SRC_PATH)/pixman/pixman -I\$(BUILD_DIR)/pixman/pixman"
-  pixman_libs="-L\$(BUILD_DIR)/pixman/pixman/.libs -lpixman-1"
+  pixman_libs="-L\$(BUILD_DIR)/pixman/pixman/.libs -Wl,-Bstatic -lpixman-1 -Wl,-Bdynamic"
 fi
 
 ##########################################
