From: Jed <lejosnej@ainfosec.com>
Date: Fri, 23 Jan 2015 16:54:42 -0500
Subject: [PATCH] qemu: add a patch for OpenXT audio policy bits, with just the audio recording flag for now

Index: qemu-1.4.0/hw/ac97.c
===================================================================
--- qemu-1.4.0.orig/hw/ac97.c	2015-04-29 16:09:03.212306480 -0400
+++ qemu-1.4.0/hw/ac97.c	2015-04-29 16:14:30.872307013 -0400
@@ -23,6 +23,8 @@
 #include "pci/pci.h"
 #include "sysemu/dma.h"
 
+extern int disable_audio_recording;
+
 enum {
     AC97_Reset                     = 0x00,
     AC97_Master_Volume_Mute        = 0x02,
@@ -1034,6 +1036,13 @@
     int to_copy = 0;
     SWVoiceIn *voice = (r - s->bm_regs) == MC_INDEX ? s->voice_mc : s->voice_pi;
 
+    /* Disable audio recording. */
+    if (disable_audio_recording)
+    {
+        *stop = 1;
+        return 0;
+    }
+
     temp = audio_MIN (temp, max);
 
     if (!temp) {
Index: qemu-1.4.0/hw/hda-audio.c
===================================================================
--- qemu-1.4.0.orig/hw/hda-audio.c	2015-04-29 16:09:03.204307200 -0400
+++ qemu-1.4.0/hw/hda-audio.c	2015-04-29 16:14:30.872307013 -0400
@@ -25,6 +25,8 @@
 
 /* -------------------------------------------------------------------------- */
 
+extern int disable_audio_recording;
+
 typedef struct desc_param {
     uint32_t id;
     uint32_t val;
@@ -604,10 +606,13 @@
                 break;
             }
         }
-        rc = hda_codec_xfer(&st->state->hda, st->stream, false,
-                            st->buf, sizeof(st->buf));
-        if (!rc) {
-            break;
+        /* Only transfer input if allowed by the policy */
+        if (!disable_audio_recording) {
+            rc = hda_codec_xfer(&st->state->hda, st->stream, false,
+                                st->buf, sizeof(st->buf));
+            if (!rc) {
+                break;
+            }
         }
         st->bpos = 0;
     }
Index: qemu-1.4.0/qemu-options.hx
===================================================================
--- qemu-1.4.0.orig/qemu-options.hx	2015-04-29 16:09:16.052306083 -0400
+++ qemu-1.4.0/qemu-options.hx	2015-04-29 16:14:30.876308660 -0400
@@ -520,6 +520,14 @@
 @end example
 ETEXI
 
+DEF("disable-audio-rec", 0, QEMU_OPTION_disable_audio_rec,
+    "-disable-audio-rec Disable audio recording\n", QEMU_ARCH_ALL)
+STEXI
+@item -disable-audio-rec
+@findex -disable-audio-rec
+Disable audio recording in the guest.
+ETEXI
+
 DEF("balloon", HAS_ARG, QEMU_OPTION_balloon,
     "-balloon none   disable balloon device\n"
     "-balloon virtio[,addr=str]\n"
Index: qemu-1.4.0/vl.c
===================================================================
--- qemu-1.4.0.orig/vl.c	2015-04-29 16:11:09.448310320 -0400
+++ qemu-1.4.0/vl.c	2015-04-29 16:14:30.876308660 -0400
@@ -236,6 +236,7 @@
 uint8_t *boot_splash_filedata;
 size_t boot_splash_filedata_size;
 uint8_t qemu_extra_params_fw[2];
+int disable_audio_recording = 0;
 
 typedef struct FWBootEntry FWBootEntry;
 
@@ -3206,6 +3207,9 @@
                 }
                 select_soundhw (optarg);
                 break;
+            case QEMU_OPTION_disable_audio_rec:
+                disable_audio_recording = 1;
+                break;
             case QEMU_OPTION_h:
                 help(0);
                 break;
