Index: qemu-1.4.0/configure
===================================================================
--- qemu-1.4.0.orig/configure	2015-05-05 09:53:06.250537018 -0400
+++ qemu-1.4.0/configure	2015-05-05 09:55:39.030068565 -0400
@@ -231,6 +231,7 @@
 virtio_blk_data_plane=""
 atapipt="no"
 atapiptv4v="no"
+stubdom="no"
 xenpt_hide_msix="yes"
 
 # parse CC options first
@@ -911,6 +912,10 @@
   ;;
   --disable-atapi-pt-v4v) atapiptv4v="no"
   ;;
+  --enable-openxt-stubdom) stubdom="yes"
+  ;;
+  --disable-openxt-stubdom) stubdom="no"
+  ;;
   --disable-glusterfs) glusterfs="no"
   ;;
   --enable-glusterfs) glusterfs="yes"
@@ -3409,6 +3414,7 @@
 echo "gcov enabled      $gcov"
 echo "ATAPI PT support  $atapipt"
 echo "ATAPI PT V4V support  $atapiptv4v"
+echo "OpenXT Stubdomain support   $stubdom"
 
 if test "$sdl_too_old" = "yes"; then
 echo "-> Your SDL version is too old - please upgrade to have SDL support"
@@ -3775,6 +3781,10 @@
   echo "CONFIG_ATAPI_PT_V4V=y" >> $config_host_mak
 fi
 
+if test "$stubdom" = "yes" ; then
+  echo "CONFIG_STUBDOM=y" >> $config_host_mak
+fi
+
 if test "$virtio_blk_data_plane" = "yes" ; then
   echo "CONFIG_VIRTIO_BLK_DATA_PLANE=y" >> $config_host_mak
 fi
