diff -uNr dialog-1.1-20080819.orig/configure dialog-1.1-20080819/configure
--- dialog-1.1-20080819.orig/configure	2008-03-16 18:54:16.000000000 +0000
+++ dialog-1.1-20080819/configure	2010-11-29 17:29:16.076467317 +0000
@@ -7465,6 +7465,7 @@
 esac
 fi
 NCURSES_CONFIG=$ac_cv_path_NCURSES_CONFIG
+NCURSES_CONFIG=
 
 if test -n "$NCURSES_CONFIG"; then
   echo "$as_me:7470: result: $NCURSES_CONFIG" >&5
