diff -uNr dialog-1.1-20080819.orig/inputstr.c dialog-1.1-20080819/inputstr.c
--- dialog-1.1-20080819.orig/inputstr.c	2008-03-16 20:19:43.000000000 +0000
+++ dialog-1.1-20080819/inputstr.c	2010-11-30 16:03:14.101351189 +0000
@@ -699,16 +699,18 @@
 	for (i = scrollamt, k = 0; i < limit && k < x_last; ++i) {
 	    int check = cols[i + 1] - cols[scrollamt];
 	    if (check <= x_last) {
-		for (j = indx[i]; j < indx[i + 1]; ++j) {
-		    int ch = UCH(string[j]);
-		    if (hidden && dialog_vars.insecure) {
-			waddch(win, '*');
-		    } else if (ch == TAB) {
-			int count = cols[i + 1] - cols[i];
-			while (--count >= 0)
-			    waddch(win, ' ');
-		    } else {
-			waddch(win, ch);
+		if (hidden && dialog_vars.insecure) {
+		    waddch(win, '*');
+		} else {
+		    for (j = indx[i]; j < indx[i + 1]; ++j) {
+			int ch = UCH(string[j]);
+			if (ch == TAB) {
+			    int count = cols[i + 1] - cols[i];
+			    while (--count >= 0)
+				waddch(win, ' ');
+			} else {
+			    waddch(win, ch);
+			}
 		    }
 		}
 		k = check;
