Index: refpolicy/policy/modules/contrib/qemu.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/qemu.te
+++ refpolicy/policy/modules/contrib/qemu.te
@@ -119,6 +119,11 @@ optional_policy(`
 # Unconfined local policy
 #
 
+# qemu emulates disk for early boot code
+optional_policy(`
+	blktap_rw_blk_file(qemu_t)
+')
+
 optional_policy(`
 	type unconfined_qemu_t;
 	typealias unconfined_qemu_t alias qemu_unconfined_t;
Index: refpolicy/policy/modules/contrib/xen.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/xen.te
+++ refpolicy/policy/modules/contrib/xen.te
@@ -257,6 +257,15 @@ qemu_signull(xend_t)
 # transition to dmidecode
 dmidecode_domtrans(xend_t)
 
+blktap_getattr_blk_file(xend_t)
+blktap_relabel_blk_file(xend_t)
+# this may be a leaked fd or something, dontaudit?
+blktap_read_blk_file(xend_t)
+
+# transition to tapdisk utilities
+tapdisk_domtrans(xend_t)
+tapctl_domtrans(xend_t)
+
 kernel_read_kernel_sysctls(xend_t)
 kernel_read_system_state(xend_t)
 kernel_write_xen_state(xend_t)
Index: refpolicy/policy/modules/roles/sysadm.te
===================================================================
--- refpolicy.orig/policy/modules/roles/sysadm.te
+++ refpolicy/policy/modules/roles/sysadm.te
@@ -1084,6 +1084,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	tapctl_run(sysadm_t, sysadm_r)
+')
+
+optional_policy(`
 	thunderbird_role(sysadm_r, sysadm_t)
 ')
 
Index: refpolicy/policy/modules/system/lvm.te
===================================================================
--- refpolicy.orig/policy/modules/system/lvm.te
+++ refpolicy/policy/modules/system/lvm.te
@@ -346,6 +346,10 @@ ifdef(`distro_redhat',`
 ')
 
 optional_policy(`
+	blktap_rw_blk_file(lvm_t)
+')
+
+optional_policy(`
 	bootloader_rw_tmp_files(lvm_t)
 ')
 
Index: refpolicy/policy/modules/system/mount.te
===================================================================
--- refpolicy.orig/policy/modules/system/mount.te
+++ refpolicy/policy/modules/system/mount.te
@@ -179,6 +179,10 @@ tunable_policy(`allow_mount_anyfile',`
 ')
 
 optional_policy(`
+	blktap_read_blk_file(mount_t)
+')
+
+optional_policy(`
 	# for nfs
 	corenet_all_recvfrom_unlabeled(mount_t)
 	corenet_all_recvfrom_netlabel(mount_t)
