--- a/policy/modules/apps/qemu.te
+++ b/policy/modules/apps/qemu.te
@@ -134,6 +134,11 @@ optional_policy(`
 # Unconfined local policy
 #
 
+# qemu emulates disk for early boot code
+optional_policy(`
+	blktap_rw_blk_file(qemu_t)
+')
+
 optional_policy(`
 	type unconfined_qemu_t;
 	typealias unconfined_qemu_t alias qemu_unconfined_t;
--- a/policy/modules/system/xen.te
+++ b/policy/modules/system/xen.te
@@ -261,6 +261,16 @@ qemu_signull(xend_t)
 # transition to dmidecode
 dmidecode_domtrans(xend_t)
 
+blktap_getattr_blk_file(xend_t)
+blktap_relabel_blk_file(xend_t)
+# this may be a leaked fd or something, dontaudit?
+blktap_read_blk_file(xend_t)
+
+# transition to tapdisk utilities
+tapdisk_domtrans(xend_t)
+tapctl_domtrans(xend_t)
+tapdisk_filetrans_control_dir(xend_t)
+
 kernel_read_kernel_sysctls(xend_t)
 kernel_read_system_state(xend_t)
 kernel_write_xen_state(xend_t)
--- a/policy/modules/roles/sysadm.te
+++ b/policy/modules/roles/sysadm.te
@@ -1112,6 +1112,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	tapctl_run(sysadm_t, sysadm_r)
+')
+
+optional_policy(`
 	thunderbird_role(sysadm_r, sysadm_t)
 ')
 
--- a/policy/modules/system/lvm.te
+++ b/policy/modules/system/lvm.te
@@ -341,6 +341,10 @@ ifdef(`distro_redhat',`
 ')
 
 optional_policy(`
+	blktap_rw_blk_file(lvm_t)
+')
+
+optional_policy(`
 	bootloader_rw_tmp_files(lvm_t)
 ')
 
--- a/policy/modules/system/mount.te
+++ b/policy/modules/system/mount.te
@@ -183,6 +183,10 @@ tunable_policy(`allow_mount_anyfile',`
 ')
 
 optional_policy(`
+	blktap_read_blk_file(mount_t)
+')
+
+optional_policy(`
 	# for nfs
 	corenet_all_recvfrom_unlabeled(mount_t)
 	corenet_all_recvfrom_netlabel(mount_t)
