Index: refpolicy/build.conf
===================================================================
--- refpolicy.orig/build.conf
+++ refpolicy/build.conf
@@ -8,17 +8,17 @@
 # version policy it supports.  Setting this will
 # override the version.  This only has an
 # effect for monolithic policies.
-#OUTPUT_POLICY = 18
+# OUTPUT_POLICY = 18
 
 # Policy Type
 # standard, mls, mcs
-TYPE = standard
+TYPE = mcs
 
 # Policy Name
 # If set, this will be used as the policy
 # name.  Otherwise the policy type will be
 # used for the name.
-NAME = refpolicy
+NAME = xc_policy
 
 # Distribution
 # Some distributions have portions of policy
@@ -27,7 +27,7 @@ NAME = refpolicy
 # for the distribution.
 # redhat, gentoo, debian, suse, and rhel4 are current options.
 # Fedora users should enable redhat.
-#DISTRO = redhat
+DISTRO = debian
 
 # Unknown Permissions Handling
 # The behavior for handling permissions defined in the
@@ -42,7 +42,7 @@ UNK_PERMS = deny
 # run init scripts, instead of requring run_init.
 # This is a build option, as role transitions do
 # not work in conditional policy.
-DIRECT_INITRC = n
+DIRECT_INITRC = y 
 
 # Systemd
 # Setting this will configure systemd as the init system.
