--- a/policy/modules/system/hostname.te
+++ b/policy/modules/system/hostname.te
@@ -70,3 +70,9 @@ optional_policy(`
 optional_policy(`
 	unconfined_dontaudit_rw_pipes(hostname_t)
 ')
+
+# linux commit: 9f834ec18defc369d73ccf9e87a2790bfa05bf46 causes SELinux
+# read/execute checks on mmap of executables to be done in the context of the
+# new domain rather than the caller. So the following line is required on
+# busybox for the hostname script to have the appropriate process domain.
+corecmd_mmap_bin_files(hostname_t)
--- a/policy/modules/system/sysnetwork.te
+++ b/policy/modules/system/sysnetwork.te
@@ -377,6 +377,12 @@ userdom_read_all_users_state(ifconfig_t)
 userdom_use_user_terminals(ifconfig_t)
 userdom_use_all_users_fds(ifconfig_t)
 
+# linux commit: 9f834ec18defc369d73ccf9e87a2790bfa05bf46 causes SELinux
+# read/execute checks on mmap of executables to be done in the context of the
+# new domain rather than the caller. So the following line is required on
+# busybox for the ifconfig script to have the appropriate process domain.
+corecmd_mmap_bin_files(ifconfig_t)
+
 ifdef(`distro_debian',`
 	term_dontaudit_use_unallocated_ttys(ifconfig_t)
 ')
