Index: refpolicy/policy/modules/system/hostname.te
===================================================================
--- refpolicy.orig/policy/modules/system/hostname.te
+++ refpolicy/policy/modules/system/hostname.te
@@ -56,6 +56,12 @@ sysnet_dontaudit_rw_dhcpc_unix_stream_so
 sysnet_read_config(hostname_t)
 sysnet_dns_name_resolve(hostname_t)
 
+# linux commit: 9f834ec18defc369d73ccf9e87a2790bfa05bf46 causes SELinux
+# read/execute checks on mmap of executables to be done in the context of the
+# new domain rather than the caller. So the following line is required on
+# busybox for the hostname script to have the appropriate process domain.
+corecmd_mmap_bin_files(hostname_t)
+
 ifdef(`distro_debian',`
 	term_dontaudit_use_unallocated_ttys(hostname_t)
 ')
Index: refpolicy/policy/modules/system/sysnetwork.te
===================================================================
--- refpolicy.orig/policy/modules/system/sysnetwork.te
+++ refpolicy/policy/modules/system/sysnetwork.te
@@ -356,6 +356,12 @@ sysnet_dontaudit_rw_dhcpc_udp_sockets(if
 userdom_use_user_terminals(ifconfig_t)
 userdom_use_all_users_fds(ifconfig_t)
 
+# linux commit: 9f834ec18defc369d73ccf9e87a2790bfa05bf46 causes SELinux
+# read/execute checks on mmap of executables to be done in the context of the
+# new domain rather than the caller. So the following line is required on
+# busybox for the ifconfig script to have the appropriate process domain.
+corecmd_mmap_bin_files(ifconfig_t)
+
 ifdef(`distro_debian',`
 	term_dontaudit_use_unallocated_ttys(ifconfig_t)
 ')
