Index: refpolicy/policy/modules/contrib/dnsmasq.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/dnsmasq.te
+++ refpolicy/policy/modules/contrib/dnsmasq.te
@@ -113,6 +113,8 @@ miscfiles_read_localization(dnsmasq_t)
 userdom_dontaudit_use_unpriv_user_fds(dnsmasq_t)
 userdom_dontaudit_search_user_home_dirs(dnsmasq_t)
 
+db_cmd_domtrans(dnsmasq_t)
+
 optional_policy(`
 	cobbler_read_lib_files(dnsmasq_t)
 ')
Index: refpolicy/policy/modules/contrib/networkmanager.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/networkmanager.te
+++ refpolicy/policy/modules/contrib/networkmanager.te
@@ -229,6 +229,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	db_cmd_domtrans(NetworkManager_t)
+')
+
+optional_policy(`
 	dbus_system_domain(NetworkManager_t, NetworkManager_exec_t)
 
 	optional_policy(`
Index: refpolicy/policy/modules/roles/sysadm.te
===================================================================
--- refpolicy.orig/policy/modules/roles/sysadm.te
+++ refpolicy/policy/modules/roles/sysadm.te
@@ -1260,6 +1260,10 @@ ifndef(`distro_redhat',`
 	')
 
 	optional_policy(`
+	       db_cmd_run(sysadm_t, sysadm_r)
+	')
+
+	optional_policy(`
 		dbus_role_template(sysadm, sysadm_r, sysadm_t)
 
 		optional_policy(`
