Index: refpolicy/policy/modules/contrib/loadkeys.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/loadkeys.te
+++ refpolicy/policy/modules/contrib/loadkeys.te
@@ -49,5 +49,9 @@ optional_policy(`
 ')
 
 optional_policy(`
+	input_server_rw_stream(loadkeys_t)
+')
+
+optional_policy(`
 	nscd_dontaudit_search_pid(loadkeys_t)
 ')
Index: refpolicy/policy/modules/contrib/qemu.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/qemu.te
+++ refpolicy/policy/modules/contrib/qemu.te
@@ -96,6 +96,10 @@ tunable_policy(`qemu_full_network',`
 ')
 
 optional_policy(`
+	input_server_stream_connect(qemu_t)
+')
+
+optional_policy(`
 	xserver_user_x_domain_template(qemu, qemu_t, qemu_tmpfs_t)
 ')
 
Index: refpolicy/policy/modules/contrib/xen.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/xen.te
+++ refpolicy/policy/modules/contrib/xen.te
@@ -257,6 +257,10 @@ qemu_signull(xend_t)
 # transition to dmidecode
 dmidecode_domtrans(xend_t)
 
+optional_policy(`
+	input_server_dbus_chat(xend_t)
+')
+
 blktap_getattr_blk_file(xend_t)
 blktap_relabel_blk_file(xend_t)
 # this may be a leaked fd or something, dontaudit?
Index: refpolicy/policy/modules/system/fstools.te
===================================================================
--- refpolicy.orig/policy/modules/system/fstools.te
+++ refpolicy/policy/modules/system/fstools.te
@@ -215,6 +215,10 @@ optional_policy(`
 	xen_rw_image_files(fsadm_t)
 ')
 
+optional_policy(`
+	input_server_write_tmp_files(fsadm_t)
+')
+
 # leaked FDs from input-server
 term_dontaudit_use_unallocated_ttys(fsadm_t)
 optional_policy(`
Index: refpolicy/policy/modules/system/lvm.te
===================================================================
--- refpolicy.orig/policy/modules/system/lvm.te
+++ refpolicy/policy/modules/system/lvm.te
@@ -369,6 +369,11 @@ optional_policy(`
 	')
 ')
 
+# redirect output from cryptsetup to a tmp file
+optional_policy(`
+	input_server_write_tmp_files(lvm_t)
+')
+
 # dontaudit FDs leaked by input-server
 optional_policy(`
 	kernel_dontaudit_read_xen_state(lvm_t)
Index: refpolicy/policy/modules/system/mount.te
===================================================================
--- refpolicy.orig/policy/modules/system/mount.te
+++ refpolicy/policy/modules/system/mount.te
@@ -250,6 +250,10 @@ optional_policy(`
 	unconfined_run_to(unconfined_mount_t, mount_exec_t)
 ')
 
+optional_policy(`
+	input_server_write_tmp_files(mount_t)
+')
+
 # leaked FDs from input-server
 optional_policy(`
 	kernel_dontaudit_read_xen_state(mount_t)
Index: refpolicy/policy/modules/system/selinuxutil.te
===================================================================
--- refpolicy.orig/policy/modules/system/selinuxutil.te
+++ refpolicy/policy/modules/system/selinuxutil.te
@@ -659,6 +659,11 @@ optional_policy(`
 	hotplug_use_fds(setfiles_t)
 ')
 
+# input-server redirects output to a tmp file
+optional_policy(`
+	input_server_write_tmp_files(setfiles_t)
+')
+
 # leaked FDs from input-server
 optional_policy(`
 	kernel_dontaudit_read_xen_state(setfiles_t)
