--- a/policy/modules/apps/loadkeys.te
+++ b/policy/modules/apps/loadkeys.te
@@ -53,5 +53,9 @@ optional_policy(`
 ')
 
 optional_policy(`
+	input_server_rw_stream(loadkeys_t)
+')
+
+optional_policy(`
 	nscd_dontaudit_search_pid(loadkeys_t)
 ')
--- a/policy/modules/apps/qemu.te
+++ b/policy/modules/apps/qemu.te
@@ -111,6 +111,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	input_server_stream_connect(qemu_t)
+')
+
+optional_policy(`
 	xserver_user_x_domain_template(qemu, qemu_t, qemu_tmpfs_t)
 ')
 
--- a/policy/modules/system/xen.te
+++ b/policy/modules/system/xen.te
@@ -261,6 +261,10 @@ qemu_signull(xend_t)
 # transition to dmidecode
 dmidecode_domtrans(xend_t)
 
+optional_policy(`
+	input_server_dbus_chat(xend_t)
+')
+
 blktap_getattr_blk_file(xend_t)
 blktap_relabel_blk_file(xend_t)
 # this may be a leaked fd or something, dontaudit?
--- a/policy/modules/system/fstools.te
+++ b/policy/modules/system/fstools.te
@@ -224,6 +224,10 @@ optional_policy(`
 	xen_rw_image_files(fsadm_t)
 ')
 
+optional_policy(`
+	input_server_write_tmp_files(fsadm_t)
+')
+
 # leaked FDs from input-server
 term_dontaudit_use_unallocated_ttys(fsadm_t)
 optional_policy(`
--- a/policy/modules/system/lvm.te
+++ b/policy/modules/system/lvm.te
@@ -396,3 +396,9 @@ optional_policy(`
 	xc_read_vhd_key_files(lvm_t)
 	xc_search_vhd_key_dirs(lvm_t)
 ')
+
+# redirect output from cryptsetup to a tmp file
+optional_policy(`
+	input_server_write_tmp_files(lvm_t)
+')
+
--- a/policy/modules/system/mount.te
+++ b/policy/modules/system/mount.te
@@ -254,6 +254,10 @@ optional_policy(`
 	unconfined_run_to(unconfined_mount_t, mount_exec_t)
 ')
 
+optional_policy(`
+	input_server_write_tmp_files(mount_t)
+')
+
 # leaked FDs from input-server
 optional_policy(`
 	kernel_dontaudit_read_xen_state(mount_t)
--- a/policy/modules/system/selinuxutil.te
+++ b/policy/modules/system/selinuxutil.te
@@ -719,6 +719,11 @@ optional_policy(`
 	hotplug_use_fds(setfiles_t)
 ')
 
+# input-server redirects output to a tmp file
+optional_policy(`
+	input_server_write_tmp_files(setfiles_t)
+')
+
 # leaked FDs from input-server
 optional_policy(`
 	kernel_dontaudit_read_xen_state(setfiles_t)
