--- a/policy/modules/contrib/brctl.te
+++ b/policy/modules/contrib/brctl.te
@@ -42,6 +42,10 @@ term_dontaudit_use_console(brctl_t)
 miscfiles_read_localization(brctl_t)
 
 optional_policy(`
+	network_slave_rw_stream_sockets(brctl_t)
+')
+
+optional_policy(`
 	xen_append_log(brctl_t)
 	xen_dontaudit_rw_unix_stream_sockets(brctl_t)
 ')
--- a/policy/modules/contrib/dnsmasq.te
+++ b/policy/modules/contrib/dnsmasq.te
@@ -133,6 +133,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	network_slave_rw_stream_sockets(dnsmasq_t)
+')
+
+optional_policy(`
 	seutil_sigchld_newrole(dnsmasq_t)
 ')
 
--- a/policy/modules/contrib/networkmanager.te
+++ b/policy/modules/contrib/networkmanager.te
@@ -303,6 +303,11 @@ optional_policy(`
 ')
 
 optional_policy(`
+	network_slave_rw_stream_sockets(NetworkManager_t)
+	network_slave_domtrans(NetworkManager_t)
+')
+
+optional_policy(`
 	nscd_domtrans(NetworkManager_t)
 	nscd_signal(NetworkManager_t)
 	nscd_signull(NetworkManager_t)
--- a/policy/modules/system/init.te
+++ b/policy/modules/system/init.te
@@ -908,6 +908,9 @@ optional_policy(`
 	optional_policy(`
 		policykit_dbus_chat(initrc_t)
 	')
+	optional_policy(`
+        	network_slave_dbus_chat(initrc_t)
+	')
 ')
 
 optional_policy(`
--- a/policy/modules/system/iptables.te
+++ b/policy/modules/system/iptables.te
@@ -125,6 +125,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	network_slave_rw_stream_sockets(iptables_t)
+')
+
+optional_policy(`
 	# for iptables -L
 	nis_use_ypbind(iptables_t)
 ')
--- a/policy/modules/system/modutils.te
+++ b/policy/modules/system/modutils.te
@@ -149,6 +149,12 @@ optional_policy(`
 ')
 
 optional_policy(`
+	network_slave_rw_stream_sockets(kmod_t)
+	network_slave_rw_fifo_files(kmod_t)
+	network_slave_use_fds(kmod_t)
+')
+
++optional_policy(`
 	nis_use_ypbind(kmod_t)
 ')
 
--- a/policy/modules/system/sysnetwork.te
+++ b/policy/modules/system/sysnetwork.te
@@ -392,6 +392,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	network_slave_rw_stream_sockets(ifconfig_t)
+')
+
+optional_policy(`
 	nis_use_ypbind(ifconfig_t)
 ')
 
--- a/policy/modules/system/udev.te
+++ b/policy/modules/system/udev.te
@@ -327,6 +327,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	network_slave_dbus_chat(udev_t)
+')
+
+optional_policy(`
 	openct_read_pid_files(udev_t)
 	openct_domtrans(udev_t)
 ')
