Index: refpolicy/policy/modules/roles/sysadm.te
===================================================================
--- refpolicy.orig/policy/modules/roles/sysadm.te
+++ refpolicy/policy/modules/roles/sysadm.te
@@ -1347,3 +1347,6 @@ optional_policy(`
 optional_policy(`
 	xsmutil_run(sysadm_t, sysadm_r)
 ')
+
+# For running lsusb from a nr shell.
+allow sysadm_t self:netlink_kobject_uevent_socket create_socket_perms;
