Index: refpolicy/policy/modules/contrib/dbus.fc
===================================================================
--- refpolicy.orig/policy/modules/contrib/dbus.fc
+++ refpolicy/policy/modules/contrib/dbus.fc
@@ -1,6 +1,7 @@
 HOME_DIR/\.dbus(/.*)?	gen_context(system_u:object_r:session_dbusd_home_t,s0)
 
 /etc/dbus-.*(/.*)?	gen_context(system_u:object_r:dbusd_etc_t,s0)
+/etc/dbus-1/event.d/.*	gen_context(system_u:object_r:bin_t,s0)
 
 /bin/dbus-daemon	--	gen_context(system_u:object_r:dbusd_exec_t,s0)
 
@@ -16,8 +17,10 @@ HOME_DIR/\.dbus(/.*)?	gen_context(system
 /usr/libexec/dbus-daemon-launch-helper	--	gen_context(system_u:object_r:dbusd_exec_t,s0)
 
 /var/lib/dbus(/.*)?	gen_context(system_u:object_r:system_dbusd_var_lib_t,s0)
+/var/volatile/dbus(/.*)		gen_context(system_u:object_r:system_dbusd_var_lib_t,s0)
 
 /run/dbus(/.*)?	gen_context(system_u:object_r:system_dbusd_var_run_t,s0)
 /run/messagebus\.pid	--	gen_context(system_u:object_r:system_dbusd_var_run_t,s0)
+/var/volatile/run/dbus(/.*)?	gen_context(system_u:object_r:system_dbusd_var_run_t,s0)
 
 /var/named/chroot/var/run/dbus(/.*)?	gen_context(system_u:object_r:system_dbusd_var_run_t,s0)
Index: refpolicy/policy/modules/contrib/dbus.if
===================================================================
--- refpolicy.orig/policy/modules/contrib/dbus.if
+++ refpolicy/policy/modules/contrib/dbus.if
@@ -121,6 +121,34 @@ interface(`dbus_system_bus_client',`
 
 	dbus_read_config($1)
 ')
+########################################
+## <summary>
+##	Allow the calling domain to create a private type in the
+##	/var/run/dbus directory.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Type creating entry in /var/run/dbus
+##	</summary>
+## </param>
+## <param name="new type">
+##	<summary>
+##	New type created.
+##	</summary>
+## </param>
+## <param name="classes">
+##	<summary>
+##	Classes of objects domtrans is valid for.
+##	</summary>
+## </param>
+#
+interface(`dbus_var_run_domtrans',`
+	gen_require(`
+		type system_dbusd_var_run_t;
+	')
+
+	filetrans_add_pattern($1, system_dbusd_var_run_t, $2, $3)
+')
 
 #######################################
 ## <summary>
@@ -626,3 +654,40 @@ interface(`dbus_unconfined',`
 
 	typeattribute $1 dbusd_unconfined;
 ')
+
+########################################
+## <summary>
+##	Dontaudit attempts to write to the
+##	dbus lib dir.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`dbus_dontaudit_write_lib_dirs',`
+	gen_require(`
+		type system_dbusd_var_lib_t;
+	')
+
+	dontaudit $1 system_dbusd_var_lib_t:dir write;
+')
+
+########################################
+## <summary>
+##	Allow caller to manage files in /var/lib/dbus.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`dbus_mange_var_lib',`
+	gen_require(`
+		type system_dbusd_var_lib_t;
+	')
+
+	manage_files_pattern($1, system_dbusd_var_lib_t, system_dbusd_var_lib_t)
+')
Index: refpolicy/policy/modules/contrib/dbus.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/dbus.te
+++ refpolicy/policy/modules/contrib/dbus.te
@@ -67,6 +67,7 @@ allow system_dbusd_t self:fifo_file rw_f
 allow system_dbusd_t self:dbus { send_msg acquire_svc };
 allow system_dbusd_t self:unix_stream_socket { accept connectto listen };
 allow system_dbusd_t self:netlink_selinux_socket { create bind read };
+allow system_dbusd_t self:netlink_kobject_uevent_socket create_socket_perms;
 
 allow system_dbusd_t dbusd_etc_t:dir list_dir_perms;
 read_files_pattern(system_dbusd_t, dbusd_etc_t, dbusd_etc_t)
@@ -81,7 +82,7 @@ read_files_pattern(system_dbusd_t, syste
 manage_dirs_pattern(system_dbusd_t, system_dbusd_var_run_t, system_dbusd_var_run_t)
 manage_files_pattern(system_dbusd_t, system_dbusd_var_run_t, system_dbusd_var_run_t)
 manage_sock_files_pattern(system_dbusd_t, system_dbusd_var_run_t, system_dbusd_var_run_t)
-files_pid_filetrans(system_dbusd_t, system_dbusd_var_run_t, { dir file })
+files_pid_filetrans(system_dbusd_t, system_dbusd_var_run_t, { file dir sock_file })
 
 can_exec(system_dbusd_t, dbusd_exec_t)
 
@@ -92,6 +93,7 @@ corecmd_list_bin(system_dbusd_t)
 corecmd_read_bin_pipes(system_dbusd_t)
 corecmd_read_bin_sockets(system_dbusd_t)
 corecmd_exec_shell(system_dbusd_t)
+corecmd_exec_bin(system_dbusd_t)
 
 dev_read_urand(system_dbusd_t)
 dev_read_sysfs(system_dbusd_t)
@@ -107,6 +109,9 @@ fs_list_inotifyfs(system_dbusd_t)
 fs_search_auto_mountpoints(system_dbusd_t)
 fs_search_cgroup_dirs(system_dbusd_t)
 fs_dontaudit_list_nfs(system_dbusd_t)
+# root fs is labeled tmpfs_t for some reason
+#   not sure dbus needs to search through it
+fs_search_tmpfs(system_dbusd_t)
 
 mls_fd_use_all_levels(system_dbusd_t)
 mls_rangetrans_target(system_dbusd_t)
@@ -130,6 +135,10 @@ auth_read_pam_console_data(system_dbusd_
 init_use_fds(system_dbusd_t)
 init_use_script_ptys(system_dbusd_t)
 init_all_labeled_script_domtrans(system_dbusd_t)
+init_domtrans_script(system_dbusd_t)
+
+files_mountpoint(system_dbusd_var_lib_t)
+networkmanager_initrc_domtrans(system_dbusd_t)
 
 logging_send_audit_msgs(system_dbusd_t)
 logging_send_syslog_msg(system_dbusd_t)
