Index: refpolicy/policy/modules/roles/sysadm.te
===================================================================
--- refpolicy.orig/policy/modules/roles/sysadm.te
+++ refpolicy/policy/modules/roles/sysadm.te
@@ -48,6 +48,8 @@ init_reload_generic_units(sysadm_t)
 userdom_manage_user_home_dirs(sysadm_t)
 userdom_home_filetrans_user_home_dir(sysadm_t)
 
+storage_rw_blktap(sysadm_t)
+
 ifdef(`direct_sysadm_daemon',`
 	optional_policy(`
 		init_run_daemon(sysadm_t, sysadm_r)
@@ -1029,6 +1031,7 @@ optional_policy(`
 
 optional_policy(`
 	ssh_role_template(sysadm, sysadm_r, sysadm_t)
+	ssh_inherit_env(sysadm_t)
 ')
 
 optional_policy(`
