Index: refpolicy/policy/modules/contrib/virt.fc
===================================================================
--- refpolicy.orig/policy/modules/contrib/virt.fc
+++ refpolicy/policy/modules/contrib/virt.fc
@@ -32,6 +32,7 @@ HOME_DIR/VirtualMachines/isos(/.*)?	gen_
 /usr/sbin/libvirt-qmf	--	gen_context(system_u:object_r:virt_qmf_exec_t,s0)
 /usr/sbin/libvirtd	--	gen_context(system_u:object_r:virtd_exec_t,s0)
 /usr/sbin/virtlockd	--	gen_context(system_u:object_r:virtlockd_exec_t,s0)
+/usr/sbin/svirt-interpose	--	gen_context(system_u:object_r:virtd_exec_t,s0)
 
 /var/cache/libvirt(/.*)?	gen_context(system_u:object_r:virt_cache_t,s0-mls_systemhigh)
 
Index: refpolicy/policy/modules/contrib/virt.if
===================================================================
--- refpolicy.orig/policy/modules/contrib/virt.if
+++ refpolicy/policy/modules/contrib/virt.if
@@ -74,6 +74,7 @@ template(`virt_domain_template',`
 	manage_files_pattern($1_t, $1_tmpfs_t, $1_tmpfs_t)
 	manage_lnk_files_pattern($1_t, $1_tmpfs_t, $1_tmpfs_t)
 	fs_tmpfs_filetrans($1_t, $1_tmpfs_t, { dir file lnk_file })
+	fs_tmpfs_rangetrans($1_t, { dir file lnk_file }, s0)
 
 	optional_policy(`
 		pulseaudio_run($1_t, virt_domain_roles)
@@ -207,6 +208,25 @@ interface(`virt_domtrans_leaseshelper',`
 ')
 
 ########################################
+## <summary>
+##	search virt config dirs.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`virt_search_dirs',`
+	gen_require(`
+		type virt_etc_t;
+	')
+
+	files_search_etc($1)
+	allow $1 virt_etc_t:dir search;
+')
+
+########################################
 ## <summary>
 ##	Execute bridgehelper in the bridgehelper
 ##	domain, and allow the specified role
