Index: refpolicy/policy/modules/system/authlogin.fc
===================================================================
--- refpolicy.orig/policy/modules/system/authlogin.fc
+++ refpolicy/policy/modules/system/authlogin.fc
@@ -3,6 +3,12 @@
 /bin/login\.shadow	--	gen_context(system_u:object_r:login_exec_t,s0)
 /bin/login\.tinylogin	--	gen_context(system_u:object_r:login_exec_t,s0)
 
+/config/etc/\.pwd.lock	--	gen_context(system_u:object_r:shadow_t,s0)
+/config/etc/group\.lock	--	gen_context(system_u:object_r:shadow_t,s0)
+/config/etc/gshadow.*	--	gen_context(system_u:object_r:shadow_t,s0)
+/config/etc/passwd\.lock	--	gen_context(system_u:object_r:shadow_t,s0)
+/config/etc/shadow.*	--	gen_context(system_u:object_r:shadow_t,s0)
+
 /etc/\.pwd\.lock	--	gen_context(system_u:object_r:shadow_t,s0)
 /etc/group\.lock	--	gen_context(system_u:object_r:shadow_t,s0)
 /etc/gshadow.*		--	gen_context(system_u:object_r:shadow_t,s0)
@@ -46,6 +52,7 @@ ifdef(`distro_suse', `
 /var/log/syslog		--	gen_context(system_u:object_r:var_log_t,s0)
 /var/log/tallylog	--	gen_context(system_u:object_r:faillog_t,s0)
 /var/log/wtmp.*		--	gen_context(system_u:object_r:wtmp_t,s0)
+/var/volatile/log/wtmp.*	--	gen_context(system_u:object_r:wtmp_t,s0)
 
 /run/console(/.*)?	 	gen_context(system_u:object_r:pam_var_console_t,s0)
 /run/faillock(/.*)?		gen_context(system_u:object_r:faillog_t,s0)
Index: refpolicy/policy/modules/system/authlogin.if
===================================================================
--- refpolicy.orig/policy/modules/system/authlogin.if
+++ refpolicy/policy/modules/system/authlogin.if
@@ -1021,6 +1021,24 @@ interface(`auth_dontaudit_read_pam_pid',
 	dontaudit $1 pam_var_run_t:file { getattr read };
 ')
 
+#######################################
+## <summary>
+##	Set attributes on PAM PID directories.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`auth_setattr_pam_pid_dirs',`
+	gen_require(`
+		type pam_var_run_t;
+	')
+
+	allow $1 pam_var_run_t:dir setattr;
+')
+
 ########################################
 ## <summary>
 ##	Create specified objects in
@@ -1620,6 +1638,7 @@ interface(`auth_setattr_login_records',`
 		type wtmp_t;
 	')
 
+	allow $1 wtmp_t:lnk_file read_lnk_file_perms;
 	allow $1 wtmp_t:file setattr;
 	logging_search_logs($1)
 ')
@@ -1642,6 +1661,7 @@ interface(`auth_read_login_records',`
 
 	logging_search_logs($1)
 	allow $1 wtmp_t:file read_file_perms;
+	allow $1 wtmp_t:lnk_file read_lnk_file_perms;
 ')
 
 ########################################
@@ -1660,7 +1680,7 @@ interface(`auth_dontaudit_read_login_rec
 	gen_require(`
 		type wtmp_t;
 	')
-
+	dontaudit $1 wtmp_t:lnk_file read_lnk_file_perms;
 	dontaudit $1 wtmp_t:file read_file_perms;
 ')
 
@@ -1680,6 +1700,7 @@ interface(`auth_dontaudit_write_login_re
 		type wtmp_t;
 	')
 
+	dontaudit $1 wtmp_t:lnk_file read_lnk_file_perms;
 	dontaudit $1 wtmp_t:file write;
 ')
 
@@ -1698,6 +1719,7 @@ interface(`auth_append_login_records',`
 		type wtmp_t;
 	')
 
+	allow $1 wtmp_t:lnk_file read_lnk_file_perms;
 	allow $1 wtmp_t:file append_file_perms;
 	logging_search_logs($1)
 ')
@@ -1717,6 +1739,7 @@ interface(`auth_write_login_records',`
 		type wtmp_t;
 	')
 
+	allow $1 wtmp_t:lnk_file read_lnk_file_perms;
 	allow $1 wtmp_t:file { write_file_perms lock };
 ')
 
@@ -1735,6 +1758,7 @@ interface(`auth_rw_login_records',`
 		type wtmp_t;
 	')
 
+	allow $1 wtmp_t:lnk_file read_lnk_file_perms;
 	allow $1 wtmp_t:file rw_file_perms;
 	logging_search_logs($1)
 ')
@@ -1755,6 +1779,7 @@ interface(`auth_log_filetrans_login_reco
 		type wtmp_t;
 	')
 
+	allow $1 wtmp_t:lnk_file read_lnk_file_perms;
 	logging_log_filetrans($1, wtmp_t, file)
 ')
 
@@ -1774,6 +1799,7 @@ interface(`auth_manage_login_records',`
 		type wtmp_t;
 	')
 
+	allow $1 wtmp_t:lnk_file read_lnk_file_perms;
 	logging_rw_generic_log_dirs($1)
 	allow $1 wtmp_t:file manage_file_perms;
 ')
