Index: refpolicy/policy/modules/contrib/dmidecode.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/dmidecode.te
+++ refpolicy/policy/modules/contrib/dmidecode.te
@@ -31,4 +31,13 @@ mls_file_read_all_levels(dmidecode_t)
 
 locallogin_use_fds(dmidecode_t)
 
+# dmidecode called from init script
+init_use_fds(dmidecode_t)
+term_read_console(dmidecode_t)
+term_write_console(dmidecode_t)
+
+seutil_use_newrole_fds(dmidecode_t)
+
 userdom_use_user_terminals(dmidecode_t)
+
+xen_dontaudit_rw_unix_stream_sockets(dmidecode_t)
