Index: refpolicy/policy/modules/contrib/firstboot.fc
===================================================================
--- refpolicy.orig/policy/modules/contrib/firstboot.fc
+++ refpolicy/policy/modules/contrib/firstboot.fc
@@ -3,3 +3,4 @@
 /usr/sbin/firstboot	--	gen_context(system_u:object_r:firstboot_exec_t,s0)
 
 /usr/share/firstboot/firstboot\.py	--	gen_context(system_u:object_r:firstboot_exec_t,s0)
+/usr/share/xenclient/firstboot\.sh	--	gen_context(system_u:object_r:firstboot_exec_t,s0)
Index: refpolicy/policy/modules/contrib/firstboot.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/firstboot.te
+++ refpolicy/policy/modules/contrib/firstboot.te
@@ -124,3 +124,7 @@ optional_policy(`
       xserver_unconfined(firstboot_t)
       xserver_stream_connect(firstboot_t)
 ')
+
+optional_policy(`
+      tapctl_domtrans(firstboot_t)
+')
