--- a/policy/modules/kernel/corenetwork.if.in
+++ b/policy/modules/kernel/corenetwork.if.in
@@ -3554,3 +3554,22 @@ interface(`corenet_unconfined',`
 
 	typeattribute $1 corenet_unconfined_type;
 ')
+
+########################################
+## <summary>
+##      Unconfined read access to network objects.
+## </summary>
+## <param name="domain">
+##      <summary>
+##      The domain allowed access.
+##      </summary>
+## </param>
+#
+interface(`corenet_unconfined_readonly',`
+        gen_require(`
+                attribute corenet_unconfined_readonly_type;
+        ')
+
+        typeattribute $1 corenet_unconfined_readonly_type;
+')
+
--- a/policy/modules/kernel/corenetwork.te.in
+++ b/policy/modules/kernel/corenetwork.te.in
@@ -22,6 +22,7 @@ attribute ibendport_type;
 attribute unreserved_port_type;
 
 attribute corenet_unconfined_type;
+attribute corenet_unconfined_readonly_type;
 
 type ppp_device_t;
 dev_node(ppp_device_t)
@@ -370,3 +371,10 @@ corenet_ib_access_all_pkeys(corenet_unco
 corenet_ib_manage_subnet_all_endports(corenet_unconfined_type)
 corenet_ib_access_unlabeled_pkeys(corenet_unconfined_type)
 corenet_ib_manage_subnet_unlabeled_endports(corenet_unconfined_type)
+
+########################################
+#
+# Unconfined readonly access to this module
+#
+allow corenet_unconfined_readonly_type port_type:tcp_socket { send_msg recv_msg };
+allow corenet_unconfined_readonly_type port_type:udp_socket { send_msg recv_msg };
