Index: refpolicy/policy/modules/system/init.if
===================================================================
--- refpolicy.orig/policy/modules/system/init.if
+++ refpolicy/policy/modules/system/init.if
@@ -2540,3 +2540,22 @@ interface(`initrc_rw_pid_chr_files',`
 
 	allow $1 initrc_var_run_t:chr_file rw_chr_file_perms;
 ')
+
+#####################################
+## <summary>
+##     Get attributes for and unlink XC init flags
+## </summary>
+## <param name="domain">
+##     <summary>
+##     Domain allowed access.
+##     </summary>
+## </param>
+#
+interface(`init_delete_flag_files',`
+	gen_require(`
+		type init_flag_t;
+	')
+
+	xc_unlink_config_objs($1)
+	allow $1 init_flag_t:file delete_file_perms;
+')
