ndex: refpolicy/policy/modules/system/udev.fc
===================================================================
Index: refpolicy/policy/modules/system/udev.fc
===================================================================
--- refpolicy.orig/policy/modules/system/udev.fc
+++ refpolicy/policy/modules/system/udev.fc
@@ -2,6 +2,7 @@
 /dev/\.udevdb	--	gen_context(system_u:object_r:udev_tbl_t,s0)
 /dev/udev\.tbl	--	gen_context(system_u:object_r:udev_tbl_t,s0)
 
+/etc/init\.d/udev	--	gen_context(system_u:object_r:udevrc_exec_t,s0)
 /etc/dev\.d/.+	--	gen_context(system_u:object_r:udev_helper_exec_t,s0)
 
 /etc/hotplug\.d/default/udev.* -- gen_context(system_u:object_r:udev_helper_exec_t,s0)
Index: refpolicy/policy/modules/system/udev.te
===================================================================
--- refpolicy.orig/policy/modules/system/udev.te
+++ refpolicy/policy/modules/system/udev.te
@@ -15,6 +15,12 @@ domain_interactive_fd(udev_t)
 init_daemon_domain(udev_t, udev_exec_t)
 init_named_socket_activation(udev_t, udev_var_run_t)
 
+type udevrc_t;
+type udevrc_exec_t;
+init_script_domain(udevrc_t, udevrc_exec_t)
+udev_domtrans(udevrc_t)
+role system_r types udevrc_t;
+
 type udev_etc_t alias etc_udev_t;
 files_config_file(udev_etc_t)
 
@@ -57,6 +63,11 @@ allow udev_t self:netlink_kobject_uevent
 allow udev_t self:netlink_generic_socket create_socket_perms;
 allow udev_t self:rawip_socket create_socket_perms;
 
+# Ignore CAP_IPC_LOCK denial triggered by mmap(MAP_LOCKED);
+# the operation will still succeed.  See the following kernel commit:
+# http://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=a5a6579db33af91f4f5134e14be758dc71c1b694
+dontaudit udev_t self:capability ipc_lock;
+
 allow udev_t udev_exec_t:file write;
 can_exec(udev_t, udev_exec_t)
 
@@ -93,6 +104,7 @@ kernel_rw_unix_dgram_sockets(udev_t)
 kernel_dgram_send(udev_t)
 kernel_signal(udev_t)
 kernel_search_debugfs(udev_t)
+kernel_load_module(udev_t)
 
 #https://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=235182
 kernel_rw_net_sysctls(udev_t)
@@ -102,6 +114,8 @@ kernel_dontaudit_search_unlabeled(udev_t
 
 corecmd_exec_all_executables(udev_t)
 
+dev_manage_sysfs_dirs(udev_t)
+dev_create_sysfs_nodes(udev_t)
 dev_rw_sysfs(udev_t)
 dev_manage_all_dev_nodes(udev_t)
 dev_rw_generic_files(udev_t)
@@ -122,10 +136,12 @@ files_read_etc_files(udev_t)
 files_exec_etc_files(udev_t)
 files_getattr_generic_locks(udev_t)
 files_search_mnt(udev_t)
+files_read_kernel_modules(udev_t)
 
 fs_getattr_all_fs(udev_t)
 fs_list_inotifyfs(udev_t)
 fs_rw_anon_inodefs_files(udev_t)
+fs_search_tracefs(udev_t)
 
 mcs_ptrace_all(udev_t)
 
@@ -178,6 +194,8 @@ sysnet_etc_filetrans_config(udev_t)
 
 userdom_dontaudit_search_user_home_content(udev_t)
 
+iptables_domtrans(udev_t)
+
 ifdef(`distro_debian',`
 	files_pid_filetrans(udev_t, udev_var_run_t, dir, "xen-hotplug")
 
@@ -335,8 +353,40 @@ optional_policy(`
 	kernel_read_xen_state(udev_t)
 	xen_manage_log(udev_t)
 	xen_read_image_files(udev_t)
+	dev_manage_xen(udev_t)
+	xen_stream_connect_xenstore(udev_t)
+	fs_rw_xenfs_files(udev_t)
 ')
 
 optional_policy(`
 	xserver_read_xdm_pid(udev_t)
 ')
+
+fs_manage_tmpfs_chr_files(udev_t)
+fs_manage_tmpfs_files(udev_t)
+fs_manage_tmpfs_symlinks(udev_t)
+virt_search_dirs(udev_t)
+
+vusbd_read_etc_files(udev_t)
+
+########################################
+#
+# Local policy for udevrc_t
+#
+
+mount_domtrans(udevrc_t)
+
+corecmd_exec_bin(udevrc_t)
+corecmd_exec_shell(udevrc_t)
+dev_search_sysfs(udevrc_t)
+files_read_etc_files(udevrc_t)
+initrc_rw_pid_chr_files(udevrc_t)
+kernel_read_system_state(udevrc_t)
+kernel_rw_hotplug_sysctls(udevrc_t)
+init_dontaudit_use_fds(udevrc_t)
+term_use_console(udevrc_t)
+dev_dontaudit_rw_generic_chr_files(udevrc_t)
+fs_search_tmpfs(udevrc_t)
+seutil_domtrans_setfiles(udevrc_t)
+
+allow udevrc_t self:fifo_file manage_fifo_file_perms;
