Index: refpolicy/policy/modules/kernel/corenetwork.if.in
===================================================================
--- refpolicy.orig/policy/modules/kernel/corenetwork.if.in
+++ refpolicy/policy/modules/kernel/corenetwork.if.in
@@ -3134,3 +3134,22 @@ interface(`corenet_unconfined',`
 
 	typeattribute $1 corenet_unconfined_type;
 ')
+
+########################################
+## <summary>
+##      Unconfined read access to network objects.
+## </summary>
+## <param name="domain">
+##      <summary>
+##      The domain allowed access.
+##      </summary>
+## </param>
+#
+interface(`corenet_unconfined_readonly',`
+        gen_require(`
+                attribute corenet_unconfined_readonly_type;
+        ')
+
+        typeattribute $1 corenet_unconfined_readonly_type;
+')
+
Index: refpolicy/policy/modules/kernel/corenetwork.te.in
===================================================================
--- refpolicy.orig/policy/modules/kernel/corenetwork.te.in
+++ refpolicy/policy/modules/kernel/corenetwork.te.in
@@ -20,6 +20,7 @@ attribute server_packet_type;
 attribute unreserved_port_type;
 
 attribute corenet_unconfined_type;
+attribute corenet_unconfined_readonly_type;
 
 type ppp_device_t;
 dev_node(ppp_device_t)
@@ -357,3 +358,10 @@ allow corenet_unconfined_type port_type:
 # Bind to any network address.
 allow corenet_unconfined_type port_type:{ tcp_socket udp_socket rawip_socket } name_bind;
 allow corenet_unconfined_type node_type:{ tcp_socket udp_socket rawip_socket } node_bind;
+
+########################################
+#
+# Unconfined readonly access to this module
+#
+allow corenet_unconfined_readonly_type port_type:tcp_socket { send_msg recv_msg };
+allow corenet_unconfined_readonly_type port_type:udp_socket { send_msg recv_msg };
