--- a/policy/modules/system/iptables.if
+++ b/policy/modules/system/iptables.if
@@ -165,6 +165,24 @@ interface(`iptables_manage_config',`
 	manage_files_pattern($1, iptables_conf_t, iptables_conf_t)
 ')
 
+###################################
+## <summary>
+##     dontaudit reading iptables_var_run_t
+## </summary>
+## <param name="domain">
+##     <summary>
+##     Domain to not audit.
+##     </summary>
+## </param>
+#
+interface(`iptables_dontaudit_read_pids',`
+       gen_require(`
+               type iptables_var_run_t;
+       ')
+
+       dontaudit $1 iptables_var_run_t:file read;
+')
+
 ########################################
 ## <summary>
 ##	All of the rules required to
