Index: refpolicy/policy/support/misc_patterns.spt
===================================================================
--- refpolicy.orig/policy/support/misc_patterns.spt
+++ refpolicy/policy/support/misc_patterns.spt
@@ -8,6 +8,7 @@
 #
 define(`domain_transition_pattern',`
 	allow $1 $2:file { getattr open read execute };
+	allow $3 $2:lnk_file read_lnk_file_perms;
 	allow $1 $3:process transition;
 	dontaudit $1 $3:process { noatsecure siginh rlimitinh };
 ')
