--- a/policy/modules/system/authlogin.fc
+++ b/policy/modules/system/authlogin.fc
@@ -41,6 +41,7 @@ ifdef(`distro_suse', `
 /var/log/lastlog	--	gen_context(system_u:object_r:lastlog_t,s0)
 /var/log/tallylog	--	gen_context(system_u:object_r:faillog_t,s0)
 /var/log/wtmp.*		--	gen_context(system_u:object_r:wtmp_t,s0)
+/var/volatile/log/wtmp.*	--	gen_context(system_u:object_r:wtmp_t,s0)
 
 /run/console(/.*)?	 	gen_context(system_u:object_r:pam_var_console_t,s0)
 /run/faillock(/.*)?		gen_context(system_u:object_r:faillog_t,s0)
@@ -54,3 +55,9 @@ ifdef(`distro_suse', `
 /run/sudo/ts/%{USERNAME}	gen_context(system_u:object_r:pam_runtime_t,s0)
 /var/(db|adm)/sudo(/.*)?	gen_context(system_u:object_r:pam_runtime_t,s0)
 /var/lib/sudo(/.*)?	gen_context(system_u:object_r:pam_runtime_t,s0)
+
+/config/etc/\.pwd\.lock	--	gen_context(system_u:object_r:shadow_t,s0)
+/config/etc/group\.lock	--	gen_context(system_u:object_r:shadow_t,s0)
+/config/etc/gshadow.*	--	gen_context(system_u:object_r:shadow_t,s0)
+/config/etc/passwd\.lock	--	gen_context(system_u:object_r:shadow_t,s0)
+/config/etc/shadow.*	--	gen_context(system_u:object_r:shadow_t,s0)
--- a/policy/modules/system/authlogin.if
+++ b/policy/modules/system/authlogin.if
@@ -1081,6 +1081,24 @@ interface(`auth_manage_var_auth',`
 	allow $1 var_auth_t:lnk_file rw_lnk_file_perms;
 ')
 
+#######################################
+## <summary>
+##	Set attributes on PAM PID directories.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`auth_setattr_pam_pid_dirs',`
+	gen_require(`
+		type pam_var_run_t;
+	')
+
+	allow $1 pam_var_run_t:dir setattr;
+')
+
 ########################################
 ## <summary>
 ##	Read PAM PID files.  (Deprecated)
@@ -1585,6 +1603,7 @@ interface(`auth_setattr_login_records',`
 		type wtmp_t;
 	')
 
+	allow $1 wtmp_t:lnk_file read_lnk_file_perms;
 	allow $1 wtmp_t:file setattr;
 	logging_search_logs($1)
 ')
@@ -1607,6 +1626,7 @@ interface(`auth_read_login_records',`
 
 	logging_search_logs($1)
 	allow $1 wtmp_t:file read_file_perms;
+	allow $1 wtmp_t:lnk_file read_lnk_file_perms;
 ')
 
 ########################################
@@ -1625,7 +1645,7 @@ interface(`auth_dontaudit_read_login_rec
 	gen_require(`
 		type wtmp_t;
 	')
-
+	dontaudit $1 wtmp_t:lnk_file read_lnk_file_perms;
 	dontaudit $1 wtmp_t:file read_file_perms;
 ')
 
@@ -1645,6 +1665,7 @@ interface(`auth_dontaudit_write_login_re
 		type wtmp_t;
 	')
 
+	dontaudit $1 wtmp_t:lnk_file read_lnk_file_perms;
 	dontaudit $1 wtmp_t:file write;
 ')
 
@@ -1663,6 +1684,7 @@ interface(`auth_append_login_records',`
 		type wtmp_t;
 	')
 
+	allow $1 wtmp_t:lnk_file read_lnk_file_perms;
 	allow $1 wtmp_t:file append_file_perms;
 	logging_search_logs($1)
 ')
@@ -1682,6 +1704,7 @@ interface(`auth_write_login_records',`
 		type wtmp_t;
 	')
 
+	allow $1 wtmp_t:lnk_file read_lnk_file_perms;
 	allow $1 wtmp_t:file { write_file_perms lock };
 ')
 
@@ -1700,6 +1723,7 @@ interface(`auth_rw_login_records',`
 		type wtmp_t;
 	')
 
+	allow $1 wtmp_t:lnk_file read_lnk_file_perms;
 	allow $1 wtmp_t:file rw_file_perms;
 	logging_search_logs($1)
 ')
@@ -1720,6 +1744,7 @@ interface(`auth_log_filetrans_login_reco
 		type wtmp_t;
 	')
 
+	allow $1 wtmp_t:lnk_file read_lnk_file_perms;
 	logging_log_filetrans($1, wtmp_t, file)
 ')
 
@@ -1739,6 +1764,7 @@ interface(`auth_manage_login_records',`
 		type wtmp_t;
 	')
 
+	allow $1 wtmp_t:lnk_file read_lnk_file_perms;
 	logging_rw_generic_log_dirs($1)
 	allow $1 wtmp_t:file manage_file_perms;
 ')
