Index: refpolicy/policy/modules/contrib/logrotate.fc
===================================================================
--- refpolicy.orig/policy/modules/contrib/logrotate.fc
+++ refpolicy/policy/modules/contrib/logrotate.fc
@@ -4,6 +4,7 @@
 # Systemd unit file
 /usr/lib/systemd/system/[^/]*logrotate.*	--	gen_context(system_u:object_r:logrotate_unit_t,s0)
 
+/sbin/logrotate        --      gen_context(system_u:object_r:logrotate_exec_t,s0)
 /usr/sbin/logrotate	--	gen_context(system_u:object_r:logrotate_exec_t,s0)
 
 /var/lib/logrotate(/.*)?	gen_context(system_u:object_r:logrotate_var_lib_t,s0)
Index: refpolicy/policy/modules/contrib/logrotate.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/logrotate.te
+++ refpolicy/policy/modules/contrib/logrotate.te
@@ -117,6 +117,7 @@ userdom_list_user_home_dirs(logrotate_t)
 userdom_use_unpriv_users_fds(logrotate_t)
 
 mta_sendmail_domtrans(logrotate_t, logrotate_mail_t)
+syslogd_initrc_domtrans(logrotate_t)
 
 ifdef(`distro_debian',`
 	allow logrotate_t logrotate_tmp_t:file relabel_file_perms;
