--- a/policy/modules/system/fstools.fc
+++ b/policy/modules/system/fstools.fc
@@ -86,6 +86,7 @@
 /usr/sbin/make_reiser4		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /usr/sbin/mkdosfs		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /usr/sbin/mke2fs		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
+/usr/sbin/mke2fs\.e2fsprogs	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /usr/sbin/mke4fs		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /usr/sbin/mkfs.*		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /usr/sbin/mkraid		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
@@ -107,6 +108,7 @@
 /usr/sbin/swapoff\.util-linux	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /usr/sbin/swapon.*		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /usr/sbin/tune2fs		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
+/usr/sbin/tune2fs\.e2fsprogs	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /usr/sbin/zdb			--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /usr/sbin/zhack			--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /usr/sbin/zinject		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
--- a/policy/modules/system/fstools.te
+++ b/policy/modules/system/fstools.te
@@ -142,6 +142,7 @@ storage_raw_read_removable_device(fsadm_
 storage_raw_write_removable_device(fsadm_t)
 storage_read_scsi_generic(fsadm_t)
 storage_swapon_fixed_disk(fsadm_t)
+storage_rw_blktap(fsadm_t)
 
 term_use_console(fsadm_t)
 
