--- a/policy/modules/admin/firstboot.fc
+++ b/policy/modules/admin/firstboot.fc
@@ -5,3 +5,4 @@
 /usr/sbin/firstboot	--	gen_context(system_u:object_r:firstboot_exec_t,s0)
 
 /usr/share/firstboot/firstboot\.py	--	gen_context(system_u:object_r:firstboot_exec_t,s0)
+/usr/share/xenclient/firstboot\.sh	--	gen_context(system_u:object_r:firstboot_exec_t,s0)
--- a/policy/modules/admin/firstboot.te
+++ b/policy/modules/admin/firstboot.te
@@ -125,3 +125,8 @@ optional_policy(`
 	xserver_unconfined(firstboot_t)
 	xserver_stream_connect(firstboot_t)
 ')
+
+optional_policy(`
+	tapctl_domtrans(firstboot_t)
+	dev_search_xen(firstboot_t)
+')
