Index: refpolicy/policy/modules/system/userdomain.fc
===================================================================
--- refpolicy.orig/policy/modules/system/userdomain.fc
+++ refpolicy/policy/modules/system/userdomain.fc
@@ -2,6 +2,12 @@ HOME_DIR	-d	gen_context(system_u:object_
 HOME_DIR/.+		gen_context(system_u:object_r:user_home_t,s0)
 HOME_DIR/\.pki(/.*)?	gen_context(system_u:object_r:user_cert_t,s0)
 
+#
+# /root
+#
+# Temporary fix to get the home directory for root user labeled right
+/root	-d	 gen_context(system_u:object_r:user_home_dir_t,s0)
+
 /tmp/gconfd-%{USERNAME} -d	gen_context(system_u:object_r:user_tmp_t,s0)
 
 /run/user		-d	gen_context(system_u:object_r:user_runtime_root_t,s0)
Index: refpolicy/policy/modules/system/userdomain.if
===================================================================
--- refpolicy.orig/policy/modules/system/userdomain.if
+++ refpolicy/policy/modules/system/userdomain.if
@@ -1531,7 +1531,6 @@ interface(`userdom_getattr_user_home_dir
 	')
 
 	allow $1 user_home_dir_t:dir getattr_dir_perms;
-	files_search_home($1)
 ')
 
 ########################################
@@ -3782,6 +3781,24 @@ interface(`userdom_signal_all_users',`
 ')
 
 ########################################
+## <summary>
+##	Do not audit signals to admin user domains.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`userdom_dontaudit_signal_admin_users',`
+	gen_require(`
+		attribute admindomain;
+	')
+
+	allow $1 admindomain:process signal;
+')
+
+########################################
 ## <summary>
 ##	Send a SIGCHLD signal to all user domains.
 ## </summary>
