Index: refpolicy/policy/modules/contrib/networkmanager.if
===================================================================
--- refpolicy.orig/policy/modules/contrib/networkmanager.if
+++ refpolicy/policy/modules/contrib/networkmanager.if
@@ -512,3 +512,32 @@ interface(`networkmanager_read_etc_rw',`
         allow $1 NetworkManager_etc_rw_t:dir list_dir_perms;
         allow $1 NetworkManager_etc_rw_t:file read_file_perms;
 ')
+
+########################################
+## <summary>
+##      Create an object in xc config  with a network manager private type.
+## </summary>
+## <param name="domain">
+##      <summary>
+##      Domain allowed access.
+##      </summary>
+## </param>
+## <param name="private type">
+##      <summary>
+##      The type of the object to be created.
+##      </summary>
+## </param>
+## <param name="object">
+##      <summary>
+##      The object class of the object being created.
+##      </summary>
+## </param>
+#
+interface(`networkmanager_xc_config_filetrans',`
+	gen_require(`
+		type NetworkManager_cert_t;
+	')
+
+	create_dirs_pattern($1, NetworkManager_cert_t, NetworkManager_cert_t)
+	xc_config_filetrans($1, NetworkManager_cert_t, { dir file })
+')
