Index: refpolicy/policy/modules/kernel/terminal.te
===================================================================
--- refpolicy.orig/policy/modules/kernel/terminal.te
+++ refpolicy/policy/modules/kernel/terminal.te
@@ -29,7 +29,7 @@ files_mountpoint(devpts_t)
 fs_associate_tmpfs(devpts_t)
 fs_xattr_type(devpts_t)
 fs_use_trans devpts gen_context(system_u:object_r:devpts_t,s0);
-
+dev_associate(devpts_t)
 #
 # devtty_t is the type of /dev/tty.
 #
