Index: refpolicy/policy/modules/kernel/storage.fc
===================================================================
--- refpolicy.orig/policy/modules/kernel/storage.fc
+++ refpolicy/policy/modules/kernel/storage.fc
@@ -83,6 +83,8 @@ ifdef(`distro_redhat', `
 
 /dev/usb/rio500		-c	gen_context(system_u:object_r:removable_device_t,s0)
 
+/dev/xen/blktap-[0-9]/.*	gen_context(system_u:object_r:blktap_device_t,s0)
+
 /lib/udev/devices/loop.* -b	gen_context(system_u:object_r:fixed_disk_device_t,mls_systemhigh)
 /lib/udev/devices/fuse	-c	gen_context(system_u:object_r:fuse_device_t,s0)
 
Index: refpolicy/policy/modules/kernel/storage.if
===================================================================
--- refpolicy.orig/policy/modules/kernel/storage.if
+++ refpolicy/policy/modules/kernel/storage.if
@@ -813,3 +813,103 @@ interface(`storage_unconfined',`
 
 	typeattribute $1 storage_unconfined_type;
 ')
+########################################
+## <summary>
+##	Get attributes of blktap devices.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`storage_getattr_blktap',`
+	gen_require(`
+		type blktap_device_t;
+	')
+
+	getattr_blk_files_pattern($1, blktap_device_t, blktap_device_t)
+	getattr_chr_files_pattern($1, blktap_device_t, blktap_device_t)
+')
+########################################
+## <summary>
+##	Read and write blktap devices.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`storage_rw_blktap',`
+	gen_require(`
+		type blktap_device_t;
+	')
+
+	rw_chr_files_pattern($1, blktap_device_t, blktap_device_t)
+	rw_blk_files_pattern($1, blktap_device_t, blktap_device_t)
+')
+########################################
+## <summary>
+##	Create, read, write, and delete blktap devices.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`storage_manage_blktap',`
+	gen_require(`
+		type blktap_device_t;
+	')
+
+	add_entry_dirs_pattern($1, blktap_device_t, blktap_device_t)
+	manage_lnk_files_pattern($1, blktap_device_t, blktap_device_t)
+	manage_chr_files_pattern($1, blktap_device_t, blktap_device_t)
+	manage_blk_files_pattern($1, blktap_device_t, blktap_device_t)
+')
+########################################
+## <summary>
+##	Automatic type transition to the type
+##	for blktap blk devices when created in /dev.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+## <param name="range">
+##	<summary>
+##	</summary>
+## </param>
+#
+interface(`storage_filetrans_blktap',`
+	gen_require(`
+		type blktap_device_t;
+	')
+
+	dev_filetrans($1, blktap_device_t, { blk_file chr_file })
+')
+########################################
+## <summary>
+##	Automatic type transition to the type
+##	for blktap blk devices when created in /dev.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+## <param name="range">
+##	<summary>
+##	</summary>
+## </param>
+#
+interface(`storage_ranged_filetrans_blktap',`
+	gen_require(`
+		type blktap_device_t;
+	')
+
+	dev_ranged_filetrans($1, blktap_device_t, { blk_file chr_file }, $2)
+')
Index: refpolicy/policy/modules/kernel/storage.te
===================================================================
--- refpolicy.orig/policy/modules/kernel/storage.te
+++ refpolicy/policy/modules/kernel/storage.te
@@ -50,6 +50,15 @@ dev_node(removable_device_t)
 type tape_device_t;
 dev_node(tape_device_t)
 
+#
+# block-tap devices
+#
+type blktap_device_t;
+dev_node(blktap_device_t)
+optional_policy(`
+	virt_image(blktap_device_t)
+')
+
 ########################################
 #
 # Unconfined access to this module
