Index: refpolicy/policy/modules/admin/su.if
===================================================================
--- refpolicy.orig/policy/modules/admin/su.if
+++ refpolicy/policy/modules/admin/su.if
@@ -160,6 +160,7 @@ template(`su_role_template',`
 	allow $1_su_t self:fifo_file rw_fifo_file_perms;
 	allow $1_su_t self:netlink_audit_socket { nlmsg_relay create_netlink_socket_perms };
 	allow $1_su_t self:key { search write };
+	allow $1_su_t self:netlink_selinux_socket { bind create };
 
 	allow $1_su_t $3:key search;
 
