--- a/policy/modules/system/lvm.te
+++ b/policy/modules/system/lvm.te
@@ -209,6 +209,7 @@ manage_dirs_pattern(lvm_t, lvm_var_run_t
 manage_files_pattern(lvm_t, lvm_var_run_t, lvm_var_run_t)
 manage_sock_files_pattern(lvm_t, lvm_var_run_t, lvm_var_run_t)
 files_pid_filetrans(lvm_t, lvm_var_run_t, { file sock_file })
+files_pid_filetrans(lvm_t, lvm_var_run_t, dir, "cryptsetup")
 
 read_files_pattern(lvm_t, lvm_etc_t, lvm_etc_t)
 allow lvm_t lvm_etc_t:file map;
@@ -218,7 +219,8 @@ read_lnk_files_pattern(lvm_t, lvm_etc_t,
 allow lvm_t lvm_metadata_t:file map;
 # Write to /etc/lvm, /etc/lvmtab, /etc/lvmtab.d
 manage_files_pattern(lvm_t, lvm_metadata_t, lvm_metadata_t)
-filetrans_pattern(lvm_t, lvm_etc_t, lvm_metadata_t, file)
+create_dirs_pattern(lvm_t, lvm_metadata_t, lvm_metadata_t)
+filetrans_pattern(lvm_t, lvm_etc_t, lvm_metadata_t, { file dir })
 files_etc_filetrans(lvm_t, lvm_metadata_t, file)
 files_search_mnt(lvm_t)
 
@@ -382,4 +384,11 @@ optional_policy(`
 optional_policy(`
 	xen_append_log(lvm_t)
 	xen_dontaudit_rw_unix_stream_sockets(lvm_t)
+	xen_sem_use(lvm_t)
+')
+
+optional_policy(`
+	xc_read_config_key_files(lvm_t)
+	xc_read_vhd_key_files(lvm_t)
+	xc_search_vhd_key_dirs(lvm_t)
 ')
