Index: refpolicy/policy/modules/contrib/apm.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/apm.te
+++ refpolicy/policy/modules/contrib/apm.te
@@ -104,6 +104,8 @@ dev_rw_apm_bios(apmd_t)
 dev_rw_sysfs(apmd_t)
 dev_dontaudit_getattr_all_chr_files(apmd_t)
 dev_dontaudit_getattr_all_blk_files(apmd_t)
+# Early devtmpfs, before udev relabel
+dev_dontaudit_rw_generic_chr_files(apmd_t)
 
 files_exec_etc_files(apmd_t)
 files_read_etc_runtime_files(apmd_t)
