Index: refpolicy/policy/modules/system/fstools.fc
===================================================================
--- refpolicy.orig/policy/modules/system/fstools.fc
+++ refpolicy/policy/modules/system/fstools.fc
@@ -48,6 +48,7 @@
 /sbin/zpios		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/zstreamdump	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/ztest		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
+/sbin/mke2fs\.e2fsprogs	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 
 /usr/bin/partition_uuid	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /usr/bin/raw		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
Index: refpolicy/policy/modules/system/fstools.te
===================================================================
--- refpolicy.orig/policy/modules/system/fstools.te
+++ refpolicy/policy/modules/system/fstools.te
@@ -140,6 +140,7 @@ storage_raw_read_removable_device(fsadm_
 storage_raw_write_removable_device(fsadm_t)
 storage_read_scsi_generic(fsadm_t)
 storage_swapon_fixed_disk(fsadm_t)
+storage_rw_blktap(fsadm_t)
 
 term_use_console(fsadm_t)
 
@@ -213,3 +214,10 @@ optional_policy(`
 	xen_append_log(fsadm_t)
 	xen_rw_image_files(fsadm_t)
 ')
+
+# leaked FDs from input-server
+term_dontaudit_use_unallocated_ttys(fsadm_t)
+optional_policy(`
+	kernel_dontaudit_read_xen_state(fsadm_t)
+	kernel_dontaudit_write_xen_state(fsadm_t)
+')
