Index: refpolicy/policy/modules/system/libraries.te
===================================================================
--- refpolicy.orig/policy/modules/system/libraries.te
+++ refpolicy/policy/modules/system/libraries.te
@@ -89,6 +89,7 @@ files_delete_etc_files(ldconfig_t)
 
 init_use_script_ptys(ldconfig_t)
 init_read_script_tmp_files(ldconfig_t)
+init_dontaudit_use_fds(ldconfig_t)
 
 miscfiles_read_localization(ldconfig_t)
 
