From: Jed <lejosnej@ainfosec.com>
Date: Mon, 16 Feb 2015 14:22:42 -0500
Subject: [PATCH] SELinux: label qemu-dm-wrapper as qemu_exec_t

Index: refpolicy/policy/modules/contrib/qemu.fc
===================================================================
--- refpolicy.orig/policy/modules/contrib/qemu.fc
+++ refpolicy/policy/modules/contrib/qemu.fc
@@ -3,6 +3,7 @@
 /usr/bin/qemu-kvm	--	gen_context(system_u:object_r:qemu_exec_t,s0)
 /usr/bin/kvm		--	gen_context(system_u:object_r:qemu_exec_t,s0)
 /usr/lib/xen/bin/qemu-dm       --      gen_context(system_u:object_r:qemu_exec_t,s0)
+/usr/bin/qemu-dm-wrapper       --      gen_context(system_u:object_r:qemu_exec_t,s0)
 /usr/bin/qemu-dm-wrapper-old       --      gen_context(system_u:object_r:qemu_exec_t,s0)
 
 /usr/lib/systemd/system/[^/]*qemu-guest-agent.*	--	gen_context(system_u:object_r:qemu_unit_t,s0)
