Index: refpolicy/Makefile
===================================================================
--- refpolicy.orig/Makefile
+++ refpolicy/Makefile
@@ -416,29 +416,29 @@ $(tunxml): $(globaltun)
 $(boolxml): $(globalbool)
 	$(verbose) $(genxml) -w -b $< > $@
 
-$(polxml): $(layerxml) $(tunxml) $(boolxml)
-	@echo "Creating $(@F)"
-	@test -d $(dir $(polxml)) || mkdir -p $(dir $(polxml))
-	@test -d $(tmpdir) || mkdir -p $(tmpdir)
-	$(verbose) echo '<?xml version="1.0" encoding="ISO-8859-1" standalone="no"?>' > $@
-	$(verbose) echo '<!DOCTYPE policy SYSTEM "$(notdir $(xmldtd))">' >> $@
-	$(verbose) echo '<policy>' >> $@
-	$(verbose) for i in $(basename $(notdir $(layerxml))); do echo "<layer name=\"$$i\">" >> $@; cat $(tmpdir)/$$i.xml >> $@; echo "</layer>" >> $@; done
-	$(verbose) cat $(tunxml) $(boolxml) >> $@
-	$(verbose) echo '</policy>' >> $@
-	$(verbose) if test -x $(XMLLINT) && test -f $(xmldtd); then \
-		$(XMLLINT) --noout --path $(dir $(xmldtd)) --dtdvalid $(xmldtd) $@ ;\
-	fi
+#$(polxml): $(layerxml) $(tunxml) $(boolxml)
+#	@echo "Creating $(@F)"
+#	@test -d $(dir $(polxml)) || mkdir -p $(dir $(polxml))
+#	@test -d $(tmpdir) || mkdir -p $(tmpdir)
+#	$(verbose) echo '<?xml version="1.0" encoding="ISO-8859-1" standalone="no"?>' > $@
+#	$(verbose) echo '<!DOCTYPE policy SYSTEM "$(notdir $(xmldtd))">' >> $@
+#	$(verbose) echo '<policy>' >> $@
+#	$(verbose) for i in $(basename $(notdir $(layerxml))); do echo "<layer name=\"$$i\">" >> $@; cat $(tmpdir)/$$i.xml >> $@; echo "</layer>" >> $@; done
+#	$(verbose) cat $(tunxml) $(boolxml) >> $@
+#	$(verbose) echo '</policy>' >> $@
+#	$(verbose) if test -x $(XMLLINT) && test -f $(xmldtd); then \
+#		$(XMLLINT) --noout --path $(dir $(xmldtd)) --dtdvalid $(xmldtd) $@ ;\
+#	fi
 
-xml: $(polxml)
+#xml: $(polxml)
 
-html $(tmpdir)/html: $(polxml)
-	@echo "Building html interface reference documentation in $(htmldir)"
-	@test -d $(htmldir) || mkdir -p $(htmldir)
-	@test -d $(tmpdir) || mkdir -p $(tmpdir)
-	$(verbose) $(gendoc) -d $(htmldir) -T $(doctemplate) -x $(polxml)
-	$(verbose) cp $(doctemplate)/*.css $(htmldir)
-	@touch $(tmpdir)/html
+#html $(tmpdir)/html: $(polxml)
+#	@echo "Building html interface reference documentation in $(htmldir)"
+#	@test -d $(htmldir) || mkdir -p $(htmldir)
+#	@test -d $(tmpdir) || mkdir -p $(tmpdir)
+#	$(verbose) $(gendoc) -d $(htmldir) -T $(doctemplate) -x $(polxml)
+#	$(verbose) cp $(doctemplate)/*.css $(htmldir)
+#	@touch $(tmpdir)/html
 
 ########################################
 #
