Index: refpolicy/policy/modules/contrib/qemu.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/qemu.te
+++ refpolicy/policy/modules/contrib/qemu.te
@@ -96,6 +96,12 @@ tunable_policy(`qemu_full_network',`
 ')
 
 optional_policy(`
+	atapi_helper_rw_lock_files(qemu_t)
+	atapi_helper_rw_tmpfs_files(qemu_t)
+	audio_helper_communicate(qemu_t, qemu_tmpfs_t)
+')
+
+optional_policy(`
 	input_server_stream_connect(qemu_t)
 ')
 
