Index: refpolicy/policy/modules/contrib/qemu.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/qemu.te
+++ refpolicy/policy/modules/contrib/qemu.te
@@ -106,6 +106,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	surfman_dbus_chat(qemu_t)
+')
+
+optional_policy(`
 	xserver_user_x_domain_template(qemu, qemu_t, qemu_tmpfs_t)
 ')
 
Index: refpolicy/policy/modules/contrib/xen.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/xen.te
+++ refpolicy/policy/modules/contrib/xen.te
@@ -275,6 +275,7 @@ tapdisk_domtrans(xend_t)
 tapctl_domtrans(xend_t)
 
 dbd_dbus_chat(xend_t)
+surfman_dbus_chat(xend_t)
 
 kernel_read_kernel_sysctls(xend_t)
 kernel_read_system_state(xend_t)
