Index: refpolicy/policy/modules/roles/sysadm.te
===================================================================
--- refpolicy.orig/policy/modules/roles/sysadm.te
+++ refpolicy/policy/modules/roles/sysadm.te
@@ -1104,6 +1104,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	tpmsetup_role(sysadm_r, sysadm_t)
+')
+
+optional_policy(`
 	tpmutil_run(sysadm_t, sysadm_r)
 	tpmutil_sa_run(sysadm_t, sysadm_r)
 ')
Index: refpolicy/policy/modules/system/lvm.te
===================================================================
--- refpolicy.orig/policy/modules/system/lvm.te
+++ refpolicy/policy/modules/system/lvm.te
@@ -393,6 +393,11 @@ optional_policy(`
 	rpm_manage_script_tmp_files(lvm_t)
 ')
 
+# leaked fds?
+optional_policy(`
+	tpmsetup_rw_semaphores(lvm_t)
+')
+
 optional_policy(`
 	udev_read_db(lvm_t)
 	udev_read_pid_files(lvm_t)
