--- a/policy/modules/services/networkmanager.te
+++ b/policy/modules/services/networkmanager.te
@@ -392,6 +392,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	uid_dbus_send(NetworkManager_t)
+')
+
+optional_policy(`
 	vpn_domtrans(NetworkManager_t)
 	vpn_kill(NetworkManager_t)
 	vpn_signal(NetworkManager_t)
--- a/policy/modules/system/xen.te
+++ b/policy/modules/system/xen.te
@@ -281,6 +281,7 @@ tapdisk_filetrans_control_dir(xend_t)
 
 dbd_dbus_chat(xend_t)
 surfman_dbus_chat(xend_t)
+uid_dbus_chat(xend_t)
 
 kernel_read_kernel_sysctls(xend_t)
 kernel_read_system_state(xend_t)
