Index: refpolicy/policy/modules/system/selinuxutil.te
===================================================================
--- refpolicy.orig/policy/modules/system/selinuxutil.te
+++ refpolicy/policy/modules/system/selinuxutil.te
@@ -230,6 +230,12 @@ ifdef(`distro_debian',`
 	')
 ')
 
+optional_policy(`
+	updatemgr_dontaudit_use_fd(load_policy_t)
+	updatemgr_dontaudit_rw_fifo_files(load_policy_t)
+	updatemgr_dontaudit_rw_stream_sockets(load_policy_t)
+')
+
 ########################################
 #
 # Newrole local policy
@@ -533,6 +539,12 @@ ifdef(`distro_ubuntu',`
 	')
 ')
 
+optional_policy(`
+	updatemgr_dontaudit_use_fd(semanage_t)
+	updatemgr_dontaudit_rw_fifo_files(semanage_t)
+	updatemgr_dontaudit_rw_stream_sockets(semanage_t)
+')
+
 ########################################
 #
 # Setfiles local policy
@@ -659,6 +671,12 @@ optional_policy(`
 	hotplug_use_fds(setfiles_t)
 ')
 
+optional_policy(`
+	updatemgr_dontaudit_use_fd(setfiles_t)
+	updatemgr_dontaudit_rw_fifo_files(setfiles_t)
+	updatemgr_dontaudit_rw_stream_sockets(setfiles_t)
+')
+
 # input-server redirects output to a tmp file
 optional_policy(`
 	input_server_write_tmp_files(setfiles_t)
Index: refpolicy/policy/modules/system/lvm.te
===================================================================
--- refpolicy.orig/policy/modules/system/lvm.te
+++ refpolicy/policy/modules/system/lvm.te
@@ -418,3 +418,10 @@ optional_policy(`
 	xc_read_vhd_key_files(lvm_t)
 	xc_search_vhd_key_dirs(lvm_t)
 ')
+
+optional_policy(`
+	updatemgr_dontaudit_use_fd(lvm_t)
+	updatemgr_dontaudit_rw_fifo_files(lvm_t)
+	updatemgr_dontaudit_rw_stream_sockets(lvm_t)
+	updatemgr_dontaudit_search_storage(lvm_t)
+')
--- a/policy/modules/system/mount.te
+++ b/policy/modules/system/mount.te
@@ -36,6 +36,12 @@ files_tmp_file(mount_tmp_t)
 type unconfined_mount_t;
 application_domain(unconfined_mount_t, mount_exec_t)
 
+optional_policy(`
+	updatemgr_dontaudit_use_fd(mount_t)
+	updatemgr_dontaudit_rw_fifo_files(mount_t)
+	updatemgr_dontaudit_rw_stream_sockets(mount_t)
+')
+
 ########################################
 #
 # mount local policy
Index: refpolicy/policy/modules/system/fstools.te
===================================================================
--- refpolicy.orig/policy/modules/system/fstools.te
+++ refpolicy/policy/modules/system/fstools.te
@@ -211,6 +211,12 @@ optional_policy(`
 ')
 
 optional_policy(`
+	updatemgr_dontaudit_use_fd(fsadm_t)
+	updatemgr_dontaudit_rw_fifo_files(fsadm_t)
+	updatemgr_dontaudit_rw_stream_sockets(fsadm_t)
+')
+
+optional_policy(`
 	xen_append_log(fsadm_t)
 	xen_rw_image_files(fsadm_t)
 ')
