Index: refpolicy/policy/modules/contrib/xen.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/xen.te
+++ refpolicy/policy/modules/contrib/xen.te
@@ -473,6 +473,10 @@ optional_policy(`
 	virt_read_config(xend_t)
 ')
 
+optional_policy(`
+	vhdutil_domtrans(xend_t)
+')
+
 ########################################
 #
 # Xen console local policy
Index: refpolicy/policy/modules/roles/sysadm.te
===================================================================
--- refpolicy.orig/policy/modules/roles/sysadm.te
+++ refpolicy/policy/modules/roles/sysadm.te
@@ -1219,6 +1219,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	vhdutil_run(sysadm_t, sysadm_r)
+')
+
+optional_policy(`
 	webalizer_run(sysadm_t, sysadm_r)
 ')
 
