--- a/policy/modules/system/xen.te
+++ b/policy/modules/system/xen.te
@@ -476,6 +476,10 @@ optional_policy(`
 	virt_read_config(xend_t)
 ')
 
+optional_policy(`
+	vhdutil_domtrans(xend_t)
+')
+
 ########################################
 #
 # Xen console local policy
--- a/policy/modules/roles/sysadm.te
+++ b/policy/modules/roles/sysadm.te
@@ -1280,6 +1280,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	vhdutil_run(sysadm_t, sysadm_r)
+')
+
+optional_policy(`
 	webalizer_run(sysadm_t, sysadm_r)
 ')
 
