--- a/policy/modules/services/dbus.te
+++ b/policy/modules/services/dbus.te
@@ -227,6 +227,10 @@ optional_policy(`
 	xserver_use_xdm_fds(system_dbusd_t)
 ')
 
+optional_policy(`
+	xc_dontaudit_search_config_dir(system_dbusd_t)
+')
+
 ########################################
 #
 # Common session bus local policy
--- a/policy/modules/admin/dmidecode.te
+++ b/policy/modules/admin/dmidecode.te
@@ -57,3 +57,7 @@ xen_dontaudit_rw_unix_stream_sockets(dmi
 optional_policy(`
 	statusreport_write_storage_files(dmidecode_t)
 ')
+
+optional_policy(`
+	xc_dontaudit_read_etc_files(dmidecode_t)
+')
--- a/policy/modules/admin/firstboot.te
+++ b/policy/modules/admin/firstboot.te
@@ -120,6 +120,11 @@ optional_policy(`
 ')
 
 optional_policy(`
+	xc_manage_config_dirs(firstboot_t)
+	xc_relabelfrom_config_dirs(firstboot_t)
+')
+
+optional_policy(`
 	xserver_domtrans(firstboot_t)
 	xserver_rw_shm(firstboot_t)
 	xserver_unconfined(firstboot_t)
--- a/policy/modules/apps/loadkeys.te
+++ b/policy/modules/apps/loadkeys.te
@@ -59,3 +59,8 @@ optional_policy(`
 optional_policy(`
 	nscd_dontaudit_search_runtime(loadkeys_t)
 ')
+
+optional_policy(`
+	xc_dontaudit_rw_argo_chr(loadkeys_t)
+	xc_search_config_dirs(loadkeys_t)
+')
--- a/policy/modules/services/networkmanager.te
+++ b/policy/modules/services/networkmanager.te
@@ -403,6 +403,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	xc_files_rw_argo_chr(NetworkManager_t)
+')
+
+optional_policy(`
 	xen_dontaudit_rw_unix_stream_sockets(NetworkManager_t)
 	xen_dbus_chat(NetworkManager_t)
 	dev_rw_xen(NetworkManager_t)
--- a/policy/modules/apps/qemu.te
+++ b/policy/modules/apps/qemu.te
@@ -116,6 +116,13 @@ optional_policy(`
 ')
 
 optional_policy(`
+	xc_dontaudit_write_iso_files(qemu_t)
+	xc_files_rw_argo_chr(qemu_t)
+	xc_read_iso_files(qemu_t)
+	xc_read_video_device(qemu_t)
+')
+
+optional_policy(`
 	xserver_user_x_domain_template(qemu, qemu_t, qemu_tmpfs_t)
 ')
 
--- a/policy/modules/system/xen.te
+++ b/policy/modules/system/xen.te
@@ -143,6 +143,11 @@ domain_type(xm_t)
 type xend_config_t;
 files_type(xend_config_t);
 
+# var/lib/ovf/* files
+type xend_ovf_iso_t;
+xc_iso_type(xend_ovf_iso_t)
+xc_ovf_var_lib_filetrans(xend_t, xend_ovf_iso_t, file)
+
 ########################################
 #
 # blktap local policy
@@ -342,6 +347,24 @@ files_rw_etc_files(xend_t)
 # accessing vmlinuz from service VM disks
 kernel_read_unlabeled_files(xend_t)
 
+xc_read_etc_files(xend_t)
+xc_getattr_config_files(xend_t)
+xc_read_config_files(xend_t)
+# never reads device.key file, just searches through /config/sec?
+xc_config_sec_dir_list(xend_t)
+xc_search_vhd_key_dirs(xend_t)
+xc_list_iso_files(xend_t)
+xc_read_iso_files(xend_t)
+xc_dontaudit_write_iso_files(xend_t)
+xc_search_service_disk_dir(xend_t)
+xc_getattr_service_disk_files(xend_t)
+xc_create_disks_dir(xend_t)
+xc_create_client_disk_files(xend_t)
+xc_delete_client_disk_files(xend_t)
+xc_write_client_disk_files(xend_t)
+xc_files_rw_argo_chr(xend_t)
+xc_config_filetrans(xend_t, xend_config_t, { file dir })
+
 fs_getattr_all_fs(xend_t)
 fs_list_auto_mountpoints(xend_t)
 fs_read_dos_files(xend_t)
@@ -480,6 +503,10 @@ optional_policy(`
 	vhdutil_domtrans(xend_t)
 ')
 
+optional_policy(`
+        xc_read_nilf_files(xend_t)
+')
+
 ########################################
 #
 # Xen console local policy
--- a/policy/modules/kernel/files.if
+++ b/policy/modules/kernel/files.if
@@ -3045,6 +3045,8 @@ interface(`files_read_etc_files',`
 	read_files_pattern($1, etc_t, etc_t)
 	allow $1 etc_t:file map;
 	read_lnk_files_pattern($1, etc_t, etc_t)
+
+	xc_search_config_dirs($1)
 ')
 
 ########################################
--- a/policy/modules/services/ssh.te
+++ b/policy/modules/services/ssh.te
@@ -136,6 +136,10 @@ allow ssh_server ssh_home_t:dir list_dir
 read_files_pattern(ssh_server, ssh_home_t, ssh_home_t)
 read_lnk_files_pattern(ssh_server, ssh_home_t, ssh_home_t)
 
+optional_policy(`
+	xc_config_dir_search(ssh_server)
+')
+
 kernel_read_kernel_sysctls(ssh_t)
 kernel_read_system_state(ssh_t)
 
@@ -211,6 +215,11 @@ optional_policy(`
 ')
 
 optional_policy(`
+	xc_files_rw_argo_chr(ssh_t)
+	xc_search_config_dirs(ssh_t)
+')
+
+optional_policy(`
 	xserver_user_x_domain_template(ssh, ssh_t, ssh_tmpfs_t)
 	xserver_domtrans_xauth(ssh_t)
 ')
@@ -327,6 +336,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	xc_files_rw_argo_chr(sshd_t)
+')
+
+optional_policy(`
 	xserver_domtrans_xauth(sshd_t)
 	xserver_link_xdm_keys(sshd_t)
 ')
@@ -378,3 +391,8 @@ optional_policy(`
 optional_policy(`
 	udev_read_db(ssh_keygen_t)
 ')
+
+# XC: ssh_keygen_t makes and stores keys in /config/ssh
+optional_policy(`
+	xc_config_dir_search(ssh_keygen_t)
+')
--- a/policy/modules/system/authlogin.te
+++ b/policy/modules/system/authlogin.te
@@ -155,6 +155,12 @@ optional_policy(`
 	nis_authenticate(chkpwd_t)
 ')
 
+# these have got to be leaked file descriptors
+optional_policy(`
+	dev_search_sysfs(chkpwd_t)
+	xc_search_config_dirs(chkpwd_t)
+')
+
 ########################################
 #
 # PAM local policy
--- a/policy/modules/system/fstools.te
+++ b/policy/modules/system/fstools.te
@@ -231,3 +231,7 @@ optional_policy(`
 	xen_append_log(fsadm_t)
 	xen_rw_image_files(fsadm_t)
 ')
+
+optional_policy(`
+	xc_dontaudit_rw_argo_chr(fsadm_t)
+')
--- a/policy/modules/system/init.te
+++ b/policy/modules/system/init.te
@@ -597,6 +597,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	xc_config_filetrans(init_t, init_flag_t, file)
+')
+
+optional_policy(`
 	unconfined_domain(init_t)
 ')
 
--- a/policy/modules/system/logging.te
+++ b/policy/modules/system/logging.te
@@ -629,6 +629,11 @@ optional_policy(`
 ')
 
 optional_policy(`
+	xc_dontaudit_search_storage(syslogd_t)
+	xc_dontaudit_search_config_dir(syslogd_t)
+')
+
+optional_policy(`
 	# log to the xconsole
 	xserver_rw_console(syslogd_t)
 ')
--- a/policy/modules/system/lvm.te
+++ b/policy/modules/system/lvm.te
@@ -405,9 +405,12 @@ optional_policy(`
 ')
 
 optional_policy(`
+	xc_dontaudit_config_sec_files_read(lvm_t)
+	xc_dontaudit_rw_argo_chr(lvm_t)
 	xc_read_config_key_files(lvm_t)
 	xc_read_vhd_key_files(lvm_t)
 	xc_search_vhd_key_dirs(lvm_t)
+	xc_secure_files_read(lvm_t)
 ')
 
 optional_policy(`
--- a/policy/modules/system/mount.te
+++ b/policy/modules/system/mount.te
@@ -262,3 +262,7 @@ optional_policy(`
 	statusreport_write_storage_files(mount_t)
 	statusreport_getattr_storage_files(mount_t)
 ')
+
+optional_policy(`
+	xc_dontaudit_rw_argo_chr(mount_t)
+')
--- a/policy/modules/system/selinuxutil.te
+++ b/policy/modules/system/selinuxutil.te
@@ -353,6 +353,12 @@ if(secure_mode) {
 	userdom_spec_domtrans_all_users(newrole_t)
 }
 
+# these have got to be leaked file descriptors
+optional_policy(`
+	dev_dontaudit_search_sysfs(newrole_t)
+	xc_dontaudit_search_config_dir(newrole_t)
+')
+
 tunable_policy(`allow_polyinstantiation',`
 	files_polyinstantiate_all(newrole_t)
 ')
@@ -593,6 +599,10 @@ optional_policy(`
 	updatemgr_dontaudit_rw_stream_sockets(semanage_t)
 ')
 
+optional_policy(`
+	xc_search_config_dirs(semanage_t)
+')
+
 ########################################
 #
 # Setfiles local policy
@@ -734,3 +744,7 @@ optional_policy(`
 	updatemgr_dontaudit_rw_fifo_files(setfiles_t)
 	updatemgr_dontaudit_rw_stream_sockets(setfiles_t)
 ')
+
+optional_policy(`
+	xc_dontaudit_rw_argo_chr(setfiles_t)
+')
--- a/policy/modules/system/sysnetwork.te
+++ b/policy/modules/system/sysnetwork.te
@@ -417,6 +417,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	xc_config_dir_search(dhcpc_t)
+')
+
+optional_policy(`
 	network_slave_rw_stream_sockets(ifconfig_t)
 ')
 
--- a/policy/modules/system/udev.te
+++ b/policy/modules/system/udev.te
@@ -406,6 +406,13 @@ optional_policy(`
 ')
 
 optional_policy(`
+	xc_list_iso_files(udev_t)
+	xc_getattr_iso_files(udev_t)
+	xc_search_storage(udev_t)
+	xc_dontaudit_search_config_dir(udev_t)
+')
+
+optional_policy(`
 	xserver_read_xdm_runtime_files(udev_t)
 ')
 
@@ -436,6 +443,10 @@ dev_dontaudit_rw_generic_chr_files(udevr
 fs_search_tmpfs(udevrc_t)
 seutil_domtrans_setfiles(udevrc_t)
 
+optional_policy(`
+	xc_config_dir_search(udevrc_t)
+')
+
 allow udevrc_t self:fifo_file manage_fifo_file_perms;
 
 ########################################
