Index: refpolicy/policy/modules/system/lvm.te
===================================================================
--- refpolicy.orig/policy/modules/system/lvm.te
+++ refpolicy/policy/modules/system/lvm.te
@@ -427,4 +427,5 @@ optional_policy(`
 	updatemgr_dontaudit_rw_fifo_files(lvm_t)
 	updatemgr_dontaudit_rw_stream_sockets(lvm_t)
 	updatemgr_dontaudit_search_storage(lvm_t)
+	xc_installer_read_tmp_files(lvm_t)
 ')
