Index: refpolicy/policy/modules/contrib/xen.te
===================================================================
--- refpolicy.orig/policy/modules/contrib/xen.te
+++ refpolicy/policy/modules/contrib/xen.te
@@ -270,6 +270,9 @@ optional_policy(`
 	statusreport_domtrans(xend_t)
 ')
 
+# transition to viptables
+viptables_domtrans(xend_t)
+
 blktap_getattr_blk_file(xend_t)
 blktap_relabel_blk_file(xend_t)
 # this may be a leaked fd or something, dontaudit?
Index: refpolicy/policy/modules/roles/sysadm.te
===================================================================
--- refpolicy.orig/policy/modules/roles/sysadm.te
+++ refpolicy/policy/modules/roles/sysadm.te
@@ -1266,6 +1266,13 @@ optional_policy(`
 	statusreport_run(sysadm_t, sysadm_r)
 ')
 
+optional_policy(`
+	nilfcreate_run(sysadm_t, sysadm_r)
+	viptables_run(sysadm_t, sysadm_r)
+	xentop_run(sysadm_t, sysadm_r)
+	xec_run(sysadm_t, sysadm_r)
+')
+
 ifndef(`distro_redhat',`
 	optional_policy(`
 		auth_role(sysadm_r, sysadm_t)
Index: refpolicy/policy/modules/system/init.te
===================================================================
--- refpolicy.orig/policy/modules/system/init.te
+++ refpolicy/policy/modules/system/init.te
@@ -1126,6 +1126,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	xec_domtrans(initrc_t)
+')
+
+optional_policy(`
 	uml_setattr_util_sockets(initrc_t)
 ')
 
