--- a/policy/modules/services/networkmanager.te
+++ b/policy/modules/services/networkmanager.te
@@ -110,6 +110,8 @@ kernel_rw_net_sysctls(NetworkManager_t)
 kernel_read_crypto_sysctls(NetworkManager_t)
 
 fs_dontaudit_manage_xenfs_files(NetworkManager_t)
+# xen 4.6 uprev
+dev_dontaudit_manage_xen(NetworkManager_t)
 
 corenet_all_recvfrom_unlabeled(NetworkManager_t)
 corenet_all_recvfrom_netlabel(NetworkManager_t)
--- a/policy/modules/kernel/devices.if
+++ b/policy/modules/kernel/devices.if
@@ -5419,6 +5419,27 @@ interface(`dev_getattr_xen',`
 
 ########################################
 ## <summary>
+##	Do not audit attempts to create,
+##	read, write /dev/xen devices nodes.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain to not audit.
+##	</summary>
+## </param>
+#
+interface(`dev_dontaudit_manage_xen',`
+	gen_require(`
+		type xen_device_t;
+	')
+
+	dontaudit $1 xen_device_t:blk_file manage_blk_file_perms;
+	dontaudit $1 xen_device_t:chr_file manage_chr_file_perms;
+	dontaudit $1 xen_device_t:lnk_file manage_lnk_file_perms;
+')
+
+########################################
+## <summary>
 ##      Unconfined readonly access to devices.
 ## </summary>
 ## <param name="domain">
