From 0ec77fb462bbb3055a28630fb3384858e3475f3f Mon Sep 17 00:00:00 2001
From: mahantesh <mahantesh.openxt@gmail.com>
Date: Tue, 3 Oct 2017 19:17:34 -0400
Subject: [PATCH] blktap3: move physical-device xenstore node creation back to
 libxl

Signed-off-by: mahantesh <mahantesh.openxt@gmail.com>
---
 scripts/block | 1 -
 1 file changed, 1 deletion(-)

diff --git a/scripts/block b/scripts/block
index 629e045..2022243 100755
--- a/scripts/block
+++ b/scripts/block
@@ -27,7 +27,6 @@ case "$1" in
 add)
         PARAMS=$(xenstore-read "${XENBUS_PATH}/params")
         MINOR=$(stat -c '%T' ${PARAMS})
-        xenstore-write "${XENBUS_PATH}/physical-device" "fe:${MINOR}"
         xenstore-write "${XENBUS_PATH}/physical-device-path" "${PARAMS}"
         xenstore-write "${XAPI}/hotplug" "online"
         ;;
-- 
2.16.1

