diff --git a/libopkg/opkg_conf.c b/libopkg/opkg_conf.c
index aa2f8c7..ceafb6b 100644
--- a/libopkg/opkg_conf.c
+++ b/libopkg/opkg_conf.c
@@ -546,7 +546,7 @@ static int opkg_lock()
     }
     free(lock_dir);
 
-    lock_fd = creat(opkg_config->lock_file, S_IRUSR | S_IWUSR | S_IRGRP);
+    lock_fd = creat(opkg_config->lock_file, S_IRUSR | S_IWUSR | S_IRGRP | O_CLOEXEC);
     if (lock_fd == -1) {
         opkg_perror(ERROR, "Could not create lock file %s",
                     opkg_config->lock_file);
