Index: trunk/libopkg/opkg_conf.c
===================================================================
--- trunk.orig/libopkg/opkg_conf.c	2012-10-04 20:42:26.511216944 -0400
+++ trunk/libopkg/opkg_conf.c	2012-10-04 20:43:12.747151301 -0400
@@ -512,7 +512,7 @@
 	else
 		sprintf_alloc (&lock_file, "%s", OPKGLOCKFILE);
 
-	lock_fd = creat(lock_file, S_IRUSR | S_IWUSR | S_IRGRP);
+	lock_fd = creat(lock_file, S_IRUSR | S_IWUSR | S_IRGRP | O_CLOEXEC);
 	if (lock_fd == -1) {
 		opkg_perror(ERROR, "Could not create lock file %s", lock_file);
 		goto err2;
