diff --git a/policy/modules/contrib/brctl.te b/policy/modules/contrib/brctl.te
index 6515970..e3fe5f7 100644
--- a/policy/modules/contrib/brctl.te
+++ b/policy/modules/contrib/brctl.te
@@ -42,6 +42,10 @@ term_dontaudit_use_console(brctl_t)
 miscfiles_read_localization(brctl_t)
 
 optional_policy(`
+	network_slave_rw_stream_sockets(brctl_t)
+')
+
+optional_policy(`
 	xen_append_log(brctl_t)
 	xen_dontaudit_rw_unix_stream_sockets(brctl_t)
 ')
diff --git a/policy/modules/contrib/dnsmasq.te b/policy/modules/contrib/dnsmasq.te
index 758f74d..1e37c45 100644
--- a/policy/modules/contrib/dnsmasq.te
+++ b/policy/modules/contrib/dnsmasq.te
@@ -127,6 +127,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	network_slave_rw_stream_sockets(dnsmasq_t)
+')
+
+optional_policy(`
 	seutil_sigchld_newrole(dnsmasq_t)
 ')
 
diff --git a/policy/modules/contrib/networkmanager.te b/policy/modules/contrib/networkmanager.te
index 6cecec6..ffadd38 100644
--- a/policy/modules/contrib/networkmanager.te
+++ b/policy/modules/contrib/networkmanager.te
@@ -287,6 +287,11 @@ optional_policy(`
 ')
 
 optional_policy(`
+	network_slave_rw_stream_sockets(NetworkManager_t)
+	network_slave_domtrans(NetworkManager_t)
+')
+
+optional_policy(`
 	nscd_domtrans(NetworkManager_t)
 	nscd_signal(NetworkManager_t)
 	nscd_signull(NetworkManager_t)
diff --git a/policy/modules/system/init.te b/policy/modules/system/init.te
index b2bf44d..755d9d4 100644
--- a/policy/modules/system/init.te
+++ b/policy/modules/system/init.te
@@ -694,6 +694,9 @@ optional_policy(`
 	optional_policy(`
 		policykit_dbus_chat(initrc_t)
 	')
+	optional_policy(`
+        	network_slave_dbus_chat(initrc_t)
+	')
 ')
 
 optional_policy(`
diff --git a/policy/modules/system/iptables.te b/policy/modules/system/iptables.te
index be8ed1e..20c46bd 100644
--- a/policy/modules/system/iptables.te
+++ b/policy/modules/system/iptables.te
@@ -114,6 +114,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	network_slave_rw_stream_sockets(iptables_t)
+')
+
+optional_policy(`
 	# for iptables -L
 	nis_use_ypbind(iptables_t)
 ')
diff --git a/policy/modules/system/modutils.te b/policy/modules/system/modutils.te
index a55dbd9..c5a7f77 100644
--- a/policy/modules/system/modutils.te
+++ b/policy/modules/system/modutils.te
@@ -203,6 +203,12 @@ optional_policy(`
 ')
 
 optional_policy(`
+	network_slave_rw_stream_sockets(insmod_t)
+	network_slave_rw_fifo_files(insmod_t)
+	network_slave_use_fds(insmod_t)
+')
+
+optional_policy(`
 	nis_use_ypbind(insmod_t)
 ')
 
diff --git a/policy/modules/system/sysnetwork.te b/policy/modules/system/sysnetwork.te
index 0fc3571..c73d06b 100644
--- a/policy/modules/system/sysnetwork.te
+++ b/policy/modules/system/sysnetwork.te
@@ -355,6 +355,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	network_slave_rw_stream_sockets(ifconfig_t)
+')
+
+optional_policy(`
 	nis_use_ypbind(ifconfig_t)
 ')
 
diff --git a/policy/modules/system/udev.te b/policy/modules/system/udev.te
index 2003fbf..7aa0bea 100644
--- a/policy/modules/system/udev.te
+++ b/policy/modules/system/udev.te
@@ -270,6 +270,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	network_slave_dbus_chat(udev_t)
+')
+
+optional_policy(`
 	openct_read_pid_files(udev_t)
 	openct_domtrans(udev_t)
 ')
