diff --git a/policy/modules/system/modutils.te b/policy/modules/system/modutils.te
index 7a363b8..a55dbd9 100644
--- a/policy/modules/system/modutils.te
+++ b/policy/modules/system/modutils.te
@@ -135,6 +135,7 @@ corecmd_exec_bin(insmod_t)
 corecmd_exec_shell(insmod_t)
 
 dev_rw_sysfs(insmod_t)
+dev_mount_usbfs(insmod_t)
 dev_search_usbfs(insmod_t)
 dev_rw_mtrr(insmod_t)
 dev_read_urand(insmod_t)
@@ -171,6 +172,7 @@ logging_search_logs(insmod_t)
 
 miscfiles_read_localization(insmod_t)
 
+selinux_mount_fs(insmod_t)
 seutil_read_file_contexts(insmod_t)
 
 userdom_use_user_terminals(insmod_t)
