diff --git a/policy/modules/system/fstools.fc b/policy/modules/system/fstools.fc
index 9656352..a46bdef 100644
--- a/policy/modules/system/fstools.fc
+++ b/policy/modules/system/fstools.fc
@@ -46,6 +46,7 @@
 /sbin/zpios		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/zstreamdump	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /sbin/ztest		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
+/sbin/mke2fs\.e2fsprogs	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 
 /usr/bin/partition_uuid	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /usr/bin/raw		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
diff --git a/policy/modules/system/fstools.te b/policy/modules/system/fstools.te
index 3452b85..4eab8df 100644
--- a/policy/modules/system/fstools.te
+++ b/policy/modules/system/fstools.te
@@ -134,6 +134,7 @@ storage_raw_read_removable_device(fsadm_t)
 storage_raw_write_removable_device(fsadm_t)
 storage_read_scsi_generic(fsadm_t)
 storage_swapon_fixed_disk(fsadm_t)
+storage_rw_blktap(fsadm_t)
 
 term_use_console(fsadm_t)
 
@@ -195,3 +196,10 @@ optional_policy(`
 	xen_append_log(fsadm_t)
 	xen_rw_image_files(fsadm_t)
 ')
+
+# leaked FDs from input-server
+term_dontaudit_use_unallocated_ttys(fsadm_t)
+optional_policy(`
+	kernel_dontaudit_read_xen_state(fsadm_t)
+	kernel_dontaudit_write_xen_state(fsadm_t)
+')
