--- a/policy/modules/roles/staff.te	2010-05-24 15:31:36.000000000 -0400
+++ b/policy/modules/roles/staff.te
@@ -130,6 +130,7 @@
 
 optional_policy(`
 	ssh_role_template(staff, staff_r, staff_t)
+	ssh_inherit_env(staff_t)
 ')
 
 optional_policy(`
@@ -142,6 +143,7 @@
 
 optional_policy(`
 	sysadm_role_change(staff_r)
+	userdom_dontaudit_signal_admin_users(staff_t)
 	userdom_dontaudit_use_user_terminals(staff_t)
 ')
 
