Index: refpolicy/policy/modules/kernel/files.fc
===================================================================
--- refpolicy.orig/policy/modules/kernel/files.fc
+++ refpolicy/policy/modules/kernel/files.fc
@@ -46,6 +46,9 @@ ifdef(`distro_suse',`
 #
 /etc			-d	gen_context(system_u:object_r:etc_t,s0)
 /etc/.*				gen_context(system_u:object_r:etc_t,s0)
+/var/volatile/etc	-d	gen_context(system_u:object_r:etc_t,s0)
+/var/volatile/etc/.*		gen_context(system_u:object_r:etc_t,s0)
+/config/system/iproute2	-d	gen_context(system_u:object_r:etc_t,s0)
 /etc/\.fstab\.hal\..+	--	gen_context(system_u:object_r:etc_runtime_t,s0)
 /etc/blkid(/.*)?		gen_context(system_u:object_r:etc_runtime_t,s0)
 /etc/cmtab		--	gen_context(system_u:object_r:etc_runtime_t,s0)
@@ -65,6 +68,8 @@ ifdef(`distro_suse',`
 /etc/ipsec\.d/examples(/.*)?	gen_context(system_u:object_r:etc_t,s0)
 
 /etc/network/ifstate	--	gen_context(system_u:object_r:etc_runtime_t,s0)
+/var/volatile/etc/ifstate	--	gen_context(system_u:object_r:etc_runtime_t,s0)
+/var/run/ifstate		--	gen_context(system_u:object_r:etc_runtime_t,s0)
 
 /etc/ptal/ptal-printd-like -- 	gen_context(system_u:object_r:etc_runtime_t,s0)
 
@@ -87,6 +92,10 @@ ifdef(`distro_suse',`
 /etc/rc\.d/init\.d/\.depend.* -- gen_context(system_u:object_r:etc_runtime_t,s0)
 ')
 
+/config/etc(/.*)		gen_context(system_u:object_r:etc_t,s0)
+/config/passwd.*	--	gen_context(system_u:object_r:etc_t,s0)
+/var/volatile/system-connections(/.*)?	gen_context(system_u:object_r:etc_t,s0)
+
 #
 # HOME_ROOT
 # expanded by genhomedircon
@@ -243,6 +252,7 @@ ifndef(`distro_redhat',`
 /var/lib/nfs/rpc_pipefs(/.*)?	<<none>>
 
 /var/lock(/.*)?			gen_context(system_u:object_r:var_lock_t,s0)
+/var/volatile/lock(/.*)?	gen_context(system_u:object_r:var_lock_t,s0)
 
 /var/log/lost\+found	-d	gen_context(system_u:object_r:lost_found_t,mls_systemhigh)
 /var/log/lost\+found/.*		<<none>>
@@ -257,6 +267,7 @@ ifndef(`distro_redhat',`
 /var/run		-l	gen_context(system_u:object_r:var_run_t,s0)
 /var/run/.*			gen_context(system_u:object_r:var_run_t,s0)
 /var/run/.*\.*pid		<<none>>
+/var/run/hosts                 gen_context(system_u:object_r:etc_t,s0)
 
 /var/spool(/.*)?		gen_context(system_u:object_r:var_spool_t,s0)
 /var/spool/postfix/etc(/.*)?	gen_context(system_u:object_r:etc_t,s0)
@@ -267,6 +278,14 @@ ifndef(`distro_redhat',`
 /var/tmp/lost\+found	-d	gen_context(system_u:object_r:lost_found_t,mls_systemhigh)
 /var/tmp/lost\+found/.*		<<none>>
 /var/tmp/vi\.recover	-d	gen_context(system_u:object_r:tmp_t,s0)
+/var/volatile/tmp      -d      gen_context(system_u:object_r:tmp_t,s0)
+/var/volatile/tmp/.*                   <<none>>
+/var/volatile/tmp/lost\+found  -d      gen_context(system_u:object_r:lost_found_t,mls_systemhigh)
+/var/volatile/tmp/lost\+found/.*               <<none>>
+/var/volatile/tmp/vi\.recover  -d      gen_context(system_u:object_r:tmp_t,s0)
+/var/volatile/run                      -d      gen_context(system_u:object_r:var_run_t,s0-mls_systemhigh)
+/var/volatile/run/.*                   gen_context(system_u:object_r:var_run_t,s0)
+/var/volatile/run/.*\.*pid             <<none>>
 
 ifdef(`distro_debian',`
 /var/run/motd		--	gen_context(system_u:object_r:initrc_var_run_t,s0)
Index: refpolicy/policy/modules/kernel/files.if
===================================================================
--- refpolicy.orig/policy/modules/kernel/files.if
+++ refpolicy/policy/modules/kernel/files.if
@@ -1178,6 +1178,7 @@ interface(`files_list_all',`
 	')
 
 	allow $1 file_type:dir list_dir_perms;
+	allow $1 file_type:lnk_file read_lnk_file_perms;
 ')
 
 ########################################
@@ -1561,6 +1562,7 @@ interface(`files_mounton_all_mountpoints
 
 	allow $1 mountpoint:dir { search_dir_perms mounton };
 	allow $1 mountpoint:file { getattr mounton };
+	allow $1 mountpoint:dir write;
 ')
 
 ########################################
@@ -2521,6 +2523,25 @@ interface(`files_read_default_symlinks',
 
 ########################################
 ## <summary>
+##	Do not audit attempts to get attributes of symbolic
+##	links with the default file type.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`files_dontaudit_getattr_default_symlinks',`
+	gen_require(`
+		type default_t;
+	')
+
+	dontaudit $1 default_t:lnk_file getattr_lnk_file_perms;
+')
+
+########################################
+## <summary>
 ##	Read sockets with the default file type.
 ## </summary>
 ## <param name="domain">
@@ -3147,6 +3168,25 @@ interface(`files_getattr_isid_type_dirs'
 
 ########################################
 ## <summary>
+##	Search directories on new filesystems
+##	that have not yet been labeled.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`files_search_isid_type_dirs',`
+	gen_require(`
+		type file_t;
+	')
+
+	allow $1 file_t:dir search_dir_perms;
+')
+
+########################################
+## <summary>
 ##	Do not audit attempts to search directories on new filesystems
 ##	that have not yet been labeled.  (Deprecated)
 ## </summary>
@@ -3387,6 +3427,25 @@ interface(`files_manage_isid_type_files'
 
 ########################################
 ## <summary>
+##	Read symbolic links on new filesystems that have
+##	not yet been labeled.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`files_read_isid_type_symlinks',`
+	gen_require(`
+		type file_t;
+	')
+
+	allow $1 file_t:lnk_file read_lnk_file_perms;
+')
+
+########################################
+## <summary>
 ##	Create, read, write, and delete symbolic links
 ##	on new filesystems that have not yet been labeled.  (Deprecated)
 ## </summary>
@@ -3863,6 +3922,24 @@ interface(`files_read_mnt_symlinks',`
 
 ########################################
 ## <summary>
+##	Do not audit attempts to getattrs of /mnt directory.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`files_dontaudit_getattr_mnt_dirs',`
+	gen_require(`
+		type mnt_t;
+	')
+
+	dontaudit $1 mnt_t:dir getattr_dir_perms;
+')
+
+########################################
+## <summary>
 ##	Create, read, write, and delete symbolic links in /mnt.
 ## </summary>
 ## <param name="domain">
@@ -4192,6 +4269,7 @@ interface(`files_getattr_tmp_dirs',`
 		type tmp_t;
 	')
 
+	allow $1 tmp_t:lnk_file read_lnk_file_perms;
 	allow $1 tmp_t:dir getattr;
 ')
 
@@ -4211,6 +4289,7 @@ interface(`files_dontaudit_getattr_tmp_d
 		type tmp_t;
 	')
 
+	allow $1 tmp_t:lnk_file read_lnk_file_perms;
 	dontaudit $1 tmp_t:dir getattr;
 ')
 
@@ -4229,6 +4308,7 @@ interface(`files_search_tmp',`
 		type tmp_t;
 	')
 
+	allow $1 tmp_t:lnk_file read_lnk_file_perms;
 	allow $1 tmp_t:dir search_dir_perms;
 	allow $1 tmp_t:lnk_file read_lnk_file_perms;
 ')
@@ -4248,6 +4328,7 @@ interface(`files_dontaudit_search_tmp',`
 		type tmp_t;
 	')
 
+	allow $1 tmp_t:lnk_file read_lnk_file_perms;
 	dontaudit $1 tmp_t:dir search_dir_perms;
 ')
 
@@ -4266,6 +4347,7 @@ interface(`files_list_tmp',`
 		type tmp_t;
 	')
 
+	allow $1 tmp_t:lnk_file read_lnk_file_perms;
 	allow $1 tmp_t:dir list_dir_perms;
 	allow $1 tmp_t:lnk_file read_lnk_file_perms;
 ')
@@ -4285,6 +4367,7 @@ interface(`files_dontaudit_list_tmp',`
 		type tmp_t;
 	')
 
+	allow $1 tmp_t:lnk_file read_lnk_file_perms;
 	dontaudit $1 tmp_t:dir list_dir_perms;
 ')
 
@@ -4303,6 +4386,7 @@ interface(`files_delete_tmp_dir_entry',`
 		type tmp_t;
 	')
 
+	allow $1 tmp_t:lnk_file read_lnk_file_perms;
 	allow $1 tmp_t:dir del_entry_dir_perms;
 	allow $1 tmp_t:lnk_file read_lnk_file_perms;
 ')
@@ -4322,6 +4406,7 @@ interface(`files_read_generic_tmp_files'
 		type tmp_t;
 	')
 
+	allow $1 tmp_t:lnk_file read_lnk_file_perms;
 	read_files_pattern($1, tmp_t, tmp_t)
 	allow $1 tmp_t:lnk_file read_lnk_file_perms;
 ')
@@ -4341,6 +4426,7 @@ interface(`files_manage_generic_tmp_dirs
 		type tmp_t;
 	')
 
+	allow $1 tmp_t:lnk_file read_lnk_file_perms;
 	manage_dirs_pattern($1, tmp_t, tmp_t)
 	allow $1 tmp_t:lnk_file read_lnk_file_perms;
 ')
@@ -4360,6 +4446,7 @@ interface(`files_manage_generic_tmp_file
 		type tmp_t;
 	')
 
+	allow $1 tmp_t:lnk_file read_lnk_file_perms;
 	manage_files_pattern($1, tmp_t, tmp_t)
 	allow $1 tmp_t:lnk_file read_lnk_file_perms;
 ')
@@ -4379,6 +4466,7 @@ interface(`files_read_generic_tmp_symlin
 		type tmp_t;
 	')
 
+	allow $1 tmp_t:lnk_file read_lnk_file_perms;
 	read_lnk_files_pattern($1, tmp_t, tmp_t)
 ')
 
@@ -4397,6 +4485,7 @@ interface(`files_rw_generic_tmp_sockets'
 		type tmp_t;
 	')
 
+	allow $1 tmp_t:lnk_file read_lnk_file_perms;
 	rw_sock_files_pattern($1, tmp_t, tmp_t)
 	allow $1 tmp_t:lnk_file read_lnk_file_perms;
 ')
@@ -4416,6 +4505,7 @@ interface(`files_setattr_all_tmp_dirs',`
 		attribute tmpfile;
 	')
 
+	allow $1 tmp_t:lnk_file read_lnk_file_perms;
 	allow $1 tmpfile:dir { search_dir_perms setattr };
 ')
 
@@ -4434,6 +4524,7 @@ interface(`files_list_all_tmp',`
 		attribute tmpfile;
 	')
 
+	allow $1 tmp_t:lnk_file read_lnk_file_perms;
 	allow $1 tmpfile:dir list_dir_perms;
 ')
 
@@ -4515,6 +4606,7 @@ interface(`files_relabel_all_tmp_files',
 		type var_t;
 	')
 
+	allow $1 var_lock_t:lnk_file read_lnk_file_perms;
 	allow $1 var_t:dir search_dir_perms;
 	relabel_files_pattern($1, tmpfile, tmpfile)
 ')
@@ -4606,6 +4698,7 @@ interface(`files_purge_tmp',`
 		attribute tmpfile;
 	')
 
+	allow $1 tmp_t:lnk_file read_lnk_file_perms;
 	allow $1 tmpfile:dir list_dir_perms;
 	delete_dirs_pattern($1, tmpfile, tmpfile)
 	delete_files_pattern($1, tmpfile, tmpfile)
@@ -5108,9 +5201,10 @@ interface(`files_delete_kernel_symbol_ta
 #
 interface(`files_search_var',`
 	gen_require(`
-		type var_t;
+		type var_t, var_lock_t;
 	')
 
+	allow $1 var_lock_t:lnk_file read_lnk_file_perms;
 	allow $1 var_t:dir search_dir_perms;
 ')
 
@@ -5134,6 +5228,24 @@ interface(`files_dontaudit_write_var_dir
 
 ########################################
 ## <summary>
+##	Do not audit attempts to write to /var/lock.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain to not audit.
+##	</summary>
+## </param>
+#
+interface(`files_dontaudit_write_lock_dirs',`
+	gen_require(`
+		type var_lock_t;
+	')
+
+	dontaudit $1 var_lock_t:dir write;
+')
+
+########################################
+## <summary>
 ##	Allow attempts to write to /var.dirs
 ## </summary>
 ## <param name="domain">
@@ -5519,9 +5631,10 @@ interface(`files_rw_var_lib_dirs',`
 #
 interface(`files_var_lib_filetrans',`
 	gen_require(`
-		type var_t, var_lib_t;
+		type var_t, var_lib_t, var_run_t;
 	')
 
+	allow $1 var_run_t:lnk_file read_lnk_file_perms;
 	allow $1 var_t:dir search_dir_perms;
 	filetrans_pattern($1, var_lib_t, $2, $3, $4)
 ')
@@ -5579,9 +5692,10 @@ interface(`files_read_var_lib_symlinks',
 #
 interface(`files_manage_urandom_seed',`
 	gen_require(`
-		type var_t, var_lib_t;
+		type var_t, var_lib_t, var_run_t;
 	')
 
+	allow $1 var_run_t:lnk_file read_lnk_file_perms;
 	allow $1 var_t:dir search_dir_perms;
 	manage_files_pattern($1, var_lib_t, var_lib_t)
 ')
@@ -5913,6 +6027,50 @@ interface(`files_lock_filetrans',`
 
 ########################################
 ## <summary>
+##     Create an object in the locks directory, with a private
+##     type using both a type and a range transition.
+## </summary>
+## <param name="domain">
+##     <summary>
+##     Domain allowed access.
+##     </summary>
+## </param>
+## <param name="private type">
+##     <summary>
+##     The type of the object to be created.
+##     </summary>
+## </param>
+## <param name="object">
+##     <summary>
+##     The object class of the object being created.
+##     </summary>
+## </param>
+## <param name="range">
+##     <summary>
+##     Range for the object.
+##     </summary>
+## </param>
+#
+interface(`files_lock_ranged_filetrans',`
+       gen_require(`
+               type var_t, var_lock_t;
+       ')
+
+	allow $1 var_t:dir search_dir_perms;
+	allow $1 var_lock_t:lnk_file read_lnk_file_perms;
+	filetrans_pattern($1, var_lock_t, $2, $3)
+
+	ifdef(`enable_mcs',`
+		range_transition $1 var_lock_t:$3 $4;
+	')
+
+	ifdef(`enable_mls',`
+		range_transition $1 var_lock_t:$3 $4;
+	')
+')
+
+########################################
+## <summary>
 ##	Do not audit attempts to get the attributes
 ##	of the /var/run directory.
 ## </summary>
@@ -5933,6 +6091,25 @@ interface(`files_dontaudit_getattr_pid_d
 
 ########################################
 ## <summary>
+##     Do not audit attempts to write to
+##     the /var/run directory.
+## </summary>
+## <param name="domain">
+##     <summary>
+##     Domain to not audit.
+##     </summary>
+## </param>
+#
+interface(`files_dontaudit_write_pid_dirs',`
+	gen_require(`
+		type var_run_t;
+	')
+
+	dontaudit $1 var_run_t:dir write;
+')
+
+########################################
+## <summary>
 ##	Set the attributes of the /var/run directory.
 ## </summary>
 ## <param name="domain">
@@ -5968,6 +6145,7 @@ interface(`files_search_pids',`
 
 	allow $1 var_run_t:lnk_file read_lnk_file_perms;
 	search_dirs_pattern($1, var_t, var_run_t)
+	read_lnk_files_pattern($1, var_t, var_run_t)
 ')
 
 ########################################
@@ -6188,6 +6366,7 @@ interface(`files_dontaudit_getattr_all_p
 interface(`files_dontaudit_write_all_pids',`
 	gen_require(`
 		attribute pidfile;
+		type var_run_t;
 	')
 
 	dontaudit $1 var_run_t:lnk_file read_lnk_file_perms;
@@ -6196,6 +6375,42 @@ interface(`files_dontaudit_write_all_pid
 
 ########################################
 ## <summary>
+##	Set attributes on all runtime data dirs.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`files_setattr_all_pid_dirs',`
+	gen_require(`
+		attribute pidfile;
+	')
+
+	allow $1 pidfile:dir setattr;
+')
+
+########################################
+## <summary>
+##	Set attributes on all runtime pid files.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`files_setattr_all_pids',`
+	gen_require(`
+		attribute pidfile;
+	')
+
+	allow $1 pidfile:file setattr;
+')
+
+########################################
+## <summary>
 ##	Do not audit attempts to ioctl daemon runtime data files.
 ## </summary>
 ## <param name="domain">
@@ -6264,6 +6479,24 @@ interface(`files_delete_all_pids',`
 
 ########################################
 ## <summary>
+##	Create process ID directories.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`files_create_pid_dirs',`
+	gen_require(`
+		type var_run_t;
+	')
+
+	create_dirs_pattern($1, var_run_t, var_run_t)
+')
+
+########################################
+## <summary>
 ##	Delete all process ID directories.
 ## </summary>
 ## <param name="domain">
@@ -6547,3 +6780,21 @@ interface(`files_unconfined',`
 
 	typeattribute $1 files_unconfined_type;
 ')
+
+########################################
+## <summary>
+##      Unconfined access to files.
+## </summary>
+## <param name="domain">
+##      <summary>
+##      Domain allowed access.
+##      </summary>
+## </param>
+#
+interface(`files_unconfined_readonly',`
+	gen_require(`
+		attribute files_unconfined_readonly_type;
+	')
+
+	typeattribute $1 files_unconfined_readonly_type;
+')
Index: refpolicy/policy/modules/kernel/files.te
===================================================================
--- refpolicy.orig/policy/modules/kernel/files.te
+++ refpolicy/policy/modules/kernel/files.te
@@ -6,6 +6,7 @@ policy_module(files, 1.19.0)
 #
 
 attribute file_type;
+attribute files_unconfined_readonly_type;
 attribute files_unconfined_type;
 attribute lockfile;
 attribute mountpoint;
@@ -60,6 +61,9 @@ files_mountpoint(default_t)
 #
 type etc_t, configfile;
 files_type(etc_t)
+# XC: this is needed to mount a tmpfs on /etc/NetworkManager/system-connections
+files_mountpoint(etc_t)
+
 # compatibility aliases for removed types:
 typealias etc_t alias automount_etc_t;
 typealias etc_t alias snmpd_etc_t;
@@ -223,3 +227,14 @@ allow files_unconfined_type file_type:fi
 tunable_policy(`allow_execmod',`
 	allow files_unconfined_type file_type:file execmod;
 ')
+
+########################################
+#
+# Unconfined readonly access to this module
+#
+
+allow files_unconfined_readonly_type file_type:{ file sock_file fifo_file blk_file } {  getattr open read lock ioctl };
+allow files_unconfined_readonly_type file_type: lnk_file { read_lnk_file_perms getattr_lnk_file_perms };
+allow files_unconfined_readonly_type file_type: sock_file write_sock_file_perms ;
+allow files_unconfined_readonly_type file_type: chr_file read_chr_file_perms ;
+allow files_unconfined_readonly_type file_type: dir { search_dir_perms getattr_dir_perms list_dir_perms };
