diff --git a/inputattach.c b/inputattach.c
index cd8b91a..273528d 100644
--- a/inputattach.c
+++ b/inputattach.c
@@ -525,6 +525,8 @@ static void show_help(void)
 	puts("");
 }
 
+#define RETRY_ERROR(x) (x == EAGAIN || x == EWOULDBLOCK || x == EINTR)
+
 int main(int argc, char **argv)
 {
 	unsigned long devt;
@@ -644,7 +646,13 @@ int main(int argc, char **argv)
 		retval = EXIT_FAILURE;
 	}
 
-	read(fd, NULL, 0);
+	do {
+                i = read(fd, NULL, 0);
+                if (i == -1) {
+                        if (RETRY_ERROR(errno))
+                                continue;
+                }
+        } while (!i);
 
 	ldisc = 0;
 	ioctl(fd, TIOCSETD, &ldisc);
