diff -uNr notification-daemon-0.3.7.orig/src/themes/standard/theme.c notification-daemon-0.3.7/src/themes/standard/theme.c
--- notification-daemon-0.3.7.orig/src/themes/standard/theme.c	2011-10-14 16:43:14.514003905 +0100
+++ notification-daemon-0.3.7/src/themes/standard/theme.c	2011-10-14 16:46:46.518204633 +0100
@@ -79,6 +79,7 @@
 #define DEFAULT_ARROW_WIDTH   28
 #define BACKGROUND_OPACITY    0.92
 #define BOTTOM_GRADIENT_HEIGHT 30
+#define WRAP_CHARS             40
 
 #if GTK_CHECK_VERSION(2, 8, 0)
 # define USE_CAIRO
@@ -824,6 +825,7 @@
 	gtk_box_pack_start(GTK_BOX(vbox), windata->body_label, TRUE, TRUE, 0);
 	gtk_misc_set_alignment(GTK_MISC(windata->body_label), 0, 0);
 	gtk_label_set_line_wrap(GTK_LABEL(windata->body_label), TRUE);
+	gtk_label_set_width_chars(GTK_LABEL(windata->body_label), WRAP_CHARS);
 	g_signal_connect_swapped(G_OBJECT(windata->body_label), "url_activated",
 							 G_CALLBACK(windata->url_clicked), win);
 
