--- NetworkManager-0.9.2.0.orig/src/nm-netlink-utils.c	2011-11-09 17:22:30.000000000 +0000
+++ NetworkManager-0.9.2.0/src/nm-netlink-utils.c	2012-07-27 16:45:55.000000000 +0000
@@ -413,7 +421,7 @@
 	info.user_data = user_data;
 	info.iface = nm_netlink_index_to_iface (ifindex);
 
-	rtnl_route_alloc_cache (nm_netlink_get_default_handle (), family, NL_AUTO_PROVIDE, &cache);
+	rtnl_route_alloc_cache (nm_netlink_get_default_handle (), family, 0, &cache);
 	g_warn_if_fail (cache != NULL);
 	if (cache) {
 		nl_cache_foreach (cache, foreach_route_cb, &info);

