--- a/bus/connection.c
+++ b/bus/connection.c
@@ -302,7 +302,9 @@ bus_connection_disconnected (DBusConnect
       if (mm != NULL)
         bus_matchmaker_disconnected (mm, connection);
 
-      _dbus_list_remove_link (&d->connections->monitors, d->link_in_monitors);
+      if (d->connections->monitors != NULL)
+        _dbus_list_remove_link (&d->connections->monitors, d->link_in_monitors);
+
       d->link_in_monitors = NULL;
     }
 
