diff --git openjdk-ecj.orig/hotspot/src/share/vm/opto/loopnode.cpp openjdk-ecj/hotspot/src/share/vm/opto/loopnode.cpp
index 7e37599..51631d7 100644
--- openjdk-ecj.orig/hotspot/src/share/vm/opto/loopnode.cpp
+++ openjdk-ecj/hotspot/src/share/vm/opto/loopnode.cpp
@@ -534,7 +534,7 @@ int CountedLoopEndNode::stride_con() const {
 Node* CountedLoopNode::match_incr_with_optional_truncation(
                       Node* expr, Node** trunc1, Node** trunc2, const TypeInt** trunc_type) {
   // Quick cutouts:
-  if (expr == NULL || expr->req() != 3)  return false;
+  if (expr == NULL || expr->req() != 3)  return NULL;
 
   Node *t1 = NULL;
   Node *t2 = NULL;
diff --git openjdk-ecj.orig/corba/make/common/shared/Compiler-gcc.gmk openjdk-ecj/corba/make/common/shared/Compiler-gcc.gmk
index c8e6779..b91447b 100644
--- openjdk-ecj.orig/corba/make/common/shared/Compiler-gcc.gmk
+++ openjdk-ecj/corba/make/common/shared/Compiler-gcc.gmk
@@ -89,7 +89,7 @@ ifeq ($(PLATFORM), linux)
   endif
   endif
   # Option used to create a shared library
-  SHARED_LIBRARY_FLAG = -shared -mimpure-text
+  SHARED_LIBRARY_FLAG = -shared
   SUN_COMP_VER := $(shell $(CC) --verbose 2>&1 )
 
 endif
diff --git openjdk-ecj.orig/jdk/make/common/shared/Compiler-gcc.gmk openjdk-ecj/jdk/make/common/shared/Compiler-gcc.gmk
index 74f9807..627c5ad 100644
--- openjdk-ecj.orig/jdk/make/common/shared/Compiler-gcc.gmk
+++ openjdk-ecj/jdk/make/common/shared/Compiler-gcc.gmk
@@ -94,7 +94,7 @@ ifeq ($(PLATFORM), linux)
   endif
   endif
   # Option used to create a shared library
-  SHARED_LIBRARY_FLAG = -shared -mimpure-text
+  SHARED_LIBRARY_FLAG = -shared
   SUN_COMP_VER := $(shell $(CC) --verbose 2>&1 )
 
 endif
