--- a/configure.ac
+++ b/configure.ac
@@ -138,7 +138,7 @@ AC_TYPE_SIZE_T
 AC_TYPE_SIGNAL
 AC_TYPE_UID_T
 
-CFLAGS="$CFLAGS -Wall -Wreturn-type -Wsign-compare"
+CFLAGS="$CFLAGS -Wall -Werror -Wreturn-type -Wsign-compare -Wno-error=unused-but-set-variable"
 
 AC_CONFIG_FILES(Makefile		\
 		po/Makefile.in		\
--- a/src/tpm_mgmt/tpm_present.c
+++ b/src/tpm_mgmt/tpm_present.c
@@ -357,5 +357,5 @@ out_close:
 out:
     if (szTpmPasswd && !isWellKnown)
 	shredPasswd( szTpmPasswd );
-	return iRc;
+    return iRc;
 }
--- a/src/tpm_mgmt/tpm_nvcommon.c
+++ b/src/tpm_mgmt/tpm_nvcommon.c
@@ -164,8 +164,8 @@ int parseStringWithValues(const char *aA
 				return -1;
 			}
 
-			if (!aArg[offset+numbytes] == '|' &&
-			    !aArg[offset+numbytes] == 0) {
+			if (aArg[offset+numbytes] != '|' &&
+			    aArg[offset+numbytes] != 0) {
 				logError(_("Illegal character following decimal "
 				           "number in %s\n"),
 					 aArg + offset);
