--- a/src/tpm_mgmt/tpm_createek.c
+++ b/src/tpm_mgmt/tpm_createek.c
@@ -60,13 +60,13 @@ static int parse(const int aOpt, const c
 	case 'i':
 		inFileSet = TRUE;
 		if (aArg){
-			strncpy(in_filename, aArg, PATH_MAX);
+			strncpy(in_filename, aArg, PATH_MAX - 1);
 		}
 		break;
 	case 'o':
 		outFileSet = TRUE;
 		if (aArg){
-			strncpy(out_filename, aArg, PATH_MAX);
+			strncpy(out_filename, aArg, PATH_MAX - 1);
 		}
 		break;
 	default:
--- a/src/tpm_mgmt/tpm_revokeek.c
+++ b/src/tpm_mgmt/tpm_revokeek.c
@@ -42,7 +42,7 @@ static int parse(const int aOpt, const c
 	case 'i':
 		isInputSet = TRUE;
 		if (aArg){
-			strncpy(in_filename, aArg, PATH_MAX);
+			strncpy(in_filename, aArg, PATH_MAX - 1);
 		}
 		break;
 	default:
--- a/src/cmds/tpm_createsignkey.c
+++ b/src/cmds/tpm_createsignkey.c
@@ -48,7 +48,7 @@ static int parse(const int aOpt, const c
 	switch (aOpt) {
 	case 'o':
 		if (aArg) {
-			strncpy(out_filename, aArg, PATH_MAX);
+			strncpy(out_filename, aArg, PATH_MAX - 1);
 			rc = 0;
 		}
 		break;
@@ -191,7 +191,7 @@ int main(int argc, char **argv)
 		BIO_set_fp(bdata, stdout, BIO_NOCLOSE);
 	else {
 		char out_filename_conf[PATH_MAX];
-		sprintf(out_filename_conf, "%s.conf", out_filename);
+		snprintf(out_filename_conf, PATH_MAX, "%.4090s.conf", out_filename);
 		if (BIO_write_filename(bdata, out_filename_conf) <= 0) {
 			logError(_("Unable to open output file: %s\n"),
 				 out_filename_conf);
--- a/src/cmds/tpm_extendpcr.c
+++ b/src/cmds/tpm_extendpcr.c
@@ -52,7 +52,7 @@ static int parse(const int aOpt, const c
 	switch (aOpt) {
 	case 'i':
 		if (aArg) {
-			strncpy(in_filename, aArg, PATH_MAX);
+			strncpy(in_filename, aArg, PATH_MAX - 1);
 			rc = 0;
 		}
 		break;
--- a/src/cmds/tpm_quote.c
+++ b/src/cmds/tpm_quote.c
@@ -64,13 +64,13 @@ static int parse(const int aOpt, const c
 	switch (aOpt) {
 	case 'i':
 		if (aArg) {
-			strncpy(in_filename, aArg, PATH_MAX);
+			strncpy(in_filename, aArg, PATH_MAX - 1);
 			rc = 0;
 		}
 		break;
 	case 'o':
 		if (aArg) {
-			strncpy(out_filename, aArg, PATH_MAX);
+			strncpy(out_filename, aArg, PATH_MAX - 1);
 			rc = 0;
 		}
 		break;
@@ -127,7 +127,6 @@ readTssKey(char* fname, unsigned char **
 	struct stat stats;
 
 	BIO *bdata = NULL, *b64 = NULL, *bmem = NULL;
-	int bioRc;
 
 	if ( tss_data == NULL || tss_size == NULL ) {
 		rc = TPMQUOTE_STD_ERROR;
@@ -223,11 +222,11 @@ readTssKey(char* fname, unsigned char **
 	bmem = BIO_pop(b64);
 	BIO_free(b64);
 	b64 = NULL;
-	bioRc = BIO_reset(bmem);
+	BIO_reset(bmem);
 
 out:
 	if (bmem) {
-		bioRc = BIO_set_close(bmem, BIO_CLOSE);
+		BIO_set_close(bmem, BIO_CLOSE);
 		BIO_free(bmem);
 	}
 	if (rc == 0) {
--- a/src/cmds/tpm_sealdata.c
+++ b/src/cmds/tpm_sealdata.c
@@ -58,13 +58,13 @@ static int parse(const int aOpt, const c
 	switch (aOpt) {
 	case 'i':
 		if (aArg) {
-			strncpy(in_filename, aArg, PATH_MAX);
+			strncpy(in_filename, aArg, PATH_MAX - 1);
 			rc = 0;
 		}
 		break;
 	case 'o':
 		if (aArg) {
-			strncpy(out_filename, aArg, PATH_MAX);
+			strncpy(out_filename, aArg, PATH_MAX - 1);
 			rc = 0;
 		}
 		break;
--- a/src/cmds/tpm_unsealdata.c
+++ b/src/cmds/tpm_unsealdata.c
@@ -47,13 +47,13 @@ static int parse(const int aOpt, const c
 	switch (aOpt) {
 	case 'i':
 		if (aArg) {
-			strncpy(in_filename, aArg, PATH_MAX);
+			strncpy(in_filename, aArg, PATH_MAX - 1);
 			rc = 0;
 		}
 		break;
 	case 'o':
 		if (aArg) {
-			strncpy(out_filename, aArg, PATH_MAX);
+			strncpy(out_filename, aArg, PATH_MAX - 1);
 			rc = 0;
 		}
 		break;
