Index: trousers-0.3.14/dist/tcsd.conf.in
===================================================================
--- trousers-0.3.14.orig/dist/tcsd.conf.in
+++ trousers-0.3.14/dist/tcsd.conf.in
@@ -26,7 +26,7 @@
 # Values: Any absolute directory path
 # Description: Path where the tcsd creates its persistent storage file.
 #
-# system_ps_file = @localstatedir@/lib/tpm/system.data
+system_ps_file = /boot/system/tpm/system.data
 #
 
 # Option: firmware_log_file
@@ -187,5 +187,4 @@
 # its TCP port. Value of 1 disables IPv6 support, so clients cannot reach
 # TCSD using that protocol.
 #
-#  disable_ipv6 = 0
-#
+disable_ipv6 = 1
