diff --git a/vnc.c b/vnc.c
index 702ee5d..ed576b3 100644
--- a/vnc.c
+++ b/vnc.c
@@ -1194,6 +1194,10 @@ static void pointer_event(VncState *vs, int button_mask, int x, int y)
     if (button_mask & 0x10)
 	dz = 1;
 
+    kbd_mouse_event_absolute(x * 0x7FFF / (ds_get_width(vs->ds) - 1),
+            y * 0x7FFF / (ds_get_height(vs->ds) - 1),
+            dz, buttons);
+#if 0
     if (vs->absolute) {
         kbd_mouse_event(x * 0x7FFF / (ds_get_width(vs->ds) - 1),
                         y * 0x7FFF / (ds_get_height(vs->ds) - 1),
@@ -1211,6 +1215,7 @@ static void pointer_event(VncState *vs, int button_mask, int x, int y)
 	vs->last_x = x;
 	vs->last_y = y;
     }
+#endif
 
     check_pointer_type_change(vs, kbd_mouse_is_absolute());
 }
