diff --git a/dbus_stubs.c b/dbus_stubs.c
index 8500f38..c4d0745 100644
--- a/dbus_stubs.c
+++ b/dbus_stubs.c
@@ -918,8 +918,13 @@ value stub_dbus_message_get_error_name(value message)
 	if (!errname)
 		error_name = Val_none;
 	else {
+            
 		int index = find_index_string(errname, __error_table);
-		caml_alloc_some(error_name, Val_int(index));
+                if (index >= 0) {
+                    caml_alloc_some(error_name, Val_int(index));
+                } else {
+                    error_name = Val_none;
+                }
 	}
 	CAMLreturn(error_name);
 }
