Index: LVM2.2.02.85/configure.in
===================================================================
--- LVM2.2.02.85.orig/configure.in	2013-11-04 17:31:11.253438608 -0500
+++ LVM2.2.02.85/configure.in	2013-11-04 17:31:27.917432542 -0500
@@ -993,7 +993,7 @@
 		AC_CHECK_HEADERS([selinux/selinux.h],, hard_bailout)
 		AC_CHECK_HEADERS([selinux/label.h])
 		AC_DEFINE([HAVE_SELINUX], 1, [Define to 1 to include support for selinux.])
-		SELINUX_LIBS="-lselinux $SELINUX_LIBS"
+		SELINUX_LIBS="-lselinux -lpcre -pthread $SELINUX_LIBS"
 		SELINUX_PC="libselinux"
 		HAVE_SELINUX=yes ], [
 		AC_MSG_WARN(Disabling selinux)
