diff --git a/xen-config-host.h b/xen-config-host.h
index f50c3aa..384c295 100644
--- a/xen-config-host.h
+++ b/xen-config-host.h
@@ -48,4 +48,8 @@ extern unsigned int xen_logdirty_enable;
 #undef HAVE_IOVEC
 #endif
 
+#ifndef CONFIG_STUBDOM
+#define CONFIG_AUDIO 1
+#endif
+
 #endif /*XEN_CONFIG_HOST_H*/
diff --git a/xen-hooks.mak b/xen-hooks.mak
index 7e6b403..4a2e199 100644
--- a/xen-hooks.mak
+++ b/xen-hooks.mak
@@ -46,6 +46,7 @@ CONFIG_SDL=
 CONFIG_AUDIO=
 OBJS += xenfbfront.o
 else
+CONFIG_AUDIO=yes
 CPPFLAGS+= -I$(XEN_ROOT)/tools/blktap/lib
 LIBS += -L$(XEN_ROOT)/tools/blktap/lib -lblktap
 OBJS += xen_blktap.o
diff --git a/xen-setup b/xen-setup
index a527e96..431d435 100755
--- a/xen-setup
+++ b/xen-setup
@@ -18,7 +18,7 @@ if test -z "${XEN_SCRIPT_DIR}"; then
 	XEN_SCRIPT_DIR="/etc/xen/scripts"
 fi
 
-${QEMU_ROOT:-.}/configure --disable-gfx-check --disable-curses --disable-slirp --disable-vnc-tls --disable-sdl --disable-aio "$@" --prefix=/usr --cc=gcc
+${QEMU_ROOT:-.}/configure --disable-gfx-check --disable-curses --disable-slirp --disable-vnc-tls --disable-sdl --disable-aio "$@" --prefix=/usr --cc=gcc --audio-drv-list=alsa
 
 if [ "x$XEN_ROOT" != x ]; then
 	echo "XEN_ROOT=$XEN_ROOT" >>config-host.mak
