diff --git a/hw/piix4acpi.c b/hw/piix4acpi.c
index 0145826..c03104a 100644
--- a/hw/piix4acpi.c
+++ b/hw/piix4acpi.c
@@ -719,7 +719,7 @@ static void gpe_acpi_init(void)
     register_ioport_write(PROC_BASE, 32, 1, gpe_cpus_writeb, s);
 
     s->gpe0_blk_address = ACPI_GPE0_BLK_ADDRESS_V1;
-    s->gpe0_blk_half_len = ACPI_GPE0_BLK_LEN_V1 / 2;
+    s->gpe0_blk_half_len = ACPI_GPE0_BLK_LEN_V0 / 2;
 
     register_ioport_read(s->gpe0_blk_address,
                          s->gpe0_blk_half_len,
