diff --git a/hw/xenfb.c b/hw/xenfb.c
index 05c51cc..b5d1163 100644
--- a/hw/xenfb.c
+++ b/hw/xenfb.c
@@ -328,12 +328,14 @@ static void xenfb_mouse_event(void *opaque,
     int dh = ds_get_height(xenfb->c.ds);
     int i;
 
+#if 0
     if (xenfb->abs_pointer_wanted)
 	xenfb_send_position(xenfb,
 			    dx * (dw - 1) / 0x7fff,
 			    dy * (dh - 1) / 0x7fff,
 			    dz);
     else
+#endif
 	xenfb_send_motion(xenfb, dx, dy, dz);
 
     for (i = 0 ; i < 8 ; i++) {
