diff --git a/hw/pass-through.c b/hw/pass-through.c
index 48a5466..3b62ce1 100644
--- a/hw/pass-through.c
+++ b/hw/pass-through.c
@@ -4092,11 +4092,6 @@ static struct pt_dev * register_real_device(PCIBus *e_bus,
     machine_bdf.bus = r_bus;
     machine_bdf.dev = r_dev;
     machine_bdf.func = r_func;
-    rc = xc_assign_device(xc_handle, domid, machine_bdf.value);
-    if ( rc < 0 ) {
-        PT_LOG("Error: xc_assign_device error %d\n", rc);
-        return NULL;
-    }
 
     /* Initialize virtualized PCI configuration (Extended 256 Bytes) */
     for ( i = 0; i < PCI_CONFIG_SIZE; i++ )
