diff --git a/hw/pass-through.c b/hw/pass-through.c
index 974e973..b3696da 100644
--- a/hw/pass-through.c
+++ b/hw/pass-through.c
@@ -1370,6 +1370,13 @@ static void pt_ioport_map(PCIDevice *d, int i,
         (uint16_t)e_phys, (uint16_t)assigned_device->bases[i].access.pio_base,
         (uint16_t)e_size, i, first_map);
 
+    if ((e_phys & 0xffff) == 0xffff) {
+        PT_LOG("e_phys=%04x pio_base=%04x len=%d index=%d first_map=%d\n",
+            (uint16_t)e_phys, (uint16_t)assigned_device->bases[i].access.pio_base,
+            (uint16_t)e_size, i, first_map);
+        return;
+    }
+
     if ( e_size == 0 )
         return;
 
