diff --git a/switcher.c b/switcher.c
index a145412..3130aaa 100644
--- a/switcher.c
+++ b/switcher.c
@@ -92,6 +92,16 @@ static void switcher_key_inject (int code, uint32_t keycode)
 
     switch (keycode)
     {
+        /* XCP-367 */
+        case KEY_MUHENKAN:
+            keycode = 0x7b;
+            break;
+        case KEY_HENKAN:
+            keycode = 0x79;
+            break;
+        case KEY_KATAKANAHIRAGANA:
+            keycode = 0x70;
+            break;
         case KEY_YEN:
             keycode = 0x7d;
             break;
