diff --git a/switcher.c b/switcher.c
index 7c0de44..b8ee278 100644
--- a/switcher.c
+++ b/switcher.c
@@ -53,6 +53,8 @@ extern int gfx_passthru;
 #define DOM0_INPUT_EVENT          'E'
 #define DOM0_INPUT_PVM            'P'
 
+#define DOM0_INPUT_ABS            'A'
+
 #define DOM0_INPUT_SOCKET       "/tmp/input.socket"
 
 int                      switcher_connected = 0;
@@ -202,6 +204,7 @@ static void switcher_key_inject (int code, uint32_t keycode)
     }
 }
 
+
 static void switcher_event(struct input_event *e)
 {
     static int mouse_button_state = 0;
@@ -321,6 +324,11 @@ void switcher_send_led_code(int led_code)
     switcher_send("%c%d", DOM0_INPUT_LEDS, led_code);
 }
 
+void switcher_abs_enabled(int enabled)
+{
+    switcher_send("%c%d", DOM0_INPUT_ABS, enabled);
+}
+
 void switcher_connect(void)
 {
     struct stat st;
