Index: openxt-20170425/build/tmp-glibc/work/core2-32-oe-linux/xenmgr/0+gitAUTOINC+8ac754b951-r0/git/xenmgr/Vm/Config.hs
===================================================================
--- xenmgr/Vm/Config.hs
+++ xenmgr/Vm/Config.hs
@@ -770,7 +770,7 @@ miscSpecs cfg = do
       timer_mode = readConfigPropertyDef uuid vmTimerMode vmTimerModeDefault >>= 
                    \ v -> return ["timer_mode=" ++ (show v)]
       nested = readConfigPropertyDef uuid vmNestedHvm False >>=
-                   \ v -> if v then return ["nested=true"] else return []
+                   \ v -> if v then return ["nestedhvm=1"] else return []
 
       acpi_table = do
           case (vmcfgAcpi cfg) of
