--- a/Makefile	2018-04-25 15:44:51.431420129 -0400
+++ b/Makefile	2018-04-25 15:36:35.327601950 -0400
@@ -12,8 +12,7 @@
 #   (at your option) any later version.
 #
 
-CC      = gcc
-CFLAGS  = -W -Wall -Wshadow -Wstrict-prototypes -Wpointer-arith -Wcast-qual \
+CFLAGS += -W -Wall -Wshadow -Wstrict-prototypes -Wpointer-arith -Wcast-qual \
           -Wcast-align -Wwrite-strings -Wmissing-prototypes -Winline -Wundef
 
 # Let lseek and mmap support 64-bit wide offsets
