From caa6264d00c3e9ee5f31467ced6c69c087ad8399 Mon Sep 17 00:00:00 2001
From: Eric Chanudet <chanudete@ainfosec.com>
Date: Mon, 13 Aug 2018 13:54:04 -0400
Subject: [PATCH 17/22] tboot: Use SHA256 by default with TPM2.0.

Signed-off-by: Chris Rogers <rogersc@ainfosec.com>
---
 tboot/common/cmdline.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/tboot/common/cmdline.c b/tboot/common/cmdline.c
index 8644d01..337fdbd 100644
--- a/tboot/common/cmdline.c
+++ b/tboot/common/cmdline.c
@@ -527,6 +527,11 @@ void get_tboot_extpol(void)
         tpm->extpol = TB_EXTPOL_FIXED;
         tpm->cur_alg = TB_HALG_SM3;
     }
+
+    if ( tpm->major == TPM20_VER_MAJOR ) {
+        tpm->extpol = TB_EXTPOL_FIXED;
+        tpm->cur_alg = TB_HALG_SHA256;
+    }
 }
 
 bool get_tboot_ignore_prev_err(void)
-- 
2.16.1

