diff --git a/dbus_stubs.c b/dbus_stubs.c
index c4d0745..9cacb77 100644
--- a/dbus_stubs.c
+++ b/dbus_stubs.c
@@ -1042,7 +1042,7 @@ static void message_append_basic(DBusMessageIter *iter, int c_type, value v)
 		}
 	case DBUS_TYPE_OBJECT_PATH:
 	case DBUS_TYPE_STRING: {
-		char *s = strdup(String_val(v));
+		char *s = String_val(v);
 		dbus_message_iter_append_basic(iter, c_type, &s);
 		break;
 		}
