diff -uNr console-tools-0.3.2.orig/kbdtools/loadkeys.y console-tools-0.3.2/kbdtools/loadkeys.y
--- console-tools-0.3.2.orig/kbdtools/loadkeys.y	1999-08-25 22:13:53.000000000 +0100
+++ console-tools-0.3.2/kbdtools/loadkeys.y	2011-10-06 18:21:31.618625324 +0100
@@ -292,6 +292,8 @@
 			{$$=($1 ^ 0xf000); unicode_used=1;}
                 | PLUS NUMBER
                         {$$=K(KT_LETTER, KVAL($2));}
+                | PLUS UNUMBER
+                        {$$=K(KT_LETTER, KVAL($2 ^ 0xf000)); unicode_used=1;}
 		| LITERAL
 			{$$=$1;}
                 | PLUS LITERAL
