Index: openssh-7.1p2/Makefile.in
===================================================================
--- openssh-7.1p2.orig/Makefile.in	2016-01-14 02:10:45.000000000 +0100
+++ openssh-7.1p2/Makefile.in	2016-07-21 10:53:03.361259750 +0200
@@ -290,10 +290,10 @@
 install-nokeys: $(CONFIGFILES) $(MANPAGES) $(TARGETS) install-files install-sysconf
 install-nosysconf: $(CONFIGFILES) $(MANPAGES) $(TARGETS) install-files
 
-check-config:
+check-config: install-files
 	-$(DESTDIR)$(sbindir)/sshd -t -f $(DESTDIR)$(sysconfdir)/sshd_config
 
-install-files:
+install-files: $(TARGETS) $(MANPAGES)
 	$(srcdir)/mkinstalldirs $(DESTDIR)$(bindir)
 	$(srcdir)/mkinstalldirs $(DESTDIR)$(sbindir)
 	$(srcdir)/mkinstalldirs $(DESTDIR)$(mandir)
@@ -332,7 +332,7 @@
 	-rm -f $(DESTDIR)$(mandir)/$(mansubdir)1/slogin.1
 	ln -s ./ssh.1 $(DESTDIR)$(mandir)/$(mansubdir)1/slogin.1
 
-install-sysconf:
+install-sysconf: $(CONFIGFILES)
 	if [ ! -d $(DESTDIR)$(sysconfdir) ]; then \
 		$(srcdir)/mkinstalldirs $(DESTDIR)$(sysconfdir); \
 	fi
