Index: openssh-7.1p2/Makefile.in
===================================================================
--- openssh-7.1p2.orig/Makefile.in	2016-07-21 10:53:03.361259750 +0200
+++ openssh-7.1p2/Makefile.in	2016-07-21 10:53:53.187444832 +0200
@@ -28,7 +28,7 @@
 SSH_PKCS11_HELPER=$(libexecdir)/ssh-pkcs11-helper
 PRIVSEP_PATH=@PRIVSEP_PATH@
 SSH_PRIVSEP_USER=@SSH_PRIVSEP_USER@
-STRIP_OPT=@STRIP_OPT@
+STRIP_OPT=
 TEST_SHELL=@TEST_SHELL@
 
 PATHS= -DSSHDIR=\"$(sysconfdir)\" \
