commit 1f9b6ef5ff38c53785ff9cf1287be22e3bded469
Author: Olivier Fourdan <fourdan@xfce.org>
Date:   Tue Apr 10 21:41:39 2012 +0200

    Resize fullscreen windows on screen size changes (bug #8622)

diff --git a/src/client.c b/src/client.c
index fb3665c..cea2de3 100644
--- a/src/client.c
+++ b/src/client.c
@@ -3553,6 +3553,10 @@ clientScreenResize(ScreenInfo *screen_info, gboolean fully_visible)
             wc.height = c->height;
             clientConfigure (c, &wc, CWX | CWY | CWWidth | CWHeight, CFG_NOTIFY);
         }
+        else if (FLAG_TEST (c->flags, CLIENT_FLAG_FULLSCREEN))
+        {
+            clientUpdateFullscreenSize (c);
+        }
         else
         {
             configure_flags = CFG_CONSTRAINED | CFG_REQUEST;

