diff -ur NetworkManager-0.9.2.0.orig/src/nm-policy.c NetworkManager-0.9.2.0/src/nm-policy.c
--- NetworkManager-0.9.2.0.orig/src/nm-policy.c	2013-04-02 17:41:57.000000000 +0000
+++ NetworkManager-0.9.2.0/src/nm-policy.c	2013-04-02 17:49:22.000000000 +0000
@@ -80,6 +80,8 @@
 #define RESET_RETRIES_TIMER 300
 #define FAILURE_REASON_TAG "failure-reason"
 
+static void schedule_activate_all (NMPolicy *policy);
+
 static gboolean 
 device_suitable_for_default_route (NMDevice *dev)
 {
@@ -972,6 +974,7 @@
 	NMPolicy *policy = (NMPolicy *) user_data;
 	GSList *connections, *iter;
 	time_t con_stamp, min_stamp, now;
+	gboolean changed = FALSE;
 
 	policy->reset_retries_id = 0;
 
@@ -984,6 +987,7 @@
 		if (con_stamp + RESET_RETRIES_TIMER <= now) {
 			set_connection_auto_retries (NM_CONNECTION (iter->data), RETRIES_DEFAULT);
 			g_object_set_data (G_OBJECT (iter->data), RESET_RETRIES_TIMESTAMP_TAG, GSIZE_TO_POINTER (0));
+			changed = TRUE;
 			continue;
 		}
 		if (con_stamp < min_stamp)
@@ -994,6 +998,11 @@
 	/* Schedule the handler again if there are some stamps left */
 	if (min_stamp != now)
 		policy->reset_retries_id = g_timeout_add_seconds (RESET_RETRIES_TIMER - (now - min_stamp), reset_connections_retries, policy);
+
+	/* If anything changed, try to activate the newly re-enabled connections */
+	if (changed)
+		schedule_activate_all (policy);
+
 	return FALSE;
 }
 
