diff -ur NetworkManager-0.9.2.0.orig/src/settings/plugins/keyfile/utils.c NetworkManager-0.9.2.0/src/settings/plugins/keyfile/utils.c
--- NetworkManager-0.9.2.0.orig/src/settings/plugins/keyfile/utils.c	2011-11-09 18:20:18.000000000 +0000
+++ NetworkManager-0.9.2.0/src/settings/plugins/keyfile/utils.c	2012-08-03 16:23:12.000000000 +0000
@@ -21,6 +21,8 @@
 #include <glib.h>
 #include <stdlib.h>
 #include <string.h>
+#include <errno.h>
+#include <sys/wait.h>
 #include "utils.h"
 
 
@@ -103,3 +105,23 @@
 	return ignore;
 }
 
+void
+xc_connections_update (char *path)
+{
+	char *argv[6] = { "/usr/bin/nm_sync.sh", "-n", "nm-connections", "-w", 0, 0 };
+	int err, stat_loc;
+
+	if ((err = fork()) == -1) {
+		g_warning("%s.%d: cannot fork to nm_sync.sh: %d", __FILE__, __LINE__, errno);
+
+	} else if (err) {
+		while ((waitpid(err, &stat_loc, 0) == -1) && (errno == EINTR))
+			;
+		if ((err = WEXITSTATUS(stat_loc))) {
+			g_warning("%s.%d: nm_sync.sh failed with code: %d", __FILE__, __LINE__, err);
+		}
+	} else {
+		argv[4] = path;
+		_exit(execvp(argv[0], argv));
+	}
+}

diff -ur NetworkManager-0.9.2.0.orig/src/settings/plugins/keyfile/utils.h NetworkManager-0.9.2.0/src/settings/plugins/keyfile/utils.h
--- NetworkManager-0.9.2.0.orig/src/settings/plugins/keyfile/utils.h	2011-11-09 18:20:18.000000000 +0000
+++ NetworkManager-0.9.2.0/src/settings/plugins/keyfile/utils.h	2012-08-03 16:22:58.000000000 +0000
@@ -25,6 +25,7 @@
 #include "common.h"
 
 gboolean nm_keyfile_plugin_utils_should_ignore_file (const char *filename);
+void        xc_connections_update (char *path);
 
 #endif  /* _UTILS_H_ */
 
diff -ur NetworkManager-0.9.2.0.orig/src/settings/plugins/keyfile//nm-keyfile-connection.c NetworkManager-0.9.2.0/src/settings/plugins/keyfile//nm-keyfile-connection.c
--- NetworkManager-0.9.2.0.orig/src/settings/plugins/keyfile//nm-keyfile-connection.c	2012-06-28 17:37:50.000000000 +0000
+++ NetworkManager-0.9.2.0/src/settings/plugins/keyfile//nm-keyfile-connection.c	2012-07-20 20:33:53.000000000 +0000
@@ -137,6 +139,7 @@
 	NM_SETTINGS_CONNECTION_CLASS (nm_keyfile_connection_parent_class)->delete (connection,
 	                                                                           callback,
 	                                                                           user_data);
+       xc_connections_update (priv->path); 
 }
 
 /* GObject */
diff -ur NetworkManager-0.9.2.0.orig/src/settings/plugins/keyfile//writer.c NetworkManager-0.9.2.0/src/settings/plugins/keyfile//writer.c
--- NetworkManager-0.9.2.0.orig/src/settings/plugins/keyfile//writer.c	2012-06-28 17:37:50.000000000 +0000
+++ NetworkManager-0.9.2.0/src/settings/plugins/keyfile//writer.c	2012-07-20 20:24:40.000000000 +0000
@@ -1049,6 +1049,7 @@
 			success = TRUE;
 		}
 	}
+       xc_connections_update (filename); 
 	g_free (path);
 
 out:
diff -ur NetworkManager-0.9.2.0.orig/src/nm-manager.c NetworkManager-0.9.2.0/src/nm-manager.c
--- NetworkManager-0.9.2.0.orig/src/nm-manager.c	2012-06-28 17:37:51.000000000 +0000
+++ NetworkManager-0.9.2.0/src/nm-manager.c	2012-08-06 14:38:37.000000000 +0000
@@ -1023,6 +1023,7 @@
 	data = g_key_file_to_data (key_file, &len, NULL);
 	if (data) {
 		ret = g_file_set_contents (filename, data, len, error);
+	        nm_log_dbg (LOGD_SETTINGS, "(%s): update state file", filename);
 		g_free (data);
 	}
 	g_key_file_free (key_file);

