diff -uNr NetworkManager-0.9.2.0.orig/src/nm-device-wifi.c NetworkManager-0.9.2.0/src/nm-device-wifi.c
--- NetworkManager-0.9.2.0.orig/src/nm-device-wifi.c	2011-11-09 17:22:30.000000000 +0000
+++ NetworkManager-0.9.2.0/src/nm-device-wifi.c	2013-03-25 19:21:42.926192369 +0000
@@ -2772,10 +2772,7 @@
 	if (setting_name) {
 		NMSettingsGetSecretsFlags flags = NM_SETTINGS_GET_SECRETS_FLAG_ALLOW_INTERACTION;
 
-		/* If the caller doesn't necessarily want completely new secrets,
-		 * only ask for new secrets after the first failure.
-		 */
-		if (new_secrets || tries)
+		if (new_secrets)
 			flags |= NM_SETTINGS_GET_SECRETS_FLAG_REQUEST_NEW;
 		nm_act_request_get_secrets (req, setting_name, flags, NULL, wifi_secrets_cb, self);
 
@@ -2860,6 +2857,9 @@
 	}
 
 	if (is_encrypted (ap, connection)) {
+		guint64 timestamp = 0;
+		gboolean new_secrets = TRUE;
+
 		/* Connection failed; either driver problems, the encryption key is
 		 * wrong, or the passwords or certificates were wrong.
 		 */
@@ -2867,7 +2867,21 @@
 		             "Activation (%s/wireless): association took too long.",
 		             nm_device_get_iface (dev));
 
-		if (handle_auth_or_fail (self, req, TRUE) == NM_ACT_STAGE_RETURN_POSTPONE) {
+		/* Ask for new secrets only if we've never activated this connection
+		 * before.  If we've connected before, don't bother the user with
+		 * dialogs, just retry or fail, and if we never connect the user can
+		 * fix the password somewhere else.
+		 */
+		timestamp = nm_settings_connection_get_timestamp (NM_SETTINGS_CONNECTION (connection));
+
+		if (timestamp) {
+			new_secrets = FALSE;
+			nm_log_warn (LOGD_DEVICE | LOGD_WIFI,
+			             "Activation (%s/wireless): previously connected (timestamp %llu)",
+			             nm_device_get_iface (dev));
+		}
+
+		if (handle_auth_or_fail (self, req, new_secrets) == NM_ACT_STAGE_RETURN_POSTPONE) {
 			nm_log_warn (LOGD_DEVICE | LOGD_WIFI,
 			             "Activation (%s/wireless): asking for new secrets",
 			             nm_device_get_iface (dev));
