Index: NetworkManager-0.9.2.0/src/nm-netlink-monitor.c
===================================================================
--- NetworkManager-0.9.2.0.orig/src/nm-netlink-monitor.c
+++ NetworkManager-0.9.2.0/src/nm-netlink-monitor.c
@@ -248,7 +248,7 @@ nlh_setup (struct nl_sock *nlh,
 {
 	int err;
 
-	nl_socket_modify_cb (nlh, NL_CB_MSG_IN, NL_CB_CUSTOM, event_msg_recv, cb_data);
+	nl_socket_modify_cb (nlh, NL_CB_MSG_IN, NL_CB_CUSTOM, event_msg_recv, nlh);
 
 	if (valid_func)
 		nl_socket_modify_cb (nlh, NL_CB_VALID, NL_CB_CUSTOM, valid_func, cb_data);
