From e4659d5bc55ca8b0a9202f0af17ecd70749d150b Mon Sep 17 00:00:00 2001
From: Gerd Hoffmann <kraxel@redhat.com>
Date: Tue, 20 Aug 2013 14:00:55 +0200
Subject: [PATCH] OvmfPkg: don't lock lock umb when running csm

---
 IntelFrameworkModulePkg/Csm/LegacyBiosDxe/LegacyBootSupport.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/IntelFrameworkModulePkg/Csm/LegacyBiosDxe/LegacyBootSupport.c b/IntelFrameworkModulePkg/Csm/LegacyBiosDxe/LegacyBootSupport.c
index 6c2688b..a391a05 100644
--- a/IntelFrameworkModulePkg/Csm/LegacyBiosDxe/LegacyBootSupport.c
+++ b/IntelFrameworkModulePkg/Csm/LegacyBiosDxe/LegacyBootSupport.c
@@ -1234,8 +1234,8 @@ GenericLegacyBoot (
   //
   Private->LegacyRegion->Lock (
                            Private->LegacyRegion,
-                           0xc0000,
-                           0x40000,
+                           0xe0000,
+                           0x10000,
                            &Granularity
                            );
 
-- 
1.8.3.1

