--- a/scripts/Makefile.in
+++ b/scripts/Makefile.in
@@ -74,6 +74,16 @@ ifeq ("@BLKDEACTIVATE@", "yes")
 	$(Q) $(INSTALL_SCRIPT) blk_availability_init_red_hat $(initdir)/blk-availability
 endif
 
+# FIXME Some are not ported yet.
+install_initscripts_yocto:
+	$(INSTALL_DIR) $(initdir)
+ifeq ("@BUILD_DMEVENTD@", "yes")
+	$(INSTALL_SCRIPT) lvm2_monitoring_init_yocto $(initdir)/lvm2-monitor
+endif
+ifeq ("@BUILD_LVMPOLLD@", "yes")
+	$(INSTALL_SCRIPT) lvm2_lvmpolld_init_yocto $(initdir)/lvm2-lvmpolld
+endif
+
 CFLAGS_lvm2_activation_generator_systemd_red_hat.o += $(EXTRA_EXEC_CFLAGS)
 
 lvm2_activation_generator_systemd_red_hat: $(OBJECTS) $(LVMINTERNAL_LIBS)
--- /dev/null
+++ b/scripts/lvm2_lvmpolld_init_yocto.in
@@ -0,0 +1,84 @@
+#!/bin/bash
+#
+# Copyright (C) 2017 Assured Information Security, Inc. All rights reserved.
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions
+# of the GNU General Public License v.2.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software Foundation,
+# Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+#
+# This file is part of LVM2.
+# It is required for the proper handling of failures of LVM2 mirror
+# devices that were created using the -m option of lvcreate.
+#
+#
+# chkconfig: 12345 02 99
+# description: Starts and stops LVM poll daemon
+#
+### BEGIN INIT INFO
+# Provides: lvm2-lvmpolld
+# Required-Start: $local_fs
+# Required-Stop: $local_fs
+# Default-Start: 1 2 3 4 5
+# Default-Stop: 0 6
+# Short-Description: A daemon that is responsible for monitoring in-progress
+#		     and possibly longer term operations on logical volumes.
+#		     It helps to reduce the number of spawned processes if same
+#                    logical volume is requested to get monitored multiple times.
+#                    Also avoids unsolicited termination due to external factors.
+### END INIT INFO
+
+. /etc/init.d/functions
+
+DAEMON=lvmpolld
+PID_FILE="@LVMPOLLD_PIDFILE@"
+
+usage()
+{
+    echo "Usage: /etc/init.d/lvm2-lvmpolld {start|stop|status}"
+}
+
+do_status()
+{
+    if start-stop-daemon -K --test --quiet --pidfile $PID_FILE ; then
+        pid=`cat $PID_FILE`
+        echo "$DAEMON (pid $pid) is running."
+    elif [ -e "$PID_FILE" ]; then
+        pid=`cat $PID_FILE`
+        echo "$DAEMON died (pid was $pid)."
+    else
+        echo "$DAEMON is stopped."
+    fi
+}
+
+if [ $# != 1 ]; then
+    usage
+    exit 1
+fi
+
+rc=0
+case "$1" in
+    start)
+        start-stop-daemon -S --oknodo --pidfile $PID_FILE --exec $DAEMON
+        rc=$?
+        ;;
+    stop|force-stop)
+        start-stop-daemon -K --oknodo --pidfile $PID_FILE
+        rm -f $PID_FILE
+        ;;
+    status)
+        do_status
+        exit 0
+        ;;
+    *)
+        usage
+        exit 1
+        ;;
+esac
+
+do_status
+
+exit $rc
--- /dev/null
+++ b/scripts/lvm2_monitoring_init_yocto.in
@@ -0,0 +1,82 @@
+#!/bin/bash
+#
+# Copyright (C) 2017 Assured Information Security, Inc. All rights reserved.
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions
+# of the GNU General Public License v.2.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software Foundation,
+# Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+#
+# This file is part of LVM2.
+# It is required for the proper handling of failures of LVM2 mirror
+# devices that were created using the -m option of lvcreate.
+#
+#
+# chkconfig: 12345 02 99
+# description: Starts and stops dmeventd monitoring for lvm2
+#
+### BEGIN INIT INFO
+# Provides: lvm2-monitor
+# Required-Start: $local_fs
+# Required-Stop: $local_fs
+# Default-Start: 1 2 3 4 5
+# Default-Stop: 0 6
+# Short-Description: A daemon to monitor device mapper events. Library plugins
+#                    can register and carry out actions triggered when
+#                    particular events occur.
+### END INIT INFO
+
+. /etc/init.d/functions
+
+DAEMON=dmeventd
+PID_FILE="@DMEVENTD_PIDFILE@"
+
+usage()
+{
+    echo "Usage: /etc/init.d/lvm2-monitor {start|stop|status}"
+}
+
+do_status()
+{
+    if start-stop-daemon -K --test --quiet --pidfile $PID_FILE ; then
+        pid=`cat $PID_FILE`
+        echo "$DAEMON (pid $pid) is running."
+    elif [ -e "$PID_FILE" ]; then
+        pid=`cat $PID_FILE`
+        echo "$DAEMON died (pid was $pid)."
+    else
+        echo "$DAEMON is stopped."
+    fi
+}
+
+if [ $# != 1 ]; then
+    usage
+    exit 1
+fi
+
+rc=0
+case "$1" in
+    start)
+        start-stop-daemon -S --oknodo --pidfile $PID_FILE --exec $DAEMON
+        rc=$?
+        ;;
+    stop|force-stop)
+        start-stop-daemon -K --oknodo --pidfile $PID_FILE
+        rm -f $PID_FILE
+        ;;
+    status)
+        do_status
+        exit 0
+        ;;
+    *)
+        usage
+        exit 1
+        ;;
+esac
+
+do_status
+
+exit $rc
--- a/Makefile.in
+++ b/Makefile.in
@@ -141,6 +141,9 @@ install_system_dirs:
 install_initscripts:
 	$(MAKE) -C scripts install_initscripts
 
+install_initscripts_yocto:
+	$(MAKE) -C scripts install_initscripts_yocto
+
 install_systemd_generators:
 	$(MAKE) -C scripts install_systemd_generators
 	$(MAKE) -C man install_systemd_generators
--- a/configure.ac
+++ b/configure.ac
@@ -1822,11 +1822,13 @@ scripts/dm_event_systemd_red_hat.socket
 scripts/lvm2_cmirrord_systemd_red_hat.service
 scripts/lvm2_lvmdbusd_systemd_red_hat.service
 scripts/lvm2_lvmpolld_init_red_hat
+scripts/lvm2_lvmpolld_init_yocto
 scripts/lvm2_lvmpolld_systemd_red_hat.service
 scripts/lvm2_lvmpolld_systemd_red_hat.socket
 scripts/lvmlockd.service
 scripts/lvmlocks.service
 scripts/lvm2_monitoring_init_red_hat
+scripts/lvm2_monitoring_init_yocto
 scripts/lvm2_monitoring_systemd_red_hat.service
 scripts/lvm2_tmpfiles_red_hat.conf
 scripts/lvmdump.sh
