diff --git a/include/libbb.h b/include/libbb.h
index 27fb50a..a8eb7ff 100644
--- a/include/libbb.h
+++ b/include/libbb.h
@@ -52,6 +52,13 @@
 #include <sys/statfs.h>
 #endif
 
+/* Don't do this here:
+ * #include <sys/sysinfo.h>
+ * Some linux/ includes pull in conflicting definition
+ * of struct sysinfo (only in some toolchanins), which breaks build.
+ * Include sys/sysinfo.h only in those files which need it.
+ */
+
 #if ENABLE_SELINUX
 #include <selinux/selinux.h>
 #include <selinux/context.h>
@@ -93,25 +100,6 @@ int klogctl(int type, char *b, int len);
  * confusing the two versions of basename.  See the dirname/basename man page
  * for details. */
 char *dirname(char *path);
-/* Include our own copy of struct sysinfo to avoid binary compatibility
- * problems with Linux 2.4, which changed things.  Grumble, grumble. */
-struct sysinfo {
-	long uptime;			/* Seconds since boot */
-	unsigned long loads[3];		/* 1, 5, and 15 minute load averages */
-	unsigned long totalram;		/* Total usable main memory size */
-	unsigned long freeram;		/* Available memory size */
-	unsigned long sharedram;	/* Amount of shared memory */
-	unsigned long bufferram;	/* Memory used by buffers */
-	unsigned long totalswap;	/* Total swap space size */
-	unsigned long freeswap;		/* swap space still available */
-	unsigned short procs;		/* Number of current processes */
-	unsigned short pad;			/* Padding needed for m68k */
-	unsigned long totalhigh;	/* Total high memory size */
-	unsigned long freehigh;		/* Available high memory size */
-	unsigned int mem_unit;		/* Memory unit size in bytes */
-	char _f[20 - 2*sizeof(long) - sizeof(int)]; /* Padding: libc5 uses this.. */
-};
-int sysinfo(struct sysinfo* info);
 
 
 /* Make all declarations hidden (-fvisibility flag only affects definitions) */
diff --git a/init/init.c b/init/init.c
index e399389..5e20011 100644
--- a/init/init.c
+++ b/init/init.c
@@ -13,6 +13,9 @@
 #include <syslog.h>
 #include <paths.h>
 #include <sys/reboot.h>
+#ifdef __linux__
+# include <sys/sysinfo.h>
+#endif
 
 #define COMMAND_SIZE 256
 #define CONSOLE_NAME_SIZE 32
diff --git a/procps/free.c b/procps/free.c
index e76dd21..b834320 100644
--- a/procps/free.c
+++ b/procps/free.c
@@ -10,6 +10,9 @@
 /* getopt not needed */
 
 #include "libbb.h"
+#ifdef __linux__
+# include <sys/sysinfo.h>
+#endif
 
 int free_main(int argc, char **argv) MAIN_EXTERNALLY_VISIBLE;
 int free_main(int argc, char **argv)
diff --git a/procps/ps.c b/procps/ps.c
index cfca851..e80e63d 100644
--- a/procps/ps.c
+++ b/procps/ps.c
@@ -16,6 +16,9 @@ enum { MAX_WIDTH = 2*1024 };
 
 #if ENABLE_DESKTOP
 
+#ifdef __linux__
+# include <sys/sysinfo.h>
+#endif
 #include <sys/times.h> /* for times() */
 //#include <sys/sysinfo.h> /* for sysinfo() */
 #ifndef AT_CLKTCK
diff --git a/procps/uptime.c b/procps/uptime.c
index b729055..607f633 100644
--- a/procps/uptime.c
+++ b/procps/uptime.c
@@ -16,6 +16,10 @@
 /* getopt not needed */
 
 #include "libbb.h"
+#ifdef __linux__
+# include <sys/sysinfo.h>
+#endif
+
 
 #ifndef FSHIFT
 # define FSHIFT 16              /* nr of bits of precision */
