Index: refpolicy/policy/modules/services/ctxusb.if
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ refpolicy/policy/modules/services/ctxusb.if	2015-07-27 12:56:05.741966817 -0400
@@ -0,0 +1,168 @@
+#############################################################################
+#
+# Copyright (C) 2014 Citrix Systems, Inc.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+#############################################################################
+
+## <summary>Policy for ctxusb.</summary>
+
+#######################################
+## <summary>
+##     Execute ctxusb client in the ctxusbc_t domain.
+## </summary>
+## <param name="domain">
+##     <summary>
+##     The type of the process forking ctxusb.
+##     </summary>
+## </param>
+#
+interface(`ctxusbc_exec_domtrans',`
+       gen_require(`
+               type ctxusbc_t, ctxusbc_exec_t;
+       ')
+       
+       domtrans_pattern($1, ctxusbc_exec_t, ctxusbc_t)
+')
+
+#######################################
+## <summary>
+##	Exchange messages with ctxusb over dbus.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	The type of the process chatting with ctxusb over dbus.
+##	</summary>
+## </param>
+#
+interface(`ctxusbd_dbus_chat',`
+	gen_require(`
+		type ctxusbd_t;
+		class dbus send_msg;
+	')
+	
+	allow $1 ctxusbd_t:dbus	send_msg;
+	allow ctxusbd_t $1:dbus	send_msg;
+')
+
+#######################################
+## <summary>
+##	Send messages to ctxusb over dbus.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	The type of the process sending messages.
+##	</summary>
+## </param>
+#
+interface(`ctxusbd_dbus_send',`
+	gen_require(`
+		type ctxusbd_t;
+		class dbus send_msg;
+	')
+	
+	allow $1 ctxusbd_t:dbus	send_msg;
+')
+
+#######################################
+## <summary>
+##    Read generic ctxusb files.
+## </summary>
+## <param name="domain">
+##      <summary>
+##      Domain allowd access.
+##	</summary>
+## </param>
+#
+interface(`ctxusbd_read_files',`
+	gen_require(`
+		type ctxusbd_t;
+	')
+	
+        allow $1 ctxusbd_t:file read_file_perms;
+')
+
+#######################################
+## <summary>
+##    Read generic ctxusb links.
+## </summary>
+## <param name="domain">
+##      <summary>
+##      Domain allowd access.
+##	</summary>
+## </param>
+#
+interface(`ctxusbd_read_links',`
+	gen_require(`
+		type ctxusbd_t;
+	')
+	
+        allow $1 ctxusbd_t:lnk_file read_lnk_file_perms;
+')
+
+#######################################
+## <summary>
+##    Search generic ctxusb directories.
+## </summary>
+## <param name="domain">
+##      <summary>
+##      Domain allowd access.
+##	</summary>
+## </param>
+#
+interface(`ctxusbd_search_dirs',`
+	gen_require(`
+		type ctxusbd_t;
+	')
+	
+        allow $1 ctxusbd_t:dir search_dir_perms;
+')
+
+#######################################
+## <summary>
+##    Send signals to ctxusb process.
+## </summary>
+## <param name="domain">
+##      <summary>
+##      Domain allowd access.
+##	</summary>
+## </param>
+#
+interface(`ctxusbd_signal',`
+	gen_require(`
+		type ctxusbd_t;
+	')
+	
+        allow $1 ctxusbd_t:process signal_perms;
+')
+
+#######################################
+## <summary>
+##    Read ctxusb configuration files.
+## </summary>
+## <param name="domain">
+##      <summary>
+##      Domain allowd access.
+##	</summary>
+## </param>
+#
+interface(`ctxusbd_read_etc_files',`
+	gen_require(`
+		type ctxusbd_etc_t;
+	')
+	
+        allow $1 ctxusbd_etc_t:file read_file_perms;
+')
