Index: refpolicy/policy/modules/services/ctxusb.te
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ refpolicy/policy/modules/services/ctxusb.te	2015-07-27 12:50:05.501964183 -0400
@@ -0,0 +1,101 @@
+#############################################################################
+#
+# Copyright (C) 2014 Citrix Systems, Inc.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+#############################################################################
+
+policy_module(ctxusb, 0.2)
+
+########################################
+#
+# Declarations
+#
+type ctxusbd_t;
+type ctxusbd_exec_t;
+init_daemon_domain(ctxusbd_t,ctxusbd_exec_t)
+
+type ctxusbc_t;
+type ctxusbc_exec_t;
+application_type(ctxusbc_t)
+domain_entry_file(ctxusbc_t,ctxusbc_exec_t)
+role system_r types ctxusbc_t;
+xen_domtrans(ctxusbc_t,ctxusbc_exec_t)
+
+type ctxusbd_etc_t;
+files_config_file(ctxusbd_etc_t)
+
+#######################################
+#
+# ctxusb daemon local policy
+#
+files_search_etc(ctxusbd_t)
+files_read_etc_files(ctxusbd_t)
+files_search_usr(ctxusbd_t)
+libs_use_ld_so(ctxusbd_t)
+dev_rw_usbfs(ctxusbd_t)
+xc_config_dir_search(ctxusbd_t)
+dbus_connect_system_bus(ctxusbd_t)
+dbus_system_bus_client(ctxusbd_t)
+dbd_dbus_chat(ctxusbd_t)
+xen_dbus_send(ctxusbd_t)
+input_server_dbus_chat(ctxusbd_t)
+qemu_sigchld(ctxusbd_t)
+dev_search_sysfs(ctxusbd_t)
+dev_rw_sysfs(ctxusbd_t)
+dev_rw_generic_usb_dev(ctxusbd_t)
+dev_read_rand(ctxusbd_t)
+logging_send_syslog_msg(ctxusbd_t)
+fs_list_inotifyfs(ctxusbd_t)
+miscfiles_read_localization(ctxusbd_t)
+kernel_search_debugfs(ctxusbd_t)
+kernel_read_system_state(ctxusbd_t)
+kernel_read_xen_state(ctxusbd_t)
+kernel_write_xen_state(ctxusbd_t)
+kernel_write_xen_state(ctxusbd_t)
+xen_stream_connect_xenstore(ctxusbd_t)
+
+corecmd_search_bin(ctxusbd_t)
+ctxusbc_exec_domtrans(ctxusbd_t)
+
+# why dac_override?
+allow ctxusbd_t self:capability dac_override;
+allow ctxusbd_t self:netlink_kobject_uevent_socket { create { ioctl read getattr write setattr append bind connect getopt setopt shutdown } };
+allow ctxusbd_t self:process { getsched signal };
+allow ctxusbd_t self:fifo_file manage_fifo_file_perms;
+
+files_etc_filetrans(ctxusbd_t, ctxusbd_etc_t, file)
+allow ctxusbd_t ctxusbd_etc_t:lnk_file read_lnk_file_perms;
+allow ctxusbd_t ctxusbd_etc_t:file manage_file_perms;
+
+#######################################
+#
+# vusb command line tool policy
+#
+dev_rw_sysfs(ctxusbc_t)
+dev_rw_generic_usb_dev(ctxusbc_t)
+fs_list_inotifyfs(ctxusbc_t)
+kernel_read_system_state(ctxusbc_t)
+logging_send_syslog_msg(ctxusbc_t)
+xen_dontaudit_rw_unix_stream_sockets(ctxusbc_t)
+xen_stream_connect_xenstore(ctxusbc_t)
+xen_write_tmp(ctxusbc_t)
+fs_rw_xenfs_files(ctxusbc_t)
+files_read_etc_files(ctxusbc_t)
+udev_read_db(ctxusbc_t)
+
+allow ctxusbc_t self:capability dac_override;
+allow ctxusbc_t self:process { getsched signal };
