Index: refpolicy/policy/modules/services/updatemgr.te
===================================================================
--- refpolicy.orig/policy/modules/services/updatemgr.te	2016-09-01 16:06:27.000000000 -0400
+++ refpolicy/policy/modules/services/updatemgr.te	2016-09-01 16:06:28.000000000 -0400
@@ -81,3 +81,5 @@
 allow updatemgr_t updatemgr_tmp_t:file manage_file_perms;
 allow updatemgr_t self:capability { dac_override chown fowner fsetid };
 
+# OTA upgrades
+xc_read_config_files(updatemgr_t)
