From 00c4fea363d12b624a7a64ceead7faf408ad7e11 Mon Sep 17 00:00:00 2001
From: Tamas K Lengyel <lengyelt@ainfosec.com>
Date: Fri, 20 Oct 2017 15:43:15 -0600
Subject: [PATCH] Fix compilation for Linux 4.4+ kernels

---
 src/linux.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/linux.c b/src/linux.c
index 57f71f3..82c3b8c 100644
--- a/src/linux.c
+++ b/src/linux.c
@@ -22,7 +22,7 @@
 #include <inttypes.h>
 #include <limits.h>
 #include <linux/ethtool.h>
-#include <linux/nvme.h>
+#include <linux/nvme_ioctl.h>
 #include <linux/sockios.h>
 #include <net/if.h>
 #include <scsi/scsi.h>
-- 
2.1.4
