--- a/src/optionroms.c
+++ b/src/optionroms.c
@@ -416,7 +416,7 @@ optionrom_setup(void)
  * VGA init
  ****************************************************************/
 
-static int S3ResumeVga;
+int S3ResumeVga;
 int ScreenAndDebug;
 struct rom_header *VgaROM;
 
--- a/src/util.h
+++ b/src/util.h
@@ -195,6 +195,7 @@ void optionrom_setup(void);
 void vgarom_setup(void);
 void s3_resume_vga(void);
 extern int ScreenAndDebug;
+extern int S3ResumeVga;
 
 // pcibios.c
 void handle_1ab1(struct bregs *regs);
--- a/src/vgahooks.c
+++ b/src/vgahooks.c
@@ -254,6 +254,14 @@ getac_setup(struct pci_device *pci)
 {
 }
 
+static void
+amd_setup(struct pci_device *pci)
+{
+    /* Don't call vga bios after resume. With windows guest, it will
+     * infinite loop */
+    S3ResumeVga = 0;
+}
+
 /****************************************************************
  * Silicon Motion hooks
  ****************************************************************/
@@ -351,4 +359,6 @@ vgahook_setup(struct pci_device *pci)
         via_setup(pci);
     else if (pci->vendor == PCI_VENDOR_ID_INTEL)
         intel_setup(pci);
+    else if (pci->vendor == PCI_VENDOR_ID_ATI) /* AMD gpu still use ATI vendor id */
+        amd_setup(pci);
 }
