--- a/scripts/layoutrom.py
+++ b/scripts/layoutrom.py
@@ -205,7 +205,7 @@ def doLayout(sections, config, genreloc)
     li.sections32fseg = getSectionsCategory(sections, '32fseg')
 
     li.sec32fseg_start, li.sec32fseg_align = setSectionsStart(
-        li.sections32fseg, li.sec32seg_start, 16
+        li.sections32fseg, li.sec32seg_start, 4096
         , segoffset=BUILD_BIOS_ADDR)
 
     # Determine 32flat runtime positions
@@ -217,7 +217,7 @@ def doLayout(sections, config, genreloc)
 
     li.sec32flat_start, li.sec32flat_align = setSectionsStart(
         textsections + rodatasections + datasections + bsssections
-        , li.sec32fseg_start, 16)
+        , li.sec32fseg_start, 4096)
 
     # Determine 32flat init positions
     li.sections32init = getSectionsCategory(sections, '32init')
@@ -229,7 +229,7 @@ def doLayout(sections, config, genreloc)
     li.sec32init_start, li.sec32init_align = setSectionsStart(
         init32_textsections + init32_rodatasections
         + init32_datasections + init32_bsssections
-        , li.sec32flat_start, 16)
+        , li.sec32flat_start, 4096)
 
     # Determine location of ZoneFSeg memory.
     li.zonefseg_end = li.sec32flat_start
@@ -242,11 +242,11 @@ def doLayout(sections, config, genreloc)
         li.zonefseg_start = li.zonefseg_end - BUILD_MIN_BIOSTABLE
         li.sec32flat_start, li.sec32flat_align = setSectionsStart(
             textsections + rodatasections + datasections + bsssections
-            , li.zonefseg_start, 16)
+            , li.zonefseg_start, 4096)
         li.sec32init_start, li.sec32init_align = setSectionsStart(
             init32_textsections + init32_rodatasections
             + init32_datasections + init32_bsssections
-            , li.sec32flat_start, 16)
+            , li.sec32flat_start, 4096)
     li.final_readonly_start = min(BUILD_BIOS_ADDR, li.sec32flat_start)
     if not genreloc:
         li.final_readonly_start = min(BUILD_BIOS_ADDR, li.sec32init_start)
@@ -263,7 +263,7 @@ def doLayout(sections, config, genreloc)
         li.zonelow_base = final_sec32low_end - 64*1024
     relocdelta = final_sec32low_end - sec32low_end
     li.sec32low_start, li.sec32low_align = setSectionsStart(
-        li.sections32low, sec32low_end, 16
+        li.sections32low, sec32low_end, 4096
         , segoffset=li.zonelow_base - relocdelta)
     li.final_sec32low_start = li.sec32low_start + relocdelta
 
