--- a/lib/route/link.c
+++ b/lib/route/link.c
@@ -1729,9 +1729,7 @@ int rtnl_link_build_change_request(struc
 	    !strcmp(orig->l_name, changes->l_name))
 		changes->ce_mask &= ~LINK_ATTR_IFNAME;
 
-	rt = af_request_type(orig->l_family);
-
-	if ((err = build_link_msg(rt, &ifi, changes, flags, result)) < 0)
+	if ((err = build_link_msg(RTM_NEWLINK, &ifi, changes, flags, result)) < 0)
 		goto errout;
 
 	return 0;
