--- a/UpdateMgr/Logic.hs
+++ b/UpdateMgr/Logic.hs
@@ -346,8 +346,8 @@ verifyUpdateMetadataSignature :: Update
 verifyUpdateMetadataSignature = void $
   handleError failed . safeShellExecuteAndLogOutput . cmd =<< allowDevRepoCert
   where
-    cmd False = "verify-repo-metadata " ++ updateDirCurrent
-    cmd True  = "verify-repo-metadata -d " ++ updateDirCurrent
+    cmd False = "verify-repo-metadata -p " ++ updateDirCurrent
+    cmd True  = "verify-repo-metadata -d -p " ++ updateDirCurrent
     failed _  = throwError $ localE FailedSignatureVerification
 
 handleError = flip catchError
