commit 612502d30abfdabb836cf0b15d02c1e0ee469526
Author: Julian Pidancet <julian.pidancet@citrix.com>
Date:   Mon Nov 14 13:20:24 2011 +0000

    iPXE: Send early debug on Xen debug port

diff --git a/src/arch/i386/prefix/libprefix.S b/src/arch/i386/prefix/libprefix.S
index 77151a6..695de1d 100644
--- a/src/arch/i386/prefix/libprefix.S
+++ b/src/arch/i386/prefix/libprefix.S
@@ -72,6 +72,7 @@ print_character:
 	pushw	%ax
 	pushw	%bx
 	pushw	%bp
+	outb    %al, $0xe9
 	/* If %di is non-zero, write character to buffer and exit */
 	testw	%di, %di
 	jz	1f
