diff --git a/carrier-detect/configure.in b/carrier-detect/configure.in
index d5753da..16ae3fa 100644
--- a/configure.in
+++ b/configure.in
@@ -10,7 +10,7 @@ dnl
 dnl
 dnl
 AC_PREREQ(2.13)
-AC_INIT()
+AC_INIT([carrier-detect], [1.0.0])
 
 AC_PROG_CC
 AM_PROG_CC_C_O
@@ -19,7 +19,7 @@ AC_SYS_LARGEFILE
 AC_CONFIG_MACRO_DIR([m4])
 
 
-AM_INIT_AUTOMAKE(carrier-detect,1.0.0)
+AM_INIT_AUTOMAKE([foreign])
 
 AM_PROG_LIBTOOL
 
