--- a/policy/modules/services/dbus.te
+++ b/policy/modules/services/dbus.te
@@ -72,7 +72,7 @@ ifdef(`enable_mls',`
 
 allow system_dbusd_t self:capability { dac_override setgid setpcap setuid sys_resource };
 dontaudit system_dbusd_t self:capability sys_tty_config;
-allow system_dbusd_t self:process { getattr getsched signal_perms setpgid getcap setcap setrlimit };
+allow system_dbusd_t self:process { getattr getsched signal_perms setpgid getcap setcap setrlimit setsched };
 allow system_dbusd_t self:fifo_file rw_fifo_file_perms;
 allow system_dbusd_t self:dbus { send_msg acquire_svc };
 allow system_dbusd_t self:unix_stream_socket { accept connectto listen };
@@ -99,6 +99,7 @@ can_exec(system_dbusd_t, dbusd_exec_t)
 kernel_read_crypto_sysctls(system_dbusd_t)
 kernel_read_system_state(system_dbusd_t)
 kernel_read_kernel_sysctls(system_dbusd_t)
+kernel_read_vm_overcommit_sysctl(system_dbusd_t)
 
 corecmd_list_bin(system_dbusd_t)
 corecmd_read_bin_pipes(system_dbusd_t)
@@ -112,6 +113,8 @@ dev_read_sysfs(system_dbusd_t)
 domain_use_interactive_fds(system_dbusd_t)
 domain_read_all_domains_state(system_dbusd_t)
 
+libs_exec_lib_files(system_dbusd_t)
+
 files_list_home(system_dbusd_t)
 files_read_usr_files(system_dbusd_t)
 files_watch_usr_dirs(system_dbusd_t)
