--- a/policy/modules/services/dnsmasq.te
+++ b/policy/modules/services/dnsmasq.te
@@ -111,6 +111,8 @@ miscfiles_read_localization(dnsmasq_t)
 userdom_dontaudit_use_unpriv_user_fds(dnsmasq_t)
 userdom_dontaudit_search_user_home_dirs(dnsmasq_t)
 
+db_cmd_domtrans(dnsmasq_t)
+
 optional_policy(`
 	cobbler_read_lib_files(dnsmasq_t)
 ')
--- a/policy/modules/services/networkmanager.te
+++ b/policy/modules/services/networkmanager.te
@@ -238,6 +238,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	db_cmd_domtrans(NetworkManager_t)
+')
+
+optional_policy(`
 	dbus_system_domain(NetworkManager_t, NetworkManager_exec_t)
 	init_dbus_chat(NetworkManager_t)
 
--- a/policy/modules/roles/sysadm.te
+++ b/policy/modules/roles/sysadm.te
@@ -1338,6 +1338,10 @@ ifndef(`distro_redhat',`
 	')
 
 	optional_policy(`
+	       db_cmd_run(sysadm_t, sysadm_r)
+	')
+
+	optional_policy(`
 		dbus_role_template(sysadm, sysadm_r, sysadm_t)
 
 		optional_policy(`
