--- a/config/file_contexts.subs_dist
+++ b/config/file_contexts.subs_dist
@@ -44,6 +44,11 @@
 /usr/lib/busybox/sbin /usr/sbin
 /usr/lib/busybox/usr /usr
 
+# config partition aliases
+# ensure the policy applied to the base filesystem objects is reflected in the
+# config partition.
+/config/etc /etc
+
 # The genhomedircon.py will expand /root home directory to /home/root
 # Add an aliase for it
 /root /home/root
