--- a/policy/modules/admin/brctl.te
+++ b/policy/modules/admin/brctl.te
@@ -43,6 +43,10 @@ term_dontaudit_use_console(brctl_t)
 miscfiles_read_localization(brctl_t)
 
 optional_policy(`
+	network_slave_rw_stream_sockets(brctl_t)
+')
+
+optional_policy(`
 	xen_append_log(brctl_t)
 	xen_dontaudit_rw_unix_stream_sockets(brctl_t)
 ')
--- a/policy/modules/services/dnsmasq.te
+++ b/policy/modules/services/dnsmasq.te
@@ -131,6 +131,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	network_slave_rw_stream_sockets(dnsmasq_t)
+')
+
+optional_policy(`
 	seutil_sigchld_newrole(dnsmasq_t)
 ')
 
--- a/policy/modules/services/networkmanager.te
+++ b/policy/modules/services/networkmanager.te
@@ -318,6 +318,11 @@ optional_policy(`
 ')
 
 optional_policy(`
+	network_slave_rw_stream_sockets(NetworkManager_t)
+	network_slave_domtrans(NetworkManager_t)
+')
+
+optional_policy(`
 	nscd_domtrans(NetworkManager_t)
 	nscd_signal(NetworkManager_t)
 	nscd_signull(NetworkManager_t)
--- a/policy/modules/system/init.te
+++ b/policy/modules/system/init.te
@@ -1201,6 +1201,9 @@ optional_policy(`
 	optional_policy(`
 		policykit_dbus_chat(initrc_t)
 	')
+	optional_policy(`
+        	network_slave_dbus_chat(initrc_t)
+	')
 ')
 
 optional_policy(`
--- a/policy/modules/system/iptables.te
+++ b/policy/modules/system/iptables.te
@@ -122,6 +122,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	network_slave_rw_stream_sockets(iptables_t)
+')
+
+optional_policy(`
 	# for iptables -L
 	nis_use_ypbind(iptables_t)
 ')
--- a/policy/modules/system/modutils.te
+++ b/policy/modules/system/modutils.te
@@ -175,6 +175,12 @@ optional_policy(`
 ')
 
 optional_policy(`
+	network_slave_rw_stream_sockets(kmod_t)
+	network_slave_rw_fifo_files(kmod_t)
+	network_slave_use_fds(kmod_t)
+')
+
+optional_policy(`
 	nis_use_ypbind(kmod_t)
 ')
 
--- a/policy/modules/system/sysnetwork.te
+++ b/policy/modules/system/sysnetwork.te
@@ -417,6 +417,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	network_slave_rw_stream_sockets(ifconfig_t)
+')
+
+optional_policy(`
 	nis_use_ypbind(ifconfig_t)
 ')
 
--- a/policy/modules/system/udev.te
+++ b/policy/modules/system/udev.te
@@ -364,6 +364,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	network_slave_dbus_chat(udev_t)
+')
+
+optional_policy(`
 	openct_read_runtime_files(udev_t)
 	openct_domtrans(udev_t)
 ')
