--- a/policy/modules/system/modutils.te
+++ b/policy/modules/system/modutils.te
@@ -76,6 +76,7 @@ corecmd_exec_bin(kmod_t)
 corecmd_exec_shell(kmod_t)
 
 dev_rw_sysfs(kmod_t)
+dev_mount_usbfs(kmod_t)
 dev_search_usbfs(kmod_t)
 dev_rw_mtrr(kmod_t)
 dev_read_urand(kmod_t)
@@ -113,12 +114,16 @@ logging_search_logs(kmod_t)
 
 miscfiles_read_localization(kmod_t)
 
+selinux_mount_fs(kmod_t)
 seutil_read_file_contexts(kmod_t)
 
 userdom_use_user_terminals(kmod_t)
 
 userdom_dontaudit_search_user_home_dirs(kmod_t)
 
+# OpenXT: the txt module searches the securityfs at init time
+selinux_search_fs(kmod_t)
+
 ifdef(`init_systemd',`
 	# for /run/tmpfiles.d/kmod.conf
 	allow kmod_t kmod_tmpfiles_conf_t:file manage_file_perms;
@@ -162,6 +167,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+       iptables_dontaudit_read_pids(kmod_t)
+')
+
+optional_policy(`
 	mount_domtrans(kmod_t)
 ')
 
