--- a/policy/modules/services/dnsmasq.fc
+++ b/policy/modules/services/dnsmasq.fc
@@ -2,6 +2,7 @@
 /etc/dnsmasq\.d(/.*)?			gen_context(system_u:object_r:dnsmasq_etc_t,s0)
 
 /etc/rc\.d/init\.d/dnsmasq	--	gen_context(system_u:object_r:dnsmasq_initrc_exec_t,s0)
+/etc/dnsmasq-config(/.*)?		gen_context(system_u:object_r:dnsmasq_etc_t, s0)
 
 /run/dnsmasq.*			--	gen_context(system_u:object_r:dnsmasq_runtime_t,s0)
 /run/libvirt/network(/.*)?		gen_context(system_u:object_r:dnsmasq_runtime_t,s0)
--- a/policy/modules/services/dnsmasq.if
+++ b/policy/modules/services/dnsmasq.if
@@ -137,6 +137,25 @@ interface(`dnsmasq_write_config',`
 
 ########################################
 ## <summary>
+##     Manage dnsmasq config files.
+## </summary>
+## <param name="domain">
+## <summary>
+##     Domain allowed.
+## </summary>
+## </param>
+#
+interface(`dnsmasq_manage_config',`
+	gen_require(`
+		type dnsmasq_etc_t;
+	')
+
+	manage_files_pattern($1, dnsmasq_etc_t, dnsmasq_etc_t)
+	files_search_etc($1)
+')
+
+########################################
+## <summary>
 ##	Delete dnsmasq pid files.  (Deprecated)
 ## </summary>
 ## <param name="domain">
--- a/policy/modules/services/dnsmasq.te
+++ b/policy/modules/services/dnsmasq.te
@@ -14,6 +14,7 @@ init_script_file(dnsmasq_initrc_exec_t)
 
 type dnsmasq_etc_t;
 files_config_file(dnsmasq_etc_t)
+files_mountpoint(dnsmasq_etc_t)
 
 type dnsmasq_lease_t;
 files_type(dnsmasq_lease_t)
@@ -27,6 +28,10 @@ init_unit_file(dnsmasq_unit_t)
 type dnsmasq_var_log_t;
 logging_log_file(dnsmasq_var_log_t)
 
+type dnsmasq_tmp_t;
+files_tmp_file(dnsmasq_tmp_t)
+files_tmp_filetrans(dnsmasq_t, dnsmasq_tmp_t, file)
+
 ########################################
 #
 # Local policy
@@ -39,6 +44,7 @@ allow dnsmasq_t self:fifo_file rw_fifo_f
 allow dnsmasq_t self:tcp_socket { accept listen };
 allow dnsmasq_t self:packet_socket create_socket_perms;
 allow dnsmasq_t self:rawip_socket create_socket_perms;
+allow dnsmasq_t dnsmasq_tmp_t:file manage_file_perms;
 
 allow dnsmasq_t dnsmasq_etc_t:dir list_dir_perms;
 allow dnsmasq_t dnsmasq_etc_t:file read_file_perms;
@@ -55,6 +61,12 @@ manage_dirs_pattern(dnsmasq_t, dnsmasq_r
 manage_files_pattern(dnsmasq_t, dnsmasq_runtime_t, dnsmasq_runtime_t)
 files_runtime_filetrans(dnsmasq_t, dnsmasq_runtime_t, { dir file })
 
+# most likely due to fd leaks from dnsmasq into db-ls-dom0 script
+dontaudit db_cmd_t dnsmasq_t:netlink_route_socket rw_socket_perms;
+
+# dnsmasq executes /etc/dnsmasq-config/*
+allow dnsmasq_t dnsmasq_etc_t:file { execute execute_no_trans };
+
 kernel_read_kernel_sysctls(dnsmasq_t)
 kernel_read_net_sysctls(dnsmasq_t)
 kernel_read_network_state(dnsmasq_t)
@@ -62,6 +74,7 @@ kernel_read_system_state(dnsmasq_t)
 kernel_request_load_module(dnsmasq_t)
 
 corecmd_exec_shell(dnsmasq_t)
+corecmd_exec_bin(dnsmasq_t)
 
 corenet_all_recvfrom_netlabel(dnsmasq_t)
 corenet_tcp_sendrecv_generic_if(dnsmasq_t)
