--- a/policy/modules/system/fstools.fc
+++ b/policy/modules/system/fstools.fc
@@ -26,6 +26,7 @@
 /usr/bin/make_reiser4		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /usr/bin/mkdosfs		--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /usr/bin/mke2fs			--	gen_context(system_u:object_r:fsadm_exec_t,s0)
+/usr/bin/mke2fs\.e2fsprogs	--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /usr/bin/mke4fs			--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /usr/bin/mkfs.*			--	gen_context(system_u:object_r:fsadm_exec_t,s0)
 /usr/bin/mkraid			--	gen_context(system_u:object_r:fsadm_exec_t,s0)
--- a/policy/modules/system/fstools.te
+++ b/policy/modules/system/fstools.te
@@ -142,6 +142,7 @@ storage_raw_read_removable_device(fsadm_
 storage_raw_write_removable_device(fsadm_t)
 storage_read_scsi_generic(fsadm_t)
 storage_swapon_fixed_disk(fsadm_t)
+storage_rw_blktap(fsadm_t)
 
 term_use_console(fsadm_t)
 
