--- a/policy/modules/admin/alsa.fc
+++ b/policy/modules/admin/alsa.fc
@@ -23,3 +23,9 @@ HOME_DIR/\.asoundrc				--	gen_context(sy
 /var/lib/alsa(/.*)?					gen_context(system_u:object_r:alsa_var_lib_t,s0)
 
 /var/lock/asound\.state\.lock			--	gen_context(system_u:object_r:alsa_var_lock_t,s0)
+
+/var/volatile/etc/asound\.conf	gen_context(system_u:object_r:alsa_etc_t,s0)
+
+/config/asound\.state(\.new)?	gen_context(system_u:object_r:alsa_etc_t,s0)
+
+/tmp/alsactl\.dump       --      gen_context(system_u:object_r:alsa_tmp_t,s0)
--- a/policy/modules/admin/alsa.if
+++ b/policy/modules/admin/alsa.if
@@ -248,3 +248,24 @@ interface(`alsa_write_lib',`
 	files_search_var_lib($1)
 	write_files_pattern($1, alsa_var_lib_t, alsa_var_lib_t)
 ')
+
+########################################
+## <summary>
+##	Manage alsa tmp files.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`alsa_manage_tmp_files',`
+	gen_require(`
+		type alsa_tmp_t;
+	')
+
+	files_search_tmp($1)
+	allow $1 alsa_tmp_t:dir list_dir_perms;
+	manage_files_pattern($1, alsa_tmp_t, alsa_tmp_t)
+	read_lnk_files_pattern($1, alsa_tmp_t, alsa_tmp_t)
+')
--- a/policy/modules/admin/alsa.te
+++ b/policy/modules/admin/alsa.te
@@ -103,6 +103,8 @@ logging_send_syslog_msg(alsa_t)
 
 miscfiles_read_localization(alsa_t)
 
+seutil_use_newrole_fds(alsa_t)
+
 userdom_manage_unpriv_user_semaphores(alsa_t)
 userdom_manage_unpriv_user_shared_mem(alsa_t)
 userdom_search_user_home_dirs(alsa_t)
@@ -111,3 +113,8 @@ optional_policy(`
 	hal_use_fds(alsa_t)
 	hal_write_log(alsa_t)
 ')
+
+xc_config_dir_getattr(alsa_t)
+xc_config_dir_search(alsa_t)
+xc_config_filetrans(alsa_t, alsa_etc_t, { dir file })
+allow alsa_t alsa_etc_t:file manage_file_perms;
