ndex: refpolicy/policy/modules/system/udev.fc
===================================================================
--- a/policy/modules/system/udev.te
+++ b/policy/modules/system/udev.te
@@ -66,6 +66,11 @@ allow udev_t self:rawip_socket create_so
 # for systemd-udevd to rename interfaces
 allow udev_t self:netlink_route_socket nlmsg_write;
 
+# Ignore CAP_IPC_LOCK denial triggered by mmap(MAP_LOCKED);
+# the operation will still succeed.  See the following kernel commit:
+# http://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=a5a6579db33af91f4f5134e14be758dc71c1b694
+dontaudit udev_t self:capability ipc_lock;
+
 can_exec(udev_t, udev_exec_t)
 
 allow udev_t udev_helper_exec_t:dir list_dir_perms;
@@ -111,6 +116,8 @@ kernel_dontaudit_search_unlabeled(udev_t
 
 corecmd_exec_all_executables(udev_t)
 
+dev_manage_sysfs_dirs(udev_t)
+dev_create_sysfs_nodes(udev_t)
 dev_rw_sysfs(udev_t)
 dev_manage_all_dev_nodes(udev_t)
 dev_rw_generic_files(udev_t)
@@ -200,6 +207,8 @@ sysnet_etc_filetrans_config(udev_t)
 
 userdom_dontaudit_search_user_home_content(udev_t)
 
+iptables_domtrans(udev_t)
+
 ifdef(`distro_debian',`
 	# for https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=851933
 	files_read_default_files(udev_t)
@@ -381,12 +390,30 @@ optional_policy(`
 	xen_manage_log(udev_t)
 	xen_read_image_files(udev_t)
 	fs_manage_xenfs_files(udev_t)
+	dev_manage_xen(udev_t)
+	xen_stream_connect_xenstore(udev_t)
+	fs_rw_xenfs_files(udev_t)
 ')
 
 optional_policy(`
 	xserver_read_xdm_runtime_files(udev_t)
 ')
 
+fs_manage_tmpfs_chr_files(udev_t)
+fs_manage_tmpfs_files(udev_t)
+fs_manage_tmpfs_symlinks(udev_t)
+virt_search_dirs(udev_t)
+
+vusbd_read_etc_files(udev_t)
+
+gen_require(`
+    type lib_t;
+')
+
+allow udev_t lib_t:dir { watch };
+allow udev_t udev_rules_t:dir { watch };
+allow udev_t udev_runtime_t:dir { watch };
+
 ########################################
 #
 # udevadm Local policy
