--- a/policy/modules/system/miscfiles.if
+++ b/policy/modules/system/miscfiles.if
@@ -488,6 +488,25 @@ interface(`miscfiles_setattr_localizatio
 
 ########################################
 ## <summary>
+##	Allow process to search localization info
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`miscfiles_search_localization',`
+	gen_require(`
+		type locale_t;
+	')
+
+	files_search_usr($1)
+	allow $1 locale_t:dir search_dir_perms;
+')
+
+########################################
+## <summary>
 ##	Allow process to read localization information.
 ## </summary>
 ## <desc>
