--- a/policy/modules/admin/dmidecode.te
+++ b/policy/modules/admin/dmidecode.te
@@ -37,3 +37,19 @@ ifdef(`init_systemd',`
 	init_use_fds(dmidecode_t)
 	init_rw_inherited_stream_socket(dmidecode_t)
 ')
+
+
+########################################
+#
+# OpenXT local policy
+#
+
+# dmidecode called from init script
+init_use_fds(dmidecode_t)
+term_read_console(dmidecode_t)
+term_write_console(dmidecode_t)
+
+# dmidecode from sysadm_r
+seutil_use_newrole_fds(dmidecode_t)
+
+xen_dontaudit_rw_unix_stream_sockets(dmidecode_t)
