--- a/policy/modules/system/sysnetwork.fc
+++ b/policy/modules/system/sysnetwork.fc
@@ -18,9 +18,11 @@ ifdef(`distro_debian',`
 /etc/hosts		--	gen_context(system_u:object_r:net_conf_t,s0)
 /etc/hosts\.allow.*	--	gen_context(system_u:object_r:net_conf_t,s0)
 /etc/hosts\.deny.*	--	gen_context(system_u:object_r:net_conf_t,s0)
+/etc/hosts		-l	gen_context(system_u:object_r:net_conf_t,s0)
+/run/hosts		--	gen_context(system_u:object_r:net_conf_t,s0)
 /etc/hostname		--	gen_context(system_u:object_r:net_conf_t,s0)
 /etc/denyhosts.*	--	gen_context(system_u:object_r:net_conf_t,s0)
-/etc/resolv\.conf.*	--	gen_context(system_u:object_r:net_conf_t,s0)
+/etc/resolv\.conf.*		gen_context(system_u:object_r:net_conf_t,s0)
 /etc/yp\.conf.*		--	gen_context(system_u:object_r:net_conf_t,s0)
 
 /etc/dhcp3(/.*)?		gen_context(system_u:object_r:dhcp_etc_t,s0)
@@ -33,6 +35,15 @@ ifdef(`distro_redhat',`
 /etc/sysconfig/networking(/.*)? gen_context(system_u:object_r:net_conf_t,s0)
 /etc/sysconfig/network-scripts(/.*)? gen_context(system_u:object_r:net_conf_t,s0)
 ')
+/etc/udhcpc.d		-d	gen_context(system_u:object_r:dhcp_etc_t,s0)
+/etc/udhcpc.d/.+	--	gen_context(system_u:object_r:dhcp_etc_t,s0)
+
+/etc/iproute2(/.*)?	--	gen_context(system_u:object_r:net_conf_t,s0)
+
+#
+# /tmp
+#
+/tmp/hosts		--	gen_context(system_u:object_r:net_conf_t,s0)
 
 #
 # /usr
@@ -71,6 +82,10 @@ ifdef(`distro_redhat',`
 /usr/sbin/mii-tool\.net-tools		--	gen_context(system_u:object_r:ifconfig_exec_t,s0)
 /usr/sbin/pump			--	gen_context(system_u:object_r:dhcpc_exec_t,s0)
 /usr/sbin/tc			--	gen_context(system_u:object_r:ifconfig_exec_t,s0)
+/usr/sbin/udhcpc		--	gen_context(system_u:object_r:dhcpc_exec_t,s0)
+
+/usr/share/udhcpc		-d	gen_context(system_u:object_r:dhcp_etc_t,s0)
+/usr/share/udhcpc/default\.script	--	gen_context(system_u:object_r:dhcp_etc_t,s0)
 
 #
 # /var
@@ -97,3 +112,10 @@ ifdef(`distro_debian',`
 /run/resolvconf/.* --	gen_context(system_u:object_r:net_conf_t,s0)
 ')
 
+/var/volatile/etc/resolv\.conf.*	--	gen_context(system_u:object_r:net_conf_t,s0)
+
+#
+# /config
+#
+/config/system/iproute2(/.*)?  --      gen_context(system_u:object_r:net_conf_t,s0)
+
--- a/policy/modules/system/sysnetwork.if
+++ b/policy/modules/system/sysnetwork.if
@@ -347,6 +347,7 @@ interface(`sysnet_read_config',`
 
 	files_search_etc($1)
 	allow $1 net_conf_t:file read_file_perms;
+	allow $1 net_conf_t:lnk_file read_lnk_file_perms;
 
 	ifdef(`distro_debian',`
 		files_search_runtime($1)
@@ -505,6 +506,7 @@ interface(`sysnet_manage_config',`
 
 	files_search_etc($1)
 	allow $1 net_conf_t:file manage_file_perms;
+	allow $1 net_conf_t:lnk_file manage_lnk_file_perms;
 
 	ifdef(`distro_debian',`
 		files_search_runtime($1)
@@ -727,6 +729,24 @@ interface(`sysnet_search_dhcp_state',`
 ')
 
 ########################################
+## <summary>
+##	Dontaudit attemps to write to state dir.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`sysnet_dontaudit_write_dhcp_state',`
+	gen_require(`
+		type dhcp_state_t;
+	')
+
+	dontaudit $1 dhcp_state_t:dir write;
+')
+
+########################################
 ## <summary>
 ##	Create DHCP state data.
 ## </summary>
--- a/policy/modules/system/sysnetwork.te
+++ b/policy/modules/system/sysnetwork.te
@@ -15,6 +15,7 @@ files_config_file(dhcp_etc_t)
 # this is shared between dhcpc and dhcpd:
 type dhcp_state_t;
 files_type(dhcp_state_t)
+files_mountpoint(dhcp_state_t)
 
 type dhcpc_t;
 type dhcpc_exec_t;
@@ -40,6 +41,7 @@ files_runtime_file(ifconfig_runtime_t)
 
 type net_conf_t;
 files_type(net_conf_t)
+files_mountpoint(net_conf_t)
 
 ifdef(`distro_debian',`
 	init_daemon_runtime_file(net_conf_t, dir, "network")
@@ -99,6 +101,7 @@ kernel_rw_net_sysctls(dhcpc_t)
 
 corecmd_exec_bin(dhcpc_t)
 corecmd_exec_shell(dhcpc_t)
+libs_exec_lib_files(dhcpc_t)
 
 corenet_all_recvfrom_netlabel(dhcpc_t)
 corenet_tcp_sendrecv_all_if(dhcpc_t)
@@ -115,6 +118,7 @@ corenet_udp_bind_all_unreserved_ports(dh
 corenet_tcp_connect_all_ports(dhcpc_t)
 corenet_sendrecv_dhcpd_client_packets(dhcpc_t)
 corenet_sendrecv_all_server_packets(dhcpc_t)
+corenet_udp_bind_all_ports(dhcpc_t)
 
 dev_read_sysfs(dhcpc_t)
 # for SSP:
@@ -125,6 +129,12 @@ domain_dontaudit_read_all_domains_state(
 
 files_read_etc_files(dhcpc_t)
 files_read_etc_runtime_files(dhcpc_t)
+# to allow reading of /var/run/resolv.conf ... should consider a better place
+#   for resolv.conf with a more descriptive type
+files_read_generic_pids(dhcpc_t)
+# to allow reading of NetworkManager configuration from
+#   /var/run/nm-dhclient-brbridged.conf
+networkmanager_read_pid_files(dhcpc_t)
 files_read_usr_files(dhcpc_t)
 files_search_home(dhcpc_t)
 files_search_var_lib(dhcpc_t)
@@ -211,11 +221,10 @@ optional_policy(`
 	modutils_run(dhcpc_t, dhcpc_roles)
 ')
 
+# run ping in the dhcpc_t domain without domtrans (busybox ip)
 # for the dhcp client to run ping to check IP addresses
-optional_policy(`
-	netutils_run_ping(dhcpc_t, dhcpc_roles)
-	netutils_run(dhcpc_t, dhcpc_roles)
-')
+netutils_run_ping(dhcpc_t, dhcpc_roles)
+netutils_run(dhcpc_t, dhcpc_roles)
 
 optional_policy(`
 	networkmanager_domtrans(dhcpc_t)
@@ -316,6 +325,7 @@ kernel_use_fds(ifconfig_t)
 kernel_read_system_state(ifconfig_t)
 kernel_read_network_state(ifconfig_t)
 kernel_request_load_module(ifconfig_t)
+kernel_load_module(ifconfig_t)
 kernel_search_network_sysctl(ifconfig_t)
 kernel_rw_net_sysctls(ifconfig_t)
 
@@ -327,6 +337,9 @@ dev_read_sysfs(ifconfig_t)
 # for IPSEC setup:
 dev_read_urand(ifconfig_t)
 
+# Early devtmpfs /dev/console access prior to udev relabel
+dev_dontaudit_rw_generic_chr_files(ifconfig_t)
+
 domain_use_interactive_fds(ifconfig_t)
 
 files_read_etc_files(ifconfig_t)
@@ -355,6 +368,7 @@ logging_send_syslog_msg(ifconfig_t)
 miscfiles_read_localization(ifconfig_t)
 
 seutil_use_runinit_fds(ifconfig_t)
+sysnet_read_config(ifconfig_t)
 
 sysnet_dontaudit_rw_dhcpc_udp_sockets(ifconfig_t)
 
