--- a/policy/modules/system/selinuxutil.te
+++ b/policy/modules/system/selinuxutil.te
@@ -120,6 +120,9 @@ type setfiles_t alias restorecon_t, can_
 type setfiles_exec_t alias restorecon_exec_t;
 init_system_domain(setfiles_t, setfiles_exec_t)
 domain_obj_id_change_exemption(setfiles_t)
+term_read_console(setfiles_t)
+dev_read_urand(setfiles_t)
+dev_rw_generic_chr_files(setfiles_t)
 
 ########################################
 #
@@ -174,6 +177,7 @@ allow load_policy_t policy_config_t:file
 
 dev_read_urand(load_policy_t)
 
+dev_dontaudit_search_sysfs(load_policy_t)
 domain_use_interactive_fds(load_policy_t)
 
 # for mcs.conf
@@ -219,6 +223,17 @@ optional_policy(`
 	portage_dontaudit_use_fds(load_policy_t)
 ')
 
+# XC init leaks fds
+init_dontaudit_use_fds(load_policy_t)
+
+# leaked by opkg
+ifdef(`distro_debian',`
+	optional_policy(`
+		dpkg_dontaudit_manage_db(load_policy_t)
+		dpkg_dontaudit_lock_db(load_policy_t)
+	')
+')
+
 ########################################
 #
 # Newrole local policy
@@ -283,6 +298,7 @@ auth_use_nsswitch(newrole_t)
 auth_run_chk_passwd(newrole_t, newrole_roles)
 auth_run_upd_passwd(newrole_t, newrole_roles)
 auth_rw_faillog(newrole_t)
+auth_read_shadow(newrole_t)
 
 # Write to utmp.
 init_rw_utmp(newrole_t)
@@ -497,6 +513,7 @@ corecmd_exec_bin(semanage_t)
 corecmd_exec_shell(semanage_t)
 
 dev_read_urand(semanage_t)
+dev_dontaudit_search_sysfs(semanage_t)
 
 domain_use_interactive_fds(semanage_t)
 
@@ -596,6 +613,14 @@ dev_relabel_all_dev_nodes(setfiles_t)
 # to handle when /dev/console needs to be relabeled
 dev_rw_generic_chr_files(setfiles_t)
 
+# leaked by opkg
+ifdef(`distro_debian',`
+	optional_policy(`
+		dpkg_dontaudit_manage_db(setfiles_t)
+		dpkg_dontaudit_lock_db(setfiles_t)
+	')
+')
+
 domain_use_interactive_fds(setfiles_t)
 domain_dontaudit_search_all_domains_state(setfiles_t)
 
