--- a/policy/modules/admin/brctl.te
+++ b/policy/modules/admin/brctl.te
@@ -47,6 +47,11 @@ optional_policy(`
 ')
 
 optional_policy(`
+	statusreport_getattr_storage_files(brctl_t)
+	statusreport_write_storage_files(brctl_t)
+')
+
+optional_policy(`
 	xen_append_log(brctl_t)
 	xen_dontaudit_rw_unix_stream_sockets(brctl_t)
 ')
--- a/policy/modules/admin/dmidecode.te
+++ b/policy/modules/admin/dmidecode.te
@@ -53,3 +53,7 @@ term_write_console(dmidecode_t)
 seutil_use_newrole_fds(dmidecode_t)
 
 xen_dontaudit_rw_unix_stream_sockets(dmidecode_t)
+
+optional_policy(`
+	statusreport_write_storage_files(dmidecode_t)
+')
--- a/policy/modules/system/xen.te
+++ b/policy/modules/system/xen.te
@@ -261,6 +261,10 @@ qemu_signull(xend_t)
 # transition to dmidecode
 dmidecode_domtrans(xend_t)
 
+optional_policy(`
+	statusreport_domtrans(xend_t)
+')
+
 blktap_getattr_blk_file(xend_t)
 blktap_relabel_blk_file(xend_t)
 # this may be a leaked fd or something, dontaudit?
--- a/policy/modules/roles/sysadm.te
+++ b/policy/modules/roles/sysadm.te
@@ -1311,6 +1311,10 @@ optional_policy(`
 	zebra_admin(sysadm_t, sysadm_r)
 ')
 
+optional_policy(`
+	statusreport_run(sysadm_t, sysadm_r)
+')
+
 ifndef(`distro_redhat',`
 	optional_policy(`
 		auth_role(sysadm_r, sysadm_t)
--- a/policy/modules/system/lvm.te
+++ b/policy/modules/system/lvm.te
@@ -372,6 +372,11 @@ optional_policy(`
 ')
 
 optional_policy(`
+	statusreport_write_storage_files(lvm_t)
+	statusreport_getattr_storage_files(lvm_t)
+')
+
+optional_policy(`
 	rpm_manage_script_tmp_files(lvm_t)
 ')
 
--- a/policy/modules/system/mount.te
+++ b/policy/modules/system/mount.te
@@ -251,3 +251,8 @@ optional_policy(`
 	unconfined_domain(unconfined_mount_t)
 	unconfined_run_to(unconfined_mount_t, mount_exec_t)
 ')
+
+optional_policy(`
+	statusreport_write_storage_files(mount_t)
+	statusreport_getattr_storage_files(mount_t)
+')
--- a/policy/modules/system/sysnetwork.te
+++ b/policy/modules/system/sysnetwork.te
@@ -429,6 +429,11 @@ optional_policy(`
 ')
 
 optional_policy(`
+	statusreport_write_storage_files(ifconfig_t)
+	statusreport_getattr_storage_files(ifconfig_t)
+')
+
+optional_policy(`
 	unconfined_dontaudit_rw_pipes(ifconfig_t)
 ')
 
--- a/policy/modules/admin/alsa.te
+++ b/policy/modules/admin/alsa.te
@@ -114,6 +114,12 @@ optional_policy(`
 	hal_write_log(alsa_t)
 ')
 
+optional_policy(`
+	statusreport_create_tmp_files(alsa_t)
+	statusreport_write_tmp_files(alsa_t)
+	statusreport_getattr_tmp_files(alsa_t)
+')
+
 xc_config_dir_getattr(alsa_t)
 xc_config_dir_search(alsa_t)
 xc_config_filetrans(alsa_t, alsa_etc_t, { dir file })
--- a/policy/modules/admin/tboot.te
+++ b/policy/modules/admin/tboot.te
@@ -20,3 +20,8 @@ role txtstat_roles types txtstat_t;
 
 domain_use_interactive_fds(txtstat_t)
 userdom_use_user_terminals(txtstat_t)
+
+# Generate data for statusreport
+statusreport_write_storage_files(txtstat_t)
+statusreport_getattr_storage_files(txtstat_t)
+
