--- a/policy/modules/roles/sysadm.te
+++ b/policy/modules/roles/sysadm.te
@@ -1337,6 +1337,7 @@ optional_policy(`
 	viptables_run(sysadm_t, sysadm_r)
 	xentop_run(sysadm_t, sysadm_r)
 	xec_run(sysadm_t, sysadm_r)
+	xen_run_xm(sysadm_t, sysadm_r)
 ')
 
 ifndef(`distro_redhat',`
--- a/policy/modules/system/xen.te
+++ b/policy/modules/system/xen.te
@@ -660,6 +660,8 @@ files_lock_filetrans(xm_t, xen_lock_t, f
 
 manage_files_pattern(xm_t, xend_var_log_t, xend_var_log_t)
 
+seutil_use_newrole_fds(xm_t)
+
 manage_files_pattern(xm_t, xend_var_lib_t, xend_var_lib_t)
 manage_fifo_files_pattern(xm_t, xend_var_lib_t, xend_var_lib_t)
 manage_sock_files_pattern(xm_t, xend_var_lib_t, xend_var_lib_t)
