--- network-manager-applet-0.9.2.0/src/wireless-security/wireless-security.c	2017-01-13 10:50:54.830968023 -0500
+++ network-manager-applet-0.9.2.0/src/wireless-security/wireless-security.c	2017-01-16 15:23:41.424841517 -0500
@@ -37,6 +37,20 @@
 #include "wireless-security.h"
 #include "eap-method.h"
 
+static void
+removeCheckboxes (GtkBuilder *builder)
+{
+	GSList *objects;
+	const char showCheckbuttonPrefix[] = "show_checkbutton_";
+
+	for (objects = gtk_builder_get_objects (builder); objects != NULL; objects = objects->next) {
+		const gchar *id = gtk_buildable_get_name (objects->data);
+		if (id && strncmp (id, showCheckbuttonPrefix, sizeof (showCheckbuttonPrefix) - 1) == 0)
+			gtk_widget_hide (GTK_WIDGET (objects->data));
+	}
+	g_slist_free(objects);
+}
+
 GType
 wireless_security_get_g_type (void)
 {
@@ -197,6 +211,7 @@
 	}
 	g_object_ref_sink (sec->ui_widget);
 
+	removeCheckboxes (sec->builder);
 	return sec;
 }
 
@@ -314,6 +329,7 @@
 		if (eap_default_widget)
 			gtk_widget_grab_focus (eap_default_widget);
 	}
+	removeCheckboxes (eap->builder);
 
 	eap_method_unref (eap);
 
