Index: qemu-1.4.0/configure
===================================================================
--- qemu-1.4.0.orig/configure	2015-05-05 15:05:13.082535065 -0400
+++ qemu-1.4.0/configure	2015-05-05 15:07:57.794070166 -0400
@@ -226,6 +226,7 @@
 seccomp=""
 glusterfs=""
 virtio_blk_data_plane=""
+stubdom="no"
 
 # parse CC options first
 for opt do
@@ -897,6 +898,10 @@
   ;;
   --enable-virtio-blk-data-plane) virtio_blk_data_plane="yes"
   ;;
+  --enable-openxt-stubdom) stubdom="yes"
+  ;;
+  --disable-openxt-stubdom) stubdom="no"
+  ;;
   *) echo "ERROR: unknown option $opt"; show_help="yes"
   ;;
   esac
@@ -3344,6 +3349,7 @@
 echo "virtio-blk-data-plane $virtio_blk_data_plane"
 echo "gcov              $gcov_tool"
 echo "gcov enabled      $gcov"
+echo "OpenXT Stubdomain support   $stubdom"
 
 if test "$sdl_too_old" = "yes"; then
 echo "-> Your SDL version is too old - please upgrade to have SDL support"
@@ -3700,6 +3706,10 @@
   echo "CONFIG_VIRTIO_BLK_DATA_PLANE=y" >> $config_host_mak
 fi
 
+if test "$stubdom" = "yes" ; then
+  echo "CONFIG_STUBDOM=y" >> $config_host_mak
+fi
+
 # USB host support
 case "$usb" in
 linux)
