Index: qemu-1.4.0/Makefile.objs
===================================================================
--- qemu-1.4.0.orig/Makefile.objs	2013-02-15 18:05:35.000000000 -0500
+++ qemu-1.4.0/Makefile.objs	2015-05-26 13:23:51.685934767 -0400
@@ -99,6 +99,10 @@
 common-obj-y += disas/
 
 ######################################################################
+# loggin-syslog
+common-obj-$(CONFIG_SYSLOG_LOGGING) += logging.o
+
+######################################################################
 # guest agent
 
 # FIXME: a few definitions from qapi-types.o/qapi-visit.o are needed
Index: qemu-1.4.0/configure
===================================================================
--- qemu-1.4.0.orig/configure	2015-05-26 13:23:51.573935310 -0400
+++ qemu-1.4.0/configure	2015-05-26 13:25:38.874556094 -0400
@@ -163,6 +163,7 @@
 xen=""
 xen_ctrl_version=""
 xen_pci_passthrough=""
+syslog="yes"
 linux_aio=""
 cap_ng=""
 attr=""
@@ -753,6 +754,10 @@
   ;;
   --enable-xen-pci-passthrough) xen_pci_passthrough="yes"
   ;;
+  --enable-syslog) syslog="yes"
+  ;;
+  --disable-syslog) syslog="no"
+  ;;
   --disable-brlapi) brlapi="no"
   ;;
   --enable-brlapi) brlapi="yes"
@@ -1074,6 +1079,8 @@
 echo "  --enable-xen             enable xen backend driver support"
 echo "  --disable-xen-pci-passthrough"
 echo "  --enable-xen-pci-passthrough"
+echo "  --enable-syslog          logging into syslog (default)"
+echo "  --disable-syslog"
 echo "  --disable-brlapi         disable BrlAPI"
 echo "  --enable-brlapi          enable BrlAPI"
 echo "  --disable-vnc-tls        disable TLS encryption for VNC server"
@@ -3440,6 +3447,9 @@
 if test "$static" = "yes" ; then
   echo "CONFIG_STATIC=y" >> $config_host_mak
 fi
+if test "$syslog" = "yes"; then
+    echo "CONFIG_SYSLOG_LOGGING=y" >> $config_host_mak
+fi
 if test "$profiler" = "yes" ; then
   echo "CONFIG_PROFILER=y" >> $config_host_mak
 fi
Index: qemu-1.4.0/logging.c
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ qemu-1.4.0/logging.c	2015-05-26 13:23:51.689935102 -0400
@@ -0,0 +1,42 @@
+#include <stdarg.h>
+#include <stdio.h>
+#include <syslog.h>
+#include "logging.h"
+
+void logging_set_prefix(const char *ident)
+{
+    closelog();
+    openlog(ident, LOG_NOWAIT | LOG_PID, LOG_DAEMON);
+}
+
+static inline void __syslog_vfprintf(const char *format, va_list ap)
+{
+    vsyslog(LOG_DAEMON | LOG_NOTICE, format, ap);
+}
+
+int qemu_log_vfprintf(FILE *stream, const char *format, va_list ap)
+{
+    __syslog_vfprintf(format, ap);
+
+    return 0;
+}
+
+int qemu_log_printf(const char *format, ...)
+{
+    va_list ap;
+
+    va_start(ap, format);
+    __syslog_vfprintf(format, ap);
+
+    return 0;
+}
+
+int qemu_log_fprintf(FILE *stream, const char *format, ...)
+{
+    va_list ap;
+
+    va_start(ap, format);
+    __syslog_vfprintf(format, ap);
+
+    return 0;
+}
Index: qemu-1.4.0/logging.h
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ qemu-1.4.0/logging.h	2015-05-26 13:23:51.689935102 -0400
@@ -0,0 +1,29 @@
+#ifndef LOGGING_H_
+# define LOGGING_H_
+
+# include <stdarg.h>
+# include <stdio.h>
+
+#ifdef printf
+# undef printf
+#endif
+#ifdef vfprintf
+# undef vfprintf
+#endif
+#ifdef fprintf
+# undef fprintf
+#endif
+
+# define printf(...) qemu_log_printf(__VA_ARGS__)
+# define vfprintf(...) qemu_log_vfprintf(__VA_ARGS__)
+# define fprintf(...) qemu_log_fprintf(__VA_ARGS__)
+
+void logging_set_prefix(const char *ident);
+int qemu_log_vfprintf(FILE *stream, const char *format, va_list ap);
+int qemu_log_printf(const char *format, ...)
+  __attribute__ ((format (printf, 1, 2)));
+int qemu_log_fprintf(FILE *stream, const char *format, ...)
+  __attribute__ ((format (printf, 2, 3)));
+
+
+#endif /* !LOGGING_H_ */
Index: qemu-1.4.0/vl.c
===================================================================
--- qemu-1.4.0.orig/vl.c	2013-02-15 18:05:35.000000000 -0500
+++ qemu-1.4.0/vl.c	2015-05-26 13:25:38.778556679 -0400
@@ -3646,6 +3646,10 @@
                 break;
             case QEMU_OPTION_name:
                 qemu_name = g_strdup(optarg);
+                /* XenClient: logging-syslog */
+#ifdef CONFIG_SYSLOG_LOGGING
+                logging_set_prefix(qemu_name);
+#endif
 		 {
 		     char *p = strchr(qemu_name, ',');
 		     if (p != NULL) {
Index: qemu-1.4.0/include/qemu-common.h
===================================================================
--- qemu-1.4.0.orig/include/qemu-common.h	2013-02-15 18:05:35.000000000 -0500
+++ qemu-1.4.0/include/qemu-common.h	2015-05-26 13:23:51.689935102 -0400
@@ -430,4 +430,8 @@
 int uleb128_encode_small(uint8_t *out, uint32_t n);
 int uleb128_decode_small(const uint8_t *in, uint32_t *n);
 
+#ifdef CONFIG_SYSLOG_LOGGING
+#include "logging.h"
+#endif
+
 #endif
Index: qemu-1.4.0/xen-all.c
===================================================================
--- qemu-1.4.0.orig/xen-all.c	2015-05-26 13:25:46.414567058 -0400
+++ qemu-1.4.0/xen-all.c	2015-05-26 13:26:16.301965171 -0400
@@ -367,7 +367,7 @@
     phys_offset = physmap->phys_offset;
     size = physmap->size;
 
-    DPRINTF("unmapping vram to %"HWADDR_PRIx" - %"HWADDR_PRIx", from ",
+    DPRINTF("unmapping vram to %"HWADDR_PRIx" - %"HWADDR_PRIx", from "
             "%"HWADDR_PRIx"\n", phys_offset, phys_offset + size, start_addr);
 
     size >>= TARGET_PAGE_BITS;
