diff --git a/lib/bb/providers.py b/lib/bb/providers.py
index bd729df..347a3dc 100644
--- a/lib/bb/providers.py
+++ b/lib/bb/providers.py
@@ -361,8 +361,8 @@ def getRuntimeProviders(dataCache, rdepend):
         return rproviders
 
     # Only search dynamic packages if we can't find anything in other variables
-    for pattern in dataCache.packages_dynamic:
-        pattern = pattern.replace('+', "\+")
+    for dynamic_package in dataCache.packages_dynamic:
+        pattern = dynamic_package.replace('+', "\+")
         if pattern in regexp_cache:
             regexp = regexp_cache[pattern]
         else:
@@ -373,6 +373,6 @@ def getRuntimeProviders(dataCache, rdepend):
                 raise
             regexp_cache[pattern] = regexp
         if regexp.match(rdepend):
-            rproviders += dataCache.packages_dynamic[pattern]
+            rproviders += dataCache.packages_dynamic[dynamic_package]
 
     return rproviders
