diff --git a/bin/bitbake b/bin/bitbake
index 478ac06..68d1691 100755
--- a/bin/bitbake
+++ b/bin/bitbake
@@ -171,6 +171,9 @@ Default BBFILES are the .bb files in the current directory.""")
 
     parser.add_option("-B", "--bind", help = "The name/address for the bitbake server to bind to",
                action = "store", dest = "bind", default = False)
+
+    parser.add_option("", "--ignore-lock-failure-and-break-my-build", help = "Allow more than one copy of bitbake to run against the same tree at the same time",
+               action = "store_true", dest = "ignore_lock_failure", default = False)
     options, args = parser.parse_args(sys.argv)
 
     configuration = BBConfiguration(options)
diff --git a/lib/bb/cooker.py b/lib/bb/cooker.py
index d1de757..c714fa4 100644
--- a/lib/bb/cooker.py
+++ b/lib/bb/cooker.py
@@ -147,7 +147,10 @@ class BBCooker:
         lockfile = self.configuration.data.expand("${TOPDIR}/bitbake.lock")
         self.lock = bb.utils.lockfile(lockfile, False, False)
         if not self.lock:
-            bb.fatal("Only one copy of bitbake should be run against a build directory")
+            if not self.configuration.ignore_lock_failure:
+                bb.fatal("Only one copy of bitbake should be run against a build directory")
+            else:
+                bb.warn("A copy of bitbake is already running here, but you have chosen to ignore this. No good will come of this. You were warned.")
 
         bbpkgs = self.configuration.data.getVar('BBPKGS', True)
         if bbpkgs and len(self.configuration.pkgs_to_build) == 0:
