diff --git a/lib/bb/build.py b/lib/bb/build.py
index 95f1dcf..72a0ae3 100644
--- a/lib/bb/build.py
+++ b/lib/bb/build.py
@@ -234,6 +234,7 @@ def exec_func_shell(function, d, runfile, cwd=None):
 
     with open(runfile, 'w') as script:
         script.write('#!/bin/sh -e\n')
+        script.write("PS4='+ $(date +%s.%N) '\n")
         data.emit_func(function, script, d)
 
         if bb.msg.loggerVerboseLogs:
