Index: network-manager-applet-0.9.8.10/src/wireless-security/eap-method-peap.c
===================================================================
--- network-manager-applet-0.9.8.10.orig/src/wireless-security/eap-method-peap.c
+++ network-manager-applet-0.9.8.10/src/wireless-security/eap-method-peap.c
@@ -320,6 +320,7 @@ eap_method_peap_new (WirelessSecurity *w
 	GtkFileFilter *filter;
 	NMSetting8021x *s_8021x = NULL;
 	const char *filename;
+	const char *default_certs_dir;
 
 	parent = eap_method_init (sizeof (EAPMethodPEAP),
 	                          validate,
@@ -345,6 +346,11 @@ eap_method_peap_new (WirelessSecurity *w
 
 	widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_peap_ca_cert_button"));
 	g_assert (widget);
+
+	default_certs_dir = getenv (DEFAULT_CERTS_DIR_ENV);
+	if (default_certs_dir)
+		gtk_file_chooser_set_current_folder (GTK_FILE_CHOOSER (widget), default_certs_dir);
+
 	gtk_file_chooser_set_local_only (GTK_FILE_CHOOSER (widget), TRUE);
 	gtk_file_chooser_button_set_title (GTK_FILE_CHOOSER_BUTTON (widget),
 	                                   _("Choose a Certificate Authority certificate..."));
Index: network-manager-applet-0.9.8.10/src/wireless-security/eap-method-tls.c
===================================================================
--- network-manager-applet-0.9.8.10.orig/src/wireless-security/eap-method-tls.c
+++ network-manager-applet-0.9.8.10/src/wireless-security/eap-method-tls.c
@@ -304,9 +304,15 @@ setup_filepicker (GtkBuilder *builder,
 	GtkWidget *widget;
 	GtkFileFilter *filter;
 	const char *filename = NULL;
+	const char *default_certs_dir;
 
 	widget = GTK_WIDGET (gtk_builder_get_object (builder, name));
 	g_assert (widget);
+
+	default_certs_dir = getenv (DEFAULT_CERTS_DIR_ENV);
+	if (default_certs_dir)
+		gtk_file_chooser_set_current_folder (GTK_FILE_CHOOSER (widget), default_certs_dir);
+
 	gtk_file_chooser_set_local_only (GTK_FILE_CHOOSER (widget), TRUE);
 	gtk_file_chooser_button_set_title (GTK_FILE_CHOOSER_BUTTON (widget), title);
 
Index: network-manager-applet-0.9.8.10/src/wireless-security/eap-method-ttls.c
===================================================================
--- network-manager-applet-0.9.8.10.orig/src/wireless-security/eap-method-ttls.c
+++ network-manager-applet-0.9.8.10/src/wireless-security/eap-method-ttls.c
@@ -320,6 +320,7 @@ eap_method_ttls_new (WirelessSecurity *w
 	GtkFileFilter *filter;
 	NMSetting8021x *s_8021x = NULL;
 	const char *filename;
+	const char *default_certs_dir;
 
 	parent = eap_method_init (sizeof (EAPMethodTTLS),
 	                          validate,
@@ -345,6 +346,11 @@ eap_method_ttls_new (WirelessSecurity *w
 
 	widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_ttls_ca_cert_button"));
 	g_assert (widget);
+
+	default_certs_dir = getenv (DEFAULT_CERTS_DIR_ENV);
+	if (default_certs_dir)
+		gtk_file_chooser_set_current_folder (GTK_FILE_CHOOSER (widget), default_certs_dir);
+
 	gtk_file_chooser_set_local_only (GTK_FILE_CHOOSER (widget), TRUE);
 	gtk_file_chooser_button_set_title (GTK_FILE_CHOOSER_BUTTON (widget),
 	                                   _("Choose a Certificate Authority certificate..."));
Index: network-manager-applet-0.9.8.10/src/wireless-security/eap-method.h
===================================================================
--- network-manager-applet-0.9.8.10.orig/src/wireless-security/eap-method.h
+++ network-manager-applet-0.9.8.10/src/wireless-security/eap-method.h
@@ -25,10 +25,13 @@
 
 #include <glib.h>
 #include <gtk/gtk.h>
+#include <stdlib.h>
 
 #include <nm-connection.h>
 #include <nm-setting-8021x.h>
 
+#define DEFAULT_CERTS_DIR_ENV "DEFAULT_CERTS_DIR"
+
 typedef struct _EAPMethod EAPMethod;
 
 typedef void        (*EMAddToSizeGroupFunc) (EAPMethod *method, GtkSizeGroup *group);
