--- a/src/applet-device-ethernet.c
+++ b/src/applet-device-ethernet.c
@@ -67,29 +67,9 @@ ethernet_new_auto_connection (NMDevice *
                               AppletNewAutoConnectionCallback callback,
                               gpointer callback_data)
 {
-	NMConnection *connection;
-	NMSettingWired *s_wired = NULL;
-	NMSettingConnection *s_con;
-	char *uuid;
-
-	connection = nm_connection_new ();
-
-	s_wired = NM_SETTING_WIRED (nm_setting_wired_new ());
-	nm_connection_add_setting (connection, NM_SETTING (s_wired));
-
-	s_con = NM_SETTING_CONNECTION (nm_setting_connection_new ());
-	uuid = nm_utils_uuid_generate ();
-	g_object_set (s_con,
-	              NM_SETTING_CONNECTION_ID, DEFAULT_ETHERNET_NAME,
-	              NM_SETTING_CONNECTION_TYPE, NM_SETTING_WIRED_SETTING_NAME,
-	              NM_SETTING_CONNECTION_AUTOCONNECT, TRUE,
-	              NM_SETTING_CONNECTION_UUID, uuid,
-	              NULL);
-	g_free (uuid);
-
-	nm_connection_add_setting (connection, NM_SETTING (s_con));
-
-	(*callback) (connection, TRUE, FALSE, callback_data);
+	/* OXT: Don't create and add AUTO connection as this will prevent
+	 * ethernet devices from being a bridged connection.
+	 */
 	return TRUE;
 }
 
@@ -238,8 +218,6 @@ ethernet_add_menu_item (NMDevice *device
 
 		if (g_slist_length (connections))
 			add_connection_items (device, connections, carrier, active, ADD_INACTIVE, menu, applet);
-		else
-			add_default_connection_item (device, carrier, menu, applet);
 	}
 
 	g_slist_free (connections);
