Index: network-manager-applet-0.9.8.10/src/connection-editor/ce-page-general.ui
===================================================================
--- network-manager-applet-0.9.8.10.orig/src/connection-editor/ce-page-general.ui
+++ network-manager-applet-0.9.8.10/src/connection-editor/ce-page-general.ui
@@ -54,12 +54,12 @@
     <child>
       <object class="GtkCheckButton" id="system_checkbutton">
         <property name="label" translatable="yes">All _users may connect to this network</property>
-        <property name="visible">True</property>
-        <property name="can_focus">True</property>
+        <property name="visible">False</property>
+        <property name="can_focus">False</property>
         <property name="receives_default">False</property>
-        <property name="use_underline">True</property>
+        <property name="use_underline">False</property>
         <property name="xalign">0</property>
-        <property name="draw_indicator">True</property>
+        <property name="draw_indicator">False</property>
       </object>
       <packing>
         <property name="top_attach">1</property>
