diff --git a/part2/stages/Functions/defaults b/part2/stages/Functions/defaults
index 003521d..ebb60f0 100644
--- a/part2/stages/Functions/defaults
+++ b/part2/stages/Functions/defaults
@@ -17,7 +17,7 @@
 # Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 #
 
-DOM0_ROOT_LV_SIZE="340M"
+DOM0_ROOT_LV_SIZE="360M"
 XC_PARTITION_TYPE="bd"
 MIN_DISK_SECTORS="$((14 * 1024 * 1024 * 1024 / 512))" # 14 GB
 MIN_PARTITION_SECTORS="$(((14 * 1024 - 8) * 1024 * 1024 / 512))" # 14 GB - 8 MB
