diff --git a/part1/stages/Configure-keyboard b/part1/stages/Configure-keyboard
index b010131..7263e84 100755
--- a/part1/stages/Configure-keyboard
+++ b/part1/stages/Configure-keyboard
@@ -39,9 +39,10 @@ choose_keyboard()
 
     DEFAULT_KEYBOARD_NAME="$(get_keyboard_name "${DEFAULT_KEYBOARD}")"
     
-    awk -F: '$2 == "y" { print "\"" $5 "\" \"\"" }' "${KEYBOARDS_LIST}" |
+    awk -F: '$2 == "y" { print "\"" $5 "\"" }' "${KEYBOARDS_LIST}" |
         xargs dialog --ok-label Select \
                      --cancel-label Previous \
+                     --no-items \
                      --default-item "${DEFAULT_KEYBOARD_NAME}" \
                      --menu "Select keyboard layout" 7 40 0 2>${CAPTURE}
     OPT=$?
diff --git a/part1/stages/Choose-install-type b/part1/stages/Choose-install-type
index 3f91d90..532d903 100755
--- a/part1/stages/Choose-install-type
+++ b/part1/stages/Choose-install-type
@@ -107,19 +107,20 @@ else
 fi
 
 local ARG=
-ARG="${ARG} \"CD-ROM\" \"\"" 
-ARG="${ARG} \"Network\" \"\"" 
+ARG="${ARG} \"CD-ROM\"" 
+ARG="${ARG} \"Network\"" 
 
 if bootfs_includes_packages ; then
-    ARG="${ARG} \"Boot media\" \"\"" 
+    ARG="${ARG} \"Boot media\"" 
 else
-    ARG="${ARG} \"USB\" \"\"" 
-    ARG="${ARG} \"iSCSI\" \"\"" 
+    ARG="${ARG} \"USB\"" 
+    ARG="${ARG} \"iSCSI\"" 
 fi
 
     local CAPTURE=$(mktemp -t)
     echo ${ARG} | xargs dialog --ok-label Select \
                                 --cancel-label Previous \
+                                --no-items \
                                 --default-item "CD-ROM" \
                                 --menu "Select repository location" \
                                 7 40 0 2>${CAPTURE}
