diff -uNr xfwm4-4.6.1.orig/src/events.c xfwm4-4.6.1/src/events.c
--- xfwm4-4.6.1.orig/src/events.c	2009-04-15 08:58:59.000000000 +0100
+++ xfwm4-4.6.1/src/events.c	2011-10-08 03:05:29.703956433 +0100
@@ -2752,7 +2752,9 @@
 {
     TRACE ("entering show_popup_cb");
 
+#if 0
     show_window_menu ((Client *) data, (gint) ev->x_root, (gint) ev->y_root, ev->button, ev->time);
+#endif
 
     return (TRUE);
 }
