Index: linux-3.18.16/drivers/tty/serial/8250/8250_pnp.c
===================================================================
--- linux-3.18.16.orig/drivers/tty/serial/8250/8250_pnp.c	2015-06-14 18:19:31.000000000 +0200
+++ linux-3.18.16/drivers/tty/serial/8250/8250_pnp.c	2015-06-19 15:18:53.439749771 +0200
@@ -444,6 +444,10 @@
 	} else if (pnp_port_valid(dev, 0)) {
 		uart.port.iobase = pnp_port_start(dev, 0);
 		uart.port.iotype = UPIO_PORT;
+               /* Xen uses port 3f8 for logging, turn off ability to disable it as it breaks suspend */
+               if (uart.port.iobase == 0x3f8) {
+                       dev->capabilities &= ~PNP_DISABLE;
+               }
 	} else if (pnp_mem_valid(dev, 0)) {
 		uart.port.mapbase = pnp_mem_start(dev, 0);
 		uart.port.iotype = UPIO_MEM;
