Index: linux-3.18.16/drivers/pci/pci.c
===================================================================
--- linux-3.18.16.orig/drivers/pci/pci.c	2015-06-14 18:19:31.000000000 +0200
+++ linux-3.18.16/drivers/pci/pci.c	2015-06-19 15:19:03.236302064 +0200
@@ -4303,6 +4303,27 @@
 static char resource_alignment_param[RESOURCE_ALIGNMENT_PARAM_SIZE] = {0};
 static DEFINE_SPINLOCK(resource_alignment_lock);
 
+static
+resource_size_t pci_auto_resource_alignment(struct pci_dev *dev)
+{
+	int i;
+	struct resource *r;
+	if (dev->hdr_type == PCI_HEADER_TYPE_BRIDGE ||
+	    (dev->hdr_type == PCI_HEADER_TYPE_NORMAL &&
+	     (dev->class >> 8) == PCI_CLASS_BRIDGE_HOST))
+		return 0;
+
+	/* move resources only if they are not page aligned */
+	for (i = 0; i < PCI_BRIDGE_RESOURCES; ++i) {
+		r = &dev->resource[i];
+		if (!(r->flags & IORESOURCE_MEM))
+			continue;
+		if (r->start & (PAGE_SIZE-1))
+			return PAGE_SIZE;
+	}
+	return 0;
+}
+
 /**
  * pci_specified_resource_alignment - get resource alignment specified by user.
  * @dev: the PCI device to get
@@ -4355,6 +4376,8 @@
 		}
 		p++;
 	}
+	if (align < PAGE_SIZE)
+		align = pci_auto_resource_alignment(dev);
 	spin_unlock(&resource_alignment_lock);
 	return align;
 }
