Index: udev-182/Makefile.am
===================================================================
--- udev-182.orig/Makefile.am	2016-01-21 18:02:47.874839168 +0100
+++ udev-182/Makefile.am	2016-01-21 18:03:23.467796306 +0100
@@ -591,7 +591,7 @@
 	src/keymap/force-release-maps/samsung-other \
 	src/keymap/force-release-maps/common-volume-keys
 
-src/keymap/keys.txt: $(INCLUDE_PREFIX)/linux/input.h
+src/keymap/keys.txt: $(INCLUDE_PREFIX)/linux/input-event-codes.h
 	$(AM_V_at)mkdir -p src/keymap
 	$(AM_V_GEN)$(AWK) '/^#define.*KEY_[^ ]+[ \t]+[0-9]/ { if ($$2 != "KEY_MAX") { print $$2 } }' < $< | sed 's/^KEY_COFFEE$$/KEY_SCREENLOCK/' > $@
 
Index: udev-182/src/mtd_probe/mtd_probe.h
===================================================================
--- udev-182.orig/src/mtd_probe/mtd_probe.h	2012-01-29 01:15:46.958827013 +0100
+++ udev-182/src/mtd_probe/mtd_probe.h	2016-01-21 18:05:06.900036373 +0100
@@ -17,6 +17,7 @@
  * Boston, MA  02110-1301  USA
  */
 
+#include <stdint.h>
 #include <mtd/mtd-user.h>
 
 /* Full oob structure as written on the flash */
