diff --git a/lcptools/Makefile b/lcptools/Makefile
index 23ed5cb..8104357 100644
--- a/lcptools/Makefile
+++ b/lcptools/Makefile
@@ -94,13 +94,13 @@ $(LCP2_LIB) : pol.o poldata.o pollist.o polelt.o lcputils2.o hash.o
 	$(AR) rc $@ $^
 
 lcp_crtpolelt : crtpolelt.o $(POLELT_PLUGINS) $(LCP2_LIB)
-	$(CC) $(CFLAGS) $(LDFLAGS) $^ -lcrypto -llcp -o $@
+	$(CC) $(CFLAGS) $(LDFLAGS) $^ -lcrypto -o $@
 
 lcp_crtpollist : crtpollist.o $(POLELT_PLUGINS) $(LCP2_LIB)
-	$(CC) $(CFLAGS) $(LDFLAGS) $^ -lcrypto -llcp -o $@
+	$(CC) $(CFLAGS) $(LDFLAGS) $^ -lcrypto -o $@
 
 lcp_crtpol2 : crtpol2.o $(POLELT_PLUGINS) $(LCP2_LIB)
-	$(CC) $(CFLAGS) $(LDFLAGS) $^ -lcrypto -llcp -o $@
+	$(CC) $(CFLAGS) $(LDFLAGS) $^ -lcrypto -o $@
 
 lcp_mlehash : mlehash.o
 	$(CC) $(CFLAGS) $(LDFLAGS) $^ -lcrypto -lz -o $@
