--- a/inputattach.c	2018-02-19 10:40:19.715360299 -0500
+++ b/inputattach.c	2018-02-19 10:42:18.615363400 -0500
@@ -789,6 +789,9 @@
 { "--twiddler-joy",	"-twidjoy",	"Handykey Twiddler used as a joystick",
 	B2400, CS8,
 	SERIO_TWIDJOY,		0x00,	0x00,	0,	twiddler_init },
+{ "--snes232",		"-pad",		"snes232",
+	B115200, CS8,
+	SERIO_SNES232,		0x00,   0x00,	0,	NULL },
 #ifdef SERIO_EGALAX
 { "--eetiegalax",	"-eeti",	"EETI eGalaxTouch",
 	B9600, CS8,
