diff --git a/hw/pt-graphics.c b/hw/pt-graphics.c
index 52069ff..225c1b2 100644
--- a/hw/pt-graphics.c
+++ b/hw/pt-graphics.c
@@ -403,9 +403,7 @@ uint32_t pch_pci_read(PCIDevice *pci_dev, uint32_t config_addr, int len)
 #undef MAP_IGD_OPREGION
 int register_vga_regions(struct pt_dev *real_device)
 {
-#ifdef MAP_IGD_OPREGION
     u32 igd_opregion;
-#endif /* !MAP_IGD_OPREGION */
     int ret = 0;
 
     if ( !gfx_passthru || real_device->pci_dev->device_class != 0x0300 )
@@ -427,18 +425,32 @@ int register_vga_regions(struct pt_dev *real_device)
         case 0x8086: // INTEL
             ret |= xc_domain_ioport_mapping(xc_handle, domid, 0x3C0,
                                             0x3C0, 0x20, DPCI_ADD_MAPPING);
-#if MAP_IGD_OPREGION // FIXME: HPs are happier without opregion ... oh well ...
-            /* 1:1 map ASL Storage register value */
-            igd_opregion = pt_pci_host_read(0, 2, 0, 0xfc, 4);
-            if (igd_opregion)
+            switch (real_device->pci_dev->device_id)
             {
-                ret |= xc_domain_memory_mapping(xc_handle, domid,
-                                                igd_opregion >> XC_PAGE_SHIFT,
-                                                igd_opregion >> XC_PAGE_SHIFT,
-                                                2, DPCI_ADD_MAPPING);
-                PT_LOG("register_vga: igd_opregion = %x\n", igd_opregion);
-            }
+                case 0x0046: // Ibex Peak-M, windows drivers fail without it.
+                   igd_opregion = pt_pci_host_read(0, 2, 0, 0xfc, 4);
+                    if (igd_opregion)
+                        ret |= xc_domain_memory_mapping(xc_handle, domid,
+                                                        igd_opregion >> XC_PAGE_SHIFT,
+                                                        igd_opregion >> XC_PAGE_SHIFT,
+                                                        2, DPCI_ADD_MAPPING);
+                    break;
+                case 0x0126: // Cougarpoint-M, windows drivers fail with it ...
+                default:
+#ifdef MAP_IGD_OPREGION
+                    /* 1:1 map ASL Storage register value */
+                    igd_opregion = pt_pci_host_read(0, 2, 0, 0xfc, 4);
+                    if (igd_opregion)
+                    {
+                        ret |= xc_domain_memory_mapping(xc_handle, domid,
+                                                        igd_opregion >> XC_PAGE_SHIFT,
+                                                        igd_opregion >> XC_PAGE_SHIFT,
+                                                        2, DPCI_ADD_MAPPING);
+                        PT_LOG("igd_opregion = %x\n", igd_opregion);
+                    }
 #endif /* !MAP_IGD_OPREGION */
+                    break;
+            }
             intel_lpc_init(real_device->dev.bus);
             break;
         default:
@@ -461,9 +473,7 @@ int register_vga_regions(struct pt_dev *real_device)
  */
 int unregister_vga_regions(struct pt_dev *real_device)
 {
-#ifdef MAP_IGD_OPREGION
     u32 igd_opregion;
-#endif /* !MAP_IGD_OPREGION */
     int ret = 0;
 
     if ( !gfx_passthru || real_device->pci_dev->device_class != 0x0300 )
@@ -485,14 +495,28 @@ int unregister_vga_regions(struct pt_dev *real_device)
         case 0x8086: // INTEL
             ret |= xc_domain_ioport_mapping(xc_handle, domid, 0x3C0,
                                             0x3C0, 0x20, DPCI_REMOVE_MAPPING);
+            switch (real_device->pci_dev->device_id)
+            {
+                case 0x0046: // Ibex Peak-M
+                   igd_opregion = pt_pci_host_read(0, 2, 0, 0xfc, 4);
+                    if (igd_opregion)
+                        ret |= xc_domain_memory_mapping(xc_handle, domid,
+                                                        igd_opregion >> XC_PAGE_SHIFT,
+                                                        igd_opregion >> XC_PAGE_SHIFT,
+                                                        2, DPCI_REMOVE_MAPPING);
+                    break;
+                case 0x0126: // Cougarpoint-M
+                default:
 #ifdef MAP_IGD_OPREGION
-        igd_opregion = pt_pci_host_read(0, 2, 0, 0xfc, 4);
-        if (igd_opregion)
-            ret |= xc_domain_memory_mapping(xc_handle, domid,
-                                            igd_opregion >> XC_PAGE_SHIFT,
-                                            igd_opregion >> XC_PAGE_SHIFT,
-                                            2, DPCI_REMOVE_MAPPING);
+                    igd_opregion = pt_pci_host_read(0, 2, 0, 0xfc, 4);
+                    if (igd_opregion)
+                        ret |= xc_domain_memory_mapping(xc_handle, domid,
+                                                        igd_opregion >> XC_PAGE_SHIFT,
+                                                        igd_opregion >> XC_PAGE_SHIFT,
+                                                        2, DPCI_REMOVE_MAPPING);
 #endif /* !MAP_IGD_OPREGION */
+                    break;
+            }
             break;
         default:
             break;
