diff --git a/hw/hda-audio.c b/hw/hda-audio.c
index 8966bf5..fdfc972 100644
--- a/hw/hda-audio.c
+++ b/hw/hda-audio.c
@@ -27,6 +27,8 @@
 
 /* -------------------------------------------------------------------------- */
 
+extern int disable_audio_recording;
+
 typedef struct desc_param {
     uint32_t id;
     uint32_t val;
@@ -491,10 +493,13 @@ static void hda_audio_input_cb(void *opaque, int avail)
                 break;
             }
         }
-        rc = hda_codec_xfer(&st->state->hda, st->stream, false,
-                            st->buf, sizeof(st->buf));
-        if (!rc) {
-            break;
+        /* Only transfer input if allowed by the policy */
+        if (!disable_audio_recording) {
+          rc = hda_codec_xfer(&st->state->hda, st->stream, false,
+                              st->buf, sizeof(st->buf));
+          if (!rc) {
+             break;
+          }
         }
         st->bpos = 0;
     }
