No vkbd device for PV domains

diff --git a/hw/xen_machine_pv.c b/hw/xen_machine_pv.c
index 65c6d2f..261413a 100644
--- a/hw/xen_machine_pv.c
+++ b/hw/xen_machine_pv.c
@@ -67,7 +67,6 @@ static void xen_init_pv(ram_addr_t ram_size, int vga_ram_size,
         exit(1);
     }
     xen_be_register("console", &xen_console_ops);
-    xen_be_register("vkbd", &xen_kbdmouse_ops);
 #ifndef CONFIG_STUBDOM
     xen_be_register("qdisk", &xen_blkdev_ops);
     domid_target = xenstore_read_target();
@@ -75,11 +74,6 @@ static void xen_init_pv(ram_addr_t ram_size, int vga_ram_size,
         xenstore_scan("qdisk", domid_target, &xen_blkdev_ops);
 #endif
 
-    kbddev = xen_be_find_xendev("vkbd", xen_domid, 0);
-    if (kbddev) {
-        xen_be_check_state(kbddev);
-    }
-
     if (surfman) {
         /* Just connecting to surfman should be enough */
         dmbus_service_connect(DMBUS_SERVICE_SURFMAN,
