diff --git a/hw/pass-through.c b/hw/pass-through.c
index 8d558a7..29a21c0 100644
--- a/hw/pass-through.c
+++ b/hw/pass-through.c
@@ -4341,7 +4341,8 @@ static struct pt_dev * register_real_device(PCIBus *e_bus,
     pt_register_regions(assigned_device);
 
     /* Setup VGA bios for passthroughed gfx */
-    if ( setup_vga_pt(assigned_device) < 0 )
+    /* When qemu is in stubdom, pci-dm-helper alreay do that */
+    if ( !running_in_stubdomain && ( setup_vga_pt(assigned_device) < 0 ) )
     {
         PT_LOG("Setup VGA BIOS of passthroughed gfx failed!\n");
         return NULL;
