diff --git a/switcher.c b/switcher.c
index 544f8d1..a145412 100644
--- a/switcher.c
+++ b/switcher.c
@@ -63,6 +63,7 @@ static int                      switcher_connected = 0;
 static struct sockaddr_un       remote;
 static QEMUTimer                *switcher_timer = NULL;
 static int                      switcher_socket = -1;
+static int                      is_abs_enabled = 0;
 
 static void switcher_recv(void *opaque);
 
@@ -332,6 +333,7 @@ void switcher_send_led_code(int led_code)
 
 void switcher_abs_enabled(int enabled)
 {
+    is_abs_enabled = enabled;
     switcher_send("%c%d", DOM0_INPUT_ABS, enabled);
 }
 
@@ -368,6 +370,7 @@ static void switcher_connect(void *opaque)
         switcher_send("%c%d", DOM0_INPUT_PVM_DOMID, domid);
     else
         switcher_send("%c%d", DOM0_INPUT_DOMID, domid);
+    switcher_abs_enabled(is_abs_enabled);
 
 out:
     if (!switcher_timer)
