diff --git a/hw/xen_blktap.c b/hw/xen_blktap.c
index 24d10a3..092a640 100644
--- a/hw/xen_blktap.c
+++ b/hw/xen_blktap.c
@@ -521,7 +521,7 @@ static void handle_blktap_ctrlmsg(void* private)
 			s = state_init();
 
 			/*Open file*/
-			if (s == NULL || open_disk(s, path, msg->readonly)) {
+			if (s == NULL || open_disk(s, path, 0)) {
 				msglen = sizeof(msg_hdr_t);
 				msg->type = CTLMSG_IMG_FAIL;
 				msg->len = msglen;
