--- a/json.c
+++ b/json.c
@@ -42,7 +42,7 @@ json_parse_string_field(json_object *jva
     json_object *jfield = NULL;
     json_bool jres = json_object_object_get_ex(jvalue, field_name, &jfield);
 
-    if (NULL == jfield || is_error(jfield) || !jres)
+    if (NULL == jfield || !jres)
         return strdup_or_null(field_default_value);
 
     if (!json_object_is_type(jfield, json_type_string))
@@ -218,7 +218,7 @@ parse_json_file(const char *json_file)
     fclose(f);
 
     json_object *jobj = json_tokener_parse(json_str);
-    if (is_error(jobj))
+    if (NULL == jobj)
     {
         fprintf(stderr, "%s is not a valid json file\n", json_file);
         return NULL;
