diff --git a/policy/modules/contrib/loadkeys.te b/policy/modules/contrib/loadkeys.te
index 9338ffa..9895358 100644
--- a/policy/modules/contrib/loadkeys.te
+++ b/policy/modules/contrib/loadkeys.te
@@ -50,6 +50,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	input_server_rw_stream(loadkeys_t)
+')
+
+optional_policy(`
 	nscd_dontaudit_search_pid(loadkeys_t)
 ')
 
diff --git a/policy/modules/contrib/qemu.te b/policy/modules/contrib/qemu.te
index 1790eff..3a792dc 100644
--- a/policy/modules/contrib/qemu.te
+++ b/policy/modules/contrib/qemu.te
@@ -71,6 +71,10 @@ tunable_policy(`qemu_full_network',`
 ')
 
 optional_policy(`
+	input_server_stream_connect(qemu_t)
+')
+
+optional_policy(`
 	xserver_user_x_domain_template(qemu, qemu_t, qemu_tmpfs_t)
 ')
 
diff --git a/policy/modules/contrib/xen.te b/policy/modules/contrib/xen.te
index cce63cd..7e13d35 100644
--- a/policy/modules/contrib/xen.te
+++ b/policy/modules/contrib/xen.te
@@ -257,6 +257,10 @@ qemu_signull(xend_t)
 # transition to dmidecode
 dmidecode_domtrans(xend_t)
 
+optional_policy(`
+	input_server_dbus_chat(xend_t)
+')
+
 blktap_getattr_blk_file(xend_t)
 blktap_relabel_blk_file(xend_t)
 # this may be a leaked fd or something, dontaudit?
diff --git a/policy/modules/system/fstools.te b/policy/modules/system/fstools.te
index 4eab8df..93539c5 100644
--- a/policy/modules/system/fstools.te
+++ b/policy/modules/system/fstools.te
@@ -197,6 +197,10 @@ optional_policy(`
 	xen_rw_image_files(fsadm_t)
 ')
 
+optional_policy(`
+	input_server_write_tmp_files(fsadm_t)
+')
+
 # leaked FDs from input-server
 term_dontaudit_use_unallocated_ttys(fsadm_t)
 optional_policy(`
diff --git a/policy/modules/system/lvm.te b/policy/modules/system/lvm.te
index a706f2b..c7f4eab 100644
--- a/policy/modules/system/lvm.te
+++ b/policy/modules/system/lvm.te
@@ -351,6 +351,11 @@ optional_policy(`
 	')
 ')
 
+# redirect output from cryptsetup to a tmp file
+optional_policy(`
+	input_server_write_tmp_files(lvm_t)
+')
+
 # dontaudit FDs leaked by input-server
 optional_policy(`
 	kernel_dontaudit_read_xen_state(lvm_t)
diff --git a/policy/modules/system/mount.te b/policy/modules/system/mount.te
index d82c83f..85c3efe 100644
--- a/policy/modules/system/mount.te
+++ b/policy/modules/system/mount.te
@@ -238,6 +238,10 @@ optional_policy(`
 	unconfined_run_to(unconfined_mount_t, mount_exec_t)
 ')
 
+optional_policy(`
+	input_server_write_tmp_files(mount_t)
+')
+
 # leaked FDs from input-server
 optional_policy(`
 	kernel_dontaudit_read_xen_state(mount_t)
diff --git a/policy/modules/system/selinuxutil.te b/policy/modules/system/selinuxutil.te
index c065484..24e1206 100644
--- a/policy/modules/system/selinuxutil.te
+++ b/policy/modules/system/selinuxutil.te
@@ -655,6 +655,11 @@ optional_policy(`
 	hotplug_use_fds(setfiles_t)
 ')
 
+# input-server redirects output to a tmp file
+optional_policy(`
+	input_server_write_tmp_files(setfiles_t)
+')
+
 # leaked FDs from input-server
 optional_policy(`
 	kernel_dontaudit_read_xen_state(setfiles_t)
