diff --git a/configure.ac b/configure.ac
index dc1feee..ac20d8d 100644
--- a/configure.ac
+++ b/configure.ac
@@ -18,13 +18,12 @@
 
 # Prelude.
 AC_PREREQ(2.13)
-AC_INIT()
+m4_include([version.m4])
+AC_INIT([dm-agent], DM_AGENT_VERSION)
 
 LT_INIT
 
-PACKAGE=dm-agent
-
-AM_INIT_AUTOMAKE($PACKAGE,$VERSION)
+AM_INIT_AUTOMAKE([foreign])
 
 # Checks for programs.
 AC_PROG_CC
diff --git a/version.m4 b/version.m4
new file mode 100644
index 0000000..8fbd942
--- /dev/null
+++ b/version.m4
@@ -0,0 +1 @@
+m4_define([DM_AGENT_VERSION], $VERSION)
