commit 612502d30abfdabb836cf0b15d02c1e0ee469526
Author: Julian Pidancet <julian.pidancet@citrix.com>
Date:   Mon Nov 14 13:20:24 2011 +0000

    iPXE: Send early debug on Xen debug port

Index: src/arch/x86/prefix/libprefix.S
===================================================================
--- src.orig/arch/x86/prefix/libprefix.S
+++ src/arch/x86/prefix/libprefix.S
@@ -79,6 +79,7 @@ print_character:
 	pushw	%ax
 	pushw	%bx
 	pushw	%bp
+	outb    %al, $0xe9
 	/* If %di is non-zero, write character to buffer and exit */
 	testw	%di, %di
 	jz	1f
