diff --git a/dojo-release-1.7.2-src/dojo/i18n.js b/dojo-release-1.7.2-src/dojo/i18n.js
index e916486..2bbe05d 100644
--- a/dojo-release-1.7.2-src/dojo/i18n.js 
+++ b/dojo-release-1.7.2-src/dojo/i18n.js
@@ -148,9 +148,10 @@ define(["./_base/kernel", "require", "./has", "./_base/array", "./_base/config",
 				// use the function ctor to keep the minifiers away and come close to global scope
 				// if bundle is an AMD bundle, then __amdResult will be defined; otherwise it's a pre-amd bundle and the bundle value is returned by eval
 				new Function("bundle, __evalError",
-					"var __amdResult, define = function(x){__amdResult= x;};" +
+					"var __amdResult, __realDefine = this.define, define = function(x){__amdResult= x; };" +
 					"return [(function(){" +
 								"try{eval(arguments[0]);}catch(e){}" +
+								"define = __realDefine;" +
 								"if(__amdResult)return 0;" +
 								"try{return eval('('+arguments[0]+')');}" +
 								"catch(e){__evalError.e = e; return __evalError;}" +
-- 
1.7.5.4

