--- a/src/nm-manager.c
+++ b/src/nm-manager.c
@@ -1357,8 +1357,10 @@ system_create_virtual_device (NMManager
 			goto out;
 		}
 
-		/* FIXME: remove when we handle bridges non-destructively */
-		if (exists && !bridge_created_by_nm (self, iface)) {
+		/* FIXME: remove when we handle bridges non-destructively
+		 * OXT: Don't ignore 'brbridged*' devices
+		 */
+		if (exists && !bridge_created_by_nm (self, iface) && strncmp(iface, "brbridged", 9)) {
 			nm_log_warn (LOGD_DEVICE, "(%s): cannot use existing bridge for '%s'",
 			             iface, nm_connection_get_id (connection));
 			goto out;
@@ -2372,8 +2374,10 @@ udev_device_added_cb (NMUdevManager *ude
 			device = nm_device_bond_new (sysfs_path, iface);
 		else if (is_bridge (ifindex)) {
 
-			/* FIXME: always create device when we handle bridges non-destructively */
-			if (bridge_created_by_nm (self, iface))
+			/* FIXME: always create device when we handle bridges non-destructively
+			 * OXT: allow creation of 'brbridged*' bridge devices
+			 */
+			if (bridge_created_by_nm (self, iface) || !strncmp(iface, "brbridged", 9))
 				device = nm_device_bridge_new (sysfs_path, iface);
 			else
 				nm_log_info (LOGD_BRIDGE, "(%s): ignoring bridge not created by NetworkManager", iface);
@@ -2398,8 +2402,16 @@ udev_device_added_cb (NMUdevManager *ude
 				nm_log_err (LOGD_HW, "(%s): failed to get VLAN parent ifindex", iface);
 		} else if (is_adsl (udev_device))
 			device = nm_device_adsl_new (sysfs_path, iface, driver);
-		else
+		else {
+			/* OXT: Don't create the device if it's not a 'eth' (wired lan) interface
+			 * This is to avoid creation of non-ethernet devices as ethernet devices
+			 */
+			if (strncmp(iface, "eth", 3)) {
+				nm_log_info (LOGD_HW, "(%s): ignoring interface as an ethernet device", iface);
+				return;
+			}
 			device = nm_device_ethernet_new (sysfs_path, iface, driver);
+		}
 	}
 
 	if (device)
--- a/src/nm-device-wired.c
+++ b/src/nm-device-wired.c
@@ -147,7 +147,10 @@ carrier_action (NMDeviceWired *self, NMD
 				nm_device_queued_state_clear (device);
 		}
 	} else if (state >= NM_DEVICE_STATE_DISCONNECTED) {
-		if (!carrier && !nm_device_get_enslaved (device))
+		/* OXT: Allow state update for enslaved devices.
+		 * Ethernet devices are enslaved to bridge devices and hence this change.
+		 */
+		if (!carrier)
 			nm_device_queue_state (device, NM_DEVICE_STATE_UNAVAILABLE, NM_DEVICE_STATE_REASON_CARRIER);
 	}
 }
