diff --git a/midori/midori-browser.c b/midori/midori-browser.c
index 28c2fb9..e19c65f 100644
--- a/midori/midori-browser.c
+++ b/midori/midori-browser.c
@@ -3188,7 +3188,7 @@ static const GtkActionEntry entries[] = {
    NULL, "<Ctrl>r",
    N_("Reload the current page"), G_CALLBACK (_action_reload_stop_activate) },
  { "Stop", GTK_STOCK_STOP,
-   NULL, "Escape",
+   NULL, "",
    N_("Stop loading the current page"), G_CALLBACK (_action_reload_stop_activate) },
  { "ReloadStop", GTK_STOCK_STOP,
    NULL, "<Ctrl>r",
