--- a/policy/modules/admin/apptool.te	1969-12-31 19:00:00.000000000 -0500
+++ b/policy/modules/admin/apptool.te	2015-01-05 16:03:12.793080030 -0500
@@ -0,0 +1,70 @@
+#############################################################################
+#
+# Copyright (C) 2014 Citrix Systems, Inc.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+#############################################################################
+
+policy_module(apptool, 0.1)
+
+########################################
+#
+# Declarations
+#
+
+type apptool_t;
+type apptool_exec_t;
+application_domain(apptool_t, apptool_exec_t)
+
+type apptool_import_t;
+files_type(apptool_import_t)
+
+type apptool_disk_t;
+xc_disk_type(apptool_disk_t)
+xc_storage_filetrans(apptool_t, apptool_disk_t, { dir file })
+
+type apptool_tmp_t;
+files_tmp_file(apptool_tmp_t)
+files_tmp_filetrans(apptool_t, apptool_tmp_t, { dir file })
+
+################################
+#
+#  Local Policy
+#
+
+allow apptool_t self:fifo_file rw_fifo_file_perms;
+allow apptool_t apptool_tmp_t:dir  manage_dir_perms;
+allow apptool_t apptool_tmp_t:file manage_file_perms;
+allow apptool_t apptool_import_t:dir search_dir_perms;
+allow apptool_t apptool_import_t:file read_file_perms;
+allow apptool_t apptool_disk_t:dir manage_dir_perms;
+allow apptool_t apptool_disk_t:file manage_file_perms;
+
+dontaudit apptool_t apptool_import_t:file write;
+
+dev_dontaudit_search_sysfs(apptool_t)
+kernel_dontaudit_read_system_state(apptool_t)
+
+corecmd_exec_bin(apptool_t)
+files_read_usr_files(apptool_t)
+kernel_read_kernel_sysctls(apptool_t)
+seutil_use_newrole_fds(apptool_t)
+userdom_use_user_ptys(apptool_t)
+xc_create_client_disk_files(apptool_t)
+
+dbus_system_bus_client(apptool_t)
+dbd_dbus_chat(apptool_t)
+xen_dbus_chat(apptool_t)
