diff --git a/policy/modules/system/userdomain.fc b/policy/modules/system/userdomain.fc
index db75976..e952a8f 100644
--- a/policy/modules/system/userdomain.fc
+++ b/policy/modules/system/userdomain.fc
@@ -1,4 +1,10 @@
 HOME_DIR	-d	gen_context(system_u:object_r:user_home_dir_t,s0-mls_systemhigh)
 HOME_DIR/.+		gen_context(system_u:object_r:user_home_t,s0)
 
+#
+# /root
+#
+# Temporary fix to get the home directory for root user labeled right
+/root	-d	 gen_context(system_u:object_r:user_home_dir_t,s0)
+
 /tmp/gconfd-USER -d	gen_context(system_u:object_r:user_tmp_t,s0)
diff --git a/policy/modules/system/userdomain.if b/policy/modules/system/userdomain.if
index 20c8bf8..61d934e 100644
--- a/policy/modules/system/userdomain.if
+++ b/policy/modules/system/userdomain.if
@@ -1479,7 +1479,6 @@ interface(`userdom_getattr_user_home_dirs',`
 	')
 
 	allow $1 user_home_dir_t:dir getattr_dir_perms;
-	files_search_home($1)
 ')
 
 ########################################
@@ -3391,6 +3390,24 @@ interface(`userdom_signal_all_users',`
 
 ########################################
 ## <summary>
+##	Do not audit signals to admin user domains.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`userdom_dontaudit_signal_admin_users',`
+	gen_require(`
+		attribute admindomain;
+	')
+
+	allow $1 admindomain:process signal;
+')
+
+########################################
+## <summary>
 ##	Send a SIGCHLD signal to all user domains.
 ## </summary>
 ## <param name="domain">
