diff --git a/policy/modules.conf b/policy/modules.conf
index fd08bf5..0374cdf 100644
--- a/policy/modules.conf
+++ b/policy/modules.conf
@@ -138,6 +138,13 @@ dmesg = module
 netutils = module
 
 # Layer: admin
+# Module: statusreport
+#
+# The Fedora hardware profiler client
+#
+statusreport = module 
+
+# Layer: admin
 # Module: su
 #
 # Run shells with substitute user and group
@@ -270,6 +277,13 @@ apache = off
 # 
 apcupsd = off
 
+# Layer: admin
+# Module: apptool
+#
+# Policy for the apptool OVF import tool
+#
+apptool = module
+
 # Layer: contrib
 # Module: apm
 #
@@ -361,6 +375,13 @@ bind = off
 # 
 bird = module
 
+#Layer: services
+# Module: blktap
+#
+# blktap off
+#
+blktap = module
+
 # Layer: contrib
 # Module: bitlbee
 #
@@ -606,6 +627,13 @@ cpufreqselector = off
 # 
 cron = module
 
+# Layer: services
+# Module: ctxusb
+#
+# Policy for ctxusb.
+#
+ctxusb = module
+
 # Layer: contrib
 # Module: ctdb
 #
@@ -655,6 +683,13 @@ daemontools = off
 # 
 dante = off
 
+# Layer: apps
+# Module: db-cmd
+#
+# Policy for db-cmd.
+#
+db-cmd = module
+
 # Layer: contrib
 # Module: dbadm
 #
@@ -662,6 +697,13 @@ dante = off
 # 
 dbadm = off
 
+# Layer: services
+# Module: dbd
+#
+# dbd daemon
+#
+dbd = module 
+
 # Layer: contrib
 # Module: dbskk
 #
@@ -669,6 +711,13 @@ dbadm = off
 # 
 dbskk = off
 
+# Layer: services
+# Module: dbusbouncer
+#
+# dbd daemon
+#
+dbusbouncer = module
+
 # Layer: contrib
 # Module: dbus
 #
@@ -753,6 +802,13 @@ djbdns = off
 # 
 dkim = off
 
+# Layer: services
+# Module: dm-agent
+#
+# XenClient agent for dm (really no idea what this daemon does).
+#
+dm-agent = module
+
 # Layer: contrib
 # Module: dmidecode
 #
@@ -900,6 +956,14 @@ games = off
 # 
 gatekeeper = off
 
+# Layer: admin
+# Module: getedid
+#
+# Tools to retrieve and interpret
+# monitor specifications using the VESA VBE DDC protocol
+#
+getedid = module
+
 # Layer: contrib
 # Module: gift
 #
@@ -1012,6 +1076,13 @@ howl = off
 # 
 i18n_input = off
 
+# Layer: services
+# Module: icbinn
+#
+# The I can't believe it's not NFS server.
+#
+icbinn = module
+
 # Layer: contrib
 # Module: icecast
 #
@@ -1047,6 +1118,13 @@ inetd = off
 # 
 inn = off
 
+# Layer: services
+# Module: input-server
+#
+# input-server daemon
+#
+input-server = module 
+
 # Layer: contrib
 # Module: iodine
 #
@@ -1180,6 +1258,13 @@ ktalk = off
 # 
 kudzu = off
 
+# Layer: services
+# Module: language-sync
+#
+# XenClient daemon for managing languages.
+#
+language-sync = module
+
 # Layer: contrib
 # Module: l2tp
 #
@@ -1425,6 +1510,13 @@ ncftool = module
 # 
 nessus = off
 
+# Layer: services
+# Module: network-daemon
+#
+# XenClient daemon for managing network backends.
+#
+network-daemon = module
+
 # Layer: contrib
 # Module: networkmanager
 #
@@ -2216,6 +2308,13 @@ sssd = off
 # 
 stunnel = off
 
+# Layer: services
+# Module: surfman
+#
+# surfman daemon
+#
+surfman = module
+
 # Layer: contrib
 # Module: svnserve
 #
@@ -2237,6 +2336,13 @@ sxid = off
 # 
 sysstat = off
 
+# Layer: admin
+# Module: sysutils
+#
+# Random system utilities.
+#
+sysutils = module
+
 # Layer: contrib
 # Module: systemtap
 #
@@ -2314,6 +2420,20 @@ tmpreaper = off
 # 
 tor = off
 
+# Layer: admin
+# Module: tpmsetup
+#
+# Policy for tpm-setup
+#
+tpmsetup = module
+
+# Layer: contrib
+# Module: tpmutil
+#
+# TPM utils/tools
+#
+tpmutil = module
+
 # Layer: contrib
 # Module: transproxy
 #
@@ -2335,6 +2455,13 @@ tripwire = off
 # 
 tuned = off
 
+# Layer: admin
+# Module: txtstat
+#
+# Utility for displaying TXT information.
+#
+txtstat = module
+
 # Layer: contrib
 # Module: tvtime
 #
@@ -2356,6 +2483,13 @@ tzdata = off
 # 
 ucspitcp = off
 
+# Layer: services
+# Module: uid
+#
+# uid daemon
+#
+uid = module
+
 # Layer: contrib
 # Module: ulogd
 #
@@ -2363,6 +2497,13 @@ ucspitcp = off
 # 
 ulogd = off
 
+# Layer: services
+# Module: updatemgr
+#
+# XenClient network daemon.
+#
+updatemgr = module
+
 # Layer: contrib
 # Module: uml
 #
@@ -2447,6 +2588,13 @@ varnishd = off
 # 
 vbetool = off 
 
+# Layer: admin
+# Module: vhdutils
+#
+# Utilities for manipulating VHD files.
+#
+vhdutils = module
+
 # Layer: contrib
 # Module: vdagent
 #
@@ -2461,6 +2609,13 @@ vdagent = module
 # 
 vhostmd = off
 
+# Layer: system
+# Module: vgmch
+#
+# XenClient utility for something having to do with graphics?
+#
+vgmch = module
+
 # Layer: contrib
 # Module: virt
 #
@@ -2545,6 +2700,27 @@ wine = off
 # 
 wireshark = off 
 
+# Layer: system
+# Module: xc-files
+#
+# Policy for XenClient system files.
+#
+xc-files = base
+
+# Layer: system
+# Module: xc-installer
+#
+# XenClient installer and updater.
+#
+xc-installer = module
+
+# Layer: apps
+# Module: xec
+#
+# XenClient administrative commands.
+#
+xec = module
+
 # Layer: contrib
 # Module: wm
 #
@@ -2552,6 +2728,13 @@ wireshark = off
 # 
 wm = off 
 
+# Layer: services
+# Module: xenpmd
+#
+# xenpmd daemon
+#
+xenpmd = module
+
 # Layer: contrib
 # Module: xen
 #
@@ -2566,6 +2749,12 @@ xen = module
 # 
 xfs = off
 
+# Layer: services
+# Module: xmlstore
+#
+# xmlstore daemon, the one that stores xml
+xmlstore = module
+
 # Layer: contrib
 # Module: xguest
 #
@@ -2580,6 +2769,13 @@ xguest = off
 # 
 xprint = off
 
+# Layer: admin
+# Module: xsm
+#
+# XSM Utilities
+#
+xsmutils = module 
+
 # Layer: contrib
 # Module: xscreensaver
 #
@@ -2820,6 +3016,13 @@ mount = module
 netlabel = module
 
 # Layer: system
+# Module: pcm-config
+#
+# XenClient PCM configuration script.
+# 
+pcm-config = module
+
+# Layer: system
 # Module: selinuxutil
 #
 # Policy for SELinux policy and userland applications.
@@ -2834,6 +3037,13 @@ selinuxutil = module
 setrans = module
 
 # Layer: system
+# Module: stubdom-helpers
+#
+# XenClient stubdom helper programs.
+# 
+stubdom-helpers = module 
+
+# Layer: system
 # Module: sysnetwork
 #
 # Policy for network configuration: ifconfig and dhcp client.
