diff --git a/policy/modules/contrib/loadkeys.te b/policy/modules/contrib/loadkeys.te
index d2f4643..9338ffa 100644
--- a/policy/modules/contrib/loadkeys.te
+++ b/policy/modules/contrib/loadkeys.te
@@ -52,3 +52,10 @@ optional_policy(`
 optional_policy(`
 	nscd_dontaudit_search_pid(loadkeys_t)
 ')
+
+# allow loadkeys to read symlink at /usr/share/keymaps/ppc
+files_read_usr_symlinks(loadkeys_t)
+kernel_read_xen_state(loadkeys_t)
+kernel_write_xen_state(loadkeys_t)
+dev_rw_input_dev(loadkeys_t)
+files_read_usr_files(loadkeys_t)
