diff --git a/policy/modules/contrib/alsa.te b/policy/modules/contrib/alsa.te
index 5356c16..e3f74b0 100644
--- a/policy/modules/contrib/alsa.te
+++ b/policy/modules/contrib/alsa.te
@@ -84,3 +84,9 @@ optional_policy(`
 	hal_use_fds(alsa_t)
 	hal_write_log(alsa_t)
 ')
+
+optional_policy(`
+	statusreport_create_tmp_files(alsa_t)
+	statusreport_write_tmp_files(alsa_t)
+	statusreport_getattr_tmp_files(alsa_t)
+')
diff --git a/policy/modules/contrib/brctl.te b/policy/modules/contrib/brctl.te
index e3fe5f7..8497b2b 100644
--- a/policy/modules/contrib/brctl.te
+++ b/policy/modules/contrib/brctl.te
@@ -46,6 +46,11 @@ optional_policy(`
 ')
 
 optional_policy(`
+	statusreport_getattr_storage_files(brctl_t)
+	statusreport_write_storage_files(brctl_t)
+')
+
+optional_policy(`
 	xen_append_log(brctl_t)
 	xen_dontaudit_rw_unix_stream_sockets(brctl_t)
 ')
diff --git a/policy/modules/contrib/dmidecode.te b/policy/modules/contrib/dmidecode.te
index bc671df..0109e10 100644
--- a/policy/modules/contrib/dmidecode.te
+++ b/policy/modules/contrib/dmidecode.te
@@ -39,3 +39,7 @@ seutil_use_newrole_fds(dmidecode_t)
 userdom_use_user_terminals(dmidecode_t)
 
 xen_dontaudit_rw_unix_stream_sockets(dmidecode_t)
+
+optional_policy(`
+	statusreport_write_storage_files(dmidecode_t)
+')
diff --git a/policy/modules/contrib/xen.te b/policy/modules/contrib/xen.te
index 7e13d35..f67b18e 100644
--- a/policy/modules/contrib/xen.te
+++ b/policy/modules/contrib/xen.te
@@ -261,6 +261,10 @@ optional_policy(`
 	input_server_dbus_chat(xend_t)
 ')
 
+optional_policy(`
+	statusreport_domtrans(xend_t)
+')
+
 blktap_getattr_blk_file(xend_t)
 blktap_relabel_blk_file(xend_t)
 # this may be a leaked fd or something, dontaudit?
diff --git a/policy/modules/roles/sysadm.te b/policy/modules/roles/sysadm.te
index 67fdd47..55292a2 100644
--- a/policy/modules/roles/sysadm.te
+++ b/policy/modules/roles/sysadm.te
@@ -434,6 +434,10 @@ optional_policy(`
 	yam_run(sysadm_t, sysadm_r)
 ')
 
+optional_policy(`
+	statusreport_run(sysadm_t, sysadm_r)
+')
+
 ifndef(`distro_redhat',`
 	optional_policy(`
 		auth_role(sysadm_r, sysadm_t)
diff --git a/policy/modules/system/lvm.te b/policy/modules/system/lvm.te
index c7f4eab..4a659c6 100644
--- a/policy/modules/system/lvm.te
+++ b/policy/modules/system/lvm.te
@@ -367,6 +367,11 @@ optional_policy(`
 ')
 
 optional_policy(`
+	statusreport_write_storage_files(lvm_t)
+	statusreport_getattr_storage_files(lvm_t)
+')
+
+optional_policy(`
 	rpm_manage_script_tmp_files(lvm_t)
 ')
 
diff --git a/policy/modules/system/mount.te b/policy/modules/system/mount.te
index 85c3efe..c72fe50 100644
--- a/policy/modules/system/mount.te
+++ b/policy/modules/system/mount.te
@@ -247,3 +247,8 @@ optional_policy(`
 	kernel_dontaudit_read_xen_state(mount_t)
 	kernel_dontaudit_write_xen_state(mount_t)
 ')
+
+optional_policy(`
+	statusreport_write_storage_files(mount_t)
+	statusreport_getattr_storage_files(mount_t)
+')
diff --git a/policy/modules/system/sysnetwork.te b/policy/modules/system/sysnetwork.te
index c73d06b..62d82c3 100644
--- a/policy/modules/system/sysnetwork.te
+++ b/policy/modules/system/sysnetwork.te
@@ -367,6 +367,11 @@ optional_policy(`
 ')
 
 optional_policy(`
+	statusreport_write_storage_files(ifconfig_t)
+	statusreport_getattr_storage_files(ifconfig_t)
+')
+
+optional_policy(`
 	unconfined_dontaudit_rw_pipes(ifconfig_t)
 ')
 
