diff --git a/policy/modules/contrib/networkmanager.te b/policy/modules/contrib/networkmanager.te
index 2d4dc98..64495aa 100644
--- a/policy/modules/contrib/networkmanager.te
+++ b/policy/modules/contrib/networkmanager.te
@@ -351,6 +351,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+	uid_dbus_send(NetworkManager_t)
+')
+
+optional_policy(`
 	vpn_domtrans(NetworkManager_t)
 	vpn_kill(NetworkManager_t)
 	vpn_signal(NetworkManager_t)
diff --git a/policy/modules/contrib/xen.te b/policy/modules/contrib/xen.te
index 680bf1c..c77eb19 100644
--- a/policy/modules/contrib/xen.te
+++ b/policy/modules/contrib/xen.te
@@ -272,6 +272,7 @@ tapctl_domtrans(xend_t)
 
 dbd_dbus_chat(xend_t)
 surfman_dbus_chat(xend_t)
+uid_dbus_chat(xend_t)
 
 kernel_read_kernel_sysctls(xend_t)
 kernel_read_system_state(xend_t)
