diff -uNr network-manager-applet-0.9.2.0.orig/src/wireless-security/eap-method-peap.c network-manager-applet-0.9.2.0/src/wireless-security/eap-method-peap.c
--- network-manager-applet-0.9.2.0.orig/src/wireless-security/eap-method-peap.c	2011-10-21 07:15:25.000000000 +0100
+++ network-manager-applet-0.9.2.0/src/wireless-security/eap-method-peap.c	2013-03-12 21:17:01.350618183 +0000
@@ -320,6 +320,7 @@
 	GtkFileFilter *filter;
 	NMSetting8021x *s_8021x = NULL;
 	const char *filename;
+	const char *default_certs_dir;
 
 	parent = eap_method_init (sizeof (EAPMethodPEAP),
 	                          validate,
@@ -345,6 +346,11 @@
 
 	widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_peap_ca_cert_button"));
 	g_assert (widget);
+
+	default_certs_dir = getenv (DEFAULT_CERTS_DIR_ENV);
+	if (default_certs_dir)
+		gtk_file_chooser_set_current_folder (GTK_FILE_CHOOSER (widget), default_certs_dir);
+
 	gtk_file_chooser_set_local_only (GTK_FILE_CHOOSER (widget), TRUE);
 	gtk_file_chooser_button_set_title (GTK_FILE_CHOOSER_BUTTON (widget),
 	                                   _("Choose a Certificate Authority certificate..."));
diff -uNr network-manager-applet-0.9.2.0.orig/src/wireless-security/eap-method-tls.c network-manager-applet-0.9.2.0/src/wireless-security/eap-method-tls.c
--- network-manager-applet-0.9.2.0.orig/src/wireless-security/eap-method-tls.c	2011-10-21 07:15:25.000000000 +0100
+++ network-manager-applet-0.9.2.0/src/wireless-security/eap-method-tls.c	2013-03-12 21:17:09.721000681 +0000
@@ -291,9 +291,15 @@
 	GtkWidget *widget;
 	GtkFileFilter *filter;
 	const char *filename = NULL;
+	const char *default_certs_dir;
 
 	widget = GTK_WIDGET (gtk_builder_get_object (builder, name));
 	g_assert (widget);
+
+	default_certs_dir = getenv (DEFAULT_CERTS_DIR_ENV);
+	if (default_certs_dir)
+		gtk_file_chooser_set_current_folder (GTK_FILE_CHOOSER (widget), default_certs_dir);
+
 	gtk_file_chooser_set_local_only (GTK_FILE_CHOOSER (widget), TRUE);
 	gtk_file_chooser_button_set_title (GTK_FILE_CHOOSER_BUTTON (widget), title);
 
diff -uNr network-manager-applet-0.9.2.0.orig/src/wireless-security/eap-method-ttls.c network-manager-applet-0.9.2.0/src/wireless-security/eap-method-ttls.c
--- network-manager-applet-0.9.2.0.orig/src/wireless-security/eap-method-ttls.c	2011-10-21 07:15:25.000000000 +0100
+++ network-manager-applet-0.9.2.0/src/wireless-security/eap-method-ttls.c	2013-03-12 21:17:24.868977293 +0000
@@ -320,6 +320,7 @@
 	GtkFileFilter *filter;
 	NMSetting8021x *s_8021x = NULL;
 	const char *filename;
+	const char *default_certs_dir;
 
 	parent = eap_method_init (sizeof (EAPMethodTTLS),
 	                          validate,
@@ -345,6 +346,11 @@
 
 	widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_ttls_ca_cert_button"));
 	g_assert (widget);
+
+	default_certs_dir = getenv (DEFAULT_CERTS_DIR_ENV);
+	if (default_certs_dir)
+		gtk_file_chooser_set_current_folder (GTK_FILE_CHOOSER (widget), default_certs_dir);
+
 	gtk_file_chooser_set_local_only (GTK_FILE_CHOOSER (widget), TRUE);
 	gtk_file_chooser_button_set_title (GTK_FILE_CHOOSER_BUTTON (widget),
 	                                   _("Choose a Certificate Authority certificate..."));
diff -uNr network-manager-applet-0.9.2.0.orig/src/wireless-security/eap-method.h network-manager-applet-0.9.2.0/src/wireless-security/eap-method.h
--- network-manager-applet-0.9.2.0.orig/src/wireless-security/eap-method.h	2011-10-21 07:15:25.000000000 +0100
+++ network-manager-applet-0.9.2.0/src/wireless-security/eap-method.h	2013-03-12 21:16:26.153071085 +0000
@@ -29,6 +29,8 @@
 #include <nm-connection.h>
 #include <nm-setting-8021x.h>
 
+#define DEFAULT_CERTS_DIR_ENV "DEFAULT_CERTS_DIR"
+
 typedef struct _EAPMethod EAPMethod;
 
 typedef void        (*EMAddToSizeGroupFunc) (EAPMethod *method, GtkSizeGroup *group);
diff --git a/src/wireless-security/eap-method.h b/src/wireless-security/eap-method.h
index 36d02db..239f061 100644
--- a/src/wireless-security/eap-method.h
+++ b/src/wireless-security/eap-method.h
@@ -25,6 +25,7 @@
 
 #include <glib.h>
 #include <gtk/gtk.h>
+#include <stdlib.h>
 
 #include <nm-connection.h>
 #include <nm-setting-8021x.h>
