--- network-manager-applet-0.9.2.0.orig/src/connection-editor/nm-connection-editor.ui	2012-07-13 16:54:53.000000000 +0000
+++ network-manager-applet-0.9.2.0/src/connection-editor/nm-connection-editor.ui	2012-07-13 17:08:49.000000000 +0000
@@ -567,10 +567,10 @@
                 <child>
                   <object class="GtkCheckButton" id="system_checkbutton">
                     <property name="label" translatable="yes">Available to all users</property>
-                    <property name="visible">True</property>
-                    <property name="can_focus">True</property>
-                    <property name="receives_default">False</property>
-                    <property name="draw_indicator">True</property>
+                    <property name="visible">False</property>
+                    <property name="can_focus">False</property>
+                    <property name="receives_default">True</property>
+                    <property name="draw_indicator">False</property>
                   </object>
                   <packing>
                     <property name="expand">False</property>
