diff --git a/tboot/common/tboot.c b/tboot/common/tboot.c
index cef5090..df79a15 100644
--- a/tboot/common/tboot.c
+++ b/tboot/common/tboot.c
@@ -193,7 +193,7 @@ static void post_launch(void)
     /* protect ourselves, MLE page table, and MLE/kernel shared page */
     base = (uint64_t)TBOOT_BASE_ADDR;
     size = (uint64_t)get_tboot_mem_end() - base;
-    uint32_t mem_type = is_kernel_linux() ? E820_RESERVED : E820_UNUSABLE;
+    uint32_t mem_type = E820_RESERVED;
     printk("protecting tboot (%Lx - %Lx) in e820 table\n", base,
            (base + size - 1));
     if ( !e820_protect_region(base, size, mem_type) )
