From b27e3c385e66594687f58a6f7eedff4374f5929c Mon Sep 17 00:00:00 2001
From: Tamas K Lengyel <tamas@tklengyel.com>
Date: Fri, 2 Mar 2018 12:59:39 -0700
Subject: [PATCH] Disable #warning

---
 libdpe/pe_updatefile.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libdpe/pe_updatefile.c b/libdpe/pe_updatefile.c
index 7caefd9..6ac4309 100644
--- a/libdpe/pe_updatefile.c
+++ b/libdpe/pe_updatefile.c
@@ -132,7 +132,7 @@ __pe_updatemmap(Pe *pe, size_t shnum __attribute__((__unused__)))
 	char *msync_end = (char *)dd + sizeof(*dd);
 	msync(msync_start, msync_end - msync_start, MS_SYNC);
 
-	#warning this is not done yet.
+	//#warning this is not done yet.
 	//struct section_header *sh = __get_last_section(pe);
 
 	size_t dd_size = sizeof (*dd) / sizeof (dd->exports);
-- 
2.11.0

