From 95ca2a91d7d72f179cb245fdbc5018bd7b57605c Mon Sep 17 00:00:00 2001
From: Tamas K Lengyel <tamas@tklengyel.com>
Date: Fri, 2 Mar 2018 13:37:18 -0700
Subject: [PATCH 2/2] Init error

---
 src/pesign.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/pesign.c b/src/pesign.c
index 6ceda34..39a5757 100644
--- a/src/pesign.c
+++ b/src/pesign.c
@@ -660,7 +660,7 @@ main(int argc, char *argv[])
 
 	if (!daemon) {
 		SECStatus status;
-		int error;
+		int error = 0;
 		if (need_db) {
 			status = NSS_Init(certdir);
 			if (status != SECSuccess) {
-- 
2.11.0

