Index: camomile-0.8.1/Makefile.in
===================================================================
--- camomile-0.8.1.orig/Makefile.in	2010-07-08 07:57:50.000000000 +0200
+++ camomile-0.8.1/Makefile.in	2016-02-24 19:57:17.623840470 +0100
@@ -406,19 +406,20 @@
 	if [ -f camomile.cma ]; then files="camomile.cma $$files"; fi&& \
 	if [ -f camomile.cmxa ]; then files="camomile.cmxa $$files"; fi&& \
 	if [ -f camomile.a ]; then files="camomile.a $$files"; fi && \
-	ocamlfind install camomile $$files META $(MLI) camomileLibrary.cmi camomileLibraryDefault.cmi camomileLibraryDyn.cmi
+	mkdir -p $(DESTDIR)$(shell ocamlfind printconf destdir) && \
+	ocamlfind install -destdir $(DESTDIR)$(shell ocamlfind printconf destdir) camomile $$files META $(MLI) camomileLibrary.cmi camomileLibraryDefault.cmi camomileLibraryDyn.cmi
 
 install-data:
-	mkdir -p '$(DATADIR)'
-	mkdir -p '$(DATADIR)'/camomile
-	mkdir -p '$(DATADIR)'/camomile/database
-	cp -f database/*.mar '$(DATADIR)'/camomile/database || true
-	mkdir -p '$(DATADIR)'/camomile/charmaps
-	cp -f charmaps/*.mar '$(DATADIR)'/camomile/charmaps || true
-	mkdir -p '$(DATADIR)'/camomile/mappings
-	cp -f mappings/*.mar '$(DATADIR)'/camomile/mappings || true
-	mkdir -p '$(DATADIR)'/camomile/locales
-	cp -f locales/*.mar '$(DATADIR)'/camomile/locales || true
+	mkdir -p '$(DESTDIR)$(DATADIR)'
+	mkdir -p '$(DESTDIR)$(DATADIR)'/camomile
+	mkdir -p '$(DESTDIR)$(DATADIR)'/camomile/database
+	cp -f database/*.mar '$(DESTDIR)$(DATADIR)'/camomile/database || true
+	mkdir -p '$(DESTDIR)$(DATADIR)'/camomile/charmaps
+	cp -f charmaps/*.mar '$(DESTDIR)$(DATADIR)'/camomile/charmaps || true
+	mkdir -p '$(DESTDIR)$(DATADIR)'/camomile/mappings
+	cp -f mappings/*.mar '$(DESTDIR)$(DATADIR)'/camomile/mappings || true
+	mkdir -p '$(DESTDIR)$(DATADIR)'/camomile/locales
+	cp -f locales/*.mar '$(DESTDIR)$(DATADIR)'/camomile/locales || true
 
 uninstall: $(UNINSTALL)
 
