--- a/cpumond/cpumond.c
+++ b/cpumond/cpumond.c
@@ -119,7 +119,7 @@ err:
     return NULL;
 }
 
-inline int statread(int statfd, long long *total, long long *idle){
+static int statread(int statfd, long long *total, long long *idle){
     long long val[10];
     char      buf[256];
     int       i, err = 0;
--- a/vhd/lib/vhd-util-read.c
+++ b/vhd/lib/vhd-util-read.c
@@ -42,7 +42,7 @@
 #include "libvhd.h"
 #include "vhd-util.h"
 
-#define nsize     15
+#define nsize     20
 static char nbuf[nsize];
 
 static inline char *
--- a/drivers/tapdisk-control.c
+++ b/drivers/tapdisk-control.c
@@ -1083,7 +1083,7 @@ static int
 tapdisk_control_stats(struct tapdisk_ctl_conn *conn,
 		      tapdisk_message_t *request, tapdisk_message_t * const response)
 {
-	td_stats_t _st, *st = &_st;
+	td_stats_t _st = { 0 }, *st = &_st;
 	td_vbd_t *vbd;
 	size_t rv;
 	void *buf;
--- a/drivers/tapdisk-image.c
+++ b/drivers/tapdisk-image.c
@@ -312,7 +312,7 @@ __tapdisk_image_open_chain(int type, con
 	list_add_tail(&image->next, &head);
 
 	if (unlikely(prt_devnum >= 0)) {
-		char dev[32];
+		char dev[35];
 		snprintf(dev, sizeof(dev),
 			 "%s%d", BLKTAP2_IO_DEVICE, prt_devnum);
 		err = tapdisk_image_open(DISK_TYPE_AIO, dev,
