diff --git a/tools/layoutrom.py b/tools/layoutrom.py
index 20b9716..4a5885a 100755
--- a/tools/layoutrom.py
+++ b/tools/layoutrom.py
@@ -201,7 +201,7 @@ def doLayout(sections):
     li.sections32fseg = getSectionsCategory(sections, '32fseg')
 
     li.sec32fseg_start, li.sec32fseg_align = setSectionsStart(
-        li.sections32fseg, li.sec32seg_start, 16
+        li.sections32fseg, li.sec32seg_start, 4096
         , segoffset=BUILD_BIOS_ADDR)
 
     # Determine 32flat runtime positions
@@ -213,16 +213,17 @@ def doLayout(sections):
 
     li.sec32flat_start, li.sec32flat_align = setSectionsStart(
         textsections + rodatasections + datasections + bsssections
-        , li.sec32fseg_start, 16)
+        , li.sec32fseg_start, 4096)
     li.zonefseg_end = li.sec32flat_start
     li.zonefseg_start = BUILD_BIOS_ADDR
+
     if li.zonefseg_start + BUILD_MIN_BIOSTABLE > li.zonefseg_end:
         # Not enough ZoneFSeg space - force a minimum space.
         li.zonefseg_end = li.sec32fseg_start
         li.zonefseg_start = li.zonefseg_end - BUILD_MIN_BIOSTABLE
         li.sec32flat_start, li.sec32flat_align = setSectionsStart(
             textsections + rodatasections + datasections + bsssections
-            , li.zonefseg_start, 16)
+            , li.zonefseg_start, 4096)
 
     # Determine 32flat init positions
     li.sections32init = getSectionsCategory(sections, '32init')
@@ -233,7 +234,7 @@ def doLayout(sections):
 
     li.sec32init_start, li.sec32init_align = setSectionsStart(
         textsections + rodatasections + datasections + bsssections
-        , li.sec32flat_start, 16)
+        , li.sec32flat_start, 4096)
 
     # Determine "low memory" data positions
     li.sections32low = getSectionsCategory(sections, '32low')
@@ -243,7 +244,7 @@ def doLayout(sections):
     zonelow_base = final_sec32low_end - 64*1024
     li.zonelow_base = max(BUILD_ROM_START, alignpos(zonelow_base, 2*1024))
     li.sec32low_start, li.sec32low_align = setSectionsStart(
-        li.sections32low, sec32low_end, 16
+        li.sections32low, sec32low_end, 4096
         , segoffset=li.zonelow_base - relocdelta)
     li.final_sec32low_start = li.sec32low_start + relocdelta
 
