diff --git a/vgasrc/vbe.c b/vgasrc/vbe.c
index d962333..0ce598c 100644
--- a/vgasrc/vbe.c
+++ b/vgasrc/vbe.c
@@ -18,6 +18,7 @@ u32 VBE_total_memory VAR16 = 256 * 1024;
 u32 VBE_capabilities VAR16;
 u32 VBE_framebuffer VAR16;
 u16 VBE_win_granularity VAR16;
+u8 VBE_linesize_align VAR16 = 1;
 
 static void
 vbe_104f00(struct bregs *regs)
@@ -100,7 +101,8 @@ vbe_104f01(struct bregs *regs)
     // Basic information about mode.
     int width = GET_GLOBAL(vmode_g->width);
     int height = GET_GLOBAL(vmode_g->height);
-    int linesize = DIV_ROUND_UP(width * vga_bpp(vmode_g), 8);
+    int linesize = ALIGN(DIV_ROUND_UP(width * vga_bpp(vmode_g), 8),
+                         GET_GLOBAL(VBE_linesize_align));
     SET_FARVAR(seg, info->bytes_per_scanline, linesize);
     SET_FARVAR(seg, info->xres, width);
     SET_FARVAR(seg, info->yres, height);
