Index: git/grub-core/loader/i386/multiboot_mbi.c
===================================================================
--- git.orig/grub-core/loader/i386/multiboot_mbi.c
+++ git/grub-core/loader/i386/multiboot_mbi.c
@@ -356,13 +356,6 @@ retrieve_video_parameters (struct multib
   grub_video_driver_id_t driv_id;
   struct grub_video_palette_data palette[256];
 
-  err = grub_multiboot_set_video_mode ();
-  if (err)
-    {
-      grub_print_error ();
-      grub_errno = GRUB_ERR_NONE;
-    }
-
   grub_video_get_palette (0, ARRAY_SIZE (palette), palette);
 
   driv_id = grub_video_get_driver_id ();
