Index: git/grub-core/normal/menu.c
===================================================================
--- git.orig/grub-core/normal/menu.c
+++ git/grub-core/normal/menu.c
@@ -580,6 +580,17 @@ run_menu (grub_menu_t menu, int nested,
   int default_entry, current_entry;
   int timeout;
   enum timeout_style timeout_style;
+  const char *tmp;
+  int edit_and_shell = 0;
+
+  tmp = grub_env_get("edit_and_shell");
+  if (tmp &&
+      (!grub_strcmp(tmp, "1") ||
+       !grub_strcmp(tmp, "true") ||
+       !grub_strcmp(tmp, "yes")))
+  {
+    edit_and_shell = 1;
+  }
 
   default_entry = get_entry_number (menu, "default");
 
@@ -772,18 +783,26 @@ run_menu (grub_menu_t menu, int nested,
 	      break;
 
 	    case 'c':
-	      menu_fini ();
-	      grub_cmdline_run (1, 0);
-	      goto refresh;
+	      if (edit_and_shell)
+	        {
+	          menu_fini ();
+	          grub_cmdline_run (1, 0);
+	          goto refresh;
+	        }
+	      break;
 
 	    case 'e':
-	      menu_fini ();
-		{
-		  grub_menu_entry_t e = grub_menu_get_entry (menu, current_entry);
-		  if (e)
-		    grub_menu_entry_run (e);
-		}
-	      goto refresh;
+	      if (edit_and_shell)
+	        {
+	          menu_fini ();
+		    {
+		      grub_menu_entry_t e = grub_menu_get_entry (menu, current_entry);
+		      if (e)
+		        grub_menu_entry_run (e);
+		    }
+	          goto refresh;
+	        }
+              break;
 
 	    default:
 	      {
