diff --git a/dbus_stubs.c b/dbus_stubs.c
index ae3f208..8500f38 100644
--- a/dbus_stubs.c
+++ b/dbus_stubs.c
@@ -573,7 +573,7 @@ value stub_dbus_connection_dispatch(value bus)
 	DBusDispatchStatus status;
 
 	status = dbus_connection_dispatch(DBusConnection_val(bus));
-	caml_alloc_variant(ret, Val_int(find_index_equal(status, __dispatch_status_table)));
+	caml_alloc_variant(ret, find_index_equal(status, __dispatch_status_table));
 	CAMLreturn(ret);
 }
 
@@ -584,7 +584,7 @@ value stub_dbus_connection_get_dispatch_status(value bus)
 	DBusDispatchStatus status;
 
 	status = dbus_connection_get_dispatch_status(DBusConnection_val(bus));
-	caml_alloc_variant(ret, Val_int(find_index_equal(status, __dispatch_status_table)));
+	caml_alloc_variant(ret, find_index_equal(status, __dispatch_status_table));
 	CAMLreturn(ret);
 }
 
