diff --git a/dbus_stubs.c b/dbus_stubs.c
index 9cacb77..d1059d5 100644
--- a/dbus_stubs.c
+++ b/dbus_stubs.c
@@ -275,6 +275,15 @@ static void raise_dbus_type_not_supported(char *s)
 	caml_raise_with_string(*dbus_err, s);
 }
 
+static void init_dbus(void)
+{
+        static int done = 0;
+        if (!done) {
+                dbus_threads_init_default();
+                done = 1;
+        }
+}
+
 /******************** BUS **********************/
 value stub_dbus_bus_get(value type)
 {
@@ -283,6 +292,7 @@ value stub_dbus_bus_get(value type)
 	DBusConnection *c_con;
 	DBusError error;
 
+        init_dbus();
 	dbus_error_init(&error);
 	c_con = dbus_bus_get(__bustype_table[Int_val(type)], &error);
 	if (!c_con)
@@ -299,6 +309,7 @@ value stub_dbus_bus_get_private(value type)
 	DBusConnection *c_con;
 	DBusError error;
 
+        init_dbus();
 	dbus_error_init(&error);
 	c_con = dbus_bus_get_private(__bustype_table[Int_val(type)], &error);
 	if (!c_con)
