diff --git a/loginutils/chpasswd.c b/loginutils/chpasswd.c
index c83d1da..33ae3ab 100644
--- a/loginutils/chpasswd.c
+++ b/loginutils/chpasswd.c
@@ -54,7 +54,7 @@ int chpasswd_main(int argc UNUSED_PARAM, char **argv)
 		 * we try to find & change his passwd in /etc/passwd */
 #if ENABLE_FEATURE_SHADOWPASSWDS
 		rc = update_passwd(bb_path_shadow_file, name, pass);
-		if (rc == 0) /* no lines updated, no errors detected */
+		if (rc < 1) /* no lines updated, or error detected (e.g. no /etc/shadow file) */
 #endif
 			rc = update_passwd(bb_path_passwd_file, name, pass);
 		/* LOGMODE_BOTH logs to syslog also */
