diff --git a/configure.in b/configure.in
index 96f05e9..be0a280 100644
--- a/configure.in
+++ b/configure.in
@@ -1134,7 +1134,7 @@ if test x$SELINUX = xyes; then
 		AC_CHECK_HEADERS([selinux/selinux.h],, hard_bailout)
 		AC_CHECK_HEADERS([selinux/label.h])
 		AC_DEFINE([HAVE_SELINUX], 1, [Define to 1 to include support for selinux.])
-		SELINUX_LIBS="-lselinux $SELINUX_LIBS"
+		SELINUX_LIBS="-lselinux -lpcre -pthread $SELINUX_LIBS"
 		SELINUX_PC="libselinux"
 		HAVE_SELINUX=yes ], [
 		AC_MSG_WARN(Disabling selinux)
