From 5210924d115ed5ed54023d7f19580de2e4bc237f Mon Sep 17 00:00:00 2001
From: Tamas K Lengyel <tamas@tklengyel.com>
Date: Fri, 22 Dec 2017 02:28:59 +0100
Subject: [PATCH] Only check for large file support when needed

---
 configure.ac | 19 ++++++++++++-------
 1 file changed, 12 insertions(+), 7 deletions(-)

diff --git a/configure.ac b/configure.ac
index edd184154..7dcac7801 100644
--- a/configure.ac
+++ b/configure.ac
@@ -368,13 +368,18 @@ case "$host_os" in
    ;;
 esac
 
-case "$host_os" in
-  cygwin | windows* | mingw32* | aros*)
-     ;;
-  *)
-     AC_CHECK_SIZEOF(off_t)
-     test x"$ac_cv_sizeof_off_t" = x8 || AC_MSG_ERROR([Large file support is required]);;
-esac
+AC_ARG_ENABLE([largefile],
+              [AS_HELP_STRING([--enable-largefile],
+                             [Enable large file support (default=guessed)])])
+if test x"$enable_largefile" = xyes ; then
+  case "$host_os" in
+    cygwin | windows* | mingw32* | aros*)
+       ;;
+    *)
+       AC_CHECK_SIZEOF(off_t)
+       test x"$ac_cv_sizeof_off_t" = x8 || AC_MSG_ERROR([Large file support is required]);;
+  esac
+fi
 
 if test x$USE_NLS = xno; then
   HOST_CFLAGS="$HOST_CFLAGS -fno-builtin-gettext"
-- 
2.11.0

