diff --git a/.gitignore b/.gitignore
new file mode 100644
index 0000000..2075a08
--- /dev/null
+++ b/.gitignore
@@ -0,0 +1,65 @@
+#build system side-effects
+.stamp_autoconfigured
+.stamp_configured
+.stamp_downloaded
+.stamp_libtool_patch
+.stamp_patched
+
+#autotools side-effects
+libxa/configure
+libxa/configure.lineno
+libxa/config.guess
+libxa/config.h.in
+libxa/config.sub
+libxa/install-sh
+libxa/missing
+libxa/compile
+libxa/config.h
+libxa/*.log
+libxa/*.status
+libxa/stamp-h*
+libxa/.stamp_*
+libxa/libtool
+libxa/depcomp
+libxa/ltmain.sh
+libxa/m4/*
+libxa/xenaccess/.deps/
+libxa/xenaccess/.deps/*
+libxa/examples/.deps/
+libxa/examples/.deps/*
+libxa/aclocal.m4
+libxa/autom4te.cache/*
+libxa/autom4te.cache/
+libxa/Makefile
+libxa/Makefile.in
+libxa/xenaccess/Makefile
+libxa/xenaccess/Makefile.in
+libxa/examples/Makefile
+libxa/examples/Makefile.in
+libxa/.applied_patches_list
+libxa/xenaccess.pc
+
+#compilation side-effects
+libxa/ylwrap
+libxa/xenaccess/xenaccess.h
+libxa/xenaccess/*.o
+libxa/xenaccess/*.lo
+libxa/xenaccess/*.la
+libxa/xenaccess/config/
+libxa/xenaccess/config/*
+libxa/xenaccess/.libs/
+libxa/xenaccess/.libs/*
+libxa/examples/*.o
+libxa/examples/dump-memory
+libxa/examples/map-addr
+libxa/examples/map-symbol
+libxa/examples/module-list
+libxa/examples/process-data
+libxa/examples/process-list
+libxa/examples/process-list-file
+
+#emacs side effects
+*~
+libxa/*~
+libxa/xenaccess/*~
+libxa/examples/*~
