Index: webkit-1.8.3/Source/WebCore/platform/gtk/GtkClickCounter.cpp
===================================================================
--- webkit-1.8.3.orig/Source/WebCore/platform/gtk/GtkClickCounter.cpp
+++ webkit-1.8.3/Source/WebCore/platform/gtk/GtkClickCounter.cpp
@@ -85,8 +85,8 @@ int GtkClickCounter::clickCountForGdkBut
     guint32 eventTime = getEventTime(event);
 
     if ((event->type == GDK_2BUTTON_PRESS || event->type == GDK_3BUTTON_PRESS)
-        || ((abs(buttonEvent->x - m_previousClickPoint.x()) < doubleClickDistance)
-            && (abs(buttonEvent->y - m_previousClickPoint.y()) < doubleClickDistance)
+        || ((fabs(buttonEvent->x - m_previousClickPoint.x()) < doubleClickDistance)
+            && (fabs(buttonEvent->y - m_previousClickPoint.y()) < doubleClickDistance)
             && (eventTime - m_previousClickTime < static_cast<guint>(doubleClickTime))
             && (buttonEvent->button == m_previousClickButton)))
         m_currentClickCount++;
