--- s/Source/WebCore/page/Settings.cpp   2012-02-02 13:58:01.438214650 +0000
+++ d/Source/WebCore/page/Settings.cpp       2012-02-02 14:00:28.277615975 +0000
@@ -223,7 +223,7 @@
     , m_scrollAnimatorEnabled(true)
 #endif
 #if ENABLE(WEB_SOCKETS)
-    , m_useHixie76WebSocketProtocol(true)
+    , m_useHixie76WebSocketProtocol(false)
 #endif
     , m_mediaPlaybackRequiresUserGesture(false)
     , m_mediaPlaybackAllowsInline(true)

