diff --git a/libxklavier/Makefile.am b/libxklavier/Makefile.am
index 1101f9a..af0ebcd 100644
--- a/libxklavier/Makefile.am
+++ b/libxklavier/Makefile.am
@@ -43,6 +43,8 @@ libxklavier_la_LIBADD = \
  $(LIBXKBFILE_PRESENT_LDFLAGS) \
  $(X_LIBS) -lX11 $(LIBICONV) 
 
+BUILT_SOURCES = $(xklavier_built_headers) $(xklavier_built_cfiles)
+
 EXTRA_DIST=marshal.list
 
 GLIB_GENMARSHAL = `$PKG_CONFIG --variable=glib_genmarshal glib-2.0`
