diff -ru microcode_ctl-1.17-orig/microcode_ctl.start microcode_ctl-1.17/microcode_ctl.start
--- microcode_ctl-1.17-orig/microcode_ctl.start	2011-01-10 15:51:32.032966856 +0000
+++ microcode_ctl-1.17/microcode_ctl.start	2011-01-10 15:46:37.529262513 +0000
@@ -25,30 +25,14 @@
 	exit 1
 fi
 
-# Sort out sourcing in the distribution specific library functions
-# and the command to run them.
-if [ -f /etc/redhat-release ]; then
-	. /etc/rc.d/init.d/functions
-	START=daemon
-	END=
-elif [ -f /etc/SuSE-release ]; then
-	START=
-	END="rc_status -v"
-	. /etc/rc.config
-	. /etc/rc.status
-elif [ -f /etc/debian_release ]; then
-	. /etc/default/rcS
-	if [ "$VERBOSE" != "no" ] ; then
-		END=debian_end
-	fi
-elif [ -f /etc/slackware-version ]; then
-	. /etc/rc.d/rc.sysvinit
-	if [ "$VERBOSE" != "no" ]; then
-		END=slack_end
-	fi
-else
-	echo "$0: Can't guess distribution, aborting!"
-	exit 1
+. /etc/default/rcS
+if [ "$VERBOSE" != "no" ] ; then
+	END=debian_end
+fi
+
+if [ ! -e $DEVICE ];
+then
+        /sbin/modprobe microcode
 fi
 
 # Lets just be sure we have a device file...
