diff --git a/src/findlib/frontend.ml b/src/findlib/frontend.ml
index 1788342..8174664 100644
--- a/src/findlib/frontend.ml
+++ b/src/findlib/frontend.ml
@@ -1441,7 +1441,6 @@ let copy_file ?(rename = (fun name -> name)) ?(append = "") src dstdir =
       output_string ch_out append;
       close_out ch_out;
       close_in ch_in;
-      Unix.utimes outpath s.Unix.st_mtime s.Unix.st_mtime;
 
       prerr_endline("Installed " ^ outpath);
     with
