--- a/Makefile.in	2012-04-03 17:58:54.400046240 +0000
+++ b/Makefile.in	2012-04-03 18:02:09.632051180 +0000
@@ -406,19 +406,20 @@
 	if [ -f camomile.cma ]; then files="camomile.cma $$files"; fi&& \
 	if [ -f camomile.cmxa ]; then files="camomile.cmxa $$files"; fi&& \
 	if [ -f camomile.a ]; then files="camomile.a $$files"; fi && \
-	ocamlfind install camomile $$files META $(MLI) camomileLibrary.cmi camomileLibraryDefault.cmi camomileLibraryDyn.cmi
+	mkdir -p $(DESTDIR)$(shell ocamlfind printconf destdir) && \
+	ocamlfind install -destdir $(DESTDIR)$(shell ocamlfind printconf destdir) camomile $$files META $(MLI) camomileLibrary.cmi camomileLibraryDefault.cmi camomileLibraryDyn.cmi
 
 install-data:
-	mkdir -p '$(DATADIR)'
-	mkdir -p '$(DATADIR)'/camomile
-	mkdir -p '$(DATADIR)'/camomile/database
-	cp -f database/*.mar '$(DATADIR)'/camomile/database || true
-	mkdir -p '$(DATADIR)'/camomile/charmaps
-	cp -f charmaps/*.mar '$(DATADIR)'/camomile/charmaps || true
-	mkdir -p '$(DATADIR)'/camomile/mappings
+	mkdir -p '$(DESTDIR)$(DATADIR)'
+	mkdir -p '$(DESTDIR)$(DATADIR)'/camomile
+	mkdir -p '$(DESTDIR)$(DATADIR)'/camomile/database
+	cp -f database/*.mar '$(DESTDIR)$(DATADIR)'/camomile/database || true
+	mkdir -p '$(DESTDIR)$(DATADIR)'/camomile/charmaps
+	cp -f charmaps/*.mar '$(DESTDIR)$(DATADIR)'/camomile/charmaps || true
+	mkdir -p '$(DESTDIR)$(DATADIR)'/camomile/mappings
 	cp -f mappings/*.mar '$(DATADIR)'/camomile/mappings || true
-	mkdir -p '$(DATADIR)'/camomile/locales
-	cp -f locales/*.mar '$(DATADIR)'/camomile/locales || true
+	mkdir -p '$(DESTDIR)$(DATADIR)'/camomile/locales
+	cp -f locales/*.mar '$(DESTDIR)$(DATADIR)'/camomile/locales || true
 
 uninstall: $(UNINSTALL)
 
