--- a/configure.in	2012-03-23 21:26:41.641279075 +0000
+++ b/configure.in	2012-03-23 21:28:20.136281590 +0000
@@ -347,7 +347,7 @@
 AC_PROG_CC
 AC_PROG_LIBTOOL
 
-CFLAGS="$CFLAGS -I../include -W -Wall -Werror -Wno-unused-parameter -Wsign-compare \
+CFLAGS="$CFLAGS -I../include -W -Wall -Werror -Wno-unused -Wsign-compare \
 	-DTCSD_DEFAULT_PORT=${TCSD_DEFAULT_PORT} -DTSS_VER_MAJOR=${TSS_VER_MAJOR} \
 	-DTSS_VER_MINOR=${TSS_VER_MINOR} -DTSS_SPEC_MAJOR=${TSS_SPEC_MAJOR} \
 	-DTSS_SPEC_MINOR=${TSS_SPEC_MINOR}"
