--- a/configure
+++ b/configure
@@ -486,6 +486,7 @@ docker="no"
 debug_mutex="no"
 libpmem=""
 libudev="no"
+libargo="no"
 
 # cross compilers defaults, can be overridden with --cross-cc-ARCH
 cross_cc_aarch64="aarch64-linux-gnu-gcc"
@@ -5686,6 +5687,16 @@ if test "$libudev" != "no" ; then
   fi
 fi
 
+##########################################
+# argo support probe
+if test "$libargo" = "yes"; then
+    libargo_cflags=$($pkg_config --cflags libargo)
+    libargo_libs=$($pkg_config --libs libargo)
+    if ! compile_prog "$libargo_cflags" "$libargo_libs"; then
+        feature_not_found "libargo" "Install libargo devel"
+    fi
+fi
+
 # Now we've finished running tests it's OK to add -Werror to the compiler flags
 if test "$werror" = "yes"; then
     QEMU_CFLAGS="-Werror $QEMU_CFLAGS"
@@ -6122,6 +6133,7 @@ echo "capstone          $capstone"
 echo "docker            $docker"
 echo "libpmem support   $libpmem"
 echo "libudev           $libudev"
+echo "libargo           $libargo"
 
 if test "$sdl_too_old" = "yes"; then
 echo "-> Your SDL version is too old - please upgrade to have SDL support"
@@ -6997,6 +7009,11 @@ if test "$libudev" != "no"; then
     echo "LIBUDEV_LIBS=$libudev_libs" >> $config_host_mak
 fi
 
+if test "$libargo" = "yes" ; then
+  echo "LIBARGO_CFLAGS=$libargo_cflags" >> $config_host_mak
+  echo "LIBARGO_LIBS=$libargo_libs" >> $config_host_mak
+fi
+
 # use included Linux headers
 if test "$linux" = "yes" ; then
   mkdir -p linux-headers
