diff --git a/hs/DBus/Wire/Unmarshal.hs b/hs/DBus/Wire/Unmarshal.hs
index 40a28a0..a06710c 100644
--- a/hs/DBus/Wire/Unmarshal.hs
+++ b/hs/DBus/Wire/Unmarshal.hs
@@ -306,7 +306,7 @@ unmarshalMessage getBytes' = E.runErrorT $ do
 
         let fieldArray = fromJust . T.fromVariant $ header !! 6
         let fieldStructures = fromJust . T.fromArray $ fieldArray
-        fields <- concat `fmap` mapM decodeField fieldStructures
+        fields <- return . concat =<< mapM decodeField fieldStructures
 
         let bodyPadding = padding (fromIntegral fieldByteCount + 16) 8
         getBytes . fromIntegral $ bodyPadding
