diff --git a/midori/main.c b/midori/main.c
index 77e2cbf..bee8985 100644
--- a/midori/main.c
+++ b/midori/main.c
@@ -54,8 +54,12 @@
     #include <libosso.h>
 #endif
 
+#include <gdk/gdkx.h>
+
 #define MIDORI_HISTORY_ERROR g_quark_from_string("MIDORI_HISTORY_ERROR")
 
+static gboolean grab_display;
+
 typedef enum
 {
     MIDORI_HISTORY_ERROR_DB_OPEN,    /* Error opening the database file */
@@ -1355,6 +1359,12 @@ midori_web_app_browser_notify_load_status_cb (MidoriBrowser* browser,
         GdkPixbuf* icon = midori_view_get_icon (MIDORI_VIEW (view));
         gtk_window_set_icon (GTK_WINDOW (browser), icon);
     }
+
+    if (grab_display)
+    {
+        gdk_x11_display_grab(gdk_display_get_default());
+        gdk_flush();
+    }
 }
 
 int
@@ -1367,6 +1377,7 @@ main (int    argc,
     gchar* snapshot;
     gboolean execute;
     gboolean version;
+    gboolean grab;
     gchar** uris;
     MidoriApp* app;
     gboolean result;
@@ -1375,6 +1386,8 @@ main (int    argc,
     {
        { "app", 'a', 0, G_OPTION_ARG_STRING, &webapp,
        N_("Run ADDRESS as a web application"), N_("ADDRESS") },
+       { "grab", 'g', 0, G_OPTION_ARG_NONE, &grab,
+           N_("XGrabServer"), NULL},
        { "config", 'c', 0, G_OPTION_ARG_FILENAME, &config,
        N_("Use FOLDER as configuration folder"), N_("FOLDER") },
        #ifdef HAVE_JSCORE
@@ -1434,6 +1447,7 @@ main (int    argc,
     #endif
 
     /* Parse cli options */
+    grab = FALSE;
     webapp = NULL;
     config = NULL;
     run = FALSE;
@@ -1449,6 +1463,7 @@ main (int    argc,
         g_error_free (error);
         return 1;
     }
+    grab_display = grab;
 
     /* libSoup uses threads, so we need to initialize threads. */
     if (!g_thread_supported ()) g_thread_init (NULL);
