diff --git a/midori/midori-browser.c b/midori/midori-browser.c
index 03da3f1..28c2fb9 100644
--- a/midori/midori-browser.c
+++ b/midori/midori-browser.c
@@ -1762,50 +1762,23 @@ static void
 _midori_browser_find (MidoriBrowser* browser,
                       gboolean       forward)
 {
-    const gchar* text;
-    gboolean case_sensitive;
-    GtkWidget* view;
-
-    if (!(view = midori_browser_get_current_tab (browser)))
-        return;
-
-    text = gtk_entry_get_text (GTK_ENTRY (browser->find_text));
-    case_sensitive = gtk_toggle_tool_button_get_active (
-        GTK_TOGGLE_TOOL_BUTTON (browser->find_case));
-    midori_view_search_text (MIDORI_VIEW (view), text, case_sensitive, forward);
 }
 
 static void
 _midori_browser_find_done (MidoriBrowser* browser)
 {
-    GtkWidget* view = midori_browser_get_current_tab (browser);
-    midori_view_unmark_text_matches (MIDORI_VIEW (view));
-    gtk_widget_hide (browser->find);
-    browser->find_typing = FALSE;
-    gtk_window_set_focus (GTK_WINDOW (browser),
-        gtk_bin_get_child (GTK_BIN (view)) ?
-        gtk_bin_get_child (GTK_BIN (view)) : view);
 }
 
 static void
 _action_find_next_activate (GtkAction*     action,
                             MidoriBrowser* browser)
 {
-    _midori_browser_find (browser, TRUE);
 }
 
 static void
 midori_browser_find_text_changed_cb (GtkWidget*     entry,
                                      MidoriBrowser* browser)
 {
-    if (browser->find_typing)
-    {
-        GtkWidget* view = midori_browser_get_current_tab (browser);
-        const gchar* text = gtk_entry_get_text (GTK_ENTRY (entry));
-        midori_view_unmark_text_matches (MIDORI_VIEW (view));
-        if (g_utf8_strlen (text, -1) > 2)
-            _midori_browser_find (browser, TRUE);
-    }
 }
 
 static gboolean
@@ -1813,8 +1786,6 @@ midori_browser_find_text_focus_out_event_cb (GtkWidget*     entry,
                                              GdkEventFocus* event,
                                              MidoriBrowser* browser)
 {
-    if (browser->find_typing)
-        _midori_browser_find_done (browser);
     return FALSE;
 }
 
@@ -1822,19 +1793,12 @@ static void
 _action_find_previous_activate (GtkAction*     action,
                                 MidoriBrowser* browser)
 {
-    _midori_browser_find (browser, FALSE);
 }
 
 static void
 _find_highlight_toggled (GtkToggleToolButton* toolitem,
                          MidoriBrowser*       browser)
 {
-    GtkWidget* view;
-    gboolean highlight;
-
-    view = midori_browser_get_current_tab (browser);
-    highlight = gtk_toggle_tool_button_get_active (toolitem);
-    midori_view_set_highlight_text_matches (MIDORI_VIEW (view), highlight);
 }
 
 static gboolean
@@ -1842,12 +1806,6 @@ midori_browser_find_key_press_event_cb (GtkWidget*     toolbar,
                                         GdkEventKey*   event,
                                         MidoriBrowser* browser)
 {
-    if (event->keyval == GDK_Escape)
-    {
-        _midori_browser_find_done (browser);
-        return TRUE;
-    }
-
     return FALSE;
 }
 
@@ -1855,7 +1813,6 @@ static void
 midori_browser_find_button_close_clicked_cb (GtkWidget*     widget,
                                              MidoriBrowser* browser)
 {
-    _midori_browser_find_done (browser);
 }
 
 static void
@@ -2220,29 +2177,18 @@ static void
 _action_zoom_in_activate (GtkAction*     action,
                           MidoriBrowser* browser)
 {
-    GtkWidget* view = midori_browser_get_current_tab (browser);
-    if (view)
-        midori_view_set_zoom_level (MIDORI_VIEW (view),
-            midori_view_get_zoom_level (MIDORI_VIEW (view)) + 0.25f);
 }
 
 static void
 _action_zoom_out_activate (GtkAction*     action,
                            MidoriBrowser* browser)
 {
-    GtkWidget* view = midori_browser_get_current_tab (browser);
-    if (view)
-        midori_view_set_zoom_level (MIDORI_VIEW (view),
-            midori_view_get_zoom_level (MIDORI_VIEW (view)) - 0.25f);
 }
 
 static void
 _action_zoom_normal_activate (GtkAction*     action,
                               MidoriBrowser* browser)
 {
-    GtkWidget* view = midori_browser_get_current_tab (browser);
-    if (view)
-        midori_view_set_zoom_level (MIDORI_VIEW (view), 1.0f);
 }
 
 static void
diff --git a/midori/midori-view.c b/midori/midori-view.c
index b4eb7b0..ff72f8d 100644
--- a/midori/midori-view.c
+++ b/midori/midori-view.c
@@ -1036,26 +1036,6 @@ gtk_widget_key_press_event_cb (WebKitWebView* web_view,
                                GdkEventKey*   event,
                                MidoriView*    view)
 {
-    guint character = gdk_unicode_to_keyval (event->keyval);
-
-    if (event->keyval == '.' || event->keyval == '/' || event->keyval == GDK_KP_Divide)
-        character = '\0';
-    else if (!view->find_while_typing)
-        return FALSE;
-
-    /* Skip control characters */
-    if (character == (event->keyval | 0x01000000))
-        return FALSE;
-
-    if (!webkit_web_view_can_cut_clipboard (web_view)
-        && !webkit_web_view_can_paste_clipboard (web_view))
-    {
-        gchar* text = character ? g_strdup_printf ("%c", character) : g_strdup ("");
-
-        g_signal_emit (view, signals[SEARCH_TEXT], 0, TRUE, text);
-        g_free (text);
-    }
-
     return FALSE;
 }
 
@@ -1066,18 +1046,7 @@ gtk_widget_scroll_event_cb (WebKitWebView*  web_view,
 {
     event->state = event->state & MIDORI_KEYS_MODIFIER_MASK;
 
-    if (event->state & GDK_CONTROL_MASK)
-    {
-        if (event->direction == GDK_SCROLL_DOWN)
-            midori_view_set_zoom_level (view,
-                midori_view_get_zoom_level (view) - 0.25f);
-        else if(event->direction == GDK_SCROLL_UP)
-            midori_view_set_zoom_level (view,
-                midori_view_get_zoom_level (view) + 0.25f);
-        return TRUE;
-    }
-    else
-        return FALSE;
+    return FALSE;
 }
 
 static void
