Index: syslinux-4.04/mtools/Makefile
===================================================================
--- syslinux-4.04.orig/mtools/Makefile	2011-04-18 23:24:17.000000000 +0200
+++ syslinux-4.04/mtools/Makefile	2016-02-16 12:16:19.203026860 +0100
@@ -4,7 +4,7 @@
 OPTFLAGS = -g -Os
 INCLUDES = -I. -I.. -I../libfat -I../libinstaller
 CFLAGS	 = $(GCCWARN) -D_FILE_OFFSET_BITS=64 $(OPTFLAGS) $(INCLUDES)
-LDFLAGS	 = -s
+LDFLAGS	 =
 
 SRCS     = syslinux.c \
 	   ../libinstaller/fat.c \
Index: syslinux-4.04/utils/Makefile
===================================================================
--- syslinux-4.04.orig/utils/Makefile	2011-04-18 23:24:17.000000000 +0200
+++ syslinux-4.04/utils/Makefile	2016-02-16 12:16:27.876262742 +0100
@@ -18,7 +18,7 @@
 include $(topdir)/MCONFIG
 
 CFLAGS   = $(GCCWARN) -Os -fomit-frame-pointer -D_FILE_OFFSET_BITS=64
-LDFLAGS  = -O2 -s
+LDFLAGS  = -O2
 
 TARGETS	 = mkdiskimage isohybrid gethostip memdiskfind
 TARGETS += isohybrid.pl  # about to be obsoleted
