diff -uNr uim-1.3.1.orig/helper/toolbar-standalone-gtk.c uim-1.3.1/helper/toolbar-standalone-gtk.c
--- uim-1.3.1.orig/helper/toolbar-standalone-gtk.c	2011-10-12 14:16:35.390443292 +0100
+++ uim-1.3.1/helper/toolbar-standalone-gtk.c	2011-10-12 14:18:03.006240043 +0100
@@ -219,6 +219,12 @@
   GtkWidget *handle;
   GtkWidget *frame;
 #endif
+  int show_toolbar;
+
+  {
+    const char *env = getenv("UIM_SHOW_TOOLBAR");
+    show_toolbar = !env || strcmp(env, "1") == 0;
+  }
 
   setlocale(LC_ALL, "");
   bindtextdomain(PACKAGE, LOCALEDIR);
@@ -284,6 +290,8 @@
   g_signal_connect(G_OBJECT(window), "size-request",
 		   G_CALLBACK(size_request_cb), NULL);
 
+  if (show_toolbar)
+  {
   gtk_widget_show_all(GTK_WIDGET(window));
 
   {
@@ -318,6 +326,7 @@
     y = sc_h - h;
     helper_win_set_position(window, x, y);
   }
+  }
 
   gtk_main();
 
