diff --git a/common/rtusb_dev_id.c b/common/rtusb_dev_id.c
index 3b000bd..e3c65c9 100755
--- a/common/rtusb_dev_id.c
+++ b/common/rtusb_dev_id.c
@@ -119,6 +119,8 @@ USB_DEVICE_ID rtusb_dev_id[] = {
 	{USB_DEVICE(0x04BB,0x0944)}, /* I-O DATA 3572 */
 	{USB_DEVICE(0x1690,0x0740)}, /* 3572 */
 	{USB_DEVICE(0x1690,0x0744)}, /* 3572 */
+	{USB_DEVICE(0x1690,0x0761)}, /* 3572 */
+	{USB_DEVICE(0x13b1,0x002f)}, /* Linksys AE1000 [Ralink 3572] */
 	{USB_DEVICE(0x5A57,0x0284)}, /* Zinwell 3572 */
 	{USB_DEVICE(0x167B,0x4001)}, /* 3572 */
 	{USB_DEVICE(0x0930,0x0A07)}, /* TOSHIBA */
