diff --git a/os/linux/config.mk b/os/linux/config.mk
index 819a345..3bbd670 100755
--- a/os/linux/config.mk
+++ b/os/linux/config.mk
@@ -11,10 +11,10 @@ HAS_QA_SUPPORT=n
 HAS_XLINK=n
 
 # Support Wpa_Supplicant
-HAS_WPA_SUPPLICANT=n
+HAS_WPA_SUPPLICANT=y
 
 # Support Native WpaSupplicant for Network Maganger
-HAS_NATIVE_WPA_SUPPLICANT_SUPPORT=n
+HAS_NATIVE_WPA_SUPPLICANT_SUPPORT=y
 
 #Support Net interface block while Tx-Sw queue full
 HAS_BLOCK_NET_IF=n
