diff --git a/drivers/tty/serial/8250/8250_pnp.c b/drivers/tty/serial/8250/8250_pnp.c
index b3455a9..c76def8 100644
--- a/drivers/tty/serial/8250/8250_pnp.c
+++ b/drivers/tty/serial/8250/8250_pnp.c
@@ -447,6 +447,10 @@ serial_pnp_probe(struct pnp_dev *dev, const struct pnp_device_id *dev_id)
 	if (pnp_resource_enabled(res)) {
 		uart.port.iobase = res->start;
 		uart.port.iotype = UPIO_PORT;
+               /* Xen uses port 3f8 for logging, turn off ability to disable it as it breaks suspend */
+               if (uart.port.iobase == 0x3f8) {
+                       dev->capabilities &= ~PNP_DISABLE;
+               }
 	} else if (pnp_mem_valid(dev, 0)) {
 		uart.port.mapbase = pnp_mem_start(dev, 0);
 		uart.port.iotype = UPIO_MEM;
