diff --git a/drivers/char/mem.c b/drivers/char/mem.c
index c6fa3bc..cf448f2 100644
--- a/drivers/char/mem.c
+++ b/drivers/char/mem.c
@@ -316,9 +316,14 @@ static int mmap_mem(struct file *file, struct vm_area_struct *vma)
 						&vma->vm_page_prot))
 		return -EINVAL;
 
-	vma->vm_page_prot = phys_mem_access_prot(file, vma->vm_pgoff,
-						 size,
-						 vma->vm_page_prot);
+	vma->vm_flags |= VM_DONTDUMP | VM_IO | VM_PFNMAP | VM_DONTEXPAND;
+	vma->vm_page_prot =  __pgprot(
+			pgprot_val(vm_get_page_prot(vma->vm_flags)) |
+			_PAGE_IOMAP |
+			pgprot_val(phys_mem_access_prot(file,
+				vma->vm_pgoff,
+				size,
+				vma->vm_page_prot)));
 
 	vma->vm_ops = &mmap_mem_ops;
 
