diff --git a/configure b/configure
index 0ee0a63..12d0f91 100755
--- a/configure
+++ b/configure
@@ -2661,7 +2661,7 @@ if test ! -f utils/ghc-pwd/ghc-pwd && test ! -f utils/ghc-pwd/ghc-pwd.exe; then
   rm -f *.hi
   rm -f ghc-pwd
   rm -f ghc-pwd.exe
-  "$WithGhc" -v0 --make ghc-pwd -o ghc-pwd
+  "$WithGhc" -v0 -optl-pthread --make ghc-pwd -o ghc-pwd
   cd ../..
 fi
 
diff --git a/libraries/unix/configure b/libraries/unix/configure
index 4b01d72..bd948a2 100755
--- a/libraries/unix/configure
+++ b/libraries/unix/configure
@@ -6642,6 +6642,9 @@ openbsd*)
 	PTHREAD_CFLAGS=-pthread
 	PTHREAD_LDFLAGS=-pthread
 	;;
+linux*)
+	PTHREAD_LIBS=pthread
+	;;
 esac
 
 
diff --git a/libraries/unix/configure.ac b/libraries/unix/configure.ac
index f2e1c39..c71be4c 100644
--- a/libraries/unix/configure.ac
+++ b/libraries/unix/configure.ac
@@ -211,6 +211,9 @@ openbsd*)
 	PTHREAD_CFLAGS=-pthread
 	PTHREAD_LDFLAGS=-pthread
 	;;
+linux*)
+	PTHREAD_LIBS=pthread
+	;;
 esac
 AC_SUBST(PTHREAD_CFLAGS)
 AC_SUBST(PTHREAD_LDFLAGS)
