diff --git a/Makefile.in b/Makefile.in
index 103ff6b..2093927 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -236,10 +236,10 @@ install: $(CONFIGFILES) $(MANPAGES) $(TARGETS) install-files install-sysconf hos
 install-nokeys: $(CONFIGFILES) $(MANPAGES) $(TARGETS) install-files install-sysconf
 install-nosysconf: $(CONFIGFILES) $(MANPAGES) $(TARGETS) install-files
 
-check-config:
+check-config: install-files
 	-$(DESTDIR)$(sbindir)/sshd -t -f $(DESTDIR)$(sysconfdir)/sshd_config
 
-install-files:
+install-files: $(TARGETS) $(MANPAGES)
 	$(srcdir)/mkinstalldirs $(DESTDIR)$(bindir)
 	$(srcdir)/mkinstalldirs $(DESTDIR)$(sbindir)
 	$(srcdir)/mkinstalldirs $(DESTDIR)$(mandir)
@@ -278,7 +278,7 @@ install-files:
 	-rm -f $(DESTDIR)$(mandir)/$(mansubdir)1/slogin.1
 	ln -s ./ssh.1 $(DESTDIR)$(mandir)/$(mansubdir)1/slogin.1
 
-install-sysconf:
+install-sysconf: $(CONFIGFILES)
 	if [ ! -d $(DESTDIR)$(sysconfdir) ]; then \
 		$(srcdir)/mkinstalldirs $(DESTDIR)$(sysconfdir); \
 	fi
