--- a/xen/arch/x86/guest/hypercall_page.S	2018-11-12 14:51:49.920565676 -0500
+++ b/xen/arch/x86/guest/hypercall_page.S	2018-11-12 14:52:39.357054001 -0500
@@ -59,8 +59,10 @@
 DECLARE_HYPERCALL(domctl)
 DECLARE_HYPERCALL(kexec_op)
 DECLARE_HYPERCALL(tmem_op)
-DECLARE_HYPERCALL(xc_reserved_op)
+DECLARE_HYPERCALL(v4v_op)
 DECLARE_HYPERCALL(xenpmu_op)
+DECLARE_HYPERCALL(dm_op)
+DECLARE_HYPERCALL(txt_op)
 
 DECLARE_HYPERCALL(arch_0)
 DECLARE_HYPERCALL(arch_1)
