diff --git a/Config.mk b/Config.mk
index 985feb4..e964064 100644
--- a/Config.mk
+++ b/Config.mk
@@ -1,8 +1,8 @@
 # -*- mode: Makefile; -*-
 
-ifeq ($(filter /%,$(XEN_ROOT)),)
-$(error XEN_ROOT must be absolute)
-endif
+#ifeq ($(filter /%,$(XEN_ROOT)),)
+#$(error XEN_ROOT must be absolute but is $(XEN_ROOT))
+#endif
 
 # fallback for older make
 realpath = $(wildcard $(foreach file,$(1),$(shell cd -P $(dir $(file)) && echo "$$PWD/$(notdir $(file))")))
