From c2b3574915adf8ef9d4863e6617b8cfb5d861c3c Mon Sep 17 00:00:00 2001
From: Tamas K Lengyel <tamas@tklengyel.com>
Date: Sun, 25 Feb 2018 13:17:01 -0700
Subject: [PATCH] Require the presence of the shim to boot under UEFI

---
 xen/common/efi/boot.c | 7 +++++++
 1 file changed, 7 insertions(+)

--- a/xen/common/efi/boot.c
+++ b/xen/common/efi/boot.c
@@ -1265,6 +1265,13 @@ efi_start(EFI_HANDLE ImageHandle, EFI_SY
         else if ( !read_file(dir_handle, cfg_file_name, &cfg, NULL) )
             blexit(L"Configuration file not found.");
 
+        if ( !shim_lock )
+        {
+            PrintStr(L"No shim found, booting xen.efi directly with OpenXT is not supported.");
+            efi_bs->Stall(10000000);
+            blexit(L"No shim found");
+        }
+
         if ( shim_lock &&
             (status = shim_lock->Measure(cfg.ptr, cfg.size, 8)) != EFI_SUCCESS )
                 PrintErrMesg(L"Configuration file could not be measured", status);
