diff --git a/xen/include/public/xsm/flask_op.h b/xen/include/public/xsm/flask_op.h
index 233de81..aa6bbb3 100644
--- a/xen/include/public/xsm/flask_op.h
+++ b/xen/include/public/xsm/flask_op.h
@@ -28,7 +28,7 @@
 #define XEN_FLASK_INTERFACE_VERSION 1
 
 struct xen_flask_load {
-    XEN_GUEST_HANDLE(char) buffer;
+    XEN_GUEST_HANDLE_64(char) buffer;
     uint32_t size;
 };
 
@@ -43,7 +43,7 @@ struct xen_flask_sid_context {
      * OUT: actual size of the output context string
      */
     uint32_t size;
-    XEN_GUEST_HANDLE(char) context;
+    XEN_GUEST_HANDLE_64(char) context;
 };
 
 struct xen_flask_access {
@@ -76,9 +76,9 @@ struct xen_flask_userlist {
     uint32_t size;
     union {
         /* IN: user to enumerate SIDs */
-        XEN_GUEST_HANDLE(char) user;
+        XEN_GUEST_HANDLE_64(char) user;
         /* OUT: SID list */
-        XEN_GUEST_HANDLE(uint32) sids;
+        XEN_GUEST_HANDLE_64(uint32) sids;
     } u;
 };
 
@@ -100,7 +100,7 @@ struct xen_flask_boolean {
     /* IN: if bool_id is -1, used to find boolean [GET/SET]
      * OUT: textual name of boolean [GET only]
      */
-    XEN_GUEST_HANDLE(char) name;
+    XEN_GUEST_HANDLE_64(char) name;
 };
 
 struct xen_flask_setavc_threshold {
