Index: xen-4.6.4/xen/arch/x86/hvm/vmx/vvmx.c
===================================================================
--- xen-4.6.4.orig/xen/arch/x86/hvm/vmx/vvmx.c
+++ xen-4.6.4/xen/arch/x86/hvm/vmx/vvmx.c
@@ -1914,6 +1914,7 @@ int nvmx_msr_read_intercept(unsigned int
         /* 1-seetings */
         data = PIN_BASED_EXT_INTR_MASK |
                PIN_BASED_NMI_EXITING |
+               PIN_BASED_VIRTUAL_NMIS |
                PIN_BASED_PREEMPT_TIMER;
         data = gen_vmx_msr(data, VMX_PINBASED_CTLS_DEFAULT1, host_data);
         break;
@@ -2278,8 +2279,10 @@ int nvmx_n2_vmexit_handler(struct cpu_us
                 if ( !++port )
                     nvcpu->nv_vmexit_pending = 1;
             } while ( !nvcpu->nv_vmexit_pending );
+#if 0
             if ( !nvcpu->nv_vmexit_pending )
                 printk(XENLOG_G_WARNING "L0 PIO %04x\n", port);
+#endif
         }
         else if ( ctrl & CPU_BASED_UNCOND_IO_EXITING )
             nvcpu->nv_vmexit_pending = 1;
