Index: xen-4.3.4/xen/arch/x86/cpu/mcheck/non-fatal.c
===================================================================
--- xen-4.3.4.orig/xen/arch/x86/cpu/mcheck/non-fatal.c
+++ xen-4.3.4/xen/arch/x86/cpu/mcheck/non-fatal.c
@@ -94,8 +94,8 @@ static int __init init_nonfatal_mce_chec
 	if (mce_disabled || !mce_available(c))
 		return -ENODEV;
 
-    if ( __get_cpu_var(poll_bankmask) == NULL )
-        return -EINVAL;
+	if ( __get_cpu_var(poll_bankmask) == NULL )
+		return -EINVAL;
 
 	/*
 	 * Check for non-fatal errors every MCE_RATE s
Index: xen-4.3.4/xen/drivers/passthrough/amd/iommu_intr.c
===================================================================
--- xen-4.3.4.orig/xen/drivers/passthrough/amd/iommu_intr.c
+++ xen-4.3.4/xen/drivers/passthrough/amd/iommu_intr.c
@@ -449,7 +449,7 @@ void* __init amd_iommu_alloc_intremap_ta
 
 int __init amd_setup_hpet_msi(struct msi_desc *msi_desc)
 {
-    if ( (!msi_desc->hpet_id != hpet_sbdf.id) ||
+    if ( (msi_desc->hpet_id != hpet_sbdf.id) ||
          (hpet_sbdf.iommu == NULL) )
     {
         AMD_IOMMU_DEBUG("Fail to setup HPET MSI remapping\n");
Index: xen-4.3.4/xen/crypto/rijndael.c
===================================================================
--- xen-4.3.4.orig/xen/crypto/rijndael.c
+++ xen-4.3.4/xen/crypto/rijndael.c
@@ -380,6 +380,7 @@ static const u32 Te4[256] = {
     0x41414141U, 0x99999999U, 0x2d2d2d2dU, 0x0f0f0f0fU,
     0xb0b0b0b0U, 0x54545454U, 0xbbbbbbbbU, 0x16161616U,
 };
+#ifdef NEED_RIJNDAEL_DECRYPT
 static const u32 Td0[256] = {
     0x51f4a750U, 0x7e416553U, 0x1a17a4c3U, 0x3a275e96U,
     0x3bab6bcbU, 0x1f9d45f1U, 0xacfa58abU, 0x4be30393U,
@@ -710,6 +711,8 @@ static const u32 Td4[256] = {
     0xe1e1e1e1U, 0x69696969U, 0x14141414U, 0x63636363U,
     0x55555555U, 0x21212121U, 0x0c0c0c0cU, 0x7d7d7d7dU,
 };
+#endif /* NEED_RIJNDAEL_DECRYPT */
+
 static const u32 rcon[] = {
 	0x01000000, 0x02000000, 0x04000000, 0x08000000,
 	0x10000000, 0x20000000, 0x40000000, 0x80000000,
