diff --git a/tools/firmware/hvmloader/pci.c b/tools/firmware/hvmloader/pci.c
index bc51431..ea24956 100644
--- a/tools/firmware/hvmloader/pci.c
+++ b/tools/firmware/hvmloader/pci.c
@@ -101,6 +101,15 @@ void pci_setup(void)
                 vga_devfn = devfn;
                 virtual_vga = VGA_cirrus;
             }
+            /* For NVIDIA VGA cards load std vga BIOS image
+             * We might want to make this generic by allowing
+             * the plugin to set a bit indicating this.
+             */
+            else if (vendor_id == 0x10DE)
+            {
+                vga_devfn = devfn;
+                virtual_vga = VGA_std;
+            }
             else if ( virtual_vga == VGA_none )
             {
                 vga_devfn = devfn;
