diff --git a/tools/firmware/hvmloader/e820.c b/tools/firmware/hvmloader/e820.c
index 295781c..ce9c2a1 100644
--- a/tools/firmware/hvmloader/e820.c
+++ b/tools/firmware/hvmloader/e820.c
@@ -165,6 +165,12 @@ int build_e820_table(struct e820entry *e820,
         e820[nr].size = (uint32_t)-e820[nr].addr;
         e820[nr].type = E820_RESERVED;
         nr++;
+
+        /* 16M for superblanker FB */
+        e820[nr].addr = 0xfd000000;
+        e820[nr].size = 0x1000000;
+        e820[nr].type = E820_RESERVED;
+        nr++;
     }
     else
     {
