diff --git a/tools/firmware/hvmloader/pci.c b/tools/firmware/hvmloader/pci.c
index c52bd77..9ca9b8a 100644
--- a/tools/firmware/hvmloader/pci.c
+++ b/tools/firmware/hvmloader/pci.c
@@ -99,6 +99,12 @@ void pci_setup(void)
             {
                 vga_devfn = devfn;
                 virtual_vga = VGA_pt;
+
+		/* Enable IO. VGA Needs IO. */
+		cmd = pci_readw(devfn, PCI_COMMAND);
+		cmd |= PCI_COMMAND_IO;
+		pci_writew(devfn, PCI_COMMAND, cmd);
+
                 if ( vendor_id == 0x8086 )
                 {
                     igd_opregion_pgbase = mem_hole_alloc(2);
