Index: xen-4.6.4/tools/firmware/hvmloader/ovmf.c
===================================================================
--- xen-4.6.4.orig/tools/firmware/hvmloader/ovmf.c
+++ xen-4.6.4/tools/firmware/hvmloader/ovmf.c
@@ -38,8 +38,14 @@
 #include "roms.inc"
 
 #define OVMF_SIZE               (sizeof(ovmf))
+#define OVMF_VARS_SIZE		0x20000ULL
+#define OVMF_CODE_SIZE		(OVMF_SIZE - OVMF_VARS_SIZE)
+
 #define OVMF_MAXOFFSET          0x000FFFFFULL
 #define OVMF_BEGIN              (0x100000000ULL - ((OVMF_SIZE + OVMF_MAXOFFSET) & ~OVMF_MAXOFFSET))
+#define OVMF_VARS_BEGIN		OVMF_BEGIN
+#define OVMF_VARS_END		(OVMF_VARS_BEGIN + OVMF_VARS_SIZE)
+#define OVMF_CODE_BEGIN		OVMF_VARS_END
 #define OVMF_END                (OVMF_BEGIN + OVMF_SIZE)
 #define LOWCHUNK_BEGIN          0x000F0000
 #define LOWCHUNK_SIZE           0x00010000
@@ -96,7 +102,7 @@ static void ovmf_finish_bios_info(void)
 static void ovmf_load(const struct bios_config *config)
 {
     xen_pfn_t mfn;
-    uint64_t addr = OVMF_BEGIN;
+    uint64_t addr = OVMF_CODE_BEGIN;
 
     /* Copy low-reset vector portion. */
     memcpy((void *) LOWCHUNK_BEGIN, (uint8_t *) config->image
@@ -113,7 +119,7 @@ static void ovmf_load(const struct bios_
     }
 
     /* Copy FD. */
-    memcpy((void *) OVMF_BEGIN, config->image, OVMF_SIZE);
+    memcpy((void *) OVMF_CODE_BEGIN, config->image + OVMF_VARS_SIZE, OVMF_CODE_SIZE);
 }
 
 static void ovmf_acpi_build_tables(void)
