--- a/tools/libxl/libxl_blktap3.c
+++ b/tools/libxl/libxl_blktap3.c
@@ -63,6 +63,17 @@ int libxl__blktap_enabled(libxl__gc *gc)
     return 1;
 }
 
+int libxl__get_tap_minor(libxl__gc *gc, libxl_disk_format format, const char *disk)
+{
+    const char *type = NULL;
+    tap_list_t tap;
+
+    memset(&tap, 0, sizeof(tap_list_t));
+
+    type = libxl__device_disk_string_of_format(format);
+    return tap_ctl_find_minor(type, disk);
+}
+
 char *libxl__blktap_devpath(libxl__gc *gc, const char *disk,
 		libxl_disk_format format,
 		char *keydir)
