Index: xen-4.9.0/tools/libxl/libxl_blktap3.c
===================================================================
--- xen-4.9.0.orig/tools/libxl/libxl_blktap3.c
+++ xen-4.9.0/tools/libxl/libxl_blktap3.c
@@ -62,6 +62,17 @@ int libxl__blktap_enabled(libxl__gc *gc)
     return 1;
 }
 
+int libxl__get_tap_minor(libxl__gc *gc, libxl_disk_format format, const char *disk)
+{
+    const char *type = NULL;
+    tap_list_t tap;
+
+    memset(&tap, 0, sizeof(tap_list_t));
+
+    type = libxl__device_disk_string_of_format(format);
+    return tap_ctl_find_minor(type, disk);
+}
+
 char *libxl__blktap_devpath(libxl__gc *gc, const char *disk,
 		libxl_disk_format format,
 		char *keydir)
@@ -103,7 +114,7 @@ char *libxl__blktap_devpath(libxl__gc *g
     return NULL;/*revert-back-changes*/
 }
 
-static bool tapdev_is_shared(libxl__gc *gc, const char *params)
+bool tapdev_is_shared(libxl__gc *gc, const char *params)
 {
     char **domids, **vbds;
     char *tp;
