Index: xen-4.6.4/tools/libxl/libxl_dm.c
===================================================================
--- xen-4.6.4.orig/tools/libxl/libxl_dm.c
+++ xen-4.6.4/tools/libxl/libxl_dm.c
@@ -1148,11 +1148,6 @@ static int libxl__build_device_model_arg
                         (gc, "file=%s,if=ide,index=%d,readonly=%s,media=cdrom,format=%s,cache=writeback,id=ide-%i",
                          disks[i].pdev_path, disk, disks[i].readwrite ? "off" : "on", format, dev_number);
             } else {
-                if (!disks[i].readwrite) {
-                    LIBXL__LOG(ctx, LIBXL__LOG_ERROR, "qemu-xen doesn't support read-only disk drivers");
-                    return ERROR_INVAL;
-                }
-
                 if (disks[i].format == LIBXL_DISK_FORMAT_EMPTY) {
                     LIBXL__LOG(ctx, LIBXL__LOG_WARNING, "cannot support"
                                " empty disk format for %s", disks[i].vdev);
@@ -1192,14 +1187,21 @@ static int libxl__build_device_model_arg
                     continue;
                 }
                 else if (disk < 4) {
-                    if (b_info->stubdomain_version == LIBXL_STUBDOMAIN_VERSION_LINUX)
-                        drive = libxl__sprintf
-                                (gc, "file=%s%c,if=ide,index=%d,media=disk,cache=writeback,format=%s",
-                                 "/dev/xvd", 'a' + disk, disk, format);
-                    else
+                    if (b_info->stubdomain_version == LIBXL_STUBDOMAIN_VERSION_LINUX) {
+                        if(disks[i].readwrite) {
+                            drive = libxl__sprintf
+                                    (gc, "file=%s%c,if=ide,index=%d,media=disk,cache=writeback,format=%s",
+                                     "/dev/xvd", 'a' + disk, disk, format);
+                        } else {
+                            drive = libxl__sprintf
+                                    (gc, "file=%s%c,if=ide,index=%d,media=disk,readonly,format=%s",
+                                     "/dev/xvd", 'a' + disk, disk, format);
+                        }
+                    } else {
                         drive = libxl__sprintf
                                 (gc, "file=%s,if=ide,index=%d,media=disk,format=%s,cache=writeback",
                                  pdev_path, disk, format);
+                    }
                 }
                 else
                     continue; /* Do not emulate this disk */
