--- xen.orig/tools/hotplug/Linux/xen-hotplug-common.sh.in	2019-08-28 14:13:57.845225343 -0400
+++ xen/tools/hotplug/Linux/xen-hotplug-common.sh.in	2019-08-28 14:17:23.495223153 -0400
@@ -20,8 +20,6 @@
 . "$dir/xen-script-common.sh"
 . "$dir/locking.sh"
 
-exec 2>>@XEN_LOG_DIR@/xen-hotplug.log
-
 export PATH="${bindir}:${sbindir}:${LIBEXEC_BIN}:/sbin:/bin:/usr/bin:/usr/sbin:$PATH"
 export LD_LIBRARY_PATH="${libdir}${LD_LIBRARY_PATH+:}$LD_LIBRARY_PATH"
 export LANG="POSIX"
