--- a/tools/libxl/libxl_qmp.c
+++ b/tools/libxl/libxl_qmp.c
@@ -955,6 +955,7 @@ int libxl__qmp_pci_add(libxl__gc *gc, in
     libxl__json_object *args = NULL;
     char *hostaddr = NULL;
     int rc = 0;
+    int i = 0;
 
     qmp = libxl__qmp_initialize(gc, domid);
     if (!qmp)
@@ -986,8 +987,13 @@ int libxl__qmp_pci_add(libxl__gc *gc, in
     if (pcidev->permissive)
         qmp_parameters_add_bool(gc, &args, "permissive", true);
 
-    rc = qmp_synchronous_send(qmp, "device_add", args,
-                              NULL, NULL, qmp->timeout);
+    do {
+        rc = qmp_synchronous_send(qmp, "device_add", args,
+                                  NULL, NULL, qmp->timeout);
+        if (rc)
+            sleep(1);
+    } while (rc && (++i < 15));
+
     if (rc == 0) {
         rc = qmp_synchronous_send(qmp, "query-pci", NULL,
                                   pci_add_callback, pcidev, qmp->timeout);
