Index: xen-4.6.4/tools/firmware/hvmloader/Makefile
===================================================================
--- xen-4.6.4.orig/tools/firmware/hvmloader/Makefile
+++ xen-4.6.4/tools/firmware/hvmloader/Makefile
@@ -25,6 +25,9 @@ SUBDIRS := acpi
 # The HVM loader is started in 32-bit mode at the address below:
 LOADADDR = 0x100000
 
+OVMF_PATH=${STAGING_LIBDIR}/../share/ovmf/OVMF.fd
+CONFIG_OVMF=y
+
 # SMBIOS spec requires format mm/dd/yyyy
 SMBIOS_REL_DATE ?= $(shell date +%m/%d/%Y)
 
