Index: xen-4.6.4/tools/libxl/libxl_dm.c
===================================================================
--- xen-4.6.4.orig/tools/libxl/libxl_dm.c
+++ xen-4.6.4/tools/libxl/libxl_dm.c
@@ -1637,6 +1637,7 @@ void libxl__spawn_stub_dm(libxl__egc *eg
                                   libxl__xs_get_dompath(gc, dm_domid)),
                    "%d", guest_domid);
     if (guest_config->b_info.stubdomain_version == LIBXL_STUBDOMAIN_VERSION_LINUX) {
+#if 0 /* LIES */
         /* qemu-xen is used as a dm in the stubdomain, so we set the bios
          * according to this */
         libxl__xs_write(gc, XBT_NULL,
@@ -1644,6 +1645,12 @@ void libxl__spawn_stub_dm(libxl__egc *eg
                                        libxl__xs_get_dompath(gc, guest_domid)),
                         "%s",
                         libxl_bios_type_to_string(LIBXL_BIOS_TYPE_SEABIOS));
+#else
+        libxl__xs_write(gc, XBT_NULL,
+                        libxl__sprintf(gc, "%s/hvmloader/bios",
+                                       libxl__xs_get_dompath(gc, guest_domid)),
+                        "%s", libxl_bios_type_to_string(guest_config->b_info.u.hvm.bios));
+#endif
         /* OpenXT: We use legacy roms, which is disabled by default in sebios */
         libxl__xs_write(gc, XBT_NULL,
                         libxl__sprintf(gc, "%s/hvmloader/seabios-legacy-load-roms",
