# HG changeset patch
# Parent c43f2fc1494c2e0e6ca66e8e06dad36677849933
diff --git a/xen/drivers/passthrough/vtd/dmar.c b/xen/drivers/passthrough/vtd/dmar.c
index 105d62d..2aa1e15 100644
--- a/xen/drivers/passthrough/vtd/dmar.c
+++ b/xen/drivers/passthrough/vtd/dmar.c
@@ -534,6 +534,14 @@ acpi_parse_one_rmrr(struct acpi_dmar_header *header)
                 base_addr, end_addr);
     }
 
+    if (end_addr <= base_addr)
+    {
+	    dprintk(XENLOG_WARNING VTDPREFIX,
+                "  The RMRR (%"PRIx64", %"PRIx64") is incorrect!\n",
+                base_addr, end_addr);
+	    return -EINVAL;
+    }
+
     rmrru = xzalloc(struct acpi_rmrr_unit);
     if ( !rmrru )
         return -ENOMEM;
