diff --git a/tools/firmware/rombios/rombios.c b/tools/firmware/rombios/rombios.c
index f6946c5..cf4b9b1 100644
--- a/tools/firmware/rombios/rombios.c
+++ b/tools/firmware/rombios/rombios.c
@@ -163,8 +163,8 @@
 
 #define BX_PMM           1   /* POST Memory Manager */
 
-#define BX_MAX_ATA_INTERFACES   4
-#define BX_MAX_ATA_DEVICES      (BX_MAX_ATA_INTERFACES*2)
+#define BX_MAX_ATA_INTERFACES   2
+#define BX_MAX_ATA_DEVICES      3
 
 #define BX_VIRTUAL_PORTS 1 /* normal output to Bochs ports */
 #define BX_DEBUG_SERIAL  0 /* output to COM1 */
@@ -11085,7 +11085,7 @@ post_default_ints:
   ;;
   ;; Floppy setup
   ;;
-  call floppy_drive_post
+//  call floppy_drive_post
 
   ;;
   ;; Hard Drive setup
@@ -11128,9 +11128,11 @@ post_default_ints:
   call _disable_rom_write_access
 #endif
 
+#if 0
 #if BX_ELTORITO_BOOT
   call _interactive_bootkey
 #endif // BX_ELTORITO_BOOT
+#endif
 
 #if BX_TCGBIOS
   call tcpa_post_part2
