diff --git a/tools/blktap2/drivers/block-vhd.c b/tools/blktap2/drivers/block-vhd.c
index 9d80824..fb6c9fc 100644
--- a/tools/blktap2/drivers/block-vhd.c
+++ b/tools/blktap2/drivers/block-vhd.c
@@ -74,7 +74,7 @@ unsigned int SPB;
 		    PRIu64", RETURNED: %" PRIu64 ", DATA_ALLOCATED: "	\
 		    "%lu, BBLK: 0x%04x\n",				\
 		    s->vhd.file, s->queued, s->completed, s->returned,	\
-		    VHD_REQS_DATA - s->vreq_free_count,			\
+		    VHD_REQS_TOTAL - s->vreq_free_count,			\
 		    s->bat.pbw_blk);					\
 	} while(0)
 
@@ -236,8 +236,8 @@ struct vhd_state {
 	struct vhd_bitmap         bitmap_list[VHD_CACHE_SIZE];
 
 	int                       vreq_free_count;
-	struct vhd_request       *vreq_free[VHD_REQS_DATA];
-	struct vhd_request        vreq_list[VHD_REQS_DATA];
+	struct vhd_request       *vreq_free[VHD_REQS_TOTAL];
+	struct vhd_request        vreq_list[VHD_REQS_TOTAL];
 
 	td_driver_t              *driver;
 
@@ -658,8 +658,8 @@ __vhd_open(td_driver_t *driver, const char *name, vhd_flag_t flags)
 
 	SPB = s->spb;
 
-	s->vreq_free_count = VHD_REQS_DATA;
-	for (i = 0; i < VHD_REQS_DATA; i++)
+	s->vreq_free_count = VHD_REQS_TOTAL;
+	for (i = 0; i < VHD_REQS_TOTAL; i++)
 		s->vreq_free[i] = s->vreq_list + i;
 
 	driver->info.size        = s->vhd.footer.curr_size >> VHD_SECTOR_SHIFT;
@@ -2344,8 +2344,8 @@ vhd_debug(td_driver_t *driver)
 	DBG(TLOG_WARN, "CACHE HITS: 0x%08"PRIx64", CACHE MISSES: 0x%08"PRIx64"\n",
 	    s->cache_hits, s->cache_misses);
 
-	DBG(TLOG_WARN, "ALLOCATED REQUESTS: (%lu total)\n", VHD_REQS_DATA);
-	for (i = 0; i < VHD_REQS_DATA; i++) {
+	DBG(TLOG_WARN, "ALLOCATED REQUESTS: (%lu total)\n", VHD_REQS_TOTAL);
+	for (i = 0; i < VHD_REQS_TOTAL; i++) {
 		struct vhd_request *r = &s->vreq_list[i];
 		td_request_t *t       = &r->treq;
 		if (t->secs)
