diff -r 0407e6ed55a3 -r 400ce3ecdaec tools/firmware/vgabios/vbe.h
--- a/tools/firmware/vgabios/vbe.h	Wed Apr 13 16:36:45 2011 -0400
+++ b/tools/firmware/vgabios/vbe.h	Thu Apr 21 13:44:36 2011 -0400
@@ -295,7 +295,7 @@
   #define VBE_DISPI_INDEX_X_OFFSET         0x8
   #define VBE_DISPI_INDEX_Y_OFFSET         0x9
 
-  #define VBE_DISPI_LFB_PHYSICAL_ADDRESS   0xE0000000
+  #define VBE_DISPI_LFB_PHYSICAL_ADDRESS   0xF0000000
   #define VBE_DISPI_ID0                    0xB0C0
   #define VBE_DISPI_ID1                    0xB0C1
   #define VBE_DISPI_ID2                    0xB0C2
