diff --git a/xen/arch/x86/mm.c b/xen/arch/x86/mm.c
index 4e2b8f0..380ef1c 100644
--- a/xen/arch/x86/mm.c
+++ b/xen/arch/x86/mm.c
@@ -4549,7 +4549,7 @@ static int xenmem_add_to_physmap_once(
     struct page_info *page = NULL;
     unsigned long gfn = 0; /* gcc ... */
     unsigned long prev_mfn, mfn = 0, gpfn, idx;
-    int rc;
+    int rc = -EINVAL;
     p2m_type_t p2mt;
     int unmap_shinfo = 0;
     unsigned long gpfn_new = xatp->gpfn;
diff --git a/xen/include/public/memory.h b/xen/include/public/memory.h
index 2e81089..c8f007c 100644
--- a/xen/include/public/memory.h
+++ b/xen/include/public/memory.h
@@ -205,7 +205,10 @@ DEFINE_XEN_GUEST_HANDLE(xen_machphys_mapping_t);
 #define XENMAPSPACE_shared_info  0 /* shared info page */
 #define XENMAPSPACE_grant_table  1 /* grant table page */
 #define XENMAPSPACE_gmfn         2 /* GMFN */
-#define XENMAPSPACE_gmfn_range   3 /* GMFN range, XENMEM_add_to_physmap only. */
+
+/* workaround old (glenn/peculier) XT tools bug by moving XENMAPSPACE_gmfn_range to another ID */
+#define XENMAPSPACE_gmfn_range   100 /* GMFN range, XENMEM_add_to_physmap only. */
+
 #define XENMAPSPACE_gmfn_foreign 4 /* GMFN from another dom,
                                     * XENMEM_add_to_physmap_range only.
                                     */
