Index: xen-4.6.1/xen/arch/x86/tboot.c
===================================================================
--- xen-4.6.1.orig/xen/arch/x86/tboot.c
+++ xen-4.6.1/xen/arch/x86/tboot.c
@@ -342,8 +342,6 @@ void tboot_shutdown(uint32_t shutdown_ty
 
     g_tboot_shared->shutdown_type = shutdown_type;
 
-    local_irq_disable();
-
     /* Create identity map for tboot shutdown code. */
     /* do before S3 integrity because mapping tboot may change xenheap */
     map_base = PFN_DOWN(g_tboot_shared->tboot_base);
@@ -358,6 +356,10 @@ void tboot_shutdown(uint32_t shutdown_ty
         return;
     }
 
+    /* Disable interrupts as early as possible but not before */
+    /* calling map_pages_to_xen */
+    local_irq_disable();
+
     /* if this is S3 then set regions to MAC */
     if ( shutdown_type == TB_SHUTDOWN_S3 )
     {
