diff --git a/xen/arch/x86/hvm/mtrr.c b/xen/arch/x86/hvm/mtrr.c
index ef51a8d..ed8e76e 100644
--- a/xen/arch/x86/hvm/mtrr.c
+++ b/xen/arch/x86/hvm/mtrr.c
@@ -702,8 +702,12 @@ uint8_t epte_get_entry_emt(struct domain *d, unsigned long gfn, mfn_t mfn,
     if ( !mfn_valid(mfn_x(mfn)) )
         return MTRR_TYPE_UNCACHABLE;
 
-    if ( hvm_get_mem_pinned_cacheattr(d, gfn, &type) )
+    if ( hvm_get_mem_pinned_cacheattr(d, gfn, &type) ) {
+        /* for cache-pinned pages, ignore PAT memory type and just use the EPT memory type specified here */
+        *ipat = 1;
+
         return type;
+    }
 
     if ( !iommu_enabled )
     {
