# HG changeset patch
# Parent af12f534c5bf5d1365be2c739a75ac0eb824a445

Index: xen-4.6.1/tools/xentrace/xentrace_format
===================================================================
--- xen-4.6.1.orig/tools/xentrace/xentrace_format
+++ xen-4.6.1/tools/xentrace/xentrace_format
@@ -213,6 +213,10 @@ while not interrupted:
         elif tsc < last_tsc[cpu] and tsc_in == 1:
             print "TSC stepped backward cpu %d !  %d %d" % (cpu,tsc,last_tsc[cpu])
 
+	# use last tsc if none provided
+	if tsc == 0:
+	    tsc = last_tsc[cpu]
+
         # provide relative TSC
         if last_tsc[cpu] > 0 and tsc_in == 1:
             reltsc = tsc - last_tsc[cpu]
@@ -223,7 +227,8 @@ while not interrupted:
             last_tsc[cpu] = tsc
 
         if mhz:
-            tsc = tsc / (mhz*1000000.0)
+            tsc = tsc / (mhz*1000.0)
+            reltsc = reltsc / (mhz*1000.0)
 
         args = {'cpu'   : cpu,
                 'tsc'   : tsc,
