diff --git a/midori/midori-browser.c b/midori/midori-browser.c
index 078713e..03da3f1 100644
--- a/midori/midori-browser.c
+++ b/midori/midori-browser.c
@@ -3759,6 +3759,11 @@ midori_browser_init (MidoriBrowser* browser)
     GtkWidget* scrolled;
     GtkWidget* icon;
 
+    rcstyle = gtk_rc_style_new ();
+    rcstyle->xthickness = rcstyle->ythickness = 0;
+    gtk_widget_modify_style (browser, rcstyle);
+    g_object_unref (rcstyle);
+    
     browser->net = katze_net_new ();
 
     browser->bookmark_popup = NULL;
diff --git a/midori/midori-view.c b/midori/midori-view.c
index a8886d4..b4eb7b0 100644
--- a/midori/midori-view.c
+++ b/midori/midori-view.c
@@ -1953,11 +1953,18 @@ midori_view_construct_web_view (MidoriView* view)
 {
     WebKitWebFrame* web_frame;
     gpointer inspector;
+    GtkRcStyle* rcstyle;
 
     g_return_if_fail (!view->web_view);
 
     view->web_view = webkit_web_view_new ();
 
+    rcstyle = gtk_rc_style_new ();
+    rcstyle->xthickness = rcstyle->ythickness = 0;
+    gtk_widget_modify_style (view->web_view, rcstyle);
+    gtk_widget_modify_style (view, rcstyle);
+    g_object_unref (rcstyle);
+
     /* Load something to avoid a bug where WebKit might not set a main frame */
     webkit_web_view_open (WEBKIT_WEB_VIEW (view->web_view), "");
     web_frame = webkit_web_view_get_main_frame (WEBKIT_WEB_VIEW (view->web_view));
