--- s/midori/midori-view.c	2012-02-20 14:15:20.282673443 +0000
+++ d/midori/midori-view.c	2012-02-20 14:18:37.297678427 +0000
@@ -30,9 +30,6 @@
 #include <gdk/gdkkeysyms.h>
 #include <webkit/webkit.h>
 
-/* This is unstable API, so we need to declare it */
-gchar*
-webkit_web_view_get_selected_text (WebKitWebView* web_view);
 /* This is public API since WebKitGTK+ 1.1.6 */
 #if !WEBKIT_CHECK_VERSION (1, 1, 6)
 void
@@ -2280,10 +2277,26 @@
 gboolean
 midori_view_has_selection (MidoriView* view)
 {
+    WebKitDOMDocument *doc;
+    WebKitDOMDOMWindow *window;
+    WebKitDOMDOMSelection *selection;
+    WebKitDOMRange *range;
+
     g_return_val_if_fail (MIDORI_IS_VIEW (view), FALSE);
 
-    katze_assign (view->selected_text, webkit_web_view_get_selected_text (
-        WEBKIT_WEB_VIEW (view->web_view)));
+    doc = webkit_web_view_get_dom_document (WEBKIT_WEB_VIEW (view->web_view));
+    window = webkit_dom_document_get_default_view (doc);
+
+    selection = webkit_dom_dom_window_get_selection (window);
+    if (selection == NULL)
+        return FALSE;
+
+    range = webkit_dom_dom_selection_get_range_at (selection, 0, NULL);
+    if (range == NULL)
+        return FALSE;
+
+    katze_assign (view->selected_text, webkit_dom_range_get_text (range));
+
     if (view->selected_text && *view->selected_text)
         return TRUE;
     else
