diff -uNr midori-0.1.8.orig/midori/midori-browser.c midori-0.1.8/midori/midori-browser.c
--- midori-0.1.8.orig/midori/midori-browser.c	2012-10-22 19:46:23.107486699 +0100
+++ midori-0.1.8/midori/midori-browser.c	2012-10-22 19:47:08.443490245 +0100
@@ -4394,6 +4394,8 @@
         }
     }
 
+    gtk_window_fullscreen (GTK_WINDOW (browser));
+
     _midori_browser_set_toolbar_style (browser, toolbar_style);
     _toggle_tabbar_smartly (browser);
     _midori_browser_set_toolbar_items (browser, toolbar_items);
