Index: src/tests/test-modem-helpers.c
diff --git a/src/tests/test-modem-helpers.c b/src/tests/test-modem-helpers.c
index 92a7af85d0914e3420247bcd25fd74c6ba343e8c..614eb38d8df2f9d5f531c3180f3562a7477e4b29 100644
--- a/src/tests/test-modem-helpers.c
+++ b/src/tests/test-modem-helpers.c
@@ -790,8 +790,11 @@ test_data_free (TestData *data)
     g_free (data);
 }
 
-
+#if GLIB_CHECK_VERSION (2, 26, 0)
+typedef GTestFixtureFunc TCFunc;
+#else
 typedef void (*TCFunc)(void);
+#endif
 
 #define TESTCASE(t, d) g_test_create_case (#t, 0, d, NULL, (TCFunc) t, NULL)
 
