Index: qemu-2.6.2/dma-helpers.c
===================================================================
--- qemu-2.6.2.orig/dma-helpers.c
+++ qemu-2.6.2/dma-helpers.c
@@ -136,7 +136,7 @@ static void dma_blk_cb(void *opaque, int
         dma_complete(dbs, ret);
         return;
     }
-    dma_blk_unmap(dbs);
+    //dma_blk_unmap(dbs);
 
     while (dbs->sg_cur_index < dbs->sg->nsg) {
         cur_addr = dbs->sg->sg[dbs->sg_cur_index].base + dbs->sg_cur_byte;
