diff --git a/Makefile b/Makefile
index 976584f..5520b3a 100755
--- a/Makefile
+++ b/Makefile
@@ -245,12 +245,12 @@ endif
 
 uninstall:
 ifeq ($(TARGET), LINUX)
-	$(MAKE) -C $(RT28xx_DIR)/os/linux -f Makefile.6 uninstall
+	$(MAKE) -C $(LINUX_SRC)/os/linux -f Makefile.6 uninstall
 endif
 
 install:
 ifeq ($(TARGET), LINUX)
-	$(MAKE) -C $(RT28xx_DIR)/os/linux -f Makefile.6 install
+	$(MAKE) -C $(LINUX_SRC) DEPMOD=echo INSTALL_MOD_PATH=$(INST_MOD_PATH) M=$(RT28xx_DIR)/os/linux modules_install
 endif
 
 # Declare the contents of the .PHONY variable as phony.  We keep that information in a variable
