Index: seabios-1.9.1/scripts/layoutrom.py
===================================================================
--- seabios-1.9.1.orig/scripts/layoutrom.py
+++ seabios-1.9.1/scripts/layoutrom.py
@@ -206,7 +206,7 @@ def doLayout(sections, config, genreloc)
 
     sections32fseg = getSectionsCategory(sections, '32fseg')
     sec32fseg_start, sec32fseg_align = setSectionsStart(
-        sections32fseg, sec32textfseg_start, 16
+        sections32fseg, sec32textfseg_start, 4096
         , segoffset=BUILD_BIOS_ADDR)
 
     # Determine 32flat runtime positions
@@ -218,7 +218,7 @@ def doLayout(sections, config, genreloc)
 
     sec32flat_start, sec32flat_align = setSectionsStart(
         textsections + rodatasections + datasections + bsssections
-        , sec32fseg_start, 16)
+        , sec32fseg_start, 4096)
 
     # Determine 32flat init positions
     sections32init = getSectionsCategory(sections, '32init')
@@ -230,7 +230,7 @@ def doLayout(sections, config, genreloc)
     sec32init_start, sec32init_align = setSectionsStart(
         init32_textsections + init32_rodatasections
         + init32_datasections + init32_bsssections
-        , sec32flat_start, 16)
+        , sec32flat_start, 4096)
 
     # Determine location of ZoneFSeg memory.
     zonefseg_end = sec32flat_start
@@ -243,11 +243,11 @@ def doLayout(sections, config, genreloc)
         zonefseg_start = zonefseg_end - BUILD_MIN_BIOSTABLE
         sec32flat_start, sec32flat_align = setSectionsStart(
             textsections + rodatasections + datasections + bsssections
-            , zonefseg_start, 16)
+            , zonefseg_start, 4096)
         sec32init_start, sec32init_align = setSectionsStart(
             init32_textsections + init32_rodatasections
             + init32_datasections + init32_bsssections
-            , sec32flat_start, 16)
+            , sec32flat_start, 4096)
     li.sec32init_start = sec32init_start
     li.sec32init_end = sec32flat_start
     li.sec32init_align = sec32init_align
@@ -270,7 +270,7 @@ def doLayout(sections, config, genreloc)
         zonelow_base = final_sec32low_end - 64*1024
     relocdelta = final_sec32low_end - sec32low_end
     li.sec32low_start, li.sec32low_align = setSectionsStart(
-        sections32low, sec32low_end, 16
+        sections32low, sec32low_end, 4096
         , segoffset=zonelow_base - relocdelta)
     li.sec32low_end = sec32low_end
     li.zonelow_base = zonelow_base
