diff --git a/surfman/src/xenstore-helper.c b/surfman/src/xenstore-helper.c
index dd548e0..bf23fff 100644
--- a/src/xenstore-helper.c
+++ b/src/xenstore-helper.c
@@ -242,7 +242,7 @@ bool vxenstore_chmod (const char *perms, unsigned int nbperm,
     if (!xs_strings_to_perms (p, nbperm, perms))
         goto xs_chmod_err;
 
-    if (!vasprintf (&buff, format, arg) == -1)
+    if (vasprintf (&buff, format, arg) == -1)
         goto xs_chmod_err;
 
     ret = xs_set_permissions (xs, xs_transaction, buff, p, nbperm);
