Index: git/fbtap.c
===================================================================
--- git.orig/fbtap.c	2015-06-21 14:06:00.151706445 +0200
+++ git/fbtap.c	2015-06-22 11:42:29.491873018 +0200
@@ -138,9 +138,12 @@
         return -EINVAL;
     }
 
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(3,18,0))
+    pgprot_val(vma->vm_page_prot) |= cachemode2protval(_PAGE_CACHE_MODE_WC);
+#else
     vma->vm_page_prot = __pgprot (((pgprot_val (vma->vm_page_prot)
                 | _PAGE_CACHE_WC) & ~_PAGE_PCD));
-
+#endif
     vma->vm_ops = &fbtap_vm_ops;
     vma->vm_private_data = fb;
 
@@ -151,7 +154,7 @@
     return 0;
 }
 
-static int
+static long
 fbtap_ioctl (struct file *filp, unsigned int cmd, unsigned long arg)
 {
     unsigned int p, num_pages;
@@ -256,6 +259,7 @@
     return -EFAULT;
 }
 
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,33))
 static int
 fbtap_ioctl_legacy (struct inode *inode, struct file *filp,
                     unsigned int cmd, unsigned long arg)
@@ -263,6 +267,8 @@
     return fbtap_ioctl (filp, cmd, arg);
 }
 
+#endif  /* (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,33)) */
+
 static ssize_t
 fbtap_read (struct file *filp, char __user *buf, size_t len, loff_t *offset)
 {
