--- a/src/connection-editor/ce-page-general.ui
+++ b/src/connection-editor/ce-page-general.ui
@@ -61,12 +61,12 @@
     <child>
       <object class="GtkCheckButton" id="system_checkbutton">
         <property name="label" translatable="yes">All _users may connect to this network</property>
-        <property name="visible">True</property>
-        <property name="can_focus">True</property>
+        <property name="visible">False</property>
+        <property name="can_focus">False</property>
         <property name="receives_default">False</property>
         <property name="margin_bottom">12</property>
-        <property name="use_underline">True</property>
-        <property name="draw_indicator">True</property>
+        <property name="use_underline">False</property>
+        <property name="draw_indicator">False</property>
       </object>
       <packing>
         <property name="top_attach">2</property>
