From 17ca712ffe926b2bc83fa5d95f23f544153b3267 Mon Sep 17 00:00:00 2001
From: Eric Chanudet <chanudete@ainfosec.com>
Date: Mon, 13 Aug 2018 13:54:04 -0400
Subject: [PATCH 08/14] tboot: Use SHA256 by default with TPM2.0.

Signed-off-by: Chris Rogers <rogersc@ainfosec.com>
---
 tboot/common/cmdline.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/tboot/common/cmdline.c b/tboot/common/cmdline.c
index b7e49c2..a3e7426 100644
--- a/tboot/common/cmdline.c
+++ b/tboot/common/cmdline.c
@@ -529,6 +529,11 @@ void get_tboot_extpol(void)
         tpm->extpol = TB_EXTPOL_FIXED;
         tpm->cur_alg = TB_HALG_SM3;
     }
+
+    if ( tpm->major == TPM20_VER_MAJOR ) {
+        tpm->extpol = TB_EXTPOL_FIXED;
+        tpm->cur_alg = TB_HALG_SHA256;
+    }
 }
 
 bool get_tboot_force_tpm2_legacy_log(void)
-- 
2.20.1

